<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="dcterms.date" content="2025-08-03">
<title>pythrash - Basic R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><link rel="stylesheet" href="../../_include/styles.css">
</head>
<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">PythRaSh</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-r" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">R</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-r">
<li>
    <a class="dropdown-item" href="../../ch/rbasics/git.html" rel="" target="">
 <span class="dropdown-text">git</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/firststeps.html" rel="" target="">
 <span class="dropdown-text">Basic R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/tidyverse.html" rel="" target="">
 <span class="dropdown-text">The tidyverse</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/correlation_regression.html" rel="" target="">
 <span class="dropdown-text">Advanced R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/baddata.html" rel="" target="">
 <span class="dropdown-text">Data Wrangling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/solutions.html" rel="" target="">
 <span class="dropdown-text">Problem Solutions</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-python" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Python</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-python">
<li>
    <a class="dropdown-item" href="../../ch/python/basics.html" rel="" target="">
 <span class="dropdown-text">Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/panda.html" rel="" target="">
 <span class="dropdown-text">Advanced Python</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/MLA.html" rel="" target="">
 <span class="dropdown-text">Machine Learning Algorithms</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/AI.html" rel="" target="">
 <span class="dropdown-text">AI</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-linux-and-clt" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Linux and CLT</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-linux-and-clt">
<li>
    <a class="dropdown-item" href="../../ch/linux-and-ctl/basics.html" rel="" target="">
 <span class="dropdown-text">Linux Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/linux-and-ctl/advanced.html" rel="" target="">
 <span class="dropdown-text">Advanced Command Line</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-probability-and-statistics" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Probability and Statistics</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-probability-and-statistics">
<li>
    <a class="dropdown-item" href="../../ch/prob_stat/prob_basic.html" rel="" target="">
 <span class="dropdown-text">Probability Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/prob_stat/stat.html" rel="" target="">
 <span class="dropdown-text">Statistics</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-genomics" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Genomics</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-genomics">
<li>
    <a class="dropdown-item" href="../../ch/genomics/basics.html" rel="" target="">
 <span class="dropdown-text">Genomics Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/genomics/advanced.html" rel="" target="">
 <span class="dropdown-text">Advanced Genomics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/genomics/pipeline.html" rel="" target="">
 <span class="dropdown-text">Pipelines</span></a>
  </li>  
    </ul>
</li>
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-blogs" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Blogs</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-blogs">
<li>
    <a class="dropdown-item" href="../../ch/blogs/en/index.html" rel="" target="">
 <span class="dropdown-text">English</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/blogs/bn/index.html" rel="" target="">
 <span class="dropdown-text">Bangla</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/MdRasheduz-zaman" rel="" target="">
 <span class="menu-text">Contact</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/MdRasheduz-zaman/PythRaSh" rel="" target=""><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#l2-data-representation" id="toc-l2-data-representation" class="nav-link active" data-scroll-target="#l2-data-representation">L2: Data Representation</a>
  <ul class="collapse">
<li><a href="#using-r-as-a-calculator" id="toc-using-r-as-a-calculator" class="nav-link" data-scroll-target="#using-r-as-a-calculator">Using R as a Calculator</a></li>
  <li>
<a href="#variables" id="toc-variables" class="nav-link" data-scroll-target="#variables">Variables</a>
  <ul class="collapse">
<li><a href="#integer-and-modulus-division-again" id="toc-integer-and-modulus-division-again" class="nav-link" data-scroll-target="#integer-and-modulus-division-again">Integer and Modulus division again</a></li>
  </ul>
</li>
  <li><a href="#rounding" id="toc-rounding" class="nav-link" data-scroll-target="#rounding">Rounding</a></li>
  <li><a href="#logical-operations" id="toc-logical-operations" class="nav-link" data-scroll-target="#logical-operations">Logical Operations</a></li>
  <li><a href="#help-and-documentation" id="toc-help-and-documentation" class="nav-link" data-scroll-target="#help-and-documentation">Help and Documentation</a></li>
  <li>
<a href="#working-with-vectors" id="toc-working-with-vectors" class="nav-link" data-scroll-target="#working-with-vectors">Working with Vectors</a>
  <ul class="collapse">
<li><a href="#vector-operations" id="toc-vector-operations" class="nav-link" data-scroll-target="#vector-operations">Vector Operations</a></li>
  </ul>
</li>
  <li>
<a href="#data-frame" id="toc-data-frame" class="nav-link" data-scroll-target="#data-frame">Data Frame</a>
  <ul class="collapse">
<li><a href="#gene-expression-table" id="toc-gene-expression-table" class="nav-link" data-scroll-target="#gene-expression-table">Gene Expression Table</a></li>
  </ul>
</li>
  <li>
<a href="#homeworks" id="toc-homeworks" class="nav-link" data-scroll-target="#homeworks">Homeworks</a>
  <ul class="collapse">
<li><a href="#deadline" id="toc-deadline" class="nav-link" data-scroll-target="#deadline">Deadline</a></li>
  </ul>
</li>
  </ul>
</li>
  <li>
<a href="#l3-data-transformation" id="toc-l3-data-transformation" class="nav-link" data-scroll-target="#l3-data-transformation">L3: Data Transformation</a>
  <ul class="collapse">
<li>
<a href="#getting-started" id="toc-getting-started" class="nav-link" data-scroll-target="#getting-started">Getting Started</a>
  <ul class="collapse">
<li><a href="#installation-of-r-markdown" id="toc-installation-of-r-markdown" class="nav-link" data-scroll-target="#installation-of-r-markdown">Installation of R Markdown</a></li>
  <li><a href="#basic-setup-for-todays-session" id="toc-basic-setup-for-todays-session" class="nav-link" data-scroll-target="#basic-setup-for-todays-session">Basic Setup for Today‚Äôs Session</a></li>
  <li><a href="#building-on-last-hw" id="toc-building-on-last-hw" class="nav-link" data-scroll-target="#building-on-last-hw">Building on Last HW:</a></li>
  <li><a href="#preamble-on-random-variables-rv" id="toc-preamble-on-random-variables-rv" class="nav-link" data-scroll-target="#preamble-on-random-variables-rv">Preamble on random variables (RV):</a></li>
  <li><a href="#basic-stuffs-atomic-vector" id="toc-basic-stuffs-atomic-vector" class="nav-link" data-scroll-target="#basic-stuffs-atomic-vector">Basic Stuffs: Atomic Vector</a></li>
  <li><a href="#basic-stuffs-matrices" id="toc-basic-stuffs-matrices" class="nav-link" data-scroll-target="#basic-stuffs-matrices">Basic Stuffs: Matrices</a></li>
  <li><a href="#basic-stuffs-list" id="toc-basic-stuffs-list" class="nav-link" data-scroll-target="#basic-stuffs-list">Basic Stuffs: List</a></li>
  </ul>
</li>
  <li>
<a href="#homeworks-matrix-and-list-operations" id="toc-homeworks-matrix-and-list-operations" class="nav-link" data-scroll-target="#homeworks-matrix-and-list-operations">üè° Homeworks: Matrix and List Operations</a>
  <ul class="collapse">
<li><a href="#protein-quantification-in-biological-samples" id="toc-protein-quantification-in-biological-samples" class="nav-link" data-scroll-target="#protein-quantification-in-biological-samples">üß™ Protein Quantification in Biological Samples</a></li>
  <li><a href="#tasks" id="toc-tasks" class="nav-link" data-scroll-target="#tasks">üîß Tasks</a></li>
  <li><a href="#interpretation-questions" id="toc-interpretation-questions" class="nav-link" data-scroll-target="#interpretation-questions">üß† Interpretation Questions</a></li>
  <li><a href="#gene-to-protein-translation" id="toc-gene-to-protein-translation" class="nav-link" data-scroll-target="#gene-to-protein-translation">üß¨ Gene-to-Protein Translation</a></li>
  <li><a href="#tasks-1" id="toc-tasks-1" class="nav-link" data-scroll-target="#tasks-1">üîß Tasks</a></li>
  <li><a href="#visualization-tasks" id="toc-visualization-tasks" class="nav-link" data-scroll-target="#visualization-tasks">üìä Visualization Tasks</a></li>
  <li><a href="#interpretation-questions-1" id="toc-interpretation-questions-1" class="nav-link" data-scroll-target="#interpretation-questions-1">üß† Interpretation Questions</a></li>
  <li><a href="#animal-breeding-economic-ranking-of-bulls-by-traits" id="toc-animal-breeding-economic-ranking-of-bulls-by-traits" class="nav-link" data-scroll-target="#animal-breeding-economic-ranking-of-bulls-by-traits">üêÑ Animal Breeding ‚Äì Economic Ranking of Bulls by Traits</a></li>
  <li><a href="#tasks-2" id="toc-tasks-2" class="nav-link" data-scroll-target="#tasks-2">üîß Tasks</a></li>
  <li><a href="#visualization-tasks-1" id="toc-visualization-tasks-1" class="nav-link" data-scroll-target="#visualization-tasks-1">üìä Visualization Tasks</a></li>
  <li><a href="#interpretation-questions-2" id="toc-interpretation-questions-2" class="nav-link" data-scroll-target="#interpretation-questions-2">üß† Interpretation Questions</a></li>
  <li><a href="#plant-breeding-trait-contributions-from-parental-lines" id="toc-plant-breeding-trait-contributions-from-parental-lines" class="nav-link" data-scroll-target="#plant-breeding-trait-contributions-from-parental-lines">üåæ Plant Breeding ‚Äì Trait Contributions from Parental Lines</a></li>
  <li><a href="#tasks-3" id="toc-tasks-3" class="nav-link" data-scroll-target="#tasks-3">üîß Tasks</a></li>
  <li><a href="#visualization-tasks-2" id="toc-visualization-tasks-2" class="nav-link" data-scroll-target="#visualization-tasks-2">üìä Visualization Tasks</a></li>
  <li><a href="#interpretation-questions-3" id="toc-interpretation-questions-3" class="nav-link" data-scroll-target="#interpretation-questions-3">üß† Interpretation Questions</a></li>
  <li><a href="#managing-matrices-and-weight-vectors-using-lists-in-r" id="toc-managing-matrices-and-weight-vectors-using-lists-in-r" class="nav-link" data-scroll-target="#managing-matrices-and-weight-vectors-using-lists-in-r">üß† Managing Matrices and Weight Vectors Using Lists in R</a></li>
  <li><a href="#step-1-create-a-master-list" id="toc-step-1-create-a-master-list" class="nav-link" data-scroll-target="#step-1-create-a-master-list">üì¶ Step 1: Create a master list</a></li>
  <li><a href="#tasks-4" id="toc-tasks-4" class="nav-link" data-scroll-target="#tasks-4">üîß Tasks</a></li>
  <li><a href="#visualization-tasks-3" id="toc-visualization-tasks-3" class="nav-link" data-scroll-target="#visualization-tasks-3">üìä Visualization Tasks</a></li>
  <li><a href="#interpretation-questions-4" id="toc-interpretation-questions-4" class="nav-link" data-scroll-target="#interpretation-questions-4">üß† Interpretation Questions</a></li>
  </ul>
</li>
  <li>
<a href="#factor-variables" id="toc-factor-variables" class="nav-link" data-scroll-target="#factor-variables">Factor Variables</a>
  <ul class="collapse">
<li><a href="#creating-factors" id="toc-creating-factors" class="nav-link" data-scroll-target="#creating-factors">Creating Factors</a></li>
  <li><a href="#factor-operations" id="toc-factor-operations" class="nav-link" data-scroll-target="#factor-operations">Factor Operations</a></li>
  </ul>
</li>
  <li>
<a href="#subsetting-data" id="toc-subsetting-data" class="nav-link" data-scroll-target="#subsetting-data">Subsetting Data</a>
  <ul class="collapse">
<li><a href="#vectors" id="toc-vectors" class="nav-link" data-scroll-target="#vectors">Vectors</a></li>
  <li><a href="#data-frames" id="toc-data-frames" class="nav-link" data-scroll-target="#data-frames">Data Frames</a></li>
  <li><a href="#row-names-in-data-frames" id="toc-row-names-in-data-frames" class="nav-link" data-scroll-target="#row-names-in-data-frames">Row Names in Data Frames</a></li>
  </ul>
</li>
  <li><a href="#homeworks-factors-subsetting-and-biological-insight" id="toc-homeworks-factors-subsetting-and-biological-insight" class="nav-link" data-scroll-target="#homeworks-factors-subsetting-and-biological-insight">üè° Homeworks: Factors, Subsetting, and Biological Insight</a></li>
  <li>
<a href="#conditionals" id="toc-conditionals" class="nav-link" data-scroll-target="#conditionals">Conditionals</a>
  <ul class="collapse">
<li><a href="#if-else-statement" id="toc-if-else-statement" class="nav-link" data-scroll-target="#if-else-statement"><code>if-else</code> statement</a></li>
  <li><a href="#ifelse-statement-for-vectors" id="toc-ifelse-statement-for-vectors" class="nav-link" data-scroll-target="#ifelse-statement-for-vectors"><code>ifelse</code> statement for vectors</a></li>
  <li><a href="#for-loop" id="toc-for-loop" class="nav-link" data-scroll-target="#for-loop"><code>for</code> loop</a></li>
  <li><a href="#while-loop" id="toc-while-loop" class="nav-link" data-scroll-target="#while-loop"><code>while</code> loop</a></li>
  <li><a href="#next-and-break" id="toc-next-and-break" class="nav-link" data-scroll-target="#next-and-break"><code>next</code> and <code>break</code></a></li>
  </ul>
</li>
  <li>
<a href="#writing-functions-in-r" id="toc-writing-functions-in-r" class="nav-link" data-scroll-target="#writing-functions-in-r">Writing Functions in R</a>
  <ul class="collapse">
<li><a href="#flag-gene-expression" id="toc-flag-gene-expression" class="nav-link" data-scroll-target="#flag-gene-expression">Flag gene expression</a></li>
  <li><a href="#function-with-multiple-arguments" id="toc-function-with-multiple-arguments" class="nav-link" data-scroll-target="#function-with-multiple-arguments">Function with multiple arguments</a></li>
  <li><a href="#return-a-list" id="toc-return-a-list" class="nav-link" data-scroll-target="#return-a-list">Return a list</a></li>
  </ul>
</li>
  <li><a href="#homeworks-control-structures-and-functions" id="toc-homeworks-control-structures-and-functions" class="nav-link" data-scroll-target="#homeworks-control-structures-and-functions">üè° Homeworks: Control Structures and Functions</a></li>
  <li>
<a href="#handling-missingwrong-values" id="toc-handling-missingwrong-values" class="nav-link" data-scroll-target="#handling-missingwrong-values">Handling Missing/Wrong Values</a>
  <ul class="collapse">
<li><a href="#identifying-issues" id="toc-identifying-issues" class="nav-link" data-scroll-target="#identifying-issues">Identifying Issues</a></li>
  <li><a href="#fixing-data" id="toc-fixing-data" class="nav-link" data-scroll-target="#fixing-data">Fixing Data</a></li>
  </ul>
</li>
  <li>
<a href="#data-transformation" id="toc-data-transformation" class="nav-link" data-scroll-target="#data-transformation">Data Transformation</a>
  <ul class="collapse">
<li><a href="#introduction-to-outliers" id="toc-introduction-to-outliers" class="nav-link" data-scroll-target="#introduction-to-outliers">Introduction to Outliers</a></li>
  <li><a href="#identifying-outliers" id="toc-identifying-outliers" class="nav-link" data-scroll-target="#identifying-outliers">Identifying Outliers</a></li>
  <li><a href="#transforming-vectors" id="toc-transforming-vectors" class="nav-link" data-scroll-target="#transforming-vectors">Transforming Vectors</a></li>
  <li><a href="#logical-expressions" id="toc-logical-expressions" class="nav-link" data-scroll-target="#logical-expressions">Logical Expressions</a></li>
  <li><a href="#logical-operators" id="toc-logical-operators" class="nav-link" data-scroll-target="#logical-operators">Logical Operators</a></li>
  <li><a href="#logical-functions" id="toc-logical-functions" class="nav-link" data-scroll-target="#logical-functions">Logical Functions</a></li>
  <li><a href="#practical-session" id="toc-practical-session" class="nav-link" data-scroll-target="#practical-session">Practical Session</a></li>
  </ul>
</li>
  <li><a href="#summary-of-the-lesson" id="toc-summary-of-the-lesson" class="nav-link" data-scroll-target="#summary-of-the-lesson">Summary of the Lesson</a></li>
  <li><a href="#homework" id="toc-homework" class="nav-link" data-scroll-target="#homework">Homework</a></li>
  </ul>
</li>
  </ul><div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="firststeps.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Basic R</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Md Rasheduzzaman </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://www.fli.de/de/institute/institut-fuer-infektionsmedizin-imed/">
            FLI Insel Riems
            </a>
          </p>
        <p class="affiliation">
            <a href="https://github.com/MdRasheduz-zaman">
            &amp; Freelancer
            </a>
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Last updated</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 3, 2025</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="abstract-title">Content summary</div>
    Data types, variables, vectors, data frame, functions
  </div>
</div>

</header><section id="l2-data-representation" class="level1"><h1>L2: Data Representation</h1>
<section id="using-r-as-a-calculator" class="level2"><h2 class="anchored" data-anchor-id="using-r-as-a-calculator">Using R as a Calculator</h2>
<p>Let‚Äôs do some basic calculation.</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">5</span><span class="op">+</span><span class="fl">3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 8</code></pre>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">3</span><span class="op">+</span><span class="fl">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 5</code></pre>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">3</span><span class="op">-</span><span class="fl">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 1</code></pre>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">3</span><span class="op">*</span><span class="fl">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 6</code></pre>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">3</span><span class="op">/</span><span class="fl">2</span> <span class="co">#normal division</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 1.5</code></pre>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">7</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html">%/%</a></span> <span class="fl">2</span> <span class="co">#integer division, only the quotient</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 3</code></pre>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">5</span> <span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html">%%</a></span> <span class="fl">3</span> <span class="co">#modulus division, the remainder</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 2</code></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="op">(</span><span class="fl">10</span><span class="op">-</span><span class="fl">5</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="fl">2</span><span class="op">+</span><span class="fl">4</span><span class="op">)</span> <span class="co">#use of parentheses</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 30</code></pre>
</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">10</span><span class="op">-</span><span class="fl">5</span><span class="op">*</span><span class="fl">2</span><span class="op">+</span><span class="fl">4</span> <span class="co">#Noticed BODMAS?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 4</code></pre>
</div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="op">(</span><span class="fl">10</span><span class="op">-</span><span class="fl">5</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="fl">2</span><span class="op">+</span><span class="fl">4</span><span class="op">)</span> <span class="co">#Noticed BODMAS</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 30</code></pre>
</div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">7</span><span class="op">/</span><span class="op">(</span><span class="fl">1</span><span class="op">+</span><span class="fl">3</span><span class="op">)</span>; <span class="fl">7</span><span class="op">/</span><span class="fl">1</span><span class="op">+</span><span class="fl">3</span> <span class="co">#multi-line codes, separated with semi-colon</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 1.75</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 10</code></pre>
</div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">1</span><span class="op">+</span><span class="fl">2</span>; <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>; <span class="fl">1</span><span class="op">/</span><span class="fl">10</span> <span class="co">#more multi-line codes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 3</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 0</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 0.1</code></pre>
</div>
</div>
</section><section id="variables" class="level2"><h2 class="anchored" data-anchor-id="variables">Variables</h2>
<p>Variables are variable. We have freedom to name them as we wish. But make any variable name meaningful and identifiable.</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fl">5</span> <span class="co">#assign value 5 to a </span></span>
<span><span class="va">b</span> <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="va">a</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 5</code></pre>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">b</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 10</code></pre>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="va">a</span> <span class="op">+</span> <span class="fl">10</span></span>
<span><span class="va">b</span> <span class="op">=</span> <span class="va">b</span> <span class="op">+</span> <span class="fl">15</span></span>
<span><span class="va">a</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 15</code></pre>
</div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span><span class="op">^</span><span class="fl">2</span> <span class="co">#a squared</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 225</code></pre>
</div>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span><span class="op">**</span><span class="fl">2</span> <span class="co">#a squared again, in a different way.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 225</code></pre>
</div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span><span class="op">^</span><span class="fl">3</span> <span class="co">#a qubed</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 3375</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>&lt;-</code> and <code>=</code> are used to assign values. It is not mathematical equality. <code>b &lt;- b + 15</code> might make better sense than <code>b = b + 15</code>.</p>
</div>
</div>
<section id="integer-and-modulus-division-again" class="level3"><h3 class="anchored" data-anchor-id="integer-and-modulus-division-again">Integer and Modulus division again</h3>
<p>Do some more practice.</p>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">7</span><span class="op">/</span><span class="fl">3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 2.333333</code></pre>
</div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">7</span><span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html">%/%</a></span><span class="fl">3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 2</code></pre>
</div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">7</span><span class="op"><a href="https://rdrr.io/r/base/Arithmetic.html">%%</a></span><span class="fl">3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 1</code></pre>
</div>
</div>
</section></section><section id="rounding" class="level2"><h2 class="anchored" data-anchor-id="rounding">Rounding</h2>
<p>Some important functions we apply on numerical values</p>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">9</span><span class="op">/</span><span class="fl">4</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">floor</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 2</code></pre>
</div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">ceiling</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 3</code></pre>
</div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 2</code></pre>
</div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">x</span>, <span class="fl">2</span><span class="op">)</span> <span class="co">#round till 2 decimal points</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 2.25</code></pre>
</div>
</div>
</section><section id="logical-operations" class="level2"><h2 class="anchored" data-anchor-id="logical-operations">Logical Operations</h2>
<p>Get to know TRUE/FALSE in R.</p>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="va">b</span> <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="va">c</span> <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="va">d</span> <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="va">a</span> <span class="op">==</span> <span class="va">b</span> <span class="co">#is a equal to b? Ans: No/FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] FALSE</code></pre>
</div>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">!=</span> <span class="va">b</span> <span class="co">#is a not equal to b? Ans: Yes/TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">&gt;</span> <span class="va">b</span> <span class="co">#is a greater than b? Ans: FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] FALSE</code></pre>
</div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;</span> <span class="va">b</span> <span class="co">#is a less than b? Ans: TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">&gt;=</span> <span class="va">b</span> <span class="co">#is a greater than or equal to b? Ans: FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] FALSE</code></pre>
</div>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;=</span> <span class="va">b</span> <span class="co">#is a less than or equal to b? Ans: TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;</span> <span class="va">b</span> <span class="op">|</span> <span class="va">d</span> <span class="op">&gt;</span> <span class="va">b</span> <span class="co">#is a less than b OR d greater than b?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#It's answer will be TRUE OR FALSE --&gt; So, TRUE</span></span>
<span><span class="va">a</span> <span class="op">&lt;</span> <span class="va">b</span> <span class="op">&amp;</span> <span class="va">c</span> <span class="op">&gt;</span> <span class="va">d</span> <span class="co">#is a less than b AND a greater than b? It's answer will be TRUE AND TRUE --&gt; So, TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;</span> <span class="va">b</span> <span class="op">&amp;</span> <span class="va">d</span> <span class="op">&gt;</span> <span class="va">c</span> <span class="co">#is a less than b AND a greater than b? It's answer will be TRUE AND FALSE --&gt; So, FALSE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] FALSE</code></pre>
</div>
</div>
</section><section id="help-and-documentation" class="level2"><h2 class="anchored" data-anchor-id="help-and-documentation">Help and Documentation</h2>
<p>But how to know more about a function? The package/library developer have written helpful documentation for us.</p>
<div class="cell">
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="op">?</span><span class="va">log</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/example.html">example</a></span><span class="op">(</span><span class="va">log</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>
log&gt; log(exp(3))
[1] 3

log&gt; log10(1e7) # = 7
[1] 7

log&gt; x &lt;- 10^-(1+2*1:9)

log&gt; cbind(deparse.level=2, # to get nice column names
log+       x, log(1+x), log1p(x), exp(x)-1, expm1(x))
          x   log(1 + x)     log1p(x)   exp(x) - 1     expm1(x)
 [1,] 1e-03 9.995003e-04 9.995003e-04 1.000500e-03 1.000500e-03
 [2,] 1e-05 9.999950e-06 9.999950e-06 1.000005e-05 1.000005e-05
 [3,] 1e-07 1.000000e-07 1.000000e-07 1.000000e-07 1.000000e-07
 [4,] 1e-09 1.000000e-09 1.000000e-09 1.000000e-09 1.000000e-09
 [5,] 1e-11 1.000000e-11 1.000000e-11 1.000000e-11 1.000000e-11
 [6,] 1e-13 9.992007e-14 1.000000e-13 9.992007e-14 1.000000e-13
 [7,] 1e-15 1.110223e-15 1.000000e-15 1.110223e-15 1.000000e-15
 [8,] 1e-17 0.000000e+00 1.000000e-17 0.000000e+00 1.000000e-17
 [9,] 1e-19 0.000000e+00 1.000000e-19 0.000000e+00 1.000000e-19</code></pre>
</div>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="op">?</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="working-with-vectors" class="level2"><h2 class="anchored" data-anchor-id="working-with-vectors">Working with Vectors</h2>
<p>What is a vector? See the example and think.</p>
<div class="cell">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span><span class="op">)</span> <span class="co">#c means concatenate</span></span>
<span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span> <span class="co">#consecutively, from 1 through 5. A short-hand notation using :</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">9</span>, <span class="fl">12</span>, <span class="fl">15</span>, <span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 5</code></pre>
</div>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mode.html">mode</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "numeric"</code></pre>
</div>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">is</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "numeric" "vector" </code></pre>
</div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="co">#first entry in vector y</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 1</code></pre>
</div>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">x</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="co">#2nd to 5th entries in vector y</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 2 3 4 5</code></pre>
</div>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">DNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"C"</span><span class="op">)</span> <span class="co">#character vector. Notice the quotation marks.</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10.0</span>, <span class="fl">20.5</span>, <span class="fl">30</span>, <span class="fl">60</span>, <span class="fl">80.9</span>, <span class="fl">90</span>, <span class="fl">100.7</span>, <span class="fl">50</span>, <span class="fl">40</span>, <span class="fl">45</span>, <span class="fl">48</span>, <span class="fl">56</span>, <span class="fl">55</span><span class="op">)</span> <span class="co">#vector of floats. All numbers became floats, it's called coercion</span></span>
<span><span class="va">dec</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">7</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">dec</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="co">#1st to 3rd and then 7th till the end of vector `dec`. Output as a vector.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code> [1]  10.0  20.5  30.0 100.7  50.0  40.0  45.0  48.0  56.0  55.0</code></pre>
</div>
</div>
<section id="vector-operations" class="level3"><h3 class="anchored" data-anchor-id="vector-operations">Vector Operations</h3>
<p>Notice the element-wise or index-wise mathematical operations (<code>+</code>, <code>/</code>, <code><a href="https://rdrr.io/r/base/Log.html">log2()</a></code>, <code><a href="https://rdrr.io/r/base/Round.html">round()</a></code>, etc.). Noticed?</p>
<div class="cell">
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">:</span><span class="fl">11</span></span>
<span><span class="co">#x and y are of same length</span></span>
<span><span class="va">x</span> <span class="op">+</span> <span class="va">y</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code> [1]  3  5  7  9 11 13 15 17 19 21</code></pre>
</div>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">y</span> <span class="op">/</span> <span class="va">x</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code> [1] 2.000000 1.500000 1.333333 1.250000 1.200000 1.166667 1.142857 1.125000
 [9] 1.111111 1.100000</code></pre>
</div>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code> [1] 0.000000 1.000000 1.584963 2.000000 2.321928 2.584963 2.807355 3.000000
 [9] 3.169925 3.321928</code></pre>
</div>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span> <span class="co">#log2 of all the values of `x`, 1 digit after decimal to round.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code> [1] 0.0 1.0 1.6 2.0 2.3 2.6 2.8 3.0 3.2 3.3</code></pre>
</div>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span> <span class="co">#same logic</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code> [1] 0.000 1.000 1.585 2.000 2.322 2.585 2.807 3.000 3.170 3.322</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Nested functions work inside out. Think again about <code>round(log2(x), 1)</code> and you will see it. At first, it is making <code>log2</code> of vector <code>x</code> and then it is rounding the log2 values to one digit after decimal. Got it?</p>
</div>
</div>
</section></section><section id="data-frame" class="level2"><h2 class="anchored" data-anchor-id="data-frame">Data Frame</h2>
<p>Now, it‚Äôs time to use vectors to make data sets‚Ä¶..</p>
<div class="cell">
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mina"</span>, <span class="st">"Raju"</span>, <span class="st">"Mithu"</span>, <span class="st">"Lali"</span><span class="op">)</span></span>
<span><span class="va">gender</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Female"</span>, <span class="st">"Male"</span>, <span class="st">"Female"</span>, <span class="st">"Female"</span><span class="op">)</span></span>
<span><span class="va">age</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">12</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">is_human</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">cartoon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">names</span>, <span class="va">gender</span>, <span class="va">age</span>, <span class="va">is_human</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">cartoon</span>, <span class="st">"cartoon.csv"</span>, sep <span class="op">=</span> <span class="st">","</span>, col.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"cartoon.csv"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="co">#`dim` means dimension. so, rows * columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 4 4</code></pre>
</div>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="co">#structure of `df`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>'data.frame':   4 obs. of  4 variables:
 $ names   : chr  "Mina" "Raju" "Mithu" "Lali"
 $ gender  : chr  "Female" "Male" "Female" "Female"
 $ age     : int  15 12 2 3
 $ is_human: logi  TRUE TRUE FALSE FALSE</code></pre>
</div>
</div>
<p>We made the vectors first, and the used them to make the <code>cartton</code> data frame or table. We learned how to export the data frame using write.table function. Also, we learned to import or read back the table using <code>read.table</code> function. What are the <code>sep</code>, <code>col.names</code>, <code>header</code> arguments there? Why do we need them? Think. Try thinking of different properties of a data set.</p>
<section id="gene-expression-table" class="level3"><h3 class="anchored" data-anchor-id="gene-expression-table">Gene Expression Table</h3>
<div class="cell">
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"MYC"</span>, <span class="st">"EGFR"</span>, <span class="st">"GAPDH"</span>, <span class="st">"CDC2"</span><span class="op">)</span>,</span>
<span>  sample1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8.2</span>, <span class="fl">6.1</span>, <span class="fl">9.5</span>, <span class="fl">7.0</span>, <span class="fl">10.0</span>, <span class="fl">12</span><span class="op">)</span>,</span>
<span>  Sample2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5.9</span>, <span class="fl">3.9</span>, <span class="fl">7.2</span>, <span class="fl">4.8</span>, <span class="fl">7.9</span>, <span class="fl">9</span><span class="op">)</span>,</span>
<span>  Sample3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8.25</span>, <span class="fl">6.15</span>, <span class="fl">9.6</span>, <span class="fl">7.1</span>, <span class="fl">10.1</span>, <span class="fl">11.9</span><span class="op">)</span>,</span>
<span>  pathways <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Apoptosis"</span>, <span class="st">"DNA Repair"</span>, <span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>, <span class="st">"Housekeeping"</span>, <span class="st">"Cell Division"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">gene_expr</span>, <span class="st">"gene_expr.csv"</span>, sep <span class="op">=</span> <span class="st">","</span>, col.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">gene_set</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"gene_expr.csv"</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, sep <span class="op">=</span> <span class="st">","</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Here, we directly used the vectors as different columns while making the data frame. Did you notice that? Also, the syntax is different here. We can‚Äôt assign the vectors with the assignment operator (means we can‚Äôt use <code>&lt;-</code> sign. We have to use the <code>=</code> sign). Try using the <code>&lt;-</code> sign. Did you notice the column names?</p>
</div>
</div>
</section></section><section id="homeworks" class="level2"><h2 class="anchored" data-anchor-id="homeworks">Homeworks</h2>
<ol type="1">
<li><p>Compute the difference between this year (2025) and the year you started at the university and divide this by the difference between this year and the year you were born. Multiply this with 100 to get the percentage of your life you have spent at the university.</p></li>
<li><p>Make different kinds of variables and vectors with the data types we learned together.</p></li>
<li><p>What are the properties of a data frame?</p></li>
</ol>
<p>Hint: Open an excel/csv/txt file you have and try to ‚Äúgeneralize‚Äù.</p>
<ol start="4" type="1">
<li><p>Can you make logical questions on the 2 small data sets we used? Try. It will help you understanding the logical operations we tried on variables. Now we are going to apply them on vectors (columns) on the data sets. For example, in the <code>cartoon</code> data set, we can ask/try to subset the data set filtering for females only, or for both females and age greater than 2 years.</p></li>
<li><p>If you are writing or practicing coding in R, write comment for each line on what it is doing. It will help to chunk it better into your brain.</p></li>
<li><p>Push the script and/or your answers to the questions (with your solutions) to one of your GitHub repo (and send me the repo link).</p></li>
</ol>
<section id="deadline" class="level3"><h3 class="anchored" data-anchor-id="deadline">Deadline</h3>
<p><strong>Friday, 10pm BD Time.</strong></p>
</section></section></section><section id="l3-data-transformation" class="level1"><h1>L3: Data Transformation</h1>
<p>Firstly, how did you solve the problems?</p>
<p>Give me your personal Mindmap. Please, send it in the chat!</p>
<section id="getting-started" class="level2"><h2 class="anchored" data-anchor-id="getting-started">Getting Started</h2>
<section id="installation-of-r-markdown" class="level3"><h3 class="anchored" data-anchor-id="installation-of-r-markdown">Installation of R Markdown</h3>
<p>We will use <code>rmarkdown</code> to have the flexibility of writing codes like the one you are reading now. If you haven‚Äôt installed the <code>rmarkdown</code> package yet, you can do so with:</p>
<div class="cell">
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Install rmarkdown package</span></span>
<span><span class="co">#install.packages("rmarkdown")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/rmarkdown">rmarkdown</a></span><span class="op">)</span></span>
<span><span class="co"># Other useful packages we might use</span></span>
<span><span class="co">#install.packages("dplyr")    # Data manipulation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#install.packages("readr")    # Reading CSV files</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Remove the hash sign before the <code>install.packages("rmarkdown")</code>, <code>install.packages("dplyr")</code>, <code>install.packages("readr")</code> if the library loading fails. That means the package is not there to be loaded. We need to download/install first.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><a href="https://r4ds.had.co.nz">Do you remember this book by Hadley Wickham?</a>. Try to follow it to get the hold on the basic R syntax and lexicon.</p>
</div>
</div>
</section><section id="basic-setup-for-todays-session" class="level3"><h3 class="anchored" data-anchor-id="basic-setup-for-todays-session">Basic Setup for Today‚Äôs Session</h3>
<div class="cell">
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Clear environment</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span>list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check working directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html">getwd</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set working directory if needed</span></span>
<span><span class="co"># setwd("path/to/your/directory")  # Uncomment and modify as needed</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="building-on-last-hw" class="level3"><h3 class="anchored" data-anchor-id="building-on-last-hw">Building on Last HW:</h3>
<div class="cell">
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cartoon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mina"</span>, <span class="st">"Raju"</span>, <span class="st">"Mithu"</span>, <span class="st">"Lali"</span><span class="op">)</span>,</span>
<span>  gender <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Female"</span>, <span class="st">"Male"</span>, <span class="st">"Female"</span>, <span class="st">"Female"</span><span class="op">)</span>,</span>
<span>  age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">12</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  is_human <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">cartoon</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  names gender age is_human
1  Mina Female  15     TRUE
2  Raju   Male  12     TRUE
3 Mithu Female   2    FALSE
4  Lali Female   3    FALSE</code></pre>
</div>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">cartoon</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 4 4</code></pre>
</div>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">cartoon</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>'data.frame':   4 obs. of  4 variables:
 $ names   : chr  "Mina" "Raju" "Mithu" "Lali"
 $ gender  : chr  "Female" "Male" "Female" "Female"
 $ age     : num  15 12 2 3
 $ is_human: logi  TRUE TRUE FALSE FALSE</code></pre>
</div>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">cartoon</span><span class="op">$</span><span class="va">names</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 4</code></pre>
</div>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">##subseting</span></span>
<span><span class="va">cartoon</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="co">#row 1-2, column 2-3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  gender age
1 Female  15
2   Male  12</code></pre>
</div>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cartoon</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">]</span> <span class="co">#row 1-3, column 1-3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  names gender age
1  Mina Female  15
3 Mithu Female   2</code></pre>
</div>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#condition for selecting only male characters</span></span>
<span><span class="va">male_df</span> <span class="op">&lt;-</span> <span class="va">cartoon</span><span class="op">[</span><span class="va">cartoon</span><span class="op">$</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"Male"</span>, <span class="op">]</span></span>
<span><span class="va">male_df</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  names gender age is_human
2  Raju   Male  12     TRUE</code></pre>
</div>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#condition for selecting female characters with age more than 2 years</span></span>
<span><span class="va">female_age</span> <span class="op">&lt;-</span> <span class="va">cartoon</span><span class="op">[</span><span class="va">cartoon</span><span class="op">$</span><span class="va">gender</span> <span class="op">==</span> <span class="st">"Female"</span> <span class="op">&amp;</span> <span class="va">cartoon</span><span class="op">$</span><span class="va">age</span> <span class="op">&gt;</span> <span class="fl">2</span>, <span class="op">]</span></span>
<span><span class="va">female_age</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  names gender age is_human
1  Mina Female  15     TRUE
4  Lali Female   3    FALSE</code></pre>
</div>
<div class="sourceCode" id="cb120"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">female_age</span><span class="op">$</span><span class="va">age</span><span class="op">)</span> <span class="co">#sum of age of female_age dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 18</code></pre>
</div>
<div class="sourceCode" id="cb122"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">cartoon</span><span class="op">$</span><span class="va">age</span><span class="op">)</span> <span class="co">#standard deviation of age of main cartoon dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 6.480741</code></pre>
</div>
<div class="sourceCode" id="cb124"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">cartoon</span><span class="op">$</span><span class="va">age</span><span class="op">)</span> <span class="co">#mean of age of main cartoon dataset</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 8</code></pre>
</div>
</div>
<p><a href="https://github.com/atiyashehreen/hw_repo3/blob/main/propertiesofdataframe.txt">Check your colleague‚Äôs repo for the Q3</a>.</p>
<p><strong>Logical Operators</strong></p>
<table class="table">
<thead><tr class="header">
<th>Operator</th>
<th>Meaning</th>
<th>Example</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>==</code></td>
<td>Equal to</td>
<td><code>x == 5</code></td>
</tr>
<tr class="even">
<td><code>!=</code></td>
<td>Not equal</td>
<td><code>x != 5</code></td>
</tr>
<tr class="odd">
<td><code>&lt;</code></td>
<td>Less than</td>
<td><code>x &lt; 5</code></td>
</tr>
<tr class="even">
<td><code>&gt;</code></td>
<td>Greater than</td>
<td><code>x &gt; 5</code></td>
</tr>
<tr class="odd">
<td><code>&lt;=</code></td>
<td>Less or equal</td>
<td><code>x &lt;= 5</code></td>
</tr>
<tr class="even">
<td><code>&gt;=</code></td>
<td>Greater or equal</td>
<td><code>x &gt;= 5</code></td>
</tr>
<tr class="odd">
<td><code>!</code></td>
<td>Not</td>
<td><code>!(x &lt; 5)</code></td>
</tr>
<tr class="even">
<td><code>|</code></td>
<td>OR</td>
<td><code>x &lt; 5 | x &gt; 10</code></td>
</tr>
<tr class="odd">
<td><code>&amp;</code></td>
<td>AND</td>
<td><code>x &gt; 5 &amp; x &lt; 10</code></td>
</tr>
</tbody>
</table></section><section id="preamble-on-random-variables-rv" class="level3"><h3 class="anchored" data-anchor-id="preamble-on-random-variables-rv">Preamble on random variables (RV):</h3>
<p>RV is so fundamental of an idea to interpret and do better in any kind of data analyses. But what is it? Let‚Äôs imagine this scenario first. You got 30 mice to do an experiment to check anti-diabetic effect of a plant extract. You randomly assigned them into 3 groups. <code>control</code>, <code>treat1</code> (meaning insulin receivers), and <code>treat2</code> (meaning your plant extract receivers). Then you kept testing and measuring. You have mean glucose level of every mouse and show whether the mean value of <code>treat1</code> is equal to <code>treat2</code> or not. So, are you done? Not really. Be fastidious about the mice. What if you got some other 30 mice? Are they the same? Will their mean glucose level be the same? No, right. We would end up with different mean value. We call this type of quantities RV. Mean, Standard deviation, median, variance, etc. all are RVs. Do you see the logic? That‚Äôs why we put this constraint and look for p-value, confidence interval (or CI), etc. by (null) hypothesis testing and sample distribution analyses. We will get into these stuffs later. But let‚Äôs check what I meant. Also ponder about <code>sample</code> vs <code>population</code>.</p>
<p>Let‚Äôs download the data first.</p>
<div class="cell">
<div class="sourceCode" id="cb126"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Download small example dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/femaleControlsPopulation.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"mice.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load data</span></span>
<span><span class="va">mice</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"mice.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let‚Äôs check now.</p>
<div class="cell">
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">control</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">mice</span><span class="op">$</span><span class="va">Bodyweight</span>,<span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">control</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 24.755</code></pre>
</div>
<div class="sourceCode" id="cb129"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">control1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">mice</span><span class="op">$</span><span class="va">Bodyweight</span>,<span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">control1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 23.6525</code></pre>
</div>
<div class="sourceCode" id="cb131"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">control2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">mice</span><span class="op">$</span><span class="va">Bodyweight</span>,<span class="fl">12</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">control2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 23.23</code></pre>
</div>
</div>
<p>Do you see the difference in the mean value now?</p>
</section><section id="basic-stuffs-atomic-vector" class="level3"><h3 class="anchored" data-anchor-id="basic-stuffs-atomic-vector">Basic Stuffs: Atomic Vector</h3>
<div class="cell">
<div class="sourceCode" id="cb133"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">atomic_vec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>Human<span class="op">=</span><span class="fl">0.5</span>, Mouse<span class="op">=</span><span class="fl">0.33</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>It is fast, but has limited access methods.</p>
<p>How to access elements here?</p>
<div class="cell">
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">atomic_vec</span><span class="op">[</span><span class="st">"Human"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Human 
  0.5 </code></pre>
</div>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">atomic_vec</span><span class="op">[</span><span class="st">"Mouse"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Mouse 
 0.33 </code></pre>
</div>
</div>
</section><section id="basic-stuffs-matrices" class="level3"><h3 class="anchored" data-anchor-id="basic-stuffs-matrices">Basic Stuffs: Matrices</h3>
<p>Matrices are essential for biologists working with expression data, distance matrices, and other numerical data.</p>
<div class="cell">
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create a gene expression matrix: rows=genes, columns=samples</span></span>
<span><span class="va">expr_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">12.3</span>, <span class="fl">8.7</span>, <span class="fl">15.2</span>, <span class="fl">6.8</span>,</span>
<span>    <span class="fl">9.5</span>, <span class="fl">11.2</span>, <span class="fl">13.7</span>, <span class="fl">7.4</span>,</span>
<span>    <span class="fl">5.6</span>, <span class="fl">6.8</span>, <span class="fl">7.9</span>, <span class="fl">6.5</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">3</span>, ncol <span class="op">=</span> <span class="fl">4</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add dimension names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"GAPDH"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control_1"</span>, <span class="st">"Control_2"</span>, <span class="st">"Treatment_1"</span>, <span class="st">"Treatment_2"</span><span class="op">)</span></span>
<span><span class="va">expr_matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Control_1 Control_2 Treatment_1 Treatment_2
BRCA1      12.3       8.7        15.2         6.8
TP53        9.5      11.2        13.7         7.4
GAPDH       5.6       6.8         7.9         6.5</code></pre>
</div>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Matrix dimensions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span>       <span class="co"># Returns rows and columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 3 4</code></pre>
</div>
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span>      <span class="co"># Number of rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 3</code></pre>
</div>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span>      <span class="co"># Number of columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 4</code></pre>
</div>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Matrix subsetting</span></span>
<span><span class="va">expr_matrix</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span>       <span class="co"># One gene, all samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  Control_1   Control_2 Treatment_1 Treatment_2 
        9.5        11.2        13.7         7.4 </code></pre>
</div>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expr_matrix</span><span class="op">[</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>     <span class="co"># All genes, treatment samples only</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Treatment_1 Treatment_2
BRCA1        15.2         6.8
TP53         13.7         7.4
GAPDH         7.9         6.5</code></pre>
</div>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expr_matrix</span><span class="op">[</span><span class="st">"TP53"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control_1"</span>, <span class="st">"Treatment_1"</span><span class="op">)</span><span class="op">]</span>  <span class="co"># Specific gene and samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  Control_1 Treatment_1 
        9.5        13.7 </code></pre>
</div>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Matrix calculations (useful for bioinformatics)</span></span>
<span><span class="co"># Mean expression per gene</span></span>
<span><span class="va">gene_means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span></span>
<span><span class="va">gene_means</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>BRCA1  TP53 GAPDH 
10.75 10.45  6.70 </code></pre>
</div>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Mean expression per sample</span></span>
<span><span class="va">sample_means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colMeans</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">)</span></span>
<span><span class="va">sample_means</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  Control_1   Control_2 Treatment_1 Treatment_2 
   9.133333    8.900000   12.266667    6.900000 </code></pre>
</div>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Calculate fold change (Treatment vs Control)</span></span>
<span><span class="va">control_means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">treatment_means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">expr_matrix</span><span class="op">[</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">fold_change</span> <span class="op">&lt;-</span> <span class="va">treatment_means</span> <span class="op">/</span> <span class="va">control_means</span></span>
<span><span class="va">fold_change</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>   BRCA1     TP53    GAPDH 
1.047619 1.019324 1.161290 </code></pre>
</div>
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Matrix visualization</span></span>
<span><span class="co"># Heatmap of expression data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">expr_matrix</span>, </span>
<span>        Colv <span class="op">=</span> <span class="cn">NA</span>,         <span class="co"># Don't cluster columns</span></span>
<span>        scale <span class="op">=</span> <span class="st">"row"</span>,     <span class="co"># Scale by row (gene)</span></span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">16</span><span class="op">)</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Gene Expression Heatmap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/matrices-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<section id="more-matrix-practice" class="level4"><h4 class="anchored" data-anchor-id="more-matrix-practice">More Matrix Practice:</h4>
<div class="cell">
<div class="sourceCode" id="cb159"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#Create a simple Gene Expression matrix (RNA-seq style)</span></span>
<span></span>
<span><span class="va">Gene_Expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="fl">5.2</span>, <span class="fl">3.1</span>, <span class="fl">8.5</span>,   <span class="co"># Sample 1</span></span>
<span>  <span class="fl">6.0</span>, <span class="fl">2.8</span>, <span class="fl">7.9</span>    <span class="co"># Sample 2</span></span>
<span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Gene_Expression</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample_1"</span>, <span class="st">"Sample_2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Gene_Expression</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Gene Expression Matrix:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Gene Expression Matrix:"</code></pre>
</div>
<div class="sourceCode" id="cb161"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Gene_Expression</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>         GeneA GeneB GeneC
Sample_1   5.2   3.1   8.5
Sample_2   6.0   2.8   7.9</code></pre>
</div>
<div class="sourceCode" id="cb163"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#1. Transpose: Genes become rows, Samples become columns</span></span>
<span></span>
<span><span class="va">Gene_Expression_T</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">Gene_Expression</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Transpose of Gene Expression Matrix:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Transpose of Gene Expression Matrix:"</code></pre>
</div>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Gene_Expression_T</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Sample_1 Sample_2
GeneA      5.2      6.0
GeneB      3.1      2.8
GeneC      8.5      7.9</code></pre>
</div>
<div class="sourceCode" id="cb167"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#2. Matrix multiplication</span></span>
<span><span class="co"># Suppose each gene has an associated "gene weight" (e.g., biological importance)</span></span>
<span></span>
<span><span class="va">Gene_Weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">1.2</span>, <span class="fl">1.0</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Gene_Weights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Gene_Weights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Total_Weighted_Expression</span> <span class="op">&lt;-</span> <span class="va">Gene_Expression</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">Gene_Weights</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Total Weighted Expression per Sample:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Total Weighted Expression per Sample:"</code></pre>
</div>
<div class="sourceCode" id="cb169"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Total_Weighted_Expression</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>         Weight
Sample_1  16.38
Sample_2  16.06</code></pre>
</div>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 3. Matrix addition</span></span>
<span><span class="co"># Hypothetically increase expression by 1 TPM everywhere (technical adjustment)</span></span>
<span></span>
<span><span class="va">Adjusted_Expression</span> <span class="op">&lt;-</span> <span class="va">Gene_Expression</span> <span class="op">+</span> <span class="fl">1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Expression Matrix after adding 1 TPM:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Expression Matrix after adding 1 TPM:"</code></pre>
</div>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Adjusted_Expression</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>         GeneA GeneB GeneC
Sample_1   6.2   4.1   9.5
Sample_2   7.0   3.8   8.9</code></pre>
</div>
<div class="sourceCode" id="cb175"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 4. Identity matrix </span></span>
<span><span class="va">I</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">I</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">I</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Identity Matrix (for genes):"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Identity Matrix (for genes):"</code></pre>
</div>
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">I</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      GeneA GeneB GeneC
GeneA     1     0     0
GeneB     0     1     0
GeneC     0     0     1</code></pre>
</div>
<div class="sourceCode" id="cb179"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Multiplying Gene Expression by Identity</span></span>
<span><span class="va">Identity_Check</span> <span class="op">&lt;-</span> <span class="va">Gene_Expression</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Gene Expression multiplied by Identity Matrix:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Gene Expression multiplied by Identity Matrix:"</code></pre>
</div>
<div class="sourceCode" id="cb181"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Identity_Check</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>         GeneA GeneB GeneC
Sample_1   5.2   3.1   8.5
Sample_2   6.0   2.8   7.9</code></pre>
</div>
<div class="sourceCode" id="cb183"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 5. Scalar multiplication </span></span>
<span><span class="co"># Suppose you want to simulate doubling expression values</span></span>
<span></span>
<span><span class="va">Doubled_Expression</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">Gene_Expression</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Doubled Gene Expression:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Doubled Gene Expression:"</code></pre>
</div>
<div class="sourceCode" id="cb185"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Doubled_Expression</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>         GeneA GeneB GeneC
Sample_1  10.4   6.2  17.0
Sample_2  12.0   5.6  15.8</code></pre>
</div>
<div class="sourceCode" id="cb187"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 6. Summations </span></span>
<span></span>
<span><span class="co"># Total expression per sample</span></span>
<span><span class="va">Total_Expression_Per_Sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">Gene_Expression</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Total Expression per Sample:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Total Expression per Sample:"</code></pre>
</div>
<div class="sourceCode" id="cb189"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Total_Expression_Per_Sample</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample_1 Sample_2 
    16.8     16.7 </code></pre>
</div>
<div class="sourceCode" id="cb191"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Total expression per gene</span></span>
<span><span class="va">Total_Expression_Per_Gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">Gene_Expression</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Total Expression per Gene:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Total Expression per Gene:"</code></pre>
</div>
<div class="sourceCode" id="cb193"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Total_Expression_Per_Gene</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>GeneA GeneB GeneC 
 11.2   5.9  16.4 </code></pre>
</div>
<div class="sourceCode" id="cb195"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 7. Simple plots </span></span>
<span></span>
<span><span class="co"># Barplot: Total expression per sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">Total_Expression_Per_Sample</span>, main<span class="op">=</span><span class="st">"Total Expression per Sample"</span>, ylab<span class="op">=</span><span class="st">"TPM"</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"skyblue"</span>, <span class="st">"salmon"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/more-mat-practice-1.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb196"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Barplot: Total expression per gene</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">Total_Expression_Per_Gene</span>, main<span class="op">=</span><span class="st">"Total Expression per Gene"</span>, ylab<span class="op">=</span><span class="st">"TPM"</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lightgreen"</span>, <span class="st">"orange"</span>, <span class="st">"violet"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/more-mat-practice-2.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb197"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Heatmap: Expression matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">Gene_Expression</span>, Rowv<span class="op">=</span><span class="cn">NA</span>, Colv<span class="op">=</span><span class="cn">NA</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>, scale<span class="op">=</span><span class="st">"column"</span>, main<span class="op">=</span><span class="st">"Gene Expression Heatmap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/more-mat-practice-3.png" class="img-fluid" width="1650"></p>
</div>
</div>
<p>Another Example: You have counts of cells in different organs for two animal species.</p>
<p>You also have a matrix with average cell sizes (micrometer, ¬µm¬≤) for each organ.</p>
<p>You can then multiply count √ó size to get total cell area for each species in each organ.</p>
<div class="cell">
<div class="sourceCode" id="cb198"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create a matrix: Cell counts</span></span>
<span><span class="va">Cell_Counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">600</span>, <span class="fl">300</span>, <span class="fl">400</span>, <span class="fl">700</span>, <span class="fl">800</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Cell_Counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mouse"</span>, <span class="st">"Rat"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Cell_Counts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Heart"</span>, <span class="st">"Liver"</span>, <span class="st">"Brain"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Cell Counts Matrix:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Cell Counts Matrix:"</code></pre>
</div>
<div class="sourceCode" id="cb200"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Cell_Counts</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Heart Liver Brain
Mouse   500   600   300
Rat     400   700   800</code></pre>
</div>
<div class="sourceCode" id="cb202"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create a matrix: Average cell size in ¬µm¬≤</span></span>
<span><span class="va">Cell_Size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">50</span>, <span class="fl">200</span>, <span class="fl">150</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Cell_Size</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Heart"</span>, <span class="st">"Liver"</span>, <span class="st">"Brain"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Cell_Size</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Avg_Cell_Size"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Cell Size Matrix (¬µm¬≤):"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Cell Size Matrix (¬µm¬≤):"</code></pre>
</div>
<div class="sourceCode" id="cb204"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Cell_Size</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Avg_Cell_Size
Heart            50
Liver           200
Brain           150</code></pre>
</div>
<div class="sourceCode" id="cb206"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 1. Transpose of Cell Counts</span></span>
<span><span class="va">Cell_Counts_T</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">Cell_Counts</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Transpose of Cell Counts:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Transpose of Cell Counts:"</code></pre>
</div>
<div class="sourceCode" id="cb208"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Cell_Counts_T</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Mouse Rat
Heart   500 400
Liver   600 700
Brain   300 800</code></pre>
</div>
<div class="sourceCode" id="cb210"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 2. Matrix multiplication: Total cell area</span></span>
<span><span class="co"># (2x3) %*% (3x1) =&gt; (2x1)</span></span>
<span><span class="va">Total_Cell_Area</span> <span class="op">&lt;-</span> <span class="va">Cell_Counts</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">Cell_Size</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Total_Cell_Area</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Cell_area"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Total Cell Area (Counts √ó Size) (¬µm¬≤):"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Total Cell Area (Counts √ó Size) (¬µm¬≤):"</code></pre>
</div>
<div class="sourceCode" id="cb212"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Total_Cell_Area</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Cell_area
Mouse    190000
Rat      280000</code></pre>
</div>
<div class="sourceCode" id="cb214"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 3. Matrix addition: Add 10 cells artificially to all counts (for example)</span></span>
<span><span class="va">Added_Cells</span> <span class="op">&lt;-</span> <span class="va">Cell_Counts</span> <span class="op">+</span> <span class="fl">10</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Cell Counts after adding 10 artificial cells:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Cell Counts after adding 10 artificial cells:"</code></pre>
</div>
<div class="sourceCode" id="cb216"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Added_Cells</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Heart Liver Brain
Mouse   510   610   310
Rat     410   710   810</code></pre>
</div>
<div class="sourceCode" id="cb218"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 4. Identity matrix</span></span>
<span><span class="va">I</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">I</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Heart"</span>, <span class="st">"Liver"</span>, <span class="st">"Brain"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">I</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Heart"</span>, <span class="st">"Liver"</span>, <span class="st">"Brain"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Identity Matrix:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Identity Matrix:"</code></pre>
</div>
<div class="sourceCode" id="cb220"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">I</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Heart Liver Brain
Heart     1     0     0
Liver     0     1     0
Brain     0     0     1</code></pre>
</div>
<div class="sourceCode" id="cb222"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 5. Multiplying Cell Counts by Identity Matrix (no real change but shows dimension rules)</span></span>
<span><span class="va">Check_Identity</span> <span class="op">&lt;-</span> <span class="va">Cell_Counts</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Cell Counts multiplied by Identity Matrix:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Cell Counts multiplied by Identity Matrix:"</code></pre>
</div>
<div class="sourceCode" id="cb224"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Check_Identity</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Heart Liver Brain
Mouse   500   600   300
Rat     400   700   800</code></pre>
</div>
<div class="sourceCode" id="cb226"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># 6. Scalar multiplication: double the counts (hypothetical growth)</span></span>
<span><span class="va">Double_Cell_Counts</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="op">*</span> <span class="va">Cell_Counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Doubled Cell Counts:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Doubled Cell Counts:"</code></pre>
</div>
<div class="sourceCode" id="cb228"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Double_Cell_Counts</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      Heart Liver Brain
Mouse  1000  1200   600
Rat     800  1400  1600</code></pre>
</div>
<div class="sourceCode" id="cb230"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Total number of cells per animal (row sums)</span></span>
<span><span class="va">Total_Cells_Per_Species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">Cell_Counts</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Total number of cells per species:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Total number of cells per species:"</code></pre>
</div>
<div class="sourceCode" id="cb232"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Total_Cells_Per_Species</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Mouse   Rat 
 1400  1900 </code></pre>
</div>
<div class="sourceCode" id="cb234"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Total number of cells per organ (column sums)</span></span>
<span><span class="va">Total_Cells_Per_Organ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">Cell_Counts</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Total number of cells per organ:"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Total number of cells per organ:"</code></pre>
</div>
<div class="sourceCode" id="cb236"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Total_Cells_Per_Organ</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Heart Liver Brain 
  900  1300  1100 </code></pre>
</div>
<div class="sourceCode" id="cb238"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># --- Simple plots ---</span></span>
<span></span>
<span><span class="co"># Bar plot of total cells per species</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">Total_Cells_Per_Species</span>, main<span class="op">=</span><span class="st">"Total Cell Counts per Species"</span>, ylab<span class="op">=</span><span class="st">"Number of Cells"</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>, <span class="st">"lightgreen"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/mat-prac-2-1.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb239"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Bar plot of total cells per organ</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">Total_Cells_Per_Organ</span>, main<span class="op">=</span><span class="st">"Total Cell Counts per Organ"</span>, ylab<span class="op">=</span><span class="st">"Number of Cells"</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"pink"</span>, <span class="st">"lightyellow"</span>, <span class="st">"lightgray"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/mat-prac-2-2.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb240"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Heatmap of the original Cell Counts matrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">Cell_Counts</span>, Rowv<span class="op">=</span><span class="cn">NA</span>, Colv<span class="op">=</span><span class="cn">NA</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>, scale<span class="op">=</span><span class="st">"column"</span>, main<span class="op">=</span><span class="st">"Heatmap of Cell Counts"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/mat-prac-2-3.png" class="img-fluid" width="1650"></p>
</div>
</div>
<table class="table">
<colgroup>
<col style="width: 27%">
<col style="width: 37%">
<col style="width: 34%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">Operation</th>
<th style="text-align: left;">Explanation</th>
<th style="text-align: left;">R Function/Example</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Matrix Creation</td>
<td style="text-align: left;">Create gene expression matrix</td>
<td style="text-align: left;"><code><a href="https://rdrr.io/r/base/matrix.html">matrix()</a></code></td>
</tr>
<tr class="even">
<td style="text-align: left;">Transpose</td>
<td style="text-align: left;">Flip genes and samples</td>
<td style="text-align: left;"><code>t(Gene_Expression)</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Matrix Multiplication</td>
<td style="text-align: left;">Calculate weighted sums</td>
<td style="text-align: left;"><code>Gene_Expression %*% Gene_Weights</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">Matrix Addition</td>
<td style="text-align: left;">Adjust counts</td>
<td style="text-align: left;"><code>Gene_Expression + 1</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Identity Matrix</td>
<td style="text-align: left;">Special neutral matrix</td>
<td style="text-align: left;"><code>diag(3)</code></td>
</tr>
<tr class="even">
<td style="text-align: left;">Scalar Multiplication</td>
<td style="text-align: left;">Simulate overall increase</td>
<td style="text-align: left;"><code>2 * Gene_Expression</code></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Row/Column Summation</td>
<td style="text-align: left;">Total per sample/gene</td>
<td style="text-align: left;">
<code><a href="https://rdrr.io/r/base/colSums.html">rowSums()</a></code>, <code><a href="https://rdrr.io/r/base/colSums.html">colSums()</a></code>
</td>
</tr>
<tr class="even">
<td style="text-align: left;">Plotting</td>
<td style="text-align: left;">Visualize expression patterns</td>
<td style="text-align: left;">
<code><a href="https://rdrr.io/r/graphics/barplot.html">barplot()</a></code>, <code><a href="https://rdrr.io/r/stats/heatmap.html">heatmap()</a></code>
</td>
</tr>
</tbody>
</table></section></section><section id="basic-stuffs-list" class="level3"><h3 class="anchored" data-anchor-id="basic-stuffs-list">Basic Stuffs: List</h3>
<p>Lists are the most flexible data structure in R - they can hold any combination of data types, including other lists! This makes them essential for biological data analysis where we often deal with mixed data types.</p>
<div class="cell">
<div class="sourceCode" id="cb241"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># A list storing different types of genomic data</span></span>
<span><span class="va">genomics_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  gene_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"MYC"</span><span class="op">)</span>,               <span class="co"># Character vector</span></span>
<span>  expression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">3.4</span>, <span class="fl">5.6</span>, <span class="fl">7.8</span>, <span class="fl">9.1</span>, <span class="fl">2.3</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">3</span><span class="op">)</span>,    <span class="co"># Numeric matrix</span></span>
<span>  is_cancer_gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span>,                 <span class="co"># Logical vector</span></span>
<span>  metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>                                       <span class="co"># Nested list!</span></span>
<span>    lab <span class="op">=</span> <span class="st">"CRG"</span>,</span>
<span>    date <span class="op">=</span> <span class="st">"2023-05-01"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How to Access Elements of a List?</p>
<div class="cell">
<div class="sourceCode" id="cb242"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Method 1: Double brackets [[ ]] for single element</span></span>
<span><span class="va">genomics_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>  <span class="co"># Returns gene_names vector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "TP53"  "BRCA1" "MYC"  </code></pre>
</div>
<div class="sourceCode" id="cb244"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Method 2: $ operator with names (when elements are named)</span></span>
<span><span class="va">genomics_data</span><span class="op">$</span><span class="va">expression</span>  <span class="co"># Returns the matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>     [,1] [,2]
[1,]  1.2  7.8
[2,]  3.4  9.1
[3,]  5.6  2.3</code></pre>
</div>
<div class="sourceCode" id="cb246"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Method 3: Single bracket [ ] returns a sublist</span></span>
<span><span class="va">genomics_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>  <span class="co"># Returns list with first two elements</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>$gene_names
[1] "TP53"  "BRCA1" "MYC"  

$expression
     [,1] [,2]
[1,]  1.2  7.8
[2,]  3.4  9.1
[3,]  5.6  2.3</code></pre>
</div>
</div>
<p>Key Difference from Vectors:</p>
<div class="cell">
<div class="sourceCode" id="cb248"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compare to your prop.table() example:</span></span>
<span><span class="va">atomic_vec</span><span class="op">[</span><span class="st">"Human"</span><span class="op">]</span>    <span class="co"># Returns named numeric (vector)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Human 
  0.5 </code></pre>
</div>
<div class="sourceCode" id="cb250"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">atomic_vec</span><span class="op">[</span><span class="st">"Mouse"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Mouse 
 0.33 </code></pre>
</div>
<div class="sourceCode" id="cb252"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">genomics_data</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="co"># Returns list containing the vector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>$gene_names
[1] "TP53"  "BRCA1" "MYC"  </code></pre>
</div>
</div>
<p>Why Biologists Need Lists?<br><code><a href="https://rdrr.io/r/stats/lm.html">lm()</a></code>, <code><a href="https://rdrr.io/r/stats/prcomp.html">prcomp()</a></code> functions, RNAseq analysis packages produces list. So, we need to learn how to handle lists.</p>
<p>See these examples:</p>
<p>A. Storing BLAST results</p>
<div class="cell">
<div class="sourceCode" id="cb254"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">blast_hits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  query_id <span class="op">=</span> <span class="st">"GeneX"</span>,</span>
<span>  hit_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NP_123"</span>, <span class="st">"NP_456"</span><span class="op">)</span>,</span>
<span>  e_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1e-50</span>, <span class="fl">3e-12</span><span class="op">)</span>,</span>
<span>  alignment <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ATG..."</span>, <span class="st">"CTA..."</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>B. Handling Mixed Data</p>
<div class="cell">
<div class="sourceCode" id="cb255"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">patient_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="st">"P1001"</span>,</span>
<span>  tests <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WBC"</span>, <span class="st">"RBC"</span><span class="op">)</span>,</span>
<span>    value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4.5</span>, <span class="fl">5.1</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  has_mutation <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Common List Operations</p>
<div class="cell">
<div class="sourceCode" id="cb256"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add new element</span></span>
<span><span class="va">genomics_data</span><span class="op">$</span><span class="va">sequencer</span> <span class="op">&lt;-</span> <span class="st">"Illumina"</span></span>
<span></span>
<span><span class="co"># Remove element</span></span>
<span><span class="va">genomics_data</span><span class="op">$</span><span class="va">is_cancer_gene</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span></span>
<span><span class="co"># Check structure (critical for complex lists)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>List of 4
 $ gene_names: chr [1:3] "TP53" "BRCA1" "MYC"
 $ expression: num [1:3, 1:2] 1.2 3.4 5.6 7.8 9.1 2.3
 $ metadata  :List of 2
  ..$ lab : chr "CRG"
  ..$ date: chr "2023-05-01"
 $ sequencer : chr "Illumina"</code></pre>
</div>
</div>
<p>By the way, how would you add more patients?</p>
<div class="cell">
<div class="sourceCode" id="cb258"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add new patient</span></span>
<span><span class="va">patient_data</span><span class="op">$</span><span class="va">P1002</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  id <span class="op">=</span> <span class="st">"P1002"</span>,</span>
<span>  tests <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>    test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WBC"</span>, <span class="st">"RBC"</span>, <span class="st">"Platelets"</span><span class="op">)</span>,</span>
<span>    value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6.2</span>, <span class="fl">4.8</span>, <span class="fl">150</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  has_mutation <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Access specific patient</span></span>
<span><span class="va">patient_data</span><span class="op">$</span><span class="va">P1001</span><span class="op">$</span><span class="va">test</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>NULL</code></pre>
</div>
</div>
<p>For Batch Processing:</p>
<div class="cell">
<div class="sourceCode" id="cb260"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">patients</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    id <span class="op">=</span> <span class="st">"P1001"</span>,</span>
<span>    tests <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WBC"</span>, <span class="st">"RBC"</span><span class="op">)</span>, value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4.5</span>, <span class="fl">5.1</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    has_mutation <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    id <span class="op">=</span> <span class="st">"P1002"</span>,</span>
<span>    tests <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WBC"</span>, <span class="st">"RBC"</span>, <span class="st">"Platelets"</span><span class="op">)</span>, value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6.2</span>, <span class="fl">4.8</span>, <span class="fl">150</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    has_mutation <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Access 2nd patient's WBC value</span></span>
<span><span class="va">patients</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">tests</span><span class="op">$</span><span class="va">value</span><span class="op">[</span><span class="va">patients</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">tests</span><span class="op">$</span><span class="va">test</span> <span class="op">==</span> <span class="st">"WBC"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 6.2</code></pre>
</div>
</div>
<p>Converting Between Structures</p>
<div class="cell">
<div class="sourceCode" id="cb262"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># List ‚Üí Vector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  gene_names1   gene_names2   gene_names3   expression1   expression2 
       "TP53"       "BRCA1"         "MYC"         "1.2"         "3.4" 
  expression3   expression4   expression5   expression6  metadata.lab 
        "5.6"         "7.8"         "9.1"         "2.3"         "CRG" 
metadata.date 
 "2023-05-01" </code></pre>
</div>
</div>
<p>Visualization</p>
<div class="cell">
<div class="sourceCode" id="cb264"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Base R plot from list data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>,</span>
<span>        names.arg <span class="op">=</span> <span class="va">genomics_data</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This code won‚Äôt work if you run. <code>unlist(genomics_data[2]</code> creates a vector of length 6 from our 3*2 matrix but <code>genomics_data[[1]]</code> has 3 things inside the <code>gene_names</code> vector. Debug like this:</p>
<div class="cell">
<div class="sourceCode" id="cb265"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">$</span><span class="va">expression</span><span class="op">)</span>  <span class="co"># e.g., 2 rows x 2 cols</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 3 2</code></pre>
</div>
<div class="sourceCode" id="cb267"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">$</span><span class="va">gene_names</span><span class="op">)</span> <span class="co"># e.g., 3 genes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 3</code></pre>
</div>
</div>
<p>A. Gene-Centric (Mean Expression)</p>
<div class="cell">
<div class="sourceCode" id="cb269"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">$</span><span class="va">expression</span><span class="op">)</span>,</span>
<span>        names.arg <span class="op">=</span> <span class="va">genomics_data</span><span class="op">$</span><span class="va">gene_names</span>,</span>
<span>        col <span class="op">=</span> <span class="st">"steelblue"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Mean Expression"</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Average Gene Expression"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<p>B. Sample-Centric (All Measurements)</p>
<div class="cell">
<div class="sourceCode" id="cb270"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">$</span><span class="va">expression</span>,</span>
<span>        beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        names.arg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Sample_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">$</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="va">genomics_data</span><span class="op">$</span><span class="va">gene_names</span>,</span>
<span>        args.legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"topright"</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"green"</span><span class="op">)</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Expression Across Samples"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This matches real-world scenarios:</p>
<p>RNA-seq: Rows=genes, cols=samples</p>
<p>rowMeans() = average expression per gene</p>
<p>beside=TRUE =&gt; compare samples within genes</p>
<p>Proteomics: Rows=proteins, cols=replicates</p>
<p>Same principles apply</p>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb271"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Calculate stats</span></span>
<span><span class="va">gene_means</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowMeans</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">$</span><span class="va">expression</span><span class="op">)</span></span>
<span><span class="va">gene_sds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">genomics_data</span><span class="op">$</span><span class="va">expression</span>, <span class="fl">1</span>, <span class="va">sd</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot with error bars</span></span>
<span><span class="va">bp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">gene_means</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">gene_means</span> <span class="op">+</span> <span class="va">gene_sds</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/arrows.html">arrows</a></span><span class="op">(</span><span class="va">bp</span>, <span class="va">gene_means</span> <span class="op">-</span> <span class="va">gene_sds</span>, <span class="va">bp</span>, <span class="va">gene_means</span> <span class="op">+</span> <span class="va">gene_sds</span>, </span>
<span>       angle <span class="op">=</span> <span class="fl">90</span>, code <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<p><strong>Task: Create a list containing:</strong></p>
<ol type="i">
<li><p>A character vector of 3 gene names</p></li>
<li><p>A numeric matrix of expression values</p></li>
<li><p>A logical vector indicating pathway membership</p></li>
<li><p>A nested list with lab metadata</p></li>
</ol></section></section><section id="homeworks-matrix-and-list-operations" class="level2"><h2 class="anchored" data-anchor-id="homeworks-matrix-and-list-operations">üè° Homeworks: Matrix and List Operations</h2>
<section id="protein-quantification-in-biological-samples" class="level3"><h3 class="anchored" data-anchor-id="protein-quantification-in-biological-samples">üß™ Protein Quantification in Biological Samples</h3>
<p>You are given the following protein concentration matrix:</p>
<p><span class="math display">\[
\text{ProteinMatrix} =
\begin{bmatrix}
  5 &amp; 3 &amp; 2 \\
  7 &amp; 6 &amp; 4 \\
\end{bmatrix}
\]</span></p>
<ul>
<li>Rows represent samples:
<ul>
<li>Sample1</li>
<li>Sample2</li>
</ul>
</li>
<li>Columns represent proteins:
<ul>
<li>ProteinX</li>
<li>ProteinY</li>
<li>ProteinZ</li>
</ul>
</li>
</ul>
<p>You are also given a weight (importance) matrix for the proteins:</p>
<p><span class="math display">\[
\text{WeightVector} =
\begin{bmatrix}
  0.5 \\
  1.0 \\
  1.5 \\
\end{bmatrix}
\]</span></p>
</section><section id="tasks" class="level3"><h3 class="anchored" data-anchor-id="tasks">üîß Tasks</h3>
<ol type="1">
<li>Make the matrices (with exact names) and multiply the ProteinMatrix by the WeightVector.</li>
</ol>
<p>That is:</p>
<p><span class="math display">\[
  \text{ProteinMatrix} \times \text{WeightVector}
  \]</span></p>
<ol start="2" type="1">
<li><p>Transpose the ProteinMatrix and show what it looks like.</p></li>
<li><p>Create the Identity matrix of compatible size and show what happens when you multiply: <span class="math display">\[
\text{ProteinMatrix} \times I
\]</span></p></li>
</ol>
<p>4.Do the calculations (rowSums, colSums, etc.) and visualization (barplot, heatmap) as shown in the class.</p>
</section><section id="interpretation-questions" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions">üß† Interpretation Questions</h3>
<ul>
<li>What does multiplying the protein levels by the weight vector mean biologically?</li>
<li>What does the result tell you about total protein burden (or total protein impact) for each sample?</li>
<li>What do the identity matrix represent in the context of protein interactions or measurement biases?</li>
<li>If you changed the weight of ProteinZ to 3.0, how would the result change?</li>
</ul></section><section id="gene-to-protein-translation" class="level3"><h3 class="anchored" data-anchor-id="gene-to-protein-translation">üß¨ Gene-to-Protein Translation</h3>
<p>You are given the following matrix representing normalized gene expression levels (e.g., TPM): <span class="math display">\[
\text{GeneExpression} =
\begin{bmatrix}
  10 &amp;  8 &amp;  5 \\
  15 &amp; 12 &amp; 10 \\
\end{bmatrix}
\]</span></p>
<ul>
<li>Rows = Samples:
<ul>
<li>Sample1</li>
<li>Sample2</li>
</ul>
</li>
<li>Columns = Genes:
<ul>
<li>GeneA</li>
<li>GeneB</li>
<li>GeneC</li>
</ul>
</li>
</ul>
<p>Each gene translates into proteins with a certain efficiency. The efficiency of translation from each gene to its corresponding protein is given by the following diagonal matrix:</p>
<p><span class="math display">\[
\text{TranslationMatrix} =
\begin{bmatrix}
  1.5 &amp; 0   &amp; 0 \\
  0   &amp; 1.2 &amp; 0 \\
  0   &amp; 0   &amp; 1.8 \\
\end{bmatrix}
\]</span></p>
<p>This means:</p>
<ul>
<li>GeneA ‚Üí ProteinA with 1.5√ó efficiency<br>
</li>
<li>GeneB ‚Üí ProteinB with 1.2√ó efficiency<br>
</li>
<li>GeneC ‚Üí ProteinC with 1.8√ó efficiency</li>
</ul></section><section id="tasks-1" class="level3"><h3 class="anchored" data-anchor-id="tasks-1">üîß Tasks</h3>
<ol type="1">
<li>Make the matrices and multiply GeneExpression √ó TranslationMatrix to compute the resulting ProteinMatrix.</li>
</ol>
<ul>
<li>Show the result step-by-step.</li>
</ul>
<ol start="2" type="1">
<li><p>Transpose the GeneExpression matrix. What does this new matrix represent?</p></li>
<li><p>Create the Identity matrix I‚ÇÉ and multiply it with the TranslationMatrix. What happens?</p></li>
<li><p>Create a new matrix containing only the expression of GeneA and GeneB across both samples. Call this submatrix A.</p></li>
</ol>
<ul>
<li>Compute the inverse A‚Åª¬π using <code><a href="https://rdrr.io/r/base/solve.html">solve()</a></code> function.</li>
<li>Then verify:<br>
A √ó A‚Åª¬π = I‚ÇÇ</li>
</ul></section><section id="visualization-tasks" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks">üìä Visualization Tasks</h3>
<ol start="5" type="1">
<li>Plot a MARplot-style scatter plot:</li>
</ol>
<ul>
<li>x-axis: Gene expression values (GeneExpression matrix, flattened)</li>
<li>y-axis: Corresponding Protein values (ProteinMatrix, flattened)</li>
<li>Label each point as ‚ÄúSample-Gene‚Äù</li>
</ul>
<ol start="6" type="1">
<li>Generate a heatmap of the ProteinMatrix using R‚Äôs <code><a href="https://rdrr.io/r/stats/heatmap.html">heatmap()</a></code> function.</li>
</ol>
<ul>
<li>Add meaningful row and column labels.</li>
<li>Enable clustering by rows and columns.</li>
</ul></section><section id="interpretation-questions-1" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-1">üß† Interpretation Questions</h3>
<ul>
<li>What does matrix multiplication represent biologically in this case?</li>
<li>Why does the diagonal TranslationMatrix make sense biologically?</li>
<li>What does it mean if Sample2 has higher protein levels even with similar gene expression?</li>
<li>How does the MARplot help interpret translation efficiency?</li>
<li>How does clustering in the heatmap reveal relationships between samples and proteins?</li>
</ul></section><section id="animal-breeding-economic-ranking-of-bulls-by-traits" class="level3"><h3 class="anchored" data-anchor-id="animal-breeding-economic-ranking-of-bulls-by-traits">üêÑ Animal Breeding ‚Äì Economic Ranking of Bulls by Traits</h3>
<p>You are evaluating two bulls for use in a dairy breeding program. Their Estimated Breeding Values (EBVs) are:</p>
<p><span class="math display">\[
\text{BullEBVs} =
\begin{bmatrix}
  400 &amp; 1.2 &amp; 0.8 \\\\
  500 &amp; 1.5 &amp; 0.6 \\\\
\end{bmatrix}
\]</span></p>
<ul>
<li>Rows:
<ul>
<li>Bull1<br>
</li>
<li>Bull2</li>
</ul>
</li>
<li>Columns:
<ul>
<li>Trait1 = Milk yield (liters/year)<br>
</li>
<li>Trait2 = Growth rate (kg/day)<br>
</li>
<li>Trait3 = Fertility (calving interval adjustment)</li>
</ul>
</li>
</ul>
<p>You assign economic weights to each trait:</p>
<p><span class="math display">\[
\text{EconomicWeights} =
\begin{bmatrix}
  0.002 \\\\
  50 \\\\
  100 \\\\
\end{bmatrix}
\]</span></p>
</section><section id="tasks-2" class="level3"><h3 class="anchored" data-anchor-id="tasks-2">üîß Tasks</h3>
<ol type="1">
<li>Compute:</li>
</ol>
<p><span class="math display">\[
  \text{TotalValue} = \text{BullEBVs} \times \text{EconomicWeights}
  \]</span></p>
<ul>
<li>What are the resulting values?</li>
<li>Which bull is more valuable economically?</li>
</ul>
<ol start="2" type="1">
<li><p>Interpret what multiplying by the economic weights means biologically.</p></li>
<li><p>Create the 3√ó3 identity matrix I‚ÇÉ and multiply it with BullEBVs.</p></li>
</ol>
<ul>
<li>What does it return?</li>
<li>What does the identity matrix mean in this case?</li>
</ul>
<ol start="4" type="1">
<li>Subset the BullEBVs matrix to remove Trait1 (milk yield) and recalculate TotalValue.</li>
</ol>
<ul>
<li>How does this change the ranking?</li>
</ul></section><section id="visualization-tasks-1" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks-1">üìä Visualization Tasks</h3>
<ol start="5" type="1">
<li><p>Create a bar plot comparing TotalValue for Bull1 and Bull2.</p></li>
<li><p>Create a heatmap of the EBVs.</p></li>
</ol>
<ul>
<li>Label rows and columns.</li>
<li>Enable clustering.</li>
</ul></section><section id="interpretation-questions-2" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-2">üß† Interpretation Questions</h3>
<ul>
<li>How do economic weights affect trait importance?</li>
<li>Why might you ignore milk yield in some breeding programs?</li>
<li>What is the value of heatmaps in visualizing multivariate trait data?</li>
<li>Can this method be extended to more bulls and more traits?</li>
</ul></section><section id="plant-breeding-trait-contributions-from-parental-lines" class="level3"><h3 class="anchored" data-anchor-id="plant-breeding-trait-contributions-from-parental-lines">üåæ Plant Breeding ‚Äì Trait Contributions from Parental Lines</h3>
<p>You are breeding a new rice variety from three parental lines. The key traits are:</p>
<ul>
<li>T1 = Drought resistance</li>
<li>T2 = Yield</li>
<li>T3 = Maturation time</li>
</ul>
<p>The following trait values (normalized 1‚Äì10) have been measured:</p>
<p><span class="math display">\[
\text{ParentTraits} =
\begin{bmatrix}
  7 &amp; 5 &amp; 3 \\\\
  6 &amp; 8 &amp; 4 \\\\
  5 &amp; 6 &amp; 6 \\\\
\end{bmatrix}
\]</span></p>
<ul>
<li>Rows:
<ul>
<li>P1 (Parent 1)</li>
<li>P2 (Parent 2)</li>
<li>P3 (Parent 3)</li>
</ul>
</li>
<li>Columns:
<ul>
<li>T1 = Drought resistance</li>
<li>T2 = Yield</li>
<li>T3 = Maturation time</li>
</ul>
</li>
</ul>
<p>You design a hybrid with contributions from each parent as follows:</p>
<p><span class="math display">\[
\text{HybridWeights} =
\begin{bmatrix}
  0.5 \\\\
  0.3 \\\\
  0.2 \\\\
\end{bmatrix}
\]</span></p>
</section><section id="tasks-3" class="level3"><h3 class="anchored" data-anchor-id="tasks-3">üîß Tasks</h3>
<ol type="1">
<li>Compute the HybridTrait vector by:</li>
</ol>
<p><span class="math display">\[
  \text{HybridTraits} = \text{HybridWeights}^T \times \text{ParentTraits}
  \]</span></p>
<p>Show the steps and result.</p>
<ol start="2" type="1">
<li><p>Explain what it means biologically when one parent contributes more to a particular trait.</p></li>
<li><p>Create an identity matrix I‚ÇÉ and multiply it with ParentTraits.</p></li>
</ol>
<ul>
<li>What do you observe?</li>
<li>What does I √ó ParentTraits represent?</li>
</ul>
<ol start="4" type="1">
<li>Subset the ParentTraits matrix to include only T1 and T2. Recalculate the hybrid traits.</li>
</ol>
<ul>
<li>Discuss how removing a trait affects your outcome.</li>
</ul></section><section id="visualization-tasks-2" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks-2">üìä Visualization Tasks</h3>
<ol start="5" type="1">
<li>Generate a heatmap of the ParentTraits matrix.</li>
</ol>
<ul>
<li>Label the rows with parent names and columns with trait names.</li>
<li>Enable row/column clustering.</li>
</ul>
<ol start="6" type="1">
<li>Create a bar plot showing the HybridTraits (T1, T2, T3).</li>
</ol>
<ul>
<li>Color code each bar by trait.</li>
<li>What trait contributes most?</li>
</ul></section><section id="interpretation-questions-3" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-3">üß† Interpretation Questions</h3>
<ul>
<li>How does the weighting of parents affect the hybrid‚Äôs performance?</li>
<li>What does the identity matrix represent here?</li>
<li>If you used equal weights (‚Öì for each), how would the hybrid traits change?</li>
<li>What real-world limitations does this simplified model ignore?</li>
</ul></section><section id="managing-matrices-and-weight-vectors-using-lists-in-r" class="level3"><h3 class="anchored" data-anchor-id="managing-matrices-and-weight-vectors-using-lists-in-r">üß† Managing Matrices and Weight Vectors Using Lists in R</h3>
<p>Now that you have completed four biological matrix problems ‚Äî Protein concentration, gene-to-protein mapping, bull breeding value ranking, and plant trait combinations ‚Äî it‚Äôs time to organize your data and weights using R‚Äôs list structure.</p>
<p>In this task, you will:</p>
<ul>
<li>Group each example‚Äôs matrix and its corresponding weight vector inside a named list.</li>
<li>Combine these named lists into a larger list called bioList.</li>
<li>Use list indexing to repeat your earlier calculations and visualizations.</li>
<li>Reflect on the benefits and challenges of using structured data objects.</li>
</ul></section><section id="step-1-create-a-master-list" class="level3"><h3 class="anchored" data-anchor-id="step-1-create-a-master-list">üì¶ Step 1: Create a master list</h3>
<p>You should now build a named list called bioList containing the following four elements:</p>
<ul>
<li>ProteinConc = list(matrix = ProteinMatrix, weights = WeightVector)</li>
<li>ProteinMap = list(matrix = ProteinMapping, weights = TranslationWeights)</li>
<li>Plant = list(matrix = ParentTraits, weights = HybridWeights)</li>
<li>Animal = list(matrix = BullEBVs, weights = EconomicWeights)</li>
</ul>
<p>Hint: Each inner list should contain both:</p>
<ul>
<li>matrix = the main data matrix</li>
<li>weights = the vector used for multiplication</li>
</ul>
<p>No R code is required here (You have them from previous part, use inside same rmd/notebook file) ‚Äî just structure your data like this in your workspace.</p>
</section><section id="tasks-4" class="level3"><h3 class="anchored" data-anchor-id="tasks-4">üîß Tasks</h3>
<ol type="1">
<li><p>List the full names of each component in bioList. What are the names of the top-level and nested components?</p></li>
<li><p>Access each matrix and its corresponding weights using list indexing.</p></li>
</ol>
<ul>
<li>How would you extract only the matrix of the Plant entry?</li>
<li>How would you extract the weights for the Protein concentration entry?</li>
</ul>
<ol start="3" type="1">
<li>Use the correct matrix and weights to perform:</li>
</ol>
<ul>
<li>ProteinConc: Weighted gene expression score</li>
<li>ProteinMap: Contribution of transcripts to each protein</li>
<li>Plant: Hybrid trait values</li>
<li>Animal: Bull total economic value</li>
</ul>
<ol start="4" type="1">
<li>Subset one matrix in each sublist (e.g., drop a trait or feature) and repeat the weighted calculation.</li>
</ol>
<ul>
<li>What changes in the results?</li>
<li>Which traits/genes have the strongest influence?</li>
</ul></section><section id="visualization-tasks-3" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks-3">üìä Visualization Tasks</h3>
<ol start="5" type="1">
<li>Generate one heatmap for any matrix stored in bioList.</li>
</ol>
<ul>
<li>Choose one (e.g., ProteinMap or Plant)</li>
<li>Apply clustering to rows and/or columns</li>
<li>Label appropriately</li>
</ul>
<ol start="6" type="1">
<li>Generate two bar plots:</li>
</ol>
<ul>
<li>One showing the result of weighted trait aggregation for the Plant hybrid</li>
<li>One showing the total breeding values for each bull</li>
</ul></section><section id="interpretation-questions-4" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-4">üß† Interpretation Questions</h3>
<ul>
<li>How does structuring your data using a list help with clarity and reproducibility?</li>
<li>What risks or challenges might occur when accessing elements from nested lists?</li>
<li>Could this structure be scaled for real datasets with many samples or traits?</li>
<li>How would you loop over all elements in bioList to apply the same function?</li>
<li>How can this list structure be useful for building automated bioinformatics pipelines?</li>
</ul>
<p>üìù Your Rmarkdown file(s) should include:</p>
<ul>
<li>All matrix calculations (tasks). Also, name the rows and columns of each matrix accordingly.</li>
<li>All interpretation answers</li>
<li>All plots (output from embedded code)</li>
<li>And your commentary blocks for each code chunk</li>
</ul>
<p><strong>knit your rmd (or Notebook) file as html/pdf file and push both the rmd (or Notebook) and html/pdf files</strong></p>
</section></section><section id="factor-variables" class="level2"><h2 class="anchored" data-anchor-id="factor-variables">Factor Variables</h2>
<p>Important for categorical data</p>
<section id="creating-factors" class="level3"><h3 class="anchored" data-anchor-id="creating-factors">Creating Factors</h3>
<p>Factors are used to represent categorical data in R. They are particularly important for biological data like genotypes, phenotypes, and experimental conditions.</p>
<div class="cell">
<div class="sourceCode" id="cb272"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Simple factor: DNA sample origins</span></span>
<span><span class="va">origins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Human"</span>, <span class="st">"Mouse"</span>, <span class="st">"Human"</span>, <span class="st">"Zebrafish"</span>, <span class="st">"Mouse"</span>, <span class="st">"Human"</span><span class="op">)</span></span>
<span><span class="va">origins_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">origins</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Human"</span>, <span class="st">"Zebrafish"</span>, <span class="st">"Mouse"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">origins_factor</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] Human     Mouse     Human     Zebrafish Mouse     Human    
Levels: Human Zebrafish Mouse</code></pre>
</div>
<div class="sourceCode" id="cb274"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check levels (categories)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">origins_factor</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Human"     "Zebrafish" "Mouse"    </code></pre>
</div>
<div class="sourceCode" id="cb276"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create a factor with predefined levels</span></span>
<span><span class="va">treatment_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Low_dose"</span>, <span class="st">"High_dose"</span>, <span class="st">"Control"</span>, <span class="st">"Low_dose"</span><span class="op">)</span>,</span>
<span>                         levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Low_dose"</span>, <span class="st">"High_dose"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">treatment_groups</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] Control   Low_dose  High_dose Control   Low_dose 
Levels: Control Low_dose High_dose</code></pre>
</div>
<div class="sourceCode" id="cb278"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Ordered factors (important for severity, stages, etc.)</span></span>
<span><span class="va">disease_severity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mild"</span>, <span class="st">"Severe"</span>, <span class="st">"Moderate"</span>, <span class="st">"Mild"</span>, <span class="st">"Critical"</span><span class="op">)</span>,</span>
<span>                         levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mild"</span>, <span class="st">"Moderate"</span>, <span class="st">"Severe"</span>, <span class="st">"Critical"</span><span class="op">)</span>,</span>
<span>                         ordered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">disease_severity</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] Mild     Severe   Moderate Mild     Critical
Levels: Mild &lt; Moderate &lt; Severe &lt; Critical</code></pre>
</div>
<div class="sourceCode" id="cb280"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compare with ordered factors</span></span>
<span><span class="va">disease_severity</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">disease_severity</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>  <span class="co"># Is Mild less severe than Severe?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] FALSE</code></pre>
</div>
</div>
</section><section id="factor-operations" class="level3"><h3 class="anchored" data-anchor-id="factor-operations">Factor Operations</h3>
<div class="cell">
<div class="sourceCode" id="cb282"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Count frequencies</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">origins_factor</span><span class="op">)</span></span>
<span><span class="va">out</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>origins_factor
    Human Zebrafish     Mouse 
        3         1         2 </code></pre>
</div>
<div class="sourceCode" id="cb284"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">out</span><span class="op">[</span><span class="st">"Human"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Human 
    3 </code></pre>
</div>
<div class="sourceCode" id="cb286"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Calculate proportions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>origins_factor
    Human Zebrafish     Mouse 
0.5000000 0.1666667 0.3333333 </code></pre>
</div>
<div class="sourceCode" id="cb288"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Change reference level (important for statistical models)</span></span>
<span><span class="va">origins_factor_relevel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">origins_factor</span>, ref <span class="op">=</span> <span class="st">"Mouse"</span><span class="op">)</span></span>
<span><span class="va">origins_factor_relevel</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] Human     Mouse     Human     Zebrafish Mouse     Human    
Levels: Mouse Human Zebrafish</code></pre>
</div>
<div class="sourceCode" id="cb290"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Convert to character</span></span>
<span><span class="va">origins_char</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">origins_factor</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot factors - Basic barplot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">origins_factor</span><span class="op">)</span>, </span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Sample Origins"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/factor-operations-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Factor <code>level</code>-ing and <code>relevel</code>-ing are different. <code>relevel</code> redefines what the reference should be. For example, in an experiment, you have <code>control</code>, <code>treatment1</code>, <code>treatment2</code> groups. Your reference might be <code>control</code>. So, all of your comparisons/statistics are on the basis of <code>control</code>. But you might change the reference (by <code>relevel</code> to <code>treatment1</code> and all of your comparison will be on the basis of <code>treatment1</code> group. Got it?</p>
</div>
</div>
<p>More advanced plot with factors:</p>
<div class="cell">
<div class="sourceCode" id="cb291"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5.2</span>, <span class="fl">7.8</span>, <span class="fl">4.5</span>, <span class="fl">12.3</span>, <span class="fl">8.1</span>, <span class="fl">3.7</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">gene_expr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">origins</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Boxplot by factor</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">gene_expr</span> <span class="op">~</span> <span class="va">origins</span>, </span>
<span>        col <span class="op">=</span> <span class="st">"lightblue"</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Gene Expression by Sample Origin"</span>, </span>
<span>        xlab <span class="op">=</span> <span class="st">"Origin"</span>, </span>
<span>        ylab <span class="op">=</span> <span class="st">"Expression Level"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Did you notice how factor <code>level</code>-ing changes the appearance of the categories in the plots? See the <code>barplot</code> and the <code>boxplot</code> again. Where are Zebrafish and Mouse now in the plots? Why are their positions on the x-axis changed?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Keep noticing the output formats. Sometimes the output is just a number, sometimes a vector or table or list, etc. Check <code>prop.table(table(origins_factor))</code>. How is it?</p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Got it?
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p><code>prop &lt;- prop.table(table(origins_factor))</code> ‚Äì is a named numeric vector (atomic vector). <code>prop$Human</code> or similar won‚Äôt work. Check this way: <code>prop</code> <code>prop["Human"]; prop["Mouse"]; prop["Zebrafish"]</code></p>
<p>Or make it a data frame (df) first, then try to use normal way of handling df.</p>
</div>
</div>
</div>
<p>Accessing the Output:</p>
<div class="cell">
<div class="sourceCode" id="cb292"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">origins_factor</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">prop</span> <span class="co">#What do you see? A data frame? No difference?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>origins_factor
    Human Zebrafish     Mouse 
0.5000000 0.1666667 0.3333333 </code></pre>
</div>
<div class="sourceCode" id="cb294"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prop</span><span class="op">[</span><span class="st">"Human"</span><span class="op">]</span>; <span class="va">prop</span><span class="op">[</span><span class="st">"Mouse"</span><span class="op">]</span>; <span class="va">prop</span><span class="op">[</span><span class="st">"Zebrafish"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Human 
  0.5 </code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>    Mouse 
0.3333333 </code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Zebrafish 
0.1666667 </code></pre>
</div>
</div>
</section></section><section id="subsetting-data" class="level2"><h2 class="anchored" data-anchor-id="subsetting-data">Subsetting Data</h2>
<section id="vectors" class="level3"><h3 class="anchored" data-anchor-id="vectors">Vectors</h3>
<div class="cell">
<div class="sourceCode" id="cb298"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create a vector</span></span>
<span><span class="va">expression_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3.2</span>, <span class="fl">4.5</span>, <span class="fl">2.1</span>, <span class="fl">6.7</span>, <span class="fl">5.9</span>, <span class="fl">3.3</span>, <span class="fl">7.8</span>, <span class="fl">2.9</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">expression_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Sample_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span></span>
<span><span class="va">expression_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample_1 Sample_2 Sample_3 Sample_4 Sample_5 Sample_6 Sample_7 Sample_8 
     3.2      4.5      2.1      6.7      5.9      3.3      7.8      2.9 </code></pre>
</div>
<div class="sourceCode" id="cb300"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Subset by position</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>             <span class="co"># Single element</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample_3 
     2.1 </code></pre>
</div>
<div class="sourceCode" id="cb302"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">5</span><span class="op">)</span><span class="op">]</span>    <span class="co"># Multiple elements</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample_1 Sample_3 Sample_5 
     3.2      2.1      5.9 </code></pre>
</div>
<div class="sourceCode" id="cb304"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_data</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>           <span class="co"># Range</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample_2 Sample_3 Sample_4 Sample_5 
     4.5      2.1      6.7      5.9 </code></pre>
</div>
<div class="sourceCode" id="cb306"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Subset by name</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="st">"Sample_6"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample_6 
     3.3 </code></pre>
</div>
<div class="sourceCode" id="cb308"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample_1"</span>, <span class="st">"Sample_8"</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample_1 Sample_8 
     3.2      2.9 </code></pre>
</div>
<div class="sourceCode" id="cb310"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Subset by condition</span></span>
<span><span class="va">expression_data</span><span class="op">[</span><span class="va">expression_data</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">]</span>              <span class="co"># Values &gt; 5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample_4 Sample_5 Sample_7 
     6.7      5.9      7.8 </code></pre>
</div>
<div class="sourceCode" id="cb312"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_data</span><span class="op">[</span><span class="va">expression_data</span> <span class="op">&gt;=</span> <span class="fl">3</span> <span class="op">&amp;</span> <span class="va">expression_data</span> <span class="op">&lt;=</span> <span class="fl">6</span><span class="op">]</span>  <span class="co"># Values between 3 and 6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample_1 Sample_2 Sample_5 Sample_6 
     3.2      4.5      5.9      3.3 </code></pre>
</div>
</div>
</section><section id="data-frames" class="level3"><h3 class="anchored" data-anchor-id="data-frames">Data Frames</h3>
<div class="cell">
<div class="sourceCode" id="cb314"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create a data frame</span></span>
<span><span class="va">gene_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"EGFR"</span>, <span class="st">"GAPDH"</span><span class="op">)</span>,</span>
<span>  expression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8.2</span>, <span class="fl">6.1</span>, <span class="fl">9.5</span>, <span class="fl">7.0</span>, <span class="fl">10.0</span><span class="op">)</span>,</span>
<span>  mutation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"No"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  pathway <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DNA Repair"</span>, <span class="st">"Apoptosis"</span>, <span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>, <span class="st">"Metabolism"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_df</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  gene_id expression mutation    pathway
1   BRCA1        8.2      Yes DNA Repair
2    TP53        6.1       No  Apoptosis
3     MYC        9.5      Yes Cell Cycle
4    EGFR        7.0       No  Signaling
5   GAPDH       10.0       No Metabolism</code></pre>
</div>
<div class="sourceCode" id="cb316"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Subsetting by row index</span></span>
<span><span class="va">gene_df</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="op">]</span>         <span class="co"># First three rows, all columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  gene_id expression mutation    pathway
1   BRCA1        8.2      Yes DNA Repair
2    TP53        6.1       No  Apoptosis
3     MYC        9.5      Yes Cell Cycle</code></pre>
</div>
<div class="sourceCode" id="cb318"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Subsetting by column index</span></span>
<span><span class="va">gene_df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>     <span class="co"># All rows, first two columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  gene_id expression
1   BRCA1        8.2
2    TP53        6.1
3     MYC        9.5
4    EGFR        7.0
5   GAPDH       10.0</code></pre>
</div>
<div class="sourceCode" id="cb320"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Subsetting by column name</span></span>
<span><span class="va">gene_df</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"mutation"</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  gene_id mutation
1   BRCA1      Yes
2    TP53       No
3     MYC      Yes
4    EGFR       No
5   GAPDH       No</code></pre>
</div>
<div class="sourceCode" id="cb322"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Using the $ operator</span></span>
<span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1]  8.2  6.1  9.5  7.0 10.0</code></pre>
</div>
<div class="sourceCode" id="cb324"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_df</span><span class="op">$</span><span class="va">mutation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] Yes No  Yes No  No 
Levels: No Yes</code></pre>
</div>
<div class="sourceCode" id="cb326"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Subsetting by condition</span></span>
<span><span class="va">gene_df</span><span class="op">[</span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">8</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  gene_id expression mutation    pathway
1   BRCA1        8.2      Yes DNA Repair
3     MYC        9.5      Yes Cell Cycle
5   GAPDH       10.0       No Metabolism</code></pre>
</div>
<div class="sourceCode" id="cb328"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_df</span><span class="op">[</span><span class="va">gene_df</span><span class="op">$</span><span class="va">mutation</span> <span class="op">==</span> <span class="st">"Yes"</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  gene_id expression mutation    pathway
1   BRCA1        8.2      Yes DNA Repair
3     MYC        9.5      Yes Cell Cycle</code></pre>
</div>
<div class="sourceCode" id="cb330"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Multiple conditions</span></span>
<span><span class="va">gene_df</span><span class="op">[</span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">7</span> <span class="op">&amp;</span> <span class="va">gene_df</span><span class="op">$</span><span class="va">mutation</span> <span class="op">==</span> <span class="st">"No"</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  gene_id expression mutation    pathway
5   GAPDH         10       No Metabolism</code></pre>
</div>
</div>
<p><strong>Logical Operators</strong></p>
<table class="table">
<thead><tr class="header">
<th>Operator</th>
<th>Meaning</th>
<th>Example</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>==</code></td>
<td>Equal to</td>
<td><code>x == 5</code></td>
</tr>
<tr class="even">
<td><code>!=</code></td>
<td>Not equal</td>
<td><code>x != 5</code></td>
</tr>
<tr class="odd">
<td><code>&lt;</code></td>
<td>Less than</td>
<td><code>x &lt; 5</code></td>
</tr>
<tr class="even">
<td><code>&gt;</code></td>
<td>Greater than</td>
<td><code>x &gt; 5</code></td>
</tr>
<tr class="odd">
<td><code>&lt;=</code></td>
<td>Less or equal</td>
<td><code>x &lt;= 5</code></td>
</tr>
<tr class="even">
<td><code>&gt;=</code></td>
<td>Greater or equal</td>
<td><code>x &gt;= 5</code></td>
</tr>
<tr class="odd">
<td><code>!</code></td>
<td>Not</td>
<td><code>!(x &lt; 5)</code></td>
</tr>
<tr class="even">
<td><code>|</code></td>
<td>OR</td>
<td><code>x &lt; 5 | x &gt; 10</code></td>
</tr>
<tr class="odd">
<td><code>&amp;</code></td>
<td>AND</td>
<td><code>x &gt; 5 &amp; x &lt; 10</code></td>
</tr>
</tbody>
</table></section><section id="row-names-in-data-frames" class="level3"><h3 class="anchored" data-anchor-id="row-names-in-data-frames">Row Names in Data Frames</h3>
<p>Row names are particularly important in bioinformatics where genes, proteins, or samples are often used as identifiers.</p>
<div class="cell">
<div class="sourceCode" id="cb332"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Setting row names for gene_df</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene_df</span><span class="op">$</span><span class="va">gene_id</span></span>
<span><span class="va">gene_df</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      gene_id expression mutation    pathway
BRCA1   BRCA1        8.2      Yes DNA Repair
TP53     TP53        6.1       No  Apoptosis
MYC       MYC        9.5      Yes Cell Cycle
EGFR     EGFR        7.0       No  Signaling
GAPDH   GAPDH       10.0       No Metabolism</code></pre>
</div>
</div>
<p>We can now drop the gene_id column, if required.</p>
<div class="cell">
<div class="sourceCode" id="cb334"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_df_clean</span> <span class="op">&lt;-</span> <span class="va">gene_df</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span>  <span class="co"># Remove the first column</span></span>
<span><span class="va">gene_df_clean</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>      expression mutation    pathway
BRCA1        8.2      Yes DNA Repair
TP53         6.1       No  Apoptosis
MYC          9.5      Yes Cell Cycle
EGFR         7.0       No  Signaling
GAPDH       10.0       No Metabolism</code></pre>
</div>
<div class="sourceCode" id="cb336"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Access rows by name</span></span>
<span><span class="va">gene_df_clean</span><span class="op">[</span><span class="st">"TP53"</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>     expression mutation   pathway
TP53        6.1       No Apoptosis</code></pre>
</div>
<div class="sourceCode" id="cb338"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check if row names are unique</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_df_clean</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] FALSE</code></pre>
</div>
<div class="sourceCode" id="cb340"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Handle potential duplicated row names</span></span>
<span><span class="co"># NOTE: R doesn't allow duplicate row names by default</span></span>
<span><span class="va">dup_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  expression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5.2</span>, <span class="fl">6.3</span>, <span class="fl">5.2</span>, <span class="fl">8.1</span><span class="op">)</span>,</span>
<span>  mutation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># This would cause an error:</span></span>
<span><span class="co">#rownames(dup_genes) &lt;- c("BRCA1", "BRCA1", "TP53", "EGFR")</span></span>
<span></span>
<span><span class="co"># Instead, we can preemptively make them unique:</span></span>
<span><span class="va">proposed_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"EGFR"</span><span class="op">)</span></span>
<span><span class="va">unique_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/make.unique.html">make.unique</a></span><span class="op">(</span><span class="va">proposed_names</span><span class="op">)</span></span>
<span><span class="va">unique_names</span>  <span class="co"># Show the generated unique names</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "BRCA1"   "BRCA1.1" "TP53"    "EGFR"   </code></pre>
</div>
<div class="sourceCode" id="cb342"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Now we can safely assign them</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">dup_genes</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">unique_names</span></span>
<span><span class="va">dup_genes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>        expression mutation
BRCA1          5.2      Yes
BRCA1.1        6.3       No
TP53           5.2      Yes
EGFR           8.1       No</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Why is unique name important for us? Imagine this meaningful biological scenario: one gene might transcribed into many transcript isoforms and hence many protein isoforms. From RNAseq data, we might get alignment count for each gene. But then we can separate the count for each transcript. One gene has one name or ID, but the transcripts are many for the same gene! So, we can denote, for example, 21 isoform of geneA like genA.1, geneA.2, geneA.3,‚Ä¶‚Ä¶., geneA.21. See this link for <a href="https://www.ncbi.nlm.nih.gov/gene?Db=gene&amp;Cmd=DetailsSearch&amp;Term=4155">MBP gene</a>. How many transcript isoforms does it have?</p>
</div>
</div>
</section></section><section id="homeworks-factors-subsetting-and-biological-insight" class="level2"><h2 class="anchored" data-anchor-id="homeworks-factors-subsetting-and-biological-insight">üè° Homeworks: Factors, Subsetting, and Biological Insight</h2>
<ol type="1">
<li><p><strong>(Factor vs Character)</strong> Explain the difference between a character vector and a factor in R. Why would <code>mutation_status</code> be a factor and not just a character vector?</p></li>
<li><p><strong>(Factor Level Order)</strong> You observed the following bacterial species in gut microbiome samples:</p></li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb344"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Lactobacillus"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Lactobacillus"</span><span class="op">)</span></span>
<span><span class="va">species_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">species</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Lactobacillus"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What will <code>levels(species_factor)</code> return? Why?</p>
<ol start="3" type="1">
<li>Given the factor:</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb345"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">disease_severity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mild"</span>, <span class="st">"Severe"</span>, <span class="st">"Moderate"</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mild"</span>, <span class="st">"Moderate"</span>, <span class="st">"Severe"</span>, <span class="st">"Critical"</span><span class="op">)</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What will be the result of <code>disease_severity[1] &lt; disease_severity[2]</code> and why?</p>
<ol start="4" type="1">
<li>You computed:</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb346"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">species_factor</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How do you extract the proportion of ‚ÄúEscherichia‚Äù samples from prop? Is <code>prop$Escherichia</code> valid?</p>
<ol start="5" type="1">
<li>Interpret what this query returns:</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb347"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_df</span><span class="op">[</span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">7</span> <span class="op">&amp;</span> <span class="va">gene_df</span><span class="op">$</span><span class="va">mutation</span> <span class="op">==</span> <span class="st">"No"</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What type of genes does it select?</p>
<ol start="6" type="1">
<li>You have:</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb348"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span><span class="op">)</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5.2</span>, <span class="fl">8.1</span>, <span class="fl">4.3</span>, <span class="fl">9.0</span>, <span class="fl">5.7</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Make a dataframe using these 2 vectors first. Then,</p>
<ol type="a">
<li>Create a factor <code>group_factor</code> for the samples.</li>
<li>Use <code><a href="https://rdrr.io/r/base/tapply.html">tapply()</a></code> to calculate mean expression per group.</li>
</ol>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use <code>?tapply()</code> to see how to use it.</p>
<p><strong>Hint:</strong> You need to provide things for <code>X</code>, <code>INDEX</code>, <code>FUN</code>. You have <code>X</code>, <code>INDEX</code> in this small dataframe. The <code>FUN</code> should be applied thinking of what you are trying to do. You are trying to get the mean or average, right?</p>
</div>
</div>
<ol start="3" type="a">
<li>Plot a <code>barplot</code> of average expression for each group.</li>
</ol>
<!-- --><ol start="7" type="1">
<li>Use the <code>gene_df</code> example. Subset the data to find genes with:</li>
</ol>
<ul>
<li>expression &gt; 8</li>
<li>pathway is either ‚ÄúCell Cycle‚Äù or ‚ÄúSignaling‚Äù</li>
</ul>
<ol start="8" type="1">
<li><p>Create an ordered factor for the disease stages: <code>c("Stage I", "Stage III", "Stage II", "Stage IV", "Stage I")</code>. Then plot the number of patients per stage using <code><a href="https://rdrr.io/r/graphics/barplot.html">barplot()</a></code>. Confirm that <code>"Stage III" &gt; "Stage I"</code> is logical in your factor.</p></li>
<li><p>Suppose <code>gene_data</code> has a column type with values ‚ÄúOncogene‚Äù, ‚ÄúTumor Suppressor‚Äù, and ‚ÄúHousekeeping‚Äù.</p></li>
</ol>
<ul>
<li>Subset all ‚ÄúOncogene‚Äù rows where expression &gt; 8.</li>
<li>Change the reference level of the factor type to ‚ÄúHousekeeping‚Äù</li>
</ul>
<ol start="10" type="1">
<li>Simulate expression data for 3 tissues (see the code chunk below): We are going to use <code><a href="https://rdrr.io/r/stats/Normal.html">rnorm()</a></code> function to generate random values from a normal distribution for this purpose. The example values inside the <code><a href="https://rdrr.io/r/stats/Normal.html">rnorm()</a></code> function means we want:</li>
</ol>
<ul>
<li>30 values in total,</li>
<li>average or mean value = 8,</li>
<li>standard deviation of expression is 2.</li>
</ul>
<p>You can play with the numbers to make your own values.</p>
<p><code><a href="https://rdrr.io/r/base/rep.html">rep()</a></code> function is to replicate things (many times). In this example, we have <code>rep(c("brain", "liver", "kidney"), each = 10)</code>. We will be having 10x ‚Äúbrains‚Äù, followed by 10x ‚Äúliver‚Äù, followed by 10x ‚Äúkidney‚Äù. So, if you have changed your values inside the <code><a href="https://rdrr.io/r/stats/Normal.html">rnorm()</a></code> function, make this value meaningful for you. Now we have 3 things, each=10. So, 3*10=30 is matching with the total value inside <code><a href="https://rdrr.io/r/stats/Normal.html">rnorm()</a></code> function. Got it?</p>
<div class="cell">
<div class="sourceCode" id="cb349"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span> <span class="co">#just for reproducibility. Not completely needed</span></span>
<span><span class="va">gene_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">30</span>, mean <span class="op">=</span> <span class="fl">8</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"brain"</span>, <span class="st">"liver"</span>, <span class="st">"kidney"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">tissue_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">tissue</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"liver"</span>, <span class="st">"brain"</span>, <span class="st">"kidney"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Make a boxplot showing expression per tissue.</li>
<li>Which tissue shows the most variable gene expression? (Use <code><a href="https://rdrr.io/r/base/tapply.html">tapply()</a></code> + <code><a href="https://rdrr.io/r/stats/sd.html">sd()</a></code>)</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Hint:</strong> Variability is an expression of measuring standard deviation (<code>sd</code>) just by squaring it. So, <code>var</code> = <code>sd^2</code>. Well, do you see how to use <code>sd</code> inside <code><a href="https://rdrr.io/r/base/tapply.html">tapply()</a></code> function? Use <code>?tapply()</code> to know how to use it.</p>
</div>
</div>
<p>Use these questions as a self-check ‚Äì reflect on why each step works before moving on to the next level (question).</p>
<p><strong>Push your <code>.Rmd</code> file and share by Friday 10PM BD Time.</strong></p>
</section><section id="conditionals" class="level2"><h2 class="anchored" data-anchor-id="conditionals">Conditionals</h2>
<section id="if-else-statement" class="level3"><h3 class="anchored" data-anchor-id="if-else-statement">
<code>if-else</code> statement</h3>
<p>General structure of <code>if-else</code> statement:</p>
<div class="cell">
<div class="sourceCode" id="cb350"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">condition1</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Code executed when condition1 is TRUE</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">condition2</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># Code executed when condition1 is FALSE but condition2 is TRUE</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="co"># Code executed when all conditions above are FALSE</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let‚Äôs use it now.</p>
<div class="cell">
<div class="sourceCode" id="cb351"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># if-else statement</span></span>
<span><span class="va">gene_value</span> <span class="op">&lt;-</span> <span class="fl">6.8</span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="va">gene_value</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"High expression"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span><span class="op">(</span><span class="va">gene_value</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Medium expression"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Low expression"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Medium expression"</code></pre>
</div>
</div>
<p>Visualize it using the image below.</p>
<div class="text-center">
<p><img src="img/if_else.png" class="img-fluid" style="width:80.0%"></p>
</div>
</section><section id="ifelse-statement-for-vectors" class="level3"><h3 class="anchored" data-anchor-id="ifelse-statement-for-vectors">
<code>ifelse</code> statement for vectors</h3>
<p><code>ifelse</code> is binary in nature. So, we can categorize only 2 things using <code>ifelse</code>. The structure is:</p>
<div class="cell">
<div class="sourceCode" id="cb353"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">test_or_condition</span>, <span class="st">"value_if_condition_is_TRUE"</span>, <span class="st">"value_if_condition_is_FALSE"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>See this example:</p>
<div class="cell">
<div class="sourceCode" id="cb354"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expression_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">12.5</span>, <span class="fl">4.3</span>, <span class="fl">8.1</span>, <span class="fl">2.2</span><span class="op">)</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">expression_values</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="st">"Upregulated"</span>, <span class="st">"Downregulated"</span><span class="op">)</span></span>
<span><span class="va">labels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Upregulated"   "Downregulated" "Upregulated"   "Downregulated"</code></pre>
</div>
</div>
<div class="text-center">
<p><img src="img/elseif.png" class="img-fluid" style="width:80.0%"></p>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>ifelse</code> has 3 things inside the parentheses, right? The first one is the condition, the second one is the category we define if the condition is met, and the third thing is the other remaining category we want to assign if the condition is not met. So, it‚Äôs usage is perfect to say if a gene/transcript is upregulated or downregulated (binary classification).</p>
</div>
</div>
<p>If we still want to categorize more than 2 categories using <code>ifelse</code>, we need to use it in a nested way. The structure will be like this:</p>
<div class="cell">
<div class="sourceCode" id="cb356"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">test1</span>, <span class="va">value1</span>,</span>
<span>       <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">test2</span>, <span class="va">value2</span>, </span>
<span>              <span class="va">value3</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>See this example:</p>
<div class="cell">
<div class="sourceCode" id="cb357"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># ifelse() for vectors</span></span>
<span><span class="va">expression_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2.5</span>, <span class="fl">5.8</span>, <span class="fl">7.2</span>, <span class="fl">3.1</span>, <span class="fl">6.9</span><span class="op">)</span></span>
<span><span class="va">expression_category</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">expression_levels</span> <span class="op">&gt;</span> <span class="fl">6</span>, </span>
<span>                             <span class="st">"High"</span>, </span>
<span>                             <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">expression_levels</span> <span class="op">&gt;</span> <span class="fl">4</span>, <span class="st">"Medium"</span>, <span class="st">"Low"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">expression_category</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Low"    "Medium" "High"   "Low"    "High"  </code></pre>
</div>
</div>
<div class="text-center">
<p><img src="img/nested_elseif.png" class="img-fluid" style="width:80.0%"></p>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>You remember the general structure of <code>ifelse</code> loop, right? the second thing after the first <code>,</code> is the assigned category if the condition is met. So, we assigned it as <code>High</code> here in this example. But then after the second <code>,</code> there is a second <code>ifelse</code> loop instead of a category. The second loop makes 2 more binary categories <code>Medium</code> and <code>Low</code>, and our task of assigning 3 categories is achieved.</p>
</div>
</div>
<p><code>dplyr</code> package has a function named <code>case-when()</code> to help us use as many categories we want. The same task would be achieved like this:</p>
<div class="cell">
<div class="sourceCode" id="cb359"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Requires dplyr package</span></span>
<span><span class="co">#install.packages("dplyr") #decomment if you need to install the package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">expression_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2.5</span>, <span class="fl">5.8</span>, <span class="fl">7.2</span>, <span class="fl">3.1</span>, <span class="fl">6.9</span><span class="op">)</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span></span>
<span>  <span class="va">expression_levels</span> <span class="op">&gt;</span> <span class="fl">6</span> <span class="op">~</span> <span class="st">"High"</span>,</span>
<span>  <span class="va">expression_levels</span> <span class="op">&gt;</span> <span class="fl">4</span> <span class="op">~</span> <span class="st">"Medium"</span>,</span>
<span>  <span class="cn">TRUE</span> <span class="op">~</span> <span class="st">"Low"</span>  <span class="co"># Default case</span></span>
<span><span class="op">)</span></span>
<span><span class="va">labels</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Low"    "Medium" "High"   "Low"    "High"  </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do you see the point how you would use the <code>ifelse</code> loop if you wanted to write a function to make 4 or 5 categories? If not, pause and re-think. You need to see the point. But anyway, categorizing more than 2 is better using <code>if-else</code> statement</p>
</div>
</div>
</section><section id="for-loop" class="level3"><h3 class="anchored" data-anchor-id="for-loop">
<code>for</code> loop</h3>
<div class="cell">
<div class="sourceCode" id="cb361"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"CDC2"</span>, <span class="st">"MBP"</span><span class="op">)</span></span>
<span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8.2</span>, <span class="fl">5.4</span>, <span class="fl">11.0</span>, <span class="fl">5.4</span>, <span class="fl">13.0</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">status</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="va">expr</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span> <span class="st">"High"</span> </span>
<span>  <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">expr</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">6</span><span class="op">)</span> <span class="st">"Moderate"</span> </span>
<span>  <span class="kw">else</span> <span class="st">"Low"</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="va">genes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, <span class="st">"has"</span>, <span class="va">status</span>, <span class="st">"expression\n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>BRCA1 has Moderate expression
TP53 has Low expression
MYC has High expression
CDC2 has Low expression
MBP has High expression</code></pre>
</div>
</div>
<p>In-class Task:</p>
<ul>
<li>Make a data frame using <code>genes</code> and <code>expr</code>.</li>
<li>Add/flag the categories <code>High</code>, <code>Moderate</code> and <code>Low</code> you get using the <code>for</code> loop in a new column named <code>expression_level</code> or similar.</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb363"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Step 1: Vectors</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"CDC2"</span>, <span class="st">"MBP"</span><span class="op">)</span></span>
<span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8.2</span>, <span class="fl">5.4</span>, <span class="fl">11.0</span>, <span class="fl">5.4</span>, <span class="fl">13.0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 2: Create a data frame</span></span>
<span><span class="va">gene_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">genes</span>, expression <span class="op">=</span> <span class="va">expr</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Step 3: Add an empty column for expression level</span></span>
<span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression_level</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span></span>
<span><span class="co"># Step 4: Use for loop to fill in the expression_level column</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">gene_df</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">gene_df</span><span class="op">$</span><span class="va">expression_level</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="kw">if</span> <span class="op">(</span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="st">"High"</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">6</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="st">"Moderate"</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="st">"Low"</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># View the final data frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">gene_df</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>   gene expression expression_level
1 BRCA1        8.2         Moderate
2  TP53        5.4              Low
3   MYC       11.0             High
4  CDC2        5.4              Low
5   MBP       13.0             High</code></pre>
</div>
</div>
</section><section id="while-loop" class="level3"><h3 class="anchored" data-anchor-id="while-loop">
<code>while</code> loop</h3>
<p>Context:</p>
<p>You are preparing biological samples (e.g., blood, DNA extracts) for analysis. You have a set of samples labeled <code>Sample 1</code> to <code>Sample 5</code>. You want to check each one in order and confirm that it‚Äôs ready for analysis. Use a while loop to process the samples sequentially.</p>
<div class="cell">
<div class="sourceCode" id="cb365"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">i</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="kw">while</span> <span class="op">(</span><span class="va">i</span> <span class="op">&lt;=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"Sample"</span>, <span class="va">i</span>, <span class="st">"is ready for analysis\n"</span><span class="op">)</span></span>
<span>  <span class="va">i</span> <span class="op">&lt;-</span> <span class="va">i</span> <span class="op">+</span> <span class="fl">1</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Sample 1 is ready for analysis
Sample 2 is ready for analysis
Sample 3 is ready for analysis
Sample 4 is ready for analysis
Sample 5 is ready for analysis</code></pre>
</div>
</div>
</section><section id="next-and-break" class="level3"><h3 class="anchored" data-anchor-id="next-and-break">
<code>next</code> and <code>break</code>
</h3>
<p>Context:</p>
<p>You are screening biological samples (e.g., tissue or blood) in a quality control process. Some samples are good, some are suboptimal (not contaminated but poor quality), and some are contaminated (must be flagged and stop further processing). Use <code><a href="https://rdrr.io/r/base/Control.html">next</a></code> to skip suboptimal samples and <code><a href="https://rdrr.io/r/base/Control.html">break</a></code> to immediately stop when a contaminated sample is found.</p>
<div class="cell">
<div class="sourceCode" id="cb367"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"good"</span>, <span class="st">"bad"</span>, <span class="st">"good"</span>, <span class="st">"contaminated"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="va">samples</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">i</span> <span class="op">==</span> <span class="st">"contaminated"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Stop! Contaminated sample."</span><span class="op">)</span></span>
<span>    <span class="kw">break</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">i</span> <span class="op">==</span> <span class="st">"bad"</span><span class="op">)</span> <span class="kw">next</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Processing"</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Processing good"
[1] "Processing good"
[1] "Stop! Contaminated sample."</code></pre>
</div>
</div>
</section></section><section id="writing-functions-in-r" class="level2"><h2 class="anchored" data-anchor-id="writing-functions-in-r">Writing Functions in R</h2>
<p>The syntax to write an R function is:</p>
<div class="cell">
<div class="sourceCode" id="cb369"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">function_name</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="op">)</span> <span class="op">{</span></span>
<span><span class="op">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let‚Äôs use it.</p>
<section id="flag-gene-expression" class="level3"><h3 class="anchored" data-anchor-id="flag-gene-expression">Flag gene expression</h3>
<div class="cell">
<div class="sourceCode" id="cb370"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">flag_expression</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">value</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">value</span> <span class="op">&gt;</span> <span class="fl">10</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="st">"High"</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">value</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="st">"Moderate"</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="st">"Low"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">flag_expression</span><span class="op">(</span><span class="fl">8.3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "Moderate"</code></pre>
</div>
</div>
<p><strong>Apply to a vector</strong></p>
<div class="cell">
<div class="sourceCode" id="cb372"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expr_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">12.2</span>, <span class="fl">4.4</span>, <span class="fl">7.5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span>X<span class="op">=</span><span class="va">expr_values</span>, FUN<span class="op">=</span><span class="va">flag_expression</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "High"     "Low"      "Moderate"</code></pre>
</div>
</div>
</section><section id="function-with-multiple-arguments" class="level3"><h3 class="anchored" data-anchor-id="function-with-multiple-arguments">Function with multiple arguments</h3>
<div class="cell">
<div class="sourceCode" id="cb374"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_status</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">gene</span>, <span class="va">expression</span>, <span class="va">threshold</span> <span class="op">=</span> <span class="fl">6</span><span class="op">)</span> <span class="op">{</span> <span class="co">#gene = "string", expression = value</span></span>
<span>  <span class="va">label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="va">threshold</span>, <span class="st">"Up"</span>, <span class="st">"Down"</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">gene</span>, <span class="st">"is"</span>, <span class="va">label</span>, <span class="st">"regulated"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">gene_status</span><span class="op">(</span><span class="st">"TP53"</span>, <span class="fl">8.1</span>, threshold <span class="op">=</span> <span class="fl">9</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] "TP53 is Down regulated"</code></pre>
</div>
</div>
</section><section id="return-a-list" class="level3"><h3 class="anchored" data-anchor-id="return-a-list">Return a list</h3>
<div class="cell">
<div class="sourceCode" id="cb376"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">calc_stats</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">values</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">values</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">calc_stats</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4.2</span>, <span class="fl">5.5</span>, <span class="fl">7.8</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>$mean
[1] 5.833333

$sd
[1] 1.823001</code></pre>
</div>
</div>
</section></section><section id="homeworks-control-structures-and-functions" class="level2"><h2 class="anchored" data-anchor-id="homeworks-control-structures-and-functions">üè° Homeworks: Control Structures and Functions</h2>
<ol type="1">
<li>
<code>for</code> Loop + Conditional: Gene Expression Classifier</li>
</ol>
<p>You have:</p>
<div class="cell">
<div class="sourceCode" id="cb378"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"EGFR"</span>, <span class="st">"MYC"</span>, <span class="st">"BRCA2"</span><span class="op">)</span></span>
<span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4.8</span>, <span class="fl">6.3</span>, <span class="fl">11.7</span>, <span class="fl">9.5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Task:</strong></p>
<p>Loop over the gene names and print: <code>&lt;Gene&gt; has &lt;Low/Moderate/High&gt; expression</code></p>
<p>Based on:</p>
<ul>
<li>High: &gt;10</li>
<li>Moderate: 6‚Äì10</li>
<li>Low: ‚â§6</li>
</ul>
<ol start="2" type="1">
<li>
<code><a href="https://rdrr.io/r/base/Control.html">next</a></code>: Sample Screening Skips Bad Samples</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb379"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"good"</span>, <span class="st">"bad"</span>, <span class="st">"good"</span>, <span class="st">"contaminated"</span>, <span class="st">"good"</span>, <span class="st">"bad"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>Context:</strong></p>
<p>Bad samples should be skipped, only good and contaminated should be reported.</p>
<p><strong>Task:</strong></p>
<p>Use a <code>for</code> loop with <code><a href="https://rdrr.io/r/base/Control.html">next</a></code> to skip <code>bad</code> samples and print a message for others: <code>Sample X is flagged for analysis</code></p>
<ol start="3" type="1">
<li>
<code><a href="https://rdrr.io/r/base/Control.html">break</a></code>: Stop at Contaminated Sample Use the same samples vector above.</li>
</ol>
<p><strong>Task:</strong></p>
<p>Print <code>Analyzing sample X</code> for each sample. But if a <code>contaminated</code> sample is found, stop processing immediately and print <code>Contamination detected! Halting...</code></p>
<ol start="4" type="1">
<li>
<code>while</code> Loop: Sample Prep Countdown</li>
</ol>
<p><strong>Context:</strong></p>
<p>You‚Äôre preparing 5 samples.</p>
<p><strong>Task:</strong></p>
<p>Use a while loop to print: <code>Sample &lt;i&gt; is ready for analysis</code> from 1 to 5.</p>
<ol start="5" type="1">
<li>
<code>while</code> + Condition: Until Threshold You‚Äôre measuring a protein level that starts at 2. With each measurement, it increases randomly between 0.5 and 1.5 units.</li>
</ol>
<p><strong>Task:</strong> Use a while loop to simulate the increase until the protein level exceeds 10. Print each step.</p>
<div class="cell">
<div class="sourceCode" id="cb380"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">level</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="co">#your code goes here</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="6" type="1">
<li>Custom Function: Classify BMI</li>
</ol>
<p>Write a function <code>bmi_category(weight, height)</code> that:</p>
<ul>
<li><p>Takes weight (kg) and height (m)</p></li>
<li><p>Calculates BMI: weight/height¬≤</p></li>
</ul>
<p>Returns <code>Underweight</code> if <code>&lt;18.5</code>, <code>Normal</code> if <code>18.5‚Äì24.9</code>, <code>Overweight</code> if <code>25‚Äì29.9</code>, <code>Obese</code> if <code>‚â•30</code>.</p>
<p>Test it:</p>
<p><code>bmi_category(65, 1.70)</code></p>
<ol start="7" type="1">
<li>Function: Gene Risk Calculator</li>
</ol>
<p>Write a function <code>gene_risk(expr, mut_status)</code>:</p>
<p><code>High</code> if <code>expression &gt; 9</code> and <code>mutation = Yes</code>.</p>
<p><code>Moderate</code> if <code>expression &gt; 6</code>.</p>
<p><code>Low</code> otherwise.</p>
<p>Test on vectors:</p>
<div class="cell">
<div class="sourceCode" id="cb381"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5.5</span>, <span class="fl">10.1</span>, <span class="fl">7.3</span><span class="op">)</span></span>
<span><span class="va">mut_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Hint: Use <code><a href="https://rdrr.io/r/base/mapply.html">mapply()</a></code>.</p>
<ol start="8" type="1">
<li><p><code>for</code> Loop + Data Frame Make a data frame of 4 genes and expression values. Add a new column category using a for loop that assigns <code>High</code>, <code>Moderate</code>, <code>Low</code> based on expression.</p></li>
<li>
<p>Function + Error Checking Write a function <code>get_expression_level(gene, df)</code> that:</p>
<p>Takes a gene name and a data frame with <code>gene</code> and <code>expr</code> columns</p>
<p>If the gene is present, returns <code>Low</code>, <code>Moderate</code> or <code>High</code>.</p>
<p>If missing, returns <code>Gene not found</code></p>
<p>Test it with:</p>
</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb382"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"BRCA1"</span><span class="op">)</span>, expr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5.5</span>, <span class="fl">10.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">get_expression_level</span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="va">df</span><span class="op">)</span></span>
<span><span class="fu">get_expression_level</span><span class="op">(</span><span class="st">"EGFR"</span>, <span class="va">df</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="handling-missingwrong-values" class="level2"><h2 class="anchored" data-anchor-id="handling-missingwrong-values">Handling Missing/Wrong Values</h2>
<section id="identifying-issues" class="level3"><h3 class="anchored" data-anchor-id="identifying-issues">Identifying Issues</h3>
<div class="cell">
<div class="sourceCode" id="cb383"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create data with missing values</span></span>
<span><span class="va">clinical_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  patient_id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,</span>
<span>  age <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">25</span>, <span class="fl">99</span>, <span class="fl">30</span>, <span class="op">-</span><span class="fl">5</span>, <span class="fl">40</span><span class="op">)</span>,    <span class="co"># -5 is wrong, 99 is suspect</span></span>
<span>  bp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">120</span>, <span class="cn">NA</span>, <span class="fl">115</span>, <span class="fl">125</span>, <span class="fl">118</span><span class="op">)</span>,  <span class="co"># NA is missing</span></span>
<span>  weight <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">65</span>, <span class="fl">70</span>, <span class="cn">NA</span>, <span class="fl">68</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>   <span class="co"># -1 is wrong</span></span>
<span><span class="op">)</span></span>
<span><span class="va">clinical_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  patient_id age  bp weight
1          1  25 120     65
2          2  99  NA     70
3          3  30 115     NA
4          4  -5 125     68
5          5  40 118     -1</code></pre>
</div>
<div class="sourceCode" id="cb385"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check for missing values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clinical_data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>     patient_id   age    bp weight
[1,]      FALSE FALSE FALSE  FALSE
[2,]      FALSE FALSE  TRUE  FALSE
[3,]      FALSE FALSE FALSE   TRUE
[4,]      FALSE FALSE FALSE  FALSE
[5,]      FALSE FALSE FALSE  FALSE</code></pre>
</div>
<div class="sourceCode" id="cb387"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clinical_data</span><span class="op">)</span><span class="op">)</span>  <span class="co"># Count NAs by column</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>patient_id        age         bp     weight 
         0          0          1          1 </code></pre>
</div>
<div class="sourceCode" id="cb389"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Check for impossible values</span></span>
<span><span class="va">clinical_data</span><span class="op">$</span><span class="va">age</span> <span class="op">&lt;</span> <span class="fl">0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] FALSE FALSE FALSE  TRUE FALSE</code></pre>
</div>
<div class="sourceCode" id="cb391"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">clinical_data</span><span class="op">$</span><span class="va">weight</span> <span class="op">&lt;</span> <span class="fl">0</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] FALSE FALSE    NA FALSE  TRUE</code></pre>
</div>
<div class="sourceCode" id="cb393"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Find indices of problematic values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">clinical_data</span><span class="op">$</span><span class="va">age</span> <span class="op">&lt;</span> <span class="fl">0</span> <span class="op">|</span> <span class="va">clinical_data</span><span class="op">$</span><span class="va">age</span> <span class="op">&gt;</span> <span class="fl">90</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 2 4</code></pre>
</div>
</div>
</section><section id="fixing-data" class="level3"><h3 class="anchored" data-anchor-id="fixing-data">Fixing Data</h3>
<div class="cell">
<div class="sourceCode" id="cb395"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Replace impossible values with NA</span></span>
<span><span class="va">clinical_data</span><span class="op">$</span><span class="va">age</span><span class="op">[</span><span class="va">clinical_data</span><span class="op">$</span><span class="va">age</span> <span class="op">&lt;</span> <span class="fl">0</span> <span class="op">|</span> <span class="va">clinical_data</span><span class="op">$</span><span class="va">age</span> <span class="op">&gt;</span> <span class="fl">90</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">clinical_data</span><span class="op">$</span><span class="va">weight</span><span class="op">[</span><span class="va">clinical_data</span><span class="op">$</span><span class="va">weight</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">clinical_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  patient_id age  bp weight
1          1  25 120     65
2          2  NA  NA     70
3          3  30 115     NA
4          4  NA 125     68
5          5  40 118     NA</code></pre>
</div>
<div class="sourceCode" id="cb397"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Replace NAs with mean (common in biological data)</span></span>
<span><span class="va">clinical_data</span><span class="op">$</span><span class="va">bp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clinical_data</span><span class="op">$</span><span class="va">bp</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">clinical_data</span><span class="op">$</span><span class="va">bp</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">clinical_data</span><span class="op">$</span><span class="va">weight</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clinical_data</span><span class="op">$</span><span class="va">weight</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">clinical_data</span><span class="op">$</span><span class="va">weight</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">clinical_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  patient_id age    bp   weight
1          1  25 120.0 65.00000
2          2  NA 119.5 70.00000
3          3  30 115.0 67.66667
4          4  NA 125.0 68.00000
5          5  40 118.0 67.66667</code></pre>
</div>
<div class="sourceCode" id="cb399"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Replace NAs with median (better for skewed data)</span></span>
<span><span class="va">clinical_data</span><span class="op">$</span><span class="va">age</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">clinical_data</span><span class="op">$</span><span class="va">age</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">clinical_data</span><span class="op">$</span><span class="va">age</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">clinical_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  patient_id age    bp   weight
1          1  25 120.0 65.00000
2          2  30 119.5 70.00000
3          3  30 115.0 67.66667
4          4  30 125.0 68.00000
5          5  40 118.0 67.66667</code></pre>
</div>
</div>
</section></section><section id="data-transformation" class="level2"><h2 class="anchored" data-anchor-id="data-transformation">Data Transformation</h2>
<section id="introduction-to-outliers" class="level3"><h3 class="anchored" data-anchor-id="introduction-to-outliers">Introduction to Outliers</h3>
<p>Outliers can significantly affect statistical analyses, especially in biological data where sample variation can be high.</p>
<div class="cell">
<div class="sourceCode" id="cb401"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create data with outliers</span></span>
<span><span class="va">expression_levels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2.3</span>, <span class="fl">2.7</span>, <span class="fl">3.1</span>, <span class="fl">2.9</span>, <span class="fl">2.5</span>, <span class="fl">3.0</span>, <span class="fl">15.2</span>, <span class="fl">2.8</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">expression_levels</span>, </span>
<span>        main <span class="op">=</span> <span class="st">"Expression Levels with Outlier"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Expression"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/outliers-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="identifying-outliers" class="level3"><h3 class="anchored" data-anchor-id="identifying-outliers">Identifying Outliers</h3>
<div class="cell">
<div class="sourceCode" id="cb402"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Statistical approach: Values beyond 1.5*IQR</span></span>
<span><span class="va">data_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">expression_levels</span><span class="op">)</span></span>
<span><span class="va">data_summary</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
  2.300   2.650   2.850   4.312   3.025  15.200 </code></pre>
</div>
<div class="sourceCode" id="cb404"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">IQR_value</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/IQR.html">IQR</a></span><span class="op">(</span><span class="va">expression_levels</span><span class="op">)</span></span>
<span><span class="va">upper_bound</span> <span class="op">&lt;-</span> <span class="va">data_summary</span><span class="op">[</span><span class="st">"3rd Qu."</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1.5</span> <span class="op">*</span> <span class="va">IQR_value</span></span>
<span><span class="va">lower_bound</span> <span class="op">&lt;-</span> <span class="va">data_summary</span><span class="op">[</span><span class="st">"1st Qu."</span><span class="op">]</span> <span class="op">-</span> <span class="fl">1.5</span> <span class="op">*</span> <span class="va">IQR_value</span></span>
<span></span>
<span><span class="co"># Find outliers</span></span>
<span><span class="va">outliers</span> <span class="op">&lt;-</span> <span class="va">expression_levels</span><span class="op">[</span><span class="va">expression_levels</span> <span class="op">&gt;</span> <span class="va">upper_bound</span> <span class="op">|</span> </span>
<span>                             <span class="va">expression_levels</span> <span class="op">&lt;</span> <span class="va">lower_bound</span><span class="op">]</span></span>
<span><span class="va">outliers</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] 15.2</code></pre>
</div>
</div>
</section><section id="transforming-vectors" class="level3"><h3 class="anchored" data-anchor-id="transforming-vectors">Transforming Vectors</h3>
<p>Mathematical transformations can normalize data, reduce outlier effects, and make data more suitable for statistical analyses.</p>
<div class="cell">
<div class="sourceCode" id="cb406"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Original data</span></span>
<span><span class="va">gene_exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">15</span>, <span class="fl">42</span>, <span class="fl">87</span>, <span class="fl">115</span>, <span class="fl">320</span>, <span class="fl">560</span>, <span class="fl">1120</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">gene_exp</span>, main <span class="op">=</span> <span class="st">"Original Expression Values"</span>, xlab <span class="op">=</span> <span class="st">"Expression"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/transformations-1.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb407"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Log transformation (common in gene expression analysis)</span></span>
<span><span class="va">log_exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">gene_exp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">log_exp</span>, main <span class="op">=</span> <span class="st">"Log2 Transformed Expression"</span>, xlab <span class="op">=</span> <span class="st">"Log2 Expression"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/transformations-2.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb408"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Square root transformation (less aggressive than log)</span></span>
<span><span class="va">sqrt_exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">gene_exp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">sqrt_exp</span>, main <span class="op">=</span> <span class="st">"Square Root Transformed Expression"</span>, xlab <span class="op">=</span> <span class="st">"Sqrt Expression"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/transformations-3.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb409"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Z-score normalization (standardization)</span></span>
<span><span class="va">z_exp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/scale.html">scale</a></span><span class="op">(</span><span class="va">gene_exp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">z_exp</span>, main <span class="op">=</span> <span class="st">"Z-score Normalized Expression"</span>, xlab <span class="op">=</span> <span class="st">"Z-score"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/transformations-4.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb410"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compare transformations</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">gene_exp</span>, main <span class="op">=</span> <span class="st">"Original"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">log_exp</span>, main <span class="op">=</span> <span class="st">"Log2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">sqrt_exp</span>, main <span class="op">=</span> <span class="st">"Square Root"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">z_exp</span>, main <span class="op">=</span> <span class="st">"Z-score"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/transformations-5.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb411"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>  <span class="co"># Reset plotting layout</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="logical-expressions" class="level3"><h3 class="anchored" data-anchor-id="logical-expressions">Logical Expressions</h3>
<div class="cell">
<div class="sourceCode" id="cb412"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Create gene expression vector</span></span>
<span><span class="va">exp_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5.2</span>, <span class="fl">3.8</span>, <span class="fl">7.1</span>, <span class="fl">2.9</span>, <span class="fl">6.5</span>, <span class="fl">8.0</span>, <span class="fl">4.3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">exp_data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Gene_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Basic comparisons</span></span>
<span><span class="va">exp_data</span> <span class="op">&gt;</span> <span class="fl">5</span>    <span class="co"># Which genes have expression &gt; 5?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Gene_1 Gene_2 Gene_3 Gene_4 Gene_5 Gene_6 Gene_7 
  TRUE  FALSE   TRUE  FALSE   TRUE   TRUE  FALSE </code></pre>
</div>
<div class="sourceCode" id="cb414"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">exp_data</span> <span class="op">&lt;=</span> <span class="fl">4</span>   <span class="co"># Which genes have expression &lt;= 4?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Gene_1 Gene_2 Gene_3 Gene_4 Gene_5 Gene_6 Gene_7 
 FALSE   TRUE  FALSE   TRUE  FALSE  FALSE  FALSE </code></pre>
</div>
<div class="sourceCode" id="cb416"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Store results in logical vector</span></span>
<span><span class="va">high_exp</span> <span class="op">&lt;-</span> <span class="va">exp_data</span> <span class="op">&gt;</span> <span class="fl">6</span></span>
<span><span class="va">high_exp</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Gene_1 Gene_2 Gene_3 Gene_4 Gene_5 Gene_6 Gene_7 
 FALSE  FALSE   TRUE  FALSE   TRUE   TRUE  FALSE </code></pre>
</div>
<div class="sourceCode" id="cb418"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Use logical vectors for subsetting</span></span>
<span><span class="va">exp_data</span><span class="op">[</span><span class="va">high_exp</span><span class="op">]</span>  <span class="co"># Get high expression values</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Gene_3 Gene_5 Gene_6 
   7.1    6.5    8.0 </code></pre>
</div>
</div>
</section><section id="logical-operators" class="level3"><h3 class="anchored" data-anchor-id="logical-operators">Logical Operators</h3>
<div class="cell">
<div class="sourceCode" id="cb420"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Combining conditions with AND (&amp;)</span></span>
<span><span class="va">exp_data</span> <span class="op">&gt;</span> <span class="fl">4</span> <span class="op">&amp;</span> <span class="va">exp_data</span> <span class="op">&lt;</span> <span class="fl">7</span>  <span class="co"># Expression between 4 and 7</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Gene_1 Gene_2 Gene_3 Gene_4 Gene_5 Gene_6 Gene_7 
  TRUE  FALSE  FALSE  FALSE   TRUE  FALSE   TRUE </code></pre>
</div>
<div class="sourceCode" id="cb422"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Combining conditions with OR (|)</span></span>
<span><span class="va">exp_data</span> <span class="op">&lt;</span> <span class="fl">4</span> <span class="op">|</span> <span class="va">exp_data</span> <span class="op">&gt;</span> <span class="fl">7</span>  <span class="co"># Expression less than 4 OR greater than 7</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Gene_1 Gene_2 Gene_3 Gene_4 Gene_5 Gene_6 Gene_7 
 FALSE   TRUE   TRUE   TRUE  FALSE   TRUE  FALSE </code></pre>
</div>
<div class="sourceCode" id="cb424"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Using NOT (!)</span></span>
<span><span class="op">!</span><span class="va">high_exp</span>  <span class="co"># Not high expression</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Gene_1 Gene_2 Gene_3 Gene_4 Gene_5 Gene_6 Gene_7 
  TRUE   TRUE  FALSE   TRUE  FALSE  FALSE   TRUE </code></pre>
</div>
<div class="sourceCode" id="cb426"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Subsetting with combined conditions</span></span>
<span><span class="va">exp_data</span><span class="op">[</span><span class="va">exp_data</span> <span class="op">&gt;</span> <span class="fl">4</span> <span class="op">&amp;</span> <span class="va">exp_data</span> <span class="op">&lt;</span> <span class="fl">7</span><span class="op">]</span>  <span class="co"># Get values between 4 and 7</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Gene_1 Gene_5 Gene_7 
   5.2    6.5    4.3 </code></pre>
</div>
</div>
</section><section id="logical-functions" class="level3"><h3 class="anchored" data-anchor-id="logical-functions">Logical Functions</h3>
<div class="cell">
<div class="sourceCode" id="cb428"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># all() - Are all values TRUE?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="va">exp_data</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>  <span class="co"># Are all expressions positive?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode" id="cb430"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># any() - Is at least one value TRUE?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/any.html">any</a></span><span class="op">(</span><span class="va">exp_data</span> <span class="op">&gt;</span> <span class="fl">7</span><span class="op">)</span>  <span class="co"># Is any expression greater than 7?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode" id="cb432"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># which() - Get indices of TRUE values</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">exp_data</span> <span class="op">&gt;</span> <span class="fl">6</span><span class="op">)</span>  <span class="co"># Which elements have expressions &gt; 6?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>Gene_3 Gene_5 Gene_6 
     3      5      6 </code></pre>
</div>
<div class="sourceCode" id="cb434"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># %in% operator - Test for membership</span></span>
<span><span class="va">test_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Gene_1"</span>, <span class="st">"Gene_5"</span>, <span class="st">"Gene_9"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">exp_data</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">test_genes</span>  <span class="co"># Which names match test_genes?</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>[1]  TRUE FALSE FALSE FALSE  TRUE FALSE FALSE</code></pre>
</div>
</div>
</section><section id="practical-session" class="level3"><h3 class="anchored" data-anchor-id="practical-session">Practical Session</h3>
<p>Check out this repo: <code>https://github.com/genomicsclass/dagdata/</code></p>
<div class="cell">
<div class="sourceCode" id="cb436"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Download small example dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="st">"https://github.com/genomicsclass/dagdata/raw/master/inst/extdata/msleep_ggplot2.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"msleep_data.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load data</span></span>
<span><span class="va">msleep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"msleep_data.csv"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol type="1">
<li>Convert ‚Äòvore‚Äô column to factor and plot its distribution.</li>
<li>Create a matrix of sleep data columns and add row names.</li>
<li>Find and handle any missing values.</li>
<li>Calculate mean sleep time by diet category (vore).</li>
<li>Identify outliers in sleep_total.</li>
</ol></section></section><section id="summary-of-the-lesson" class="level2"><h2 class="anchored" data-anchor-id="summary-of-the-lesson">Summary of the Lesson</h2>
<p>In this lesson, we covered:</p>
<ol type="1">
<li>
<strong>Factor Variables</strong>: Essential for categorical data in biology (genotypes, treatments, etc.)</li>
</ol>
<ul>
<li>Creation, levels, ordering, and visualization</li>
</ul>
<ol start="2" type="1">
<li>
<strong>Subsetting Techniques</strong>: Critical for data extraction and analysis</li>
</ol>
<ul>
<li>Vector and data frame subsetting with various methods</li>
<li>Using row names effectively for biological identifiers</li>
</ul>
<ol start="3" type="1">
<li>
<strong>Matrix Operations</strong>: Fundamental for expression data</li>
</ol>
<ul>
<li>Creation, manipulation, and biological applications</li>
<li>Calculating fold changes and other common operations</li>
</ul>
<ol start="4" type="1">
<li>
<strong>Missing Values</strong>: Practical approaches for real-world biological data</li>
</ol>
<ul>
<li>Identification and appropriate replacement methods</li>
</ul>
<ol start="5" type="1">
<li>
<strong>Data Transformation</strong>: Making data suitable for statistical analysis</li>
</ol>
<ul>
<li>Log, square root, and z-score transformations</li>
<li>Outlier identification and handling</li>
</ul>
<ol start="6" type="1">
<li>
<strong>Logical Operations</strong>: For data filtering and decision making</li>
</ol>
<ul>
<li>
<p>Conditions, combinations, and applications</p>
<p>These skills form the foundation for the more advanced visualization techniques we‚Äôll cover in future lessons.</p>
</li>
</ul>
<ol start="7" type="1">
<li>
<strong>List</strong>: Fundamental for many biological data and packages‚Äô output.</li>
</ol>
<ul>
<li>Properties, accessing, and applications</li>
</ul>
<ol start="8" type="1">
<li>We will know more about conditionals, R packages to handle data and visualization in a better and efficient way.</li>
</ol></section><section id="homework" class="level2"><h2 class="anchored" data-anchor-id="homework">Homework</h2>
<ol type="1">
<li>
<strong>Matrix Operations</strong>:</li>
</ol>
<ul>
<li>Create a gene expression matrix with 8 genes and 4 conditions</li>
<li>Calculate the mean expression for each gene</li>
<li>Calculate fold change between condition 4 and condition 1</li>
<li>Create a heatmap of your matrix</li>
</ul>
<ol start="2" type="1">
<li>
<strong>Factor Analysis</strong>:</li>
</ol>
<ul>
<li>Using the <code>iris</code> dataset, convert Species to an ordered factor</li>
<li>Create boxplots showing Sepal.Length by Species</li>
<li>Calculate mean petal length for each species level</li>
</ul>
<ol start="3" type="1">
<li>
<strong>Data Cleaning Challenge</strong>:</li>
</ol>
<ul>
<li>In the downloaded <code>msleep_data.csv</code>:</li>
<li>Identify all columns with missing values</li>
<li>Replace missing values appropriately</li>
<li>Create a new categorical variable ‚Äúsleep_duration‚Äù with levels ‚ÄúShort‚Äù, ‚ÄúMedium‚Äù, ‚ÄúLong‚Äù</li>
</ul>
<ol start="4" type="1">
<li>
<strong>List challenge</strong>:</li>
</ol>
<ul>
<li>Make your own lists</li>
<li>Replicate all the tasks we did</li>
<li>You may ask AI to give you beginner-level questions but don‚Äôt ask to solve the questions programmatically. Tell AI not to provide answers.</li>
</ul>
<ol start="5" type="1">
<li>
<strong>Complete Documentation</strong>:</li>
</ol>
<ul>
<li>Write all code in R Markdown</li>
<li>Include comments explaining your approach</li>
<li>Push to GitHub</li>
</ul>
<section id="due-date-friday-10pm-bd-time" class="level4"><h4 class="anchored" data-anchor-id="due-date-friday-10pm-bd-time">Due date: Friday 10pm BD Time</h4>
<div class="cell">
<div class="sourceCode" id="cb437"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span> <span class="co"># For reproducibility</span></span>
<span><span class="va">gene_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">30</span>, mean <span class="op">=</span> <span class="fl">8</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"brain"</span>, <span class="st">"liver"</span>, <span class="st">"kidney"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">tissue_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">tissue</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"liver"</span>, <span class="st">"brain"</span>, <span class="st">"kidney"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tissue_factor</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code> [1] brain  brain  brain  brain  brain  brain  brain  brain  brain  brain 
[11] liver  liver  liver  liver  liver  liver  liver  liver  liver  liver 
[21] kidney kidney kidney kidney kidney kidney kidney kidney kidney kidney
Levels: liver brain kidney</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb439"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expr_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  tissue <span class="op">=</span> <span class="va">tissue_factor</span>,</span>
<span>  expression <span class="op">=</span> <span class="va">gene_expr</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#expr_data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">expr_data</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>  tissue expression
1  brain  10.741917
2  brain   6.870604
3  brain   8.726257
4  brain   9.265725
5  brain   8.808537
6  brain   7.787751</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb441"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">expression</span> <span class="op">~</span> <span class="va">tissue</span>, </span>
<span>        data <span class="op">=</span> <span class="va">expr_data</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Gene Expression by Tissue"</span>,</span>
<span>        xlab <span class="op">=</span> <span class="st">"Tissue"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Expression"</span>,</span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>, <span class="st">"lightgreen"</span>, <span class="st">"lightpink"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="firststeps_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb442"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">sd_expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span><span class="op">(</span><span class="va">expr_data</span><span class="op">$</span><span class="va">expression</span>, <span class="va">expr_data</span><span class="op">$</span><span class="va">tissue</span>, <span class="va">sd</span><span class="op">)</span></span>
<span><span class="va">sd_expression</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>   liver    brain   kidney 
3.261169 1.670898 2.312116 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb444"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">max_sd_tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">sd_expression</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span><span class="op">(</span><span class="va">sd_expression</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">max_sd_value</span> <span class="op">&lt;-</span> <span class="va">sd_expression</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.max</a></span><span class="op">(</span><span class="va">sd_expression</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"The tissue with the highest sd is:"</span>, <span class="va">max_sd_tissue</span>,<span class="st">", with SD ="</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">max_sd_value</span>, <span class="fl">2</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre class="custom-output"><code>The tissue with the highest sd is: liver , with SD = 3.26 </code></pre>
</div>
</div>


<!-- -->

</section></section></section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents"><h2 class="anchored quarto-appendix-heading">Citation</h2><div><div class="quarto-appendix-secondary-label">BibTeX citation:</div><pre class="sourceCode code-with-copy quarto-appendix-bibtex"><code class="sourceCode bibtex">@online{rasheduzzaman2025,
  author = {Md Rasheduzzaman},
  title = {Basic {R}},
  date = {2025-08-03},
  langid = {en},
  abstract = {Data types, variables, vectors, data frame, functions}
}
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre><div class="quarto-appendix-secondary-label">For attribution, please cite this work as:</div><div id="ref-rasheduzzaman2025" class="csl-entry quarto-appendix-citeas" role="listitem">
Md Rasheduzzaman. 2025. <span>‚ÄúBasic R.‚Äù</span> August 3, 2025.
</div></div></section></div></main><!-- /main --><hr>
<div style="margin-top: 2em;">
  <p style="font-weight: 500; font-size: 1.1em;">
    üí¨ Have thoughts or questions? Join the discussion below using your GitHub account!
  </p>
  <p style="font-size: 0.95em; color: #666;">
    You can <strong>edit or delete</strong> your own comments. Reactions like üëç ‚ù§Ô∏è üöÄ are also supported.
  </p>
</div>

<div id="giscus_thread" style="margin-top: 1.5em;"></div>

<script src="https://giscus.app/client.js" data-repo="MdRasheduz-zaman/PythRaSh" data-repo-id="R_kgDOOZN9Xw" data-category-id="DIC_kwDOOZN9X84CpE6N" data-mapping="pathname" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="preferred_color_scheme" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb446" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb446-1"><a href="#cb446-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb446-2"><a href="#cb446-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Basic R"</span></span>
<span id="cb446-3"><a href="#cb446-3" aria-hidden="true" tabindex="-1"></a><span class="an">abstract:</span><span class="co"> "Data types, variables, vectors, data frame, functions"</span></span>
<span id="cb446-4"><a href="#cb446-4" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb446-5"><a href="#cb446-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-8"><a href="#cb446-8" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-9"><a href="#cb446-9" aria-hidden="true" tabindex="-1"></a><span class="co">#| include: false</span></span>
<span id="cb446-10"><a href="#cb446-10" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(here<span class="sc">::</span><span class="fu">here</span>(<span class="st">"src/helpersrc.R"</span>))</span>
<span id="cb446-11"><a href="#cb446-11" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-12"><a href="#cb446-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-13"><a href="#cb446-13" aria-hidden="true" tabindex="-1"></a><span class="fu"># L2: Data Representation</span></span>
<span id="cb446-14"><a href="#cb446-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-15"><a href="#cb446-15" aria-hidden="true" tabindex="-1"></a><span class="fu">## Using R as a Calculator</span></span>
<span id="cb446-16"><a href="#cb446-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-17"><a href="#cb446-17" aria-hidden="true" tabindex="-1"></a>Let's do some basic calculation.</span>
<span id="cb446-18"><a href="#cb446-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-21"><a href="#cb446-21" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-22"><a href="#cb446-22" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span><span class="sc">+</span><span class="dv">3</span></span>
<span id="cb446-23"><a href="#cb446-23" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span><span class="sc">+</span><span class="dv">2</span></span>
<span id="cb446-24"><a href="#cb446-24" aria-hidden="true" tabindex="-1"></a><span class="dv">3-2</span></span>
<span id="cb446-25"><a href="#cb446-25" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span><span class="sc">*</span><span class="dv">2</span></span>
<span id="cb446-26"><a href="#cb446-26" aria-hidden="true" tabindex="-1"></a><span class="dv">3</span><span class="sc">/</span><span class="dv">2</span> <span class="co">#normal division</span></span>
<span id="cb446-27"><a href="#cb446-27" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span> <span class="sc">%/%</span> <span class="dv">2</span> <span class="co">#integer division, only the quotient</span></span>
<span id="cb446-28"><a href="#cb446-28" aria-hidden="true" tabindex="-1"></a><span class="dv">5</span> <span class="sc">%%</span> <span class="dv">3</span> <span class="co">#modulus division, the remainder</span></span>
<span id="cb446-29"><a href="#cb446-29" aria-hidden="true" tabindex="-1"></a>(<span class="dv">10-5</span>)<span class="sc">*</span>(<span class="dv">2</span><span class="sc">+</span><span class="dv">4</span>) <span class="co">#use of parentheses</span></span>
<span id="cb446-30"><a href="#cb446-30" aria-hidden="true" tabindex="-1"></a><span class="dv">10-5</span><span class="sc">*</span><span class="dv">2</span><span class="sc">+</span><span class="dv">4</span> <span class="co">#Noticed BODMAS?</span></span>
<span id="cb446-31"><a href="#cb446-31" aria-hidden="true" tabindex="-1"></a>(<span class="dv">10-5</span>)<span class="sc">*</span>(<span class="dv">2</span><span class="sc">+</span><span class="dv">4</span>) <span class="co">#Noticed BODMAS</span></span>
<span id="cb446-32"><a href="#cb446-32" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span><span class="sc">/</span>(<span class="dv">1</span><span class="sc">+</span><span class="dv">3</span>); <span class="dv">7</span><span class="sc">/</span><span class="dv">1</span><span class="sc">+</span><span class="dv">3</span> <span class="co">#multi-line codes, separated with semi-colon</span></span>
<span id="cb446-33"><a href="#cb446-33" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span><span class="sc">+</span><span class="dv">2</span>; <span class="fu">log</span>(<span class="dv">1</span>); <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span> <span class="co">#more multi-line codes</span></span>
<span id="cb446-34"><a href="#cb446-34" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-35"><a href="#cb446-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-36"><a href="#cb446-36" aria-hidden="true" tabindex="-1"></a><span class="fu">## Variables</span></span>
<span id="cb446-37"><a href="#cb446-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-38"><a href="#cb446-38" aria-hidden="true" tabindex="-1"></a>Variables are variable. We have freedom to name them as we wish. But make any variable name meaningful and identifiable.</span>
<span id="cb446-39"><a href="#cb446-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-42"><a href="#cb446-42" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-43"><a href="#cb446-43" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="dv">5</span> <span class="co">#assign value 5 to a </span></span>
<span id="cb446-44"><a href="#cb446-44" aria-hidden="true" tabindex="-1"></a>b <span class="ot">=</span> <span class="dv">10</span></span>
<span id="cb446-45"><a href="#cb446-45" aria-hidden="true" tabindex="-1"></a>a</span>
<span id="cb446-46"><a href="#cb446-46" aria-hidden="true" tabindex="-1"></a>b</span>
<span id="cb446-47"><a href="#cb446-47" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> a <span class="sc">+</span> <span class="dv">10</span></span>
<span id="cb446-48"><a href="#cb446-48" aria-hidden="true" tabindex="-1"></a>b <span class="ot">=</span> b <span class="sc">+</span> <span class="dv">15</span></span>
<span id="cb446-49"><a href="#cb446-49" aria-hidden="true" tabindex="-1"></a>a</span>
<span id="cb446-50"><a href="#cb446-50" aria-hidden="true" tabindex="-1"></a>a<span class="sc">^</span><span class="dv">2</span> <span class="co">#a squared</span></span>
<span id="cb446-51"><a href="#cb446-51" aria-hidden="true" tabindex="-1"></a>a<span class="sc">**</span><span class="dv">2</span> <span class="co">#a squared again, in a different way.</span></span>
<span id="cb446-52"><a href="#cb446-52" aria-hidden="true" tabindex="-1"></a>a<span class="sc">^</span><span class="dv">3</span> <span class="co">#a qubed</span></span>
<span id="cb446-53"><a href="#cb446-53" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-54"><a href="#cb446-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-55"><a href="#cb446-55" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-56"><a href="#cb446-56" aria-hidden="true" tabindex="-1"></a><span class="in">`&lt;-`</span> and <span class="in">`=`</span> are used to assign values. It is not mathematical equality. <span class="in">`b &lt;- b + 15`</span> might make better sense than <span class="in">`b = b + 15`</span>.</span>
<span id="cb446-57"><a href="#cb446-57" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-58"><a href="#cb446-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-59"><a href="#cb446-59" aria-hidden="true" tabindex="-1"></a><span class="fu">### Integer and Modulus division again</span></span>
<span id="cb446-60"><a href="#cb446-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-61"><a href="#cb446-61" aria-hidden="true" tabindex="-1"></a>Do some more practice.</span>
<span id="cb446-62"><a href="#cb446-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-65"><a href="#cb446-65" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-66"><a href="#cb446-66" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span><span class="sc">/</span><span class="dv">3</span></span>
<span id="cb446-67"><a href="#cb446-67" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span><span class="sc">%/%</span><span class="dv">3</span></span>
<span id="cb446-68"><a href="#cb446-68" aria-hidden="true" tabindex="-1"></a><span class="dv">7</span><span class="sc">%%</span><span class="dv">3</span></span>
<span id="cb446-69"><a href="#cb446-69" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-70"><a href="#cb446-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-71"><a href="#cb446-71" aria-hidden="true" tabindex="-1"></a><span class="fu">## Rounding</span></span>
<span id="cb446-72"><a href="#cb446-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-73"><a href="#cb446-73" aria-hidden="true" tabindex="-1"></a>Some important functions we apply on numerical values</span>
<span id="cb446-74"><a href="#cb446-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-77"><a href="#cb446-77" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-78"><a href="#cb446-78" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">9</span><span class="sc">/</span><span class="dv">4</span></span>
<span id="cb446-79"><a href="#cb446-79" aria-hidden="true" tabindex="-1"></a><span class="fu">floor</span>(x)</span>
<span id="cb446-80"><a href="#cb446-80" aria-hidden="true" tabindex="-1"></a><span class="fu">ceiling</span>(x)</span>
<span id="cb446-81"><a href="#cb446-81" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(x)</span>
<span id="cb446-82"><a href="#cb446-82" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(x, <span class="dv">2</span>) <span class="co">#round till 2 decimal points</span></span>
<span id="cb446-83"><a href="#cb446-83" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-84"><a href="#cb446-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-85"><a href="#cb446-85" aria-hidden="true" tabindex="-1"></a><span class="fu">## Logical Operations</span></span>
<span id="cb446-86"><a href="#cb446-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-87"><a href="#cb446-87" aria-hidden="true" tabindex="-1"></a>Get to know TRUE/FALSE in R.</span>
<span id="cb446-88"><a href="#cb446-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-91"><a href="#cb446-91" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-92"><a href="#cb446-92" aria-hidden="true" tabindex="-1"></a>a <span class="ot">=</span> <span class="dv">5</span></span>
<span id="cb446-93"><a href="#cb446-93" aria-hidden="true" tabindex="-1"></a>b <span class="ot">=</span> <span class="dv">7</span></span>
<span id="cb446-94"><a href="#cb446-94" aria-hidden="true" tabindex="-1"></a>c <span class="ot">=</span> <span class="dv">10</span></span>
<span id="cb446-95"><a href="#cb446-95" aria-hidden="true" tabindex="-1"></a>d <span class="ot">=</span> <span class="dv">3</span></span>
<span id="cb446-96"><a href="#cb446-96" aria-hidden="true" tabindex="-1"></a>a <span class="sc">==</span> b <span class="co">#is a equal to b? Ans: No/FALSE</span></span>
<span id="cb446-97"><a href="#cb446-97" aria-hidden="true" tabindex="-1"></a>a <span class="sc">!=</span> b <span class="co">#is a not equal to b? Ans: Yes/TRUE</span></span>
<span id="cb446-98"><a href="#cb446-98" aria-hidden="true" tabindex="-1"></a>a <span class="sc">&gt;</span> b <span class="co">#is a greater than b? Ans: FALSE</span></span>
<span id="cb446-99"><a href="#cb446-99" aria-hidden="true" tabindex="-1"></a>a <span class="sc">&lt;</span> b <span class="co">#is a less than b? Ans: TRUE</span></span>
<span id="cb446-100"><a href="#cb446-100" aria-hidden="true" tabindex="-1"></a>a <span class="sc">&gt;=</span> b <span class="co">#is a greater than or equal to b? Ans: FALSE</span></span>
<span id="cb446-101"><a href="#cb446-101" aria-hidden="true" tabindex="-1"></a>a <span class="sc">&lt;=</span> b <span class="co">#is a less than or equal to b? Ans: TRUE</span></span>
<span id="cb446-102"><a href="#cb446-102" aria-hidden="true" tabindex="-1"></a>a <span class="sc">&lt;</span> b <span class="sc">|</span> d <span class="sc">&gt;</span> b <span class="co">#is a less than b OR d greater than b?</span></span>
<span id="cb446-103"><a href="#cb446-103" aria-hidden="true" tabindex="-1"></a><span class="co">#It's answer will be TRUE OR FALSE --&gt; So, TRUE</span></span>
<span id="cb446-104"><a href="#cb446-104" aria-hidden="true" tabindex="-1"></a>a <span class="sc">&lt;</span> b <span class="sc">&amp;</span> c <span class="sc">&gt;</span> d <span class="co">#is a less than b AND a greater than b? It's answer will be TRUE AND TRUE --&gt; So, TRUE</span></span>
<span id="cb446-105"><a href="#cb446-105" aria-hidden="true" tabindex="-1"></a>a <span class="sc">&lt;</span> b <span class="sc">&amp;</span> d <span class="sc">&gt;</span> c <span class="co">#is a less than b AND a greater than b? It's answer will be TRUE AND FALSE --&gt; So, FALSE</span></span>
<span id="cb446-106"><a href="#cb446-106" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-107"><a href="#cb446-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-108"><a href="#cb446-108" aria-hidden="true" tabindex="-1"></a><span class="fu">## Help and Documentation</span></span>
<span id="cb446-109"><a href="#cb446-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-110"><a href="#cb446-110" aria-hidden="true" tabindex="-1"></a>But how to know more about a function? The package/library developer have written helpful documentation for us.</span>
<span id="cb446-111"><a href="#cb446-111" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-114"><a href="#cb446-114" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-115"><a href="#cb446-115" aria-hidden="true" tabindex="-1"></a>?log</span>
<span id="cb446-116"><a href="#cb446-116" aria-hidden="true" tabindex="-1"></a><span class="fu">example</span>(log)</span>
<span id="cb446-117"><a href="#cb446-117" aria-hidden="true" tabindex="-1"></a>?<span class="fu">log</span>()</span>
<span id="cb446-118"><a href="#cb446-118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-119"><a href="#cb446-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-120"><a href="#cb446-120" aria-hidden="true" tabindex="-1"></a><span class="fu">## Working with Vectors</span></span>
<span id="cb446-121"><a href="#cb446-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-122"><a href="#cb446-122" aria-hidden="true" tabindex="-1"></a>What is a vector? See the example and think.</span>
<span id="cb446-123"><a href="#cb446-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-126"><a href="#cb446-126" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-127"><a href="#cb446-127" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">5</span>) <span class="co">#c means concatenate</span></span>
<span id="cb446-128"><a href="#cb446-128" aria-hidden="true" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span> <span class="co">#consecutively, from 1 through 5. A short-hand notation using :</span></span>
<span id="cb446-129"><a href="#cb446-129" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">6</span>, <span class="dv">9</span>, <span class="dv">12</span>, <span class="dv">15</span>, <span class="dv">20</span>)</span>
<span id="cb446-130"><a href="#cb446-130" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(x)</span>
<span id="cb446-131"><a href="#cb446-131" aria-hidden="true" tabindex="-1"></a><span class="fu">mode</span>(x)</span>
<span id="cb446-132"><a href="#cb446-132" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(x)</span>
<span id="cb446-133"><a href="#cb446-133" aria-hidden="true" tabindex="-1"></a>x[<span class="dv">1</span>] <span class="co">#first entry in vector y</span></span>
<span id="cb446-134"><a href="#cb446-134" aria-hidden="true" tabindex="-1"></a>x[<span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>] <span class="co">#2nd to 5th entries in vector y</span></span>
<span id="cb446-135"><a href="#cb446-135" aria-hidden="true" tabindex="-1"></a>DNA <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>) <span class="co">#character vector. Notice the quotation marks.</span></span>
<span id="cb446-136"><a href="#cb446-136" aria-hidden="true" tabindex="-1"></a>dec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">10.0</span>, <span class="fl">20.5</span>, <span class="dv">30</span>, <span class="dv">60</span>, <span class="fl">80.9</span>, <span class="dv">90</span>, <span class="fl">100.7</span>, <span class="dv">50</span>, <span class="dv">40</span>, <span class="dv">45</span>, <span class="dv">48</span>, <span class="dv">56</span>, <span class="dv">55</span>) <span class="co">#vector of floats. All numbers became floats, it's called coercion</span></span>
<span id="cb446-137"><a href="#cb446-137" aria-hidden="true" tabindex="-1"></a>dec[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">7</span><span class="sc">:</span><span class="fu">length</span>(dec))] <span class="co">#1st to 3rd and then 7th till the end of vector `dec`. Output as a vector.</span></span>
<span id="cb446-138"><a href="#cb446-138" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-139"><a href="#cb446-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-140"><a href="#cb446-140" aria-hidden="true" tabindex="-1"></a><span class="fu">### Vector Operations</span></span>
<span id="cb446-141"><a href="#cb446-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-142"><a href="#cb446-142" aria-hidden="true" tabindex="-1"></a>Notice the element-wise or index-wise mathematical operations (<span class="in">`+`</span>, <span class="in">`/`</span>, <span class="in">`log2()`</span>, <span class="in">`round()`</span>, etc.). Noticed?</span>
<span id="cb446-143"><a href="#cb446-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-146"><a href="#cb446-146" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-147"><a href="#cb446-147" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span></span>
<span id="cb446-148"><a href="#cb446-148" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">11</span></span>
<span id="cb446-149"><a href="#cb446-149" aria-hidden="true" tabindex="-1"></a><span class="co">#x and y are of same length</span></span>
<span id="cb446-150"><a href="#cb446-150" aria-hidden="true" tabindex="-1"></a>x <span class="sc">+</span> y</span>
<span id="cb446-151"><a href="#cb446-151" aria-hidden="true" tabindex="-1"></a>y <span class="sc">/</span> x</span>
<span id="cb446-152"><a href="#cb446-152" aria-hidden="true" tabindex="-1"></a><span class="fu">log2</span>(x)</span>
<span id="cb446-153"><a href="#cb446-153" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">log2</span>(x), <span class="dv">1</span>) <span class="co">#log2 of all the values of `x`, 1 digit after decimal to round.</span></span>
<span id="cb446-154"><a href="#cb446-154" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">log2</span>(x), <span class="dv">3</span>) <span class="co">#same logic</span></span>
<span id="cb446-155"><a href="#cb446-155" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-156"><a href="#cb446-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-157"><a href="#cb446-157" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-158"><a href="#cb446-158" aria-hidden="true" tabindex="-1"></a>Nested functions work inside out. Think again about <span class="in">`round(log2(x), 1)`</span> and you will see it. At first, it is making <span class="in">`log2`</span> of vector <span class="in">`x`</span> and then it is rounding the log2 values to one digit after decimal. Got it?</span>
<span id="cb446-159"><a href="#cb446-159" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-160"><a href="#cb446-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-161"><a href="#cb446-161" aria-hidden="true" tabindex="-1"></a><span class="fu">## Data Frame</span></span>
<span id="cb446-162"><a href="#cb446-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-163"><a href="#cb446-163" aria-hidden="true" tabindex="-1"></a>Now, it's time to use vectors to make data sets.....</span>
<span id="cb446-164"><a href="#cb446-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-167"><a href="#cb446-167" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-168"><a href="#cb446-168" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Mina"</span>, <span class="st">"Raju"</span>, <span class="st">"Mithu"</span>, <span class="st">"Lali"</span>)</span>
<span id="cb446-169"><a href="#cb446-169" aria-hidden="true" tabindex="-1"></a>gender <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Female"</span>, <span class="st">"Male"</span>, <span class="st">"Female"</span>, <span class="st">"Female"</span>)</span>
<span id="cb446-170"><a href="#cb446-170" aria-hidden="true" tabindex="-1"></a>age <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15</span>, <span class="dv">12</span>, <span class="dv">2</span>, <span class="dv">3</span>)</span>
<span id="cb446-171"><a href="#cb446-171" aria-hidden="true" tabindex="-1"></a>is_human <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>)</span>
<span id="cb446-172"><a href="#cb446-172" aria-hidden="true" tabindex="-1"></a>cartoon <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(names, gender, age, is_human)</span>
<span id="cb446-173"><a href="#cb446-173" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(cartoon, <span class="st">"cartoon.csv"</span>, <span class="at">sep =</span> <span class="st">","</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-174"><a href="#cb446-174" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"cartoon.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">","</span>)</span>
<span id="cb446-175"><a href="#cb446-175" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(df) <span class="co">#`dim` means dimension. so, rows * columns</span></span>
<span id="cb446-176"><a href="#cb446-176" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(df) <span class="co">#structure of `df`</span></span>
<span id="cb446-177"><a href="#cb446-177" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-178"><a href="#cb446-178" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-179"><a href="#cb446-179" aria-hidden="true" tabindex="-1"></a>We made the vectors first, and the used them to make the <span class="in">`cartton`</span> data frame or table. We learned how to export the data frame using write.table function. Also, we learned to import or read back the table using <span class="in">`read.table`</span> function. What are the <span class="in">`sep`</span>, <span class="in">`col.names`</span>, <span class="in">`header`</span> arguments there? Why do we need them? Think. Try thinking of different properties of a data set.</span>
<span id="cb446-180"><a href="#cb446-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-181"><a href="#cb446-181" aria-hidden="true" tabindex="-1"></a><span class="fu">### Gene Expression Table</span></span>
<span id="cb446-182"><a href="#cb446-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-185"><a href="#cb446-185" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-186"><a href="#cb446-186" aria-hidden="true" tabindex="-1"></a>gene_expr <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb446-187"><a href="#cb446-187" aria-hidden="true" tabindex="-1"></a>  <span class="at">genes =</span> <span class="fu">c</span>(<span class="st">"TP53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"MYC"</span>, <span class="st">"EGFR"</span>, <span class="st">"GAPDH"</span>, <span class="st">"CDC2"</span>),</span>
<span id="cb446-188"><a href="#cb446-188" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample1 =</span> <span class="fu">c</span>(<span class="fl">8.2</span>, <span class="fl">6.1</span>, <span class="fl">9.5</span>, <span class="fl">7.0</span>, <span class="fl">10.0</span>, <span class="dv">12</span>),</span>
<span id="cb446-189"><a href="#cb446-189" aria-hidden="true" tabindex="-1"></a>  <span class="at">Sample2 =</span> <span class="fu">c</span>(<span class="fl">5.9</span>, <span class="fl">3.9</span>, <span class="fl">7.2</span>, <span class="fl">4.8</span>, <span class="fl">7.9</span>, <span class="dv">9</span>),</span>
<span id="cb446-190"><a href="#cb446-190" aria-hidden="true" tabindex="-1"></a>  <span class="at">Sample3 =</span> <span class="fu">c</span>(<span class="fl">8.25</span>, <span class="fl">6.15</span>, <span class="fl">9.6</span>, <span class="fl">7.1</span>, <span class="fl">10.1</span>, <span class="fl">11.9</span>),</span>
<span id="cb446-191"><a href="#cb446-191" aria-hidden="true" tabindex="-1"></a>  <span class="at">pathways =</span> <span class="fu">c</span>(<span class="st">"Apoptosis"</span>, <span class="st">"DNA Repair"</span>, <span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>, <span class="st">"Housekeeping"</span>, <span class="st">"Cell Division"</span>)</span>
<span id="cb446-192"><a href="#cb446-192" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-193"><a href="#cb446-193" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(gene_expr, <span class="st">"gene_expr.csv"</span>, <span class="at">sep =</span> <span class="st">","</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-194"><a href="#cb446-194" aria-hidden="true" tabindex="-1"></a>gene_set <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"gene_expr.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">","</span>)</span>
<span id="cb446-195"><a href="#cb446-195" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-196"><a href="#cb446-196" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-197"><a href="#cb446-197" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-198"><a href="#cb446-198" aria-hidden="true" tabindex="-1"></a>Here, we directly used the vectors as different columns while making the data frame. Did you notice that? Also, the syntax is different here. We can't assign the vectors with the assignment operator (means we can't use <span class="in">`&lt;-`</span> sign. We have to use the <span class="in">`=`</span> sign). Try using the <span class="in">`&lt;-`</span> sign. Did you notice the column names?</span>
<span id="cb446-199"><a href="#cb446-199" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-200"><a href="#cb446-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-201"><a href="#cb446-201" aria-hidden="true" tabindex="-1"></a><span class="fu">## Homeworks</span></span>
<span id="cb446-202"><a href="#cb446-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-203"><a href="#cb446-203" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Compute the difference between this year (2025) and the year you started at the university and divide this by the difference between this year and the year you were born. Multiply this with 100 to get the percentage of your life you have spent at the university.</span>
<span id="cb446-204"><a href="#cb446-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-205"><a href="#cb446-205" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Make different kinds of variables and vectors with the data types we learned together.</span>
<span id="cb446-206"><a href="#cb446-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-207"><a href="#cb446-207" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>What are the properties of a data frame?</span>
<span id="cb446-208"><a href="#cb446-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-209"><a href="#cb446-209" aria-hidden="true" tabindex="-1"></a>Hint: Open an excel/csv/txt file you have and try to "generalize".</span>
<span id="cb446-210"><a href="#cb446-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-211"><a href="#cb446-211" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Can you make logical questions on the 2 small data sets we used? Try. It will help you understanding the logical operations we tried on variables. Now we are going to apply them on vectors (columns) on the data sets. For example, in the <span class="in">`cartoon`</span> data set, we can ask/try to subset the data set filtering for females only, or for both females and age greater than 2 years.</span>
<span id="cb446-212"><a href="#cb446-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-213"><a href="#cb446-213" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>If you are writing or practicing coding in R, write comment for each line on what it is doing. It will help to chunk it better into your brain.</span>
<span id="cb446-214"><a href="#cb446-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-215"><a href="#cb446-215" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>Push the script and/or your answers to the questions (with your solutions) to one of your GitHub repo (and send me the repo link).</span>
<span id="cb446-216"><a href="#cb446-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-217"><a href="#cb446-217" aria-hidden="true" tabindex="-1"></a><span class="fu">### Deadline</span></span>
<span id="cb446-218"><a href="#cb446-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-219"><a href="#cb446-219" aria-hidden="true" tabindex="-1"></a>**Friday, 10pm BD Time.**</span>
<span id="cb446-220"><a href="#cb446-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-221"><a href="#cb446-221" aria-hidden="true" tabindex="-1"></a><span class="fu"># L3: Data Transformation</span></span>
<span id="cb446-222"><a href="#cb446-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-223"><a href="#cb446-223" aria-hidden="true" tabindex="-1"></a>Firstly, how did you solve the problems?</span>
<span id="cb446-224"><a href="#cb446-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-225"><a href="#cb446-225" aria-hidden="true" tabindex="-1"></a>Give me your personal Mindmap. Please, send it in the chat!</span>
<span id="cb446-226"><a href="#cb446-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-227"><a href="#cb446-227" aria-hidden="true" tabindex="-1"></a><span class="fu">## Getting Started</span></span>
<span id="cb446-228"><a href="#cb446-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-229"><a href="#cb446-229" aria-hidden="true" tabindex="-1"></a><span class="fu">### Installation of R Markdown</span></span>
<span id="cb446-230"><a href="#cb446-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-231"><a href="#cb446-231" aria-hidden="true" tabindex="-1"></a>We will use <span class="in">`rmarkdown`</span> to have the flexibility of writing codes like the one you are reading now. If you haven't installed the <span class="in">`rmarkdown`</span> package yet, you can do so with:</span>
<span id="cb446-232"><a href="#cb446-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-233"><a href="#cb446-233" aria-hidden="true" tabindex="-1"></a><span class="in">```{r install-packages, eval=FALSE}</span></span>
<span id="cb446-234"><a href="#cb446-234" aria-hidden="true" tabindex="-1"></a><span class="co"># Install rmarkdown package</span></span>
<span id="cb446-235"><a href="#cb446-235" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("rmarkdown")</span></span>
<span id="cb446-236"><a href="#cb446-236" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rmarkdown)</span>
<span id="cb446-237"><a href="#cb446-237" aria-hidden="true" tabindex="-1"></a><span class="co"># Other useful packages we might use</span></span>
<span id="cb446-238"><a href="#cb446-238" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("dplyr")    # Data manipulation</span></span>
<span id="cb446-239"><a href="#cb446-239" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb446-240"><a href="#cb446-240" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("readr")    # Reading CSV files</span></span>
<span id="cb446-241"><a href="#cb446-241" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb446-242"><a href="#cb446-242" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-243"><a href="#cb446-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-244"><a href="#cb446-244" aria-hidden="true" tabindex="-1"></a>Remove the hash sign before the <span class="in">`install.packages("rmarkdown")`</span>, <span class="in">`install.packages("dplyr")`</span>, <span class="in">`install.packages("readr")`</span> if the library loading fails. That means the package is not there to be loaded. We need to download/install first.</span>
<span id="cb446-245"><a href="#cb446-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-246"><a href="#cb446-246" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-247"><a href="#cb446-247" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Do you remember this book by Hadley Wickham?</span><span class="co">](https://r4ds.had.co.nz)</span>. Try to follow it to get the hold on the basic R syntax and lexicon.</span>
<span id="cb446-248"><a href="#cb446-248" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-249"><a href="#cb446-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-250"><a href="#cb446-250" aria-hidden="true" tabindex="-1"></a><span class="fu">### Basic Setup for Today's Session</span></span>
<span id="cb446-251"><a href="#cb446-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-252"><a href="#cb446-252" aria-hidden="true" tabindex="-1"></a><span class="in">```{r initial-setup, eval=FALSE}</span></span>
<span id="cb446-253"><a href="#cb446-253" aria-hidden="true" tabindex="-1"></a><span class="co"># Clear environment</span></span>
<span id="cb446-254"><a href="#cb446-254" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(<span class="at">list =</span> <span class="fu">ls</span>())</span>
<span id="cb446-255"><a href="#cb446-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-256"><a href="#cb446-256" aria-hidden="true" tabindex="-1"></a><span class="co"># Check working directory</span></span>
<span id="cb446-257"><a href="#cb446-257" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span>
<span id="cb446-258"><a href="#cb446-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-259"><a href="#cb446-259" aria-hidden="true" tabindex="-1"></a><span class="co"># Set working directory if needed</span></span>
<span id="cb446-260"><a href="#cb446-260" aria-hidden="true" tabindex="-1"></a><span class="co"># setwd("path/to/your/directory")  # Uncomment and modify as needed</span></span>
<span id="cb446-261"><a href="#cb446-261" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-262"><a href="#cb446-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-263"><a href="#cb446-263" aria-hidden="true" tabindex="-1"></a><span class="fu">### Building on Last HW:</span></span>
<span id="cb446-264"><a href="#cb446-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-267"><a href="#cb446-267" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-268"><a href="#cb446-268" aria-hidden="true" tabindex="-1"></a>cartoon <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb446-269"><a href="#cb446-269" aria-hidden="true" tabindex="-1"></a>  <span class="at">names =</span> <span class="fu">c</span>(<span class="st">"Mina"</span>, <span class="st">"Raju"</span>, <span class="st">"Mithu"</span>, <span class="st">"Lali"</span>),</span>
<span id="cb446-270"><a href="#cb446-270" aria-hidden="true" tabindex="-1"></a>  <span class="at">gender =</span> <span class="fu">c</span>(<span class="st">"Female"</span>, <span class="st">"Male"</span>, <span class="st">"Female"</span>, <span class="st">"Female"</span>),</span>
<span id="cb446-271"><a href="#cb446-271" aria-hidden="true" tabindex="-1"></a>  <span class="at">age =</span> <span class="fu">c</span>(<span class="dv">15</span>, <span class="dv">12</span>, <span class="dv">2</span>, <span class="dv">3</span>),</span>
<span id="cb446-272"><a href="#cb446-272" aria-hidden="true" tabindex="-1"></a>  <span class="at">is_human =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>)</span>
<span id="cb446-273"><a href="#cb446-273" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-274"><a href="#cb446-274" aria-hidden="true" tabindex="-1"></a>cartoon</span>
<span id="cb446-275"><a href="#cb446-275" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(cartoon)</span>
<span id="cb446-276"><a href="#cb446-276" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(cartoon)</span>
<span id="cb446-277"><a href="#cb446-277" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(cartoon<span class="sc">$</span>names)</span>
<span id="cb446-278"><a href="#cb446-278" aria-hidden="true" tabindex="-1"></a><span class="do">##subseting</span></span>
<span id="cb446-279"><a href="#cb446-279" aria-hidden="true" tabindex="-1"></a>cartoon[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>] <span class="co">#row 1-2, column 2-3</span></span>
<span id="cb446-280"><a href="#cb446-280" aria-hidden="true" tabindex="-1"></a>cartoon[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>), <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)] <span class="co">#row 1-3, column 1-3</span></span>
<span id="cb446-281"><a href="#cb446-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-282"><a href="#cb446-282" aria-hidden="true" tabindex="-1"></a><span class="co">#condition for selecting only male characters</span></span>
<span id="cb446-283"><a href="#cb446-283" aria-hidden="true" tabindex="-1"></a>male_df <span class="ot">&lt;-</span> cartoon[cartoon<span class="sc">$</span>gender <span class="sc">==</span> <span class="st">"Male"</span>, ]</span>
<span id="cb446-284"><a href="#cb446-284" aria-hidden="true" tabindex="-1"></a>male_df</span>
<span id="cb446-285"><a href="#cb446-285" aria-hidden="true" tabindex="-1"></a><span class="co">#condition for selecting female characters with age more than 2 years</span></span>
<span id="cb446-286"><a href="#cb446-286" aria-hidden="true" tabindex="-1"></a>female_age <span class="ot">&lt;-</span> cartoon[cartoon<span class="sc">$</span>gender <span class="sc">==</span> <span class="st">"Female"</span> <span class="sc">&amp;</span> cartoon<span class="sc">$</span>age <span class="sc">&gt;</span> <span class="dv">2</span>, ]</span>
<span id="cb446-287"><a href="#cb446-287" aria-hidden="true" tabindex="-1"></a>female_age</span>
<span id="cb446-288"><a href="#cb446-288" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(female_age<span class="sc">$</span>age) <span class="co">#sum of age of female_age dataset</span></span>
<span id="cb446-289"><a href="#cb446-289" aria-hidden="true" tabindex="-1"></a><span class="fu">sd</span>(cartoon<span class="sc">$</span>age) <span class="co">#standard deviation of age of main cartoon dataset</span></span>
<span id="cb446-290"><a href="#cb446-290" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(cartoon<span class="sc">$</span>age) <span class="co">#mean of age of main cartoon dataset</span></span>
<span id="cb446-291"><a href="#cb446-291" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-292"><a href="#cb446-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-293"><a href="#cb446-293" aria-hidden="true" tabindex="-1"></a><span class="co">[</span><span class="ot">Check your colleague's repo for the Q3</span><span class="co">](https://github.com/atiyashehreen/hw_repo3/blob/main/propertiesofdataframe.txt)</span>.</span>
<span id="cb446-294"><a href="#cb446-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-295"><a href="#cb446-295" aria-hidden="true" tabindex="-1"></a>**Logical Operators**</span>
<span id="cb446-296"><a href="#cb446-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-297"><a href="#cb446-297" aria-hidden="true" tabindex="-1"></a>| Operator | Meaning          | Example          |</span>
<span id="cb446-298"><a href="#cb446-298" aria-hidden="true" tabindex="-1"></a>|----------|------------------|------------------|</span>
<span id="cb446-299"><a href="#cb446-299" aria-hidden="true" tabindex="-1"></a>| <span class="in">`==`</span>     | Equal to         | <span class="in">`x == 5`</span>         |</span>
<span id="cb446-300"><a href="#cb446-300" aria-hidden="true" tabindex="-1"></a>| <span class="in">`!=`</span>     | Not equal        | <span class="in">`x != 5`</span>         |</span>
<span id="cb446-301"><a href="#cb446-301" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&lt;`</span>      | Less than        | <span class="in">`x &lt; 5`</span>          |</span>
<span id="cb446-302"><a href="#cb446-302" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&gt;`</span>      | Greater than     | <span class="in">`x &gt; 5`</span>          |</span>
<span id="cb446-303"><a href="#cb446-303" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&lt;=`</span>     | Less or equal    | <span class="in">`x &lt;= 5`</span>         |</span>
<span id="cb446-304"><a href="#cb446-304" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&gt;=`</span>     | Greater or equal | <span class="in">`x &gt;= 5`</span>         |</span>
<span id="cb446-305"><a href="#cb446-305" aria-hidden="true" tabindex="-1"></a>| <span class="in">`!`</span>      | Not              | <span class="in">`!(x &lt; 5)`</span>       |</span>
<span id="cb446-306"><a href="#cb446-306" aria-hidden="true" tabindex="-1"></a>| <span class="in">`|`</span>      | OR               | <span class="in">`x &lt; 5 | x &gt; 10`</span> |</span>
<span id="cb446-307"><a href="#cb446-307" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&amp;`</span>      | AND              | <span class="in">`x &gt; 5 &amp; x &lt; 10`</span> |</span>
<span id="cb446-308"><a href="#cb446-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-309"><a href="#cb446-309" aria-hidden="true" tabindex="-1"></a><span class="fu">### Preamble on random variables (RV):</span></span>
<span id="cb446-310"><a href="#cb446-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-311"><a href="#cb446-311" aria-hidden="true" tabindex="-1"></a>RV is so fundamental of an idea to interpret and do better in any kind of data analyses. But what is it? Let's imagine this scenario first. You got 30 mice to do an experiment to check anti-diabetic effect of a plant extract. You randomly assigned them into 3 groups. <span class="in">`control`</span>, <span class="in">`treat1`</span> (meaning insulin receivers), and <span class="in">`treat2`</span> (meaning your plant extract receivers). Then you kept testing and measuring. You have mean glucose level of every mouse and show whether the mean value of <span class="in">`treat1`</span> is equal to <span class="in">`treat2`</span> or not. So, are you done? Not really. Be fastidious about the mice. What if you got some other 30 mice? Are they the same? Will their mean glucose level be the same? No, right. We would end up with different mean value. We call this type of quantities RV. Mean, Standard deviation, median, variance, etc. all are RVs. Do you see the logic? That's why we put this constraint and look for p-value, confidence interval (or CI), etc. by (null) hypothesis testing and sample distribution analyses. We will get into these stuffs later. But let's check what I meant. Also ponder about <span class="in">`sample`</span> vs <span class="in">`population`</span>.</span>
<span id="cb446-312"><a href="#cb446-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-313"><a href="#cb446-313" aria-hidden="true" tabindex="-1"></a>Let's download the data first.</span>
<span id="cb446-314"><a href="#cb446-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-315"><a href="#cb446-315" aria-hidden="true" tabindex="-1"></a><span class="in">```{r RV}</span></span>
<span id="cb446-316"><a href="#cb446-316" aria-hidden="true" tabindex="-1"></a><span class="co"># Download small example dataset</span></span>
<span id="cb446-317"><a href="#cb446-317" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/femaleControlsPopulation.csv"</span>,</span>
<span id="cb446-318"><a href="#cb446-318" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"mice.csv"</span>)</span>
<span id="cb446-319"><a href="#cb446-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-320"><a href="#cb446-320" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb446-321"><a href="#cb446-321" aria-hidden="true" tabindex="-1"></a>mice <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"mice.csv"</span>)</span>
<span id="cb446-322"><a href="#cb446-322" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-323"><a href="#cb446-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-324"><a href="#cb446-324" aria-hidden="true" tabindex="-1"></a>Let's check now.</span>
<span id="cb446-325"><a href="#cb446-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-326"><a href="#cb446-326" aria-hidden="true" tabindex="-1"></a><span class="in">```{r rv-check}</span></span>
<span id="cb446-327"><a href="#cb446-327" aria-hidden="true" tabindex="-1"></a>control <span class="ot">&lt;-</span> <span class="fu">sample</span>(mice<span class="sc">$</span>Bodyweight,<span class="dv">12</span>)</span>
<span id="cb446-328"><a href="#cb446-328" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(control)</span>
<span id="cb446-329"><a href="#cb446-329" aria-hidden="true" tabindex="-1"></a>control1 <span class="ot">&lt;-</span> <span class="fu">sample</span>(mice<span class="sc">$</span>Bodyweight,<span class="dv">12</span>)</span>
<span id="cb446-330"><a href="#cb446-330" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(control1)</span>
<span id="cb446-331"><a href="#cb446-331" aria-hidden="true" tabindex="-1"></a>control2 <span class="ot">&lt;-</span> <span class="fu">sample</span>(mice<span class="sc">$</span>Bodyweight,<span class="dv">12</span>)</span>
<span id="cb446-332"><a href="#cb446-332" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(control2)</span>
<span id="cb446-333"><a href="#cb446-333" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-334"><a href="#cb446-334" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-335"><a href="#cb446-335" aria-hidden="true" tabindex="-1"></a>Do you see the difference in the mean value now?</span>
<span id="cb446-336"><a href="#cb446-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-337"><a href="#cb446-337" aria-hidden="true" tabindex="-1"></a><span class="fu">### Basic Stuffs: Atomic Vector</span></span>
<span id="cb446-338"><a href="#cb446-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-339"><a href="#cb446-339" aria-hidden="true" tabindex="-1"></a><span class="in">```{r atomic-vec}</span></span>
<span id="cb446-340"><a href="#cb446-340" aria-hidden="true" tabindex="-1"></a>atomic_vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="at">Human=</span><span class="fl">0.5</span>, <span class="at">Mouse=</span><span class="fl">0.33</span>)</span>
<span id="cb446-341"><a href="#cb446-341" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-342"><a href="#cb446-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-343"><a href="#cb446-343" aria-hidden="true" tabindex="-1"></a>It is fast, but has limited access methods.</span>
<span id="cb446-344"><a href="#cb446-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-345"><a href="#cb446-345" aria-hidden="true" tabindex="-1"></a>How to access elements here?</span>
<span id="cb446-346"><a href="#cb446-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-349"><a href="#cb446-349" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-350"><a href="#cb446-350" aria-hidden="true" tabindex="-1"></a>atomic_vec[<span class="st">"Human"</span>]</span>
<span id="cb446-351"><a href="#cb446-351" aria-hidden="true" tabindex="-1"></a>atomic_vec[<span class="st">"Mouse"</span>]</span>
<span id="cb446-352"><a href="#cb446-352" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-353"><a href="#cb446-353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-354"><a href="#cb446-354" aria-hidden="true" tabindex="-1"></a><span class="fu">### Basic Stuffs: Matrices</span></span>
<span id="cb446-355"><a href="#cb446-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-356"><a href="#cb446-356" aria-hidden="true" tabindex="-1"></a>Matrices are essential for biologists working with expression data, distance matrices, and other numerical data.</span>
<span id="cb446-357"><a href="#cb446-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-358"><a href="#cb446-358" aria-hidden="true" tabindex="-1"></a><span class="in">```{r matrices}</span></span>
<span id="cb446-359"><a href="#cb446-359" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a gene expression matrix: rows=genes, columns=samples</span></span>
<span id="cb446-360"><a href="#cb446-360" aria-hidden="true" tabindex="-1"></a>expr_matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb446-361"><a href="#cb446-361" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">12.3</span>, <span class="fl">8.7</span>, <span class="fl">15.2</span>, <span class="fl">6.8</span>,</span>
<span id="cb446-362"><a href="#cb446-362" aria-hidden="true" tabindex="-1"></a>    <span class="fl">9.5</span>, <span class="fl">11.2</span>, <span class="fl">13.7</span>, <span class="fl">7.4</span>,</span>
<span id="cb446-363"><a href="#cb446-363" aria-hidden="true" tabindex="-1"></a>    <span class="fl">5.6</span>, <span class="fl">6.8</span>, <span class="fl">7.9</span>, <span class="fl">6.5</span>),</span>
<span id="cb446-364"><a href="#cb446-364" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">ncol =</span> <span class="dv">4</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb446-365"><a href="#cb446-365" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-366"><a href="#cb446-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-367"><a href="#cb446-367" aria-hidden="true" tabindex="-1"></a><span class="co"># Add dimension names</span></span>
<span id="cb446-368"><a href="#cb446-368" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(expr_matrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"GAPDH"</span>)</span>
<span id="cb446-369"><a href="#cb446-369" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(expr_matrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Control_1"</span>, <span class="st">"Control_2"</span>, <span class="st">"Treatment_1"</span>, <span class="st">"Treatment_2"</span>)</span>
<span id="cb446-370"><a href="#cb446-370" aria-hidden="true" tabindex="-1"></a>expr_matrix</span>
<span id="cb446-371"><a href="#cb446-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-372"><a href="#cb446-372" aria-hidden="true" tabindex="-1"></a><span class="co"># Matrix dimensions</span></span>
<span id="cb446-373"><a href="#cb446-373" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(expr_matrix)       <span class="co"># Returns rows and columns</span></span>
<span id="cb446-374"><a href="#cb446-374" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(expr_matrix)      <span class="co"># Number of rows</span></span>
<span id="cb446-375"><a href="#cb446-375" aria-hidden="true" tabindex="-1"></a><span class="fu">ncol</span>(expr_matrix)      <span class="co"># Number of columns</span></span>
<span id="cb446-376"><a href="#cb446-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-377"><a href="#cb446-377" aria-hidden="true" tabindex="-1"></a><span class="co"># Matrix subsetting</span></span>
<span id="cb446-378"><a href="#cb446-378" aria-hidden="true" tabindex="-1"></a>expr_matrix[<span class="dv">2</span>, ]       <span class="co"># One gene, all samples</span></span>
<span id="cb446-379"><a href="#cb446-379" aria-hidden="true" tabindex="-1"></a>expr_matrix[, <span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>]     <span class="co"># All genes, treatment samples only</span></span>
<span id="cb446-380"><a href="#cb446-380" aria-hidden="true" tabindex="-1"></a>expr_matrix[<span class="st">"TP53"</span>, <span class="fu">c</span>(<span class="st">"Control_1"</span>, <span class="st">"Treatment_1"</span>)]  <span class="co"># Specific gene and samples</span></span>
<span id="cb446-381"><a href="#cb446-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-382"><a href="#cb446-382" aria-hidden="true" tabindex="-1"></a><span class="co"># Matrix calculations (useful for bioinformatics)</span></span>
<span id="cb446-383"><a href="#cb446-383" aria-hidden="true" tabindex="-1"></a><span class="co"># Mean expression per gene</span></span>
<span id="cb446-384"><a href="#cb446-384" aria-hidden="true" tabindex="-1"></a>gene_means <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(expr_matrix)</span>
<span id="cb446-385"><a href="#cb446-385" aria-hidden="true" tabindex="-1"></a>gene_means</span>
<span id="cb446-386"><a href="#cb446-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-387"><a href="#cb446-387" aria-hidden="true" tabindex="-1"></a><span class="co"># Mean expression per sample</span></span>
<span id="cb446-388"><a href="#cb446-388" aria-hidden="true" tabindex="-1"></a>sample_means <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(expr_matrix)</span>
<span id="cb446-389"><a href="#cb446-389" aria-hidden="true" tabindex="-1"></a>sample_means</span>
<span id="cb446-390"><a href="#cb446-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-391"><a href="#cb446-391" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate fold change (Treatment vs Control)</span></span>
<span id="cb446-392"><a href="#cb446-392" aria-hidden="true" tabindex="-1"></a>control_means <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(expr_matrix[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])</span>
<span id="cb446-393"><a href="#cb446-393" aria-hidden="true" tabindex="-1"></a>treatment_means <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(expr_matrix[, <span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb446-394"><a href="#cb446-394" aria-hidden="true" tabindex="-1"></a>fold_change <span class="ot">&lt;-</span> treatment_means <span class="sc">/</span> control_means</span>
<span id="cb446-395"><a href="#cb446-395" aria-hidden="true" tabindex="-1"></a>fold_change</span>
<span id="cb446-396"><a href="#cb446-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-397"><a href="#cb446-397" aria-hidden="true" tabindex="-1"></a><span class="co"># Matrix visualization</span></span>
<span id="cb446-398"><a href="#cb446-398" aria-hidden="true" tabindex="-1"></a><span class="co"># Heatmap of expression data</span></span>
<span id="cb446-399"><a href="#cb446-399" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(expr_matrix, </span>
<span id="cb446-400"><a href="#cb446-400" aria-hidden="true" tabindex="-1"></a>        <span class="at">Colv =</span> <span class="cn">NA</span>,         <span class="co"># Don't cluster columns</span></span>
<span id="cb446-401"><a href="#cb446-401" aria-hidden="true" tabindex="-1"></a>        <span class="at">scale =</span> <span class="st">"row"</span>,     <span class="co"># Scale by row (gene)</span></span>
<span id="cb446-402"><a href="#cb446-402" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">16</span>),</span>
<span id="cb446-403"><a href="#cb446-403" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Gene Expression Heatmap"</span>)</span>
<span id="cb446-404"><a href="#cb446-404" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-405"><a href="#cb446-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-406"><a href="#cb446-406" aria-hidden="true" tabindex="-1"></a><span class="fu">#### More Matrix Practice:</span></span>
<span id="cb446-407"><a href="#cb446-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-408"><a href="#cb446-408" aria-hidden="true" tabindex="-1"></a><span class="in">```{r more-mat-practice}</span></span>
<span id="cb446-409"><a href="#cb446-409" aria-hidden="true" tabindex="-1"></a><span class="co">#Create a simple Gene Expression matrix (RNA-seq style)</span></span>
<span id="cb446-410"><a href="#cb446-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-411"><a href="#cb446-411" aria-hidden="true" tabindex="-1"></a>Gene_Expression <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb446-412"><a href="#cb446-412" aria-hidden="true" tabindex="-1"></a>  <span class="fl">5.2</span>, <span class="fl">3.1</span>, <span class="fl">8.5</span>,   <span class="co"># Sample 1</span></span>
<span id="cb446-413"><a href="#cb446-413" aria-hidden="true" tabindex="-1"></a>  <span class="fl">6.0</span>, <span class="fl">2.8</span>, <span class="fl">7.9</span>    <span class="co"># Sample 2</span></span>
<span id="cb446-414"><a href="#cb446-414" aria-hidden="true" tabindex="-1"></a>), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-415"><a href="#cb446-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-416"><a href="#cb446-416" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(Gene_Expression) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Sample_1"</span>, <span class="st">"Sample_2"</span>)</span>
<span id="cb446-417"><a href="#cb446-417" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Gene_Expression) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb446-418"><a href="#cb446-418" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-419"><a href="#cb446-419" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Gene Expression Matrix:"</span>)</span>
<span id="cb446-420"><a href="#cb446-420" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Gene_Expression)</span>
<span id="cb446-421"><a href="#cb446-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-422"><a href="#cb446-422" aria-hidden="true" tabindex="-1"></a><span class="co">#1. Transpose: Genes become rows, Samples become columns</span></span>
<span id="cb446-423"><a href="#cb446-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-424"><a href="#cb446-424" aria-hidden="true" tabindex="-1"></a>Gene_Expression_T <span class="ot">&lt;-</span> <span class="fu">t</span>(Gene_Expression)</span>
<span id="cb446-425"><a href="#cb446-425" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Transpose of Gene Expression Matrix:"</span>)</span>
<span id="cb446-426"><a href="#cb446-426" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Gene_Expression_T)</span>
<span id="cb446-427"><a href="#cb446-427" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-428"><a href="#cb446-428" aria-hidden="true" tabindex="-1"></a><span class="co">#2. Matrix multiplication</span></span>
<span id="cb446-429"><a href="#cb446-429" aria-hidden="true" tabindex="-1"></a><span class="co"># Suppose each gene has an associated "gene weight" (e.g., biological importance)</span></span>
<span id="cb446-430"><a href="#cb446-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-431"><a href="#cb446-431" aria-hidden="true" tabindex="-1"></a>Gene_Weights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.8</span>, <span class="fl">1.2</span>, <span class="fl">1.0</span>), <span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-432"><a href="#cb446-432" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(Gene_Weights) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb446-433"><a href="#cb446-433" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Gene_Weights) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb446-434"><a href="#cb446-434" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-435"><a href="#cb446-435" aria-hidden="true" tabindex="-1"></a>Total_Weighted_Expression <span class="ot">&lt;-</span> Gene_Expression <span class="sc">%*%</span> Gene_Weights</span>
<span id="cb446-436"><a href="#cb446-436" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Total Weighted Expression per Sample:"</span>)</span>
<span id="cb446-437"><a href="#cb446-437" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Total_Weighted_Expression)</span>
<span id="cb446-438"><a href="#cb446-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-439"><a href="#cb446-439" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Matrix addition</span></span>
<span id="cb446-440"><a href="#cb446-440" aria-hidden="true" tabindex="-1"></a><span class="co"># Hypothetically increase expression by 1 TPM everywhere (technical adjustment)</span></span>
<span id="cb446-441"><a href="#cb446-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-442"><a href="#cb446-442" aria-hidden="true" tabindex="-1"></a>Adjusted_Expression <span class="ot">&lt;-</span> Gene_Expression <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb446-443"><a href="#cb446-443" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Expression Matrix after adding 1 TPM:"</span>)</span>
<span id="cb446-444"><a href="#cb446-444" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Adjusted_Expression)</span>
<span id="cb446-445"><a href="#cb446-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-446"><a href="#cb446-446" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. Identity matrix </span></span>
<span id="cb446-447"><a href="#cb446-447" aria-hidden="true" tabindex="-1"></a>I <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb446-448"><a href="#cb446-448" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(I) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb446-449"><a href="#cb446-449" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(I) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb446-450"><a href="#cb446-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-451"><a href="#cb446-451" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Identity Matrix (for genes):"</span>)</span>
<span id="cb446-452"><a href="#cb446-452" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(I)</span>
<span id="cb446-453"><a href="#cb446-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-454"><a href="#cb446-454" aria-hidden="true" tabindex="-1"></a><span class="co"># Multiplying Gene Expression by Identity</span></span>
<span id="cb446-455"><a href="#cb446-455" aria-hidden="true" tabindex="-1"></a>Identity_Check <span class="ot">&lt;-</span> Gene_Expression <span class="sc">%*%</span> I</span>
<span id="cb446-456"><a href="#cb446-456" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Gene Expression multiplied by Identity Matrix:"</span>)</span>
<span id="cb446-457"><a href="#cb446-457" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Identity_Check)</span>
<span id="cb446-458"><a href="#cb446-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-459"><a href="#cb446-459" aria-hidden="true" tabindex="-1"></a><span class="co"># 5. Scalar multiplication </span></span>
<span id="cb446-460"><a href="#cb446-460" aria-hidden="true" tabindex="-1"></a><span class="co"># Suppose you want to simulate doubling expression values</span></span>
<span id="cb446-461"><a href="#cb446-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-462"><a href="#cb446-462" aria-hidden="true" tabindex="-1"></a>Doubled_Expression <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> Gene_Expression</span>
<span id="cb446-463"><a href="#cb446-463" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Doubled Gene Expression:"</span>)</span>
<span id="cb446-464"><a href="#cb446-464" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Doubled_Expression)</span>
<span id="cb446-465"><a href="#cb446-465" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-466"><a href="#cb446-466" aria-hidden="true" tabindex="-1"></a><span class="co"># 6. Summations </span></span>
<span id="cb446-467"><a href="#cb446-467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-468"><a href="#cb446-468" aria-hidden="true" tabindex="-1"></a><span class="co"># Total expression per sample</span></span>
<span id="cb446-469"><a href="#cb446-469" aria-hidden="true" tabindex="-1"></a>Total_Expression_Per_Sample <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(Gene_Expression)</span>
<span id="cb446-470"><a href="#cb446-470" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Total Expression per Sample:"</span>)</span>
<span id="cb446-471"><a href="#cb446-471" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Total_Expression_Per_Sample)</span>
<span id="cb446-472"><a href="#cb446-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-473"><a href="#cb446-473" aria-hidden="true" tabindex="-1"></a><span class="co"># Total expression per gene</span></span>
<span id="cb446-474"><a href="#cb446-474" aria-hidden="true" tabindex="-1"></a>Total_Expression_Per_Gene <span class="ot">&lt;-</span> <span class="fu">colSums</span>(Gene_Expression)</span>
<span id="cb446-475"><a href="#cb446-475" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Total Expression per Gene:"</span>)</span>
<span id="cb446-476"><a href="#cb446-476" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Total_Expression_Per_Gene)</span>
<span id="cb446-477"><a href="#cb446-477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-478"><a href="#cb446-478" aria-hidden="true" tabindex="-1"></a><span class="co"># 7. Simple plots </span></span>
<span id="cb446-479"><a href="#cb446-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-480"><a href="#cb446-480" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot: Total expression per sample</span></span>
<span id="cb446-481"><a href="#cb446-481" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(Total_Expression_Per_Sample, <span class="at">main=</span><span class="st">"Total Expression per Sample"</span>, <span class="at">ylab=</span><span class="st">"TPM"</span>, <span class="at">col=</span><span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"salmon"</span>))</span>
<span id="cb446-482"><a href="#cb446-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-483"><a href="#cb446-483" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot: Total expression per gene</span></span>
<span id="cb446-484"><a href="#cb446-484" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(Total_Expression_Per_Gene, <span class="at">main=</span><span class="st">"Total Expression per Gene"</span>, <span class="at">ylab=</span><span class="st">"TPM"</span>, <span class="at">col=</span><span class="fu">c</span>(<span class="st">"lightgreen"</span>, <span class="st">"orange"</span>, <span class="st">"violet"</span>))</span>
<span id="cb446-485"><a href="#cb446-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-486"><a href="#cb446-486" aria-hidden="true" tabindex="-1"></a><span class="co"># Heatmap: Expression matrix</span></span>
<span id="cb446-487"><a href="#cb446-487" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(Gene_Expression, <span class="at">Rowv=</span><span class="cn">NA</span>, <span class="at">Colv=</span><span class="cn">NA</span>, <span class="at">col=</span><span class="fu">heat.colors</span>(<span class="dv">256</span>), <span class="at">scale=</span><span class="st">"column"</span>, <span class="at">main=</span><span class="st">"Gene Expression Heatmap"</span>)</span>
<span id="cb446-488"><a href="#cb446-488" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-489"><a href="#cb446-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-490"><a href="#cb446-490" aria-hidden="true" tabindex="-1"></a>Another Example: You have counts of cells in different organs for two animal species.</span>
<span id="cb446-491"><a href="#cb446-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-492"><a href="#cb446-492" aria-hidden="true" tabindex="-1"></a>You also have a matrix with average cell sizes (micrometer, ¬µm¬≤) for each organ.</span>
<span id="cb446-493"><a href="#cb446-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-494"><a href="#cb446-494" aria-hidden="true" tabindex="-1"></a>You can then multiply count √ó size to get total cell area for each species in each organ.</span>
<span id="cb446-495"><a href="#cb446-495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-496"><a href="#cb446-496" aria-hidden="true" tabindex="-1"></a><span class="in">```{r mat-prac-2}</span></span>
<span id="cb446-497"><a href="#cb446-497" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a matrix: Cell counts</span></span>
<span id="cb446-498"><a href="#cb446-498" aria-hidden="true" tabindex="-1"></a>Cell_Counts <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">500</span>, <span class="dv">600</span>, <span class="dv">300</span>, <span class="dv">400</span>, <span class="dv">700</span>, <span class="dv">800</span>), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-499"><a href="#cb446-499" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(Cell_Counts) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Mouse"</span>, <span class="st">"Rat"</span>)</span>
<span id="cb446-500"><a href="#cb446-500" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Cell_Counts) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Heart"</span>, <span class="st">"Liver"</span>, <span class="st">"Brain"</span>)</span>
<span id="cb446-501"><a href="#cb446-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-502"><a href="#cb446-502" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Cell Counts Matrix:"</span>)</span>
<span id="cb446-503"><a href="#cb446-503" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Cell_Counts)</span>
<span id="cb446-504"><a href="#cb446-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-505"><a href="#cb446-505" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a matrix: Average cell size in ¬µm¬≤</span></span>
<span id="cb446-506"><a href="#cb446-506" aria-hidden="true" tabindex="-1"></a>Cell_Size <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="dv">50</span>, <span class="dv">200</span>, <span class="dv">150</span>), <span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-507"><a href="#cb446-507" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(Cell_Size) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Heart"</span>, <span class="st">"Liver"</span>, <span class="st">"Brain"</span>)</span>
<span id="cb446-508"><a href="#cb446-508" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Cell_Size) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Avg_Cell_Size"</span>)</span>
<span id="cb446-509"><a href="#cb446-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-510"><a href="#cb446-510" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Cell Size Matrix (¬µm¬≤):"</span>)</span>
<span id="cb446-511"><a href="#cb446-511" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Cell_Size)</span>
<span id="cb446-512"><a href="#cb446-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-513"><a href="#cb446-513" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Transpose of Cell Counts</span></span>
<span id="cb446-514"><a href="#cb446-514" aria-hidden="true" tabindex="-1"></a>Cell_Counts_T <span class="ot">&lt;-</span> <span class="fu">t</span>(Cell_Counts)</span>
<span id="cb446-515"><a href="#cb446-515" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Transpose of Cell Counts:"</span>)</span>
<span id="cb446-516"><a href="#cb446-516" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Cell_Counts_T)</span>
<span id="cb446-517"><a href="#cb446-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-518"><a href="#cb446-518" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Matrix multiplication: Total cell area</span></span>
<span id="cb446-519"><a href="#cb446-519" aria-hidden="true" tabindex="-1"></a><span class="co"># (2x3) %*% (3x1) =&gt; (2x1)</span></span>
<span id="cb446-520"><a href="#cb446-520" aria-hidden="true" tabindex="-1"></a>Total_Cell_Area <span class="ot">&lt;-</span> Cell_Counts <span class="sc">%*%</span> Cell_Size</span>
<span id="cb446-521"><a href="#cb446-521" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Total_Cell_Area) <span class="ot">&lt;-</span> <span class="st">"Cell_area"</span></span>
<span id="cb446-522"><a href="#cb446-522" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Total Cell Area (Counts √ó Size) (¬µm¬≤):"</span>)</span>
<span id="cb446-523"><a href="#cb446-523" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Total_Cell_Area)</span>
<span id="cb446-524"><a href="#cb446-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-525"><a href="#cb446-525" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Matrix addition: Add 10 cells artificially to all counts (for example)</span></span>
<span id="cb446-526"><a href="#cb446-526" aria-hidden="true" tabindex="-1"></a>Added_Cells <span class="ot">&lt;-</span> Cell_Counts <span class="sc">+</span> <span class="dv">10</span></span>
<span id="cb446-527"><a href="#cb446-527" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Cell Counts after adding 10 artificial cells:"</span>)</span>
<span id="cb446-528"><a href="#cb446-528" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Added_Cells)</span>
<span id="cb446-529"><a href="#cb446-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-530"><a href="#cb446-530" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. Identity matrix</span></span>
<span id="cb446-531"><a href="#cb446-531" aria-hidden="true" tabindex="-1"></a>I <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb446-532"><a href="#cb446-532" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(I) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Heart"</span>, <span class="st">"Liver"</span>, <span class="st">"Brain"</span>)</span>
<span id="cb446-533"><a href="#cb446-533" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(I) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Heart"</span>, <span class="st">"Liver"</span>, <span class="st">"Brain"</span>)</span>
<span id="cb446-534"><a href="#cb446-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-535"><a href="#cb446-535" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Identity Matrix:"</span>)</span>
<span id="cb446-536"><a href="#cb446-536" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(I)</span>
<span id="cb446-537"><a href="#cb446-537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-538"><a href="#cb446-538" aria-hidden="true" tabindex="-1"></a><span class="co"># 5. Multiplying Cell Counts by Identity Matrix (no real change but shows dimension rules)</span></span>
<span id="cb446-539"><a href="#cb446-539" aria-hidden="true" tabindex="-1"></a>Check_Identity <span class="ot">&lt;-</span> Cell_Counts <span class="sc">%*%</span> I</span>
<span id="cb446-540"><a href="#cb446-540" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Cell Counts multiplied by Identity Matrix:"</span>)</span>
<span id="cb446-541"><a href="#cb446-541" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Check_Identity)</span>
<span id="cb446-542"><a href="#cb446-542" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-543"><a href="#cb446-543" aria-hidden="true" tabindex="-1"></a><span class="co"># 6. Scalar multiplication: double the counts (hypothetical growth)</span></span>
<span id="cb446-544"><a href="#cb446-544" aria-hidden="true" tabindex="-1"></a>Double_Cell_Counts <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> Cell_Counts</span>
<span id="cb446-545"><a href="#cb446-545" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Doubled Cell Counts:"</span>)</span>
<span id="cb446-546"><a href="#cb446-546" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Double_Cell_Counts)</span>
<span id="cb446-547"><a href="#cb446-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-548"><a href="#cb446-548" aria-hidden="true" tabindex="-1"></a><span class="co"># Total number of cells per animal (row sums)</span></span>
<span id="cb446-549"><a href="#cb446-549" aria-hidden="true" tabindex="-1"></a>Total_Cells_Per_Species <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(Cell_Counts)</span>
<span id="cb446-550"><a href="#cb446-550" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Total number of cells per species:"</span>)</span>
<span id="cb446-551"><a href="#cb446-551" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Total_Cells_Per_Species)</span>
<span id="cb446-552"><a href="#cb446-552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-553"><a href="#cb446-553" aria-hidden="true" tabindex="-1"></a><span class="co"># Total number of cells per organ (column sums)</span></span>
<span id="cb446-554"><a href="#cb446-554" aria-hidden="true" tabindex="-1"></a>Total_Cells_Per_Organ <span class="ot">&lt;-</span> <span class="fu">colSums</span>(Cell_Counts)</span>
<span id="cb446-555"><a href="#cb446-555" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="st">"Total number of cells per organ:"</span>)</span>
<span id="cb446-556"><a href="#cb446-556" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Total_Cells_Per_Organ)</span>
<span id="cb446-557"><a href="#cb446-557" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-558"><a href="#cb446-558" aria-hidden="true" tabindex="-1"></a><span class="co"># --- Simple plots ---</span></span>
<span id="cb446-559"><a href="#cb446-559" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-560"><a href="#cb446-560" aria-hidden="true" tabindex="-1"></a><span class="co"># Bar plot of total cells per species</span></span>
<span id="cb446-561"><a href="#cb446-561" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(Total_Cells_Per_Species, <span class="at">main=</span><span class="st">"Total Cell Counts per Species"</span>, <span class="at">ylab=</span><span class="st">"Number of Cells"</span>, <span class="at">col=</span><span class="fu">c</span>(<span class="st">"lightblue"</span>, <span class="st">"lightgreen"</span>))</span>
<span id="cb446-562"><a href="#cb446-562" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-563"><a href="#cb446-563" aria-hidden="true" tabindex="-1"></a><span class="co"># Bar plot of total cells per organ</span></span>
<span id="cb446-564"><a href="#cb446-564" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(Total_Cells_Per_Organ, <span class="at">main=</span><span class="st">"Total Cell Counts per Organ"</span>, <span class="at">ylab=</span><span class="st">"Number of Cells"</span>, <span class="at">col=</span><span class="fu">c</span>(<span class="st">"pink"</span>, <span class="st">"lightyellow"</span>, <span class="st">"lightgray"</span>))</span>
<span id="cb446-565"><a href="#cb446-565" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-566"><a href="#cb446-566" aria-hidden="true" tabindex="-1"></a><span class="co"># Heatmap of the original Cell Counts matrix</span></span>
<span id="cb446-567"><a href="#cb446-567" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(Cell_Counts, <span class="at">Rowv=</span><span class="cn">NA</span>, <span class="at">Colv=</span><span class="cn">NA</span>, <span class="at">col=</span><span class="fu">heat.colors</span>(<span class="dv">256</span>), <span class="at">scale=</span><span class="st">"column"</span>, <span class="at">main=</span><span class="st">"Heatmap of Cell Counts"</span>)</span>
<span id="cb446-568"><a href="#cb446-568" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-569"><a href="#cb446-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-570"><a href="#cb446-570" aria-hidden="true" tabindex="-1"></a>| Operation             | Explanation                   | R Function/Example                 |</span>
<span id="cb446-571"><a href="#cb446-571" aria-hidden="true" tabindex="-1"></a>|:-------------------|:--------------------------|:------------------------|</span>
<span id="cb446-572"><a href="#cb446-572" aria-hidden="true" tabindex="-1"></a>| Matrix Creation       | Create gene expression matrix | <span class="in">`matrix()`</span>                         |</span>
<span id="cb446-573"><a href="#cb446-573" aria-hidden="true" tabindex="-1"></a>| Transpose             | Flip genes and samples        | <span class="in">`t(Gene_Expression)`</span>               |</span>
<span id="cb446-574"><a href="#cb446-574" aria-hidden="true" tabindex="-1"></a>| Matrix Multiplication | Calculate weighted sums       | <span class="in">`Gene_Expression %*% Gene_Weights`</span> |</span>
<span id="cb446-575"><a href="#cb446-575" aria-hidden="true" tabindex="-1"></a>| Matrix Addition       | Adjust counts                 | <span class="in">`Gene_Expression + 1`</span>              |</span>
<span id="cb446-576"><a href="#cb446-576" aria-hidden="true" tabindex="-1"></a>| Identity Matrix       | Special neutral matrix        | <span class="in">`diag(3)`</span>                          |</span>
<span id="cb446-577"><a href="#cb446-577" aria-hidden="true" tabindex="-1"></a>| Scalar Multiplication | Simulate overall increase     | <span class="in">`2 * Gene_Expression`</span>              |</span>
<span id="cb446-578"><a href="#cb446-578" aria-hidden="true" tabindex="-1"></a>| Row/Column Summation  | Total per sample/gene         | <span class="in">`rowSums()`</span>, <span class="in">`colSums()`</span>           |</span>
<span id="cb446-579"><a href="#cb446-579" aria-hidden="true" tabindex="-1"></a>| Plotting              | Visualize expression patterns | <span class="in">`barplot()`</span>, <span class="in">`heatmap()`</span>           |</span>
<span id="cb446-580"><a href="#cb446-580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-581"><a href="#cb446-581" aria-hidden="true" tabindex="-1"></a><span class="fu">### Basic Stuffs: List</span></span>
<span id="cb446-582"><a href="#cb446-582" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-583"><a href="#cb446-583" aria-hidden="true" tabindex="-1"></a>Lists are the most flexible data structure in R - they can hold any combination of data types, including other lists! This makes them essential for biological data analysis where we often deal with mixed data types.</span>
<span id="cb446-584"><a href="#cb446-584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-585"><a href="#cb446-585" aria-hidden="true" tabindex="-1"></a><span class="in">```{r list}</span></span>
<span id="cb446-586"><a href="#cb446-586" aria-hidden="true" tabindex="-1"></a><span class="co"># A list storing different types of genomic data</span></span>
<span id="cb446-587"><a href="#cb446-587" aria-hidden="true" tabindex="-1"></a>genomics_data <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb446-588"><a href="#cb446-588" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_names =</span> <span class="fu">c</span>(<span class="st">"TP53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"MYC"</span>),               <span class="co"># Character vector</span></span>
<span id="cb446-589"><a href="#cb446-589" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression =</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">1.2</span>, <span class="fl">3.4</span>, <span class="fl">5.6</span>, <span class="fl">7.8</span>, <span class="fl">9.1</span>, <span class="fl">2.3</span>), <span class="at">nrow=</span><span class="dv">3</span>),    <span class="co"># Numeric matrix</span></span>
<span id="cb446-590"><a href="#cb446-590" aria-hidden="true" tabindex="-1"></a>  <span class="at">is_cancer_gene =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>),                 <span class="co"># Logical vector</span></span>
<span id="cb446-591"><a href="#cb446-591" aria-hidden="true" tabindex="-1"></a>  <span class="at">metadata =</span> <span class="fu">list</span>(                                       <span class="co"># Nested list!</span></span>
<span id="cb446-592"><a href="#cb446-592" aria-hidden="true" tabindex="-1"></a>    <span class="at">lab =</span> <span class="st">"CRG"</span>,</span>
<span id="cb446-593"><a href="#cb446-593" aria-hidden="true" tabindex="-1"></a>    <span class="at">date =</span> <span class="st">"2023-05-01"</span></span>
<span id="cb446-594"><a href="#cb446-594" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb446-595"><a href="#cb446-595" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-596"><a href="#cb446-596" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-597"><a href="#cb446-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-598"><a href="#cb446-598" aria-hidden="true" tabindex="-1"></a>How to Access Elements of a List?</span>
<span id="cb446-599"><a href="#cb446-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-600"><a href="#cb446-600" aria-hidden="true" tabindex="-1"></a><span class="in">```{r access-list}</span></span>
<span id="cb446-601"><a href="#cb446-601" aria-hidden="true" tabindex="-1"></a><span class="co"># Method 1: Double brackets [[ ]] for single element</span></span>
<span id="cb446-602"><a href="#cb446-602" aria-hidden="true" tabindex="-1"></a>genomics_data[[<span class="dv">1</span>]]  <span class="co"># Returns gene_names vector</span></span>
<span id="cb446-603"><a href="#cb446-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-604"><a href="#cb446-604" aria-hidden="true" tabindex="-1"></a><span class="co"># Method 2: $ operator with names (when elements are named)</span></span>
<span id="cb446-605"><a href="#cb446-605" aria-hidden="true" tabindex="-1"></a>genomics_data<span class="sc">$</span>expression  <span class="co"># Returns the matrix</span></span>
<span id="cb446-606"><a href="#cb446-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-607"><a href="#cb446-607" aria-hidden="true" tabindex="-1"></a><span class="co"># Method 3: Single bracket [ ] returns a sublist</span></span>
<span id="cb446-608"><a href="#cb446-608" aria-hidden="true" tabindex="-1"></a>genomics_data[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]  <span class="co"># Returns list with first two elements</span></span>
<span id="cb446-609"><a href="#cb446-609" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-610"><a href="#cb446-610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-611"><a href="#cb446-611" aria-hidden="true" tabindex="-1"></a>Key Difference from Vectors:</span>
<span id="cb446-612"><a href="#cb446-612" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-613"><a href="#cb446-613" aria-hidden="true" tabindex="-1"></a><span class="in">```{r vecVSlist}</span></span>
<span id="cb446-614"><a href="#cb446-614" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare to your prop.table() example:</span></span>
<span id="cb446-615"><a href="#cb446-615" aria-hidden="true" tabindex="-1"></a>atomic_vec[<span class="st">"Human"</span>]    <span class="co"># Returns named numeric (vector)</span></span>
<span id="cb446-616"><a href="#cb446-616" aria-hidden="true" tabindex="-1"></a>atomic_vec[<span class="st">"Mouse"</span>]</span>
<span id="cb446-617"><a href="#cb446-617" aria-hidden="true" tabindex="-1"></a>genomics_data[<span class="dv">1</span>] <span class="co"># Returns list containing the vector</span></span>
<span id="cb446-618"><a href="#cb446-618" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-619"><a href="#cb446-619" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-620"><a href="#cb446-620" aria-hidden="true" tabindex="-1"></a>Why Biologists Need Lists?\</span>
<span id="cb446-621"><a href="#cb446-621" aria-hidden="true" tabindex="-1"></a><span class="in">`lm()`</span>, <span class="in">`prcomp()`</span> functions, RNAseq analysis packages produces list. So, we need to learn how to handle lists.</span>
<span id="cb446-622"><a href="#cb446-622" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-623"><a href="#cb446-623" aria-hidden="true" tabindex="-1"></a>See these examples:</span>
<span id="cb446-624"><a href="#cb446-624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-625"><a href="#cb446-625" aria-hidden="true" tabindex="-1"></a>A. Storing BLAST results</span>
<span id="cb446-626"><a href="#cb446-626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-627"><a href="#cb446-627" aria-hidden="true" tabindex="-1"></a><span class="in">```{r list-ex1}</span></span>
<span id="cb446-628"><a href="#cb446-628" aria-hidden="true" tabindex="-1"></a>blast_hits <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb446-629"><a href="#cb446-629" aria-hidden="true" tabindex="-1"></a>  <span class="at">query_id =</span> <span class="st">"GeneX"</span>,</span>
<span id="cb446-630"><a href="#cb446-630" aria-hidden="true" tabindex="-1"></a>  <span class="at">hit_ids =</span> <span class="fu">c</span>(<span class="st">"NP_123"</span>, <span class="st">"NP_456"</span>),</span>
<span id="cb446-631"><a href="#cb446-631" aria-hidden="true" tabindex="-1"></a>  <span class="at">e_values =</span> <span class="fu">c</span>(<span class="fl">1e-50</span>, <span class="fl">3e-12</span>),</span>
<span id="cb446-632"><a href="#cb446-632" aria-hidden="true" tabindex="-1"></a>  <span class="at">alignment =</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="st">"ATG..."</span>, <span class="st">"CTA..."</span>), <span class="at">ncol=</span><span class="dv">1</span>))</span>
<span id="cb446-633"><a href="#cb446-633" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-634"><a href="#cb446-634" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-635"><a href="#cb446-635" aria-hidden="true" tabindex="-1"></a>B. Handling Mixed Data</span>
<span id="cb446-636"><a href="#cb446-636" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-637"><a href="#cb446-637" aria-hidden="true" tabindex="-1"></a><span class="in">```{r list-ex2}</span></span>
<span id="cb446-638"><a href="#cb446-638" aria-hidden="true" tabindex="-1"></a>patient_data <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb446-639"><a href="#cb446-639" aria-hidden="true" tabindex="-1"></a>  <span class="at">id =</span> <span class="st">"P1001"</span>,</span>
<span id="cb446-640"><a href="#cb446-640" aria-hidden="true" tabindex="-1"></a>  <span class="at">tests =</span> <span class="fu">data.frame</span>(</span>
<span id="cb446-641"><a href="#cb446-641" aria-hidden="true" tabindex="-1"></a>    <span class="at">test =</span> <span class="fu">c</span>(<span class="st">"WBC"</span>, <span class="st">"RBC"</span>),</span>
<span id="cb446-642"><a href="#cb446-642" aria-hidden="true" tabindex="-1"></a>    <span class="at">value =</span> <span class="fu">c</span>(<span class="fl">4.5</span>, <span class="fl">5.1</span>)</span>
<span id="cb446-643"><a href="#cb446-643" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb446-644"><a href="#cb446-644" aria-hidden="true" tabindex="-1"></a>  <span class="at">has_mutation =</span> <span class="cn">TRUE</span></span>
<span id="cb446-645"><a href="#cb446-645" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-646"><a href="#cb446-646" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-647"><a href="#cb446-647" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-648"><a href="#cb446-648" aria-hidden="true" tabindex="-1"></a>Common List Operations</span>
<span id="cb446-649"><a href="#cb446-649" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-650"><a href="#cb446-650" aria-hidden="true" tabindex="-1"></a><span class="in">```{r list-op}</span></span>
<span id="cb446-651"><a href="#cb446-651" aria-hidden="true" tabindex="-1"></a><span class="co"># Add new element</span></span>
<span id="cb446-652"><a href="#cb446-652" aria-hidden="true" tabindex="-1"></a>genomics_data<span class="sc">$</span>sequencer <span class="ot">&lt;-</span> <span class="st">"Illumina"</span></span>
<span id="cb446-653"><a href="#cb446-653" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-654"><a href="#cb446-654" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove element</span></span>
<span id="cb446-655"><a href="#cb446-655" aria-hidden="true" tabindex="-1"></a>genomics_data<span class="sc">$</span>is_cancer_gene <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb446-656"><a href="#cb446-656" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-657"><a href="#cb446-657" aria-hidden="true" tabindex="-1"></a><span class="co"># Check structure (critical for complex lists)</span></span>
<span id="cb446-658"><a href="#cb446-658" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(genomics_data)</span>
<span id="cb446-659"><a href="#cb446-659" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-660"><a href="#cb446-660" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-661"><a href="#cb446-661" aria-hidden="true" tabindex="-1"></a>By the way, how would you add more patients?</span>
<span id="cb446-662"><a href="#cb446-662" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-665"><a href="#cb446-665" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-666"><a href="#cb446-666" aria-hidden="true" tabindex="-1"></a><span class="co"># Add new patient</span></span>
<span id="cb446-667"><a href="#cb446-667" aria-hidden="true" tabindex="-1"></a>patient_data<span class="sc">$</span>P1002 <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb446-668"><a href="#cb446-668" aria-hidden="true" tabindex="-1"></a>  <span class="at">id =</span> <span class="st">"P1002"</span>,</span>
<span id="cb446-669"><a href="#cb446-669" aria-hidden="true" tabindex="-1"></a>  <span class="at">tests =</span> <span class="fu">data.frame</span>(</span>
<span id="cb446-670"><a href="#cb446-670" aria-hidden="true" tabindex="-1"></a>    <span class="at">test =</span> <span class="fu">c</span>(<span class="st">"WBC"</span>, <span class="st">"RBC"</span>, <span class="st">"Platelets"</span>),</span>
<span id="cb446-671"><a href="#cb446-671" aria-hidden="true" tabindex="-1"></a>    <span class="at">value =</span> <span class="fu">c</span>(<span class="fl">6.2</span>, <span class="fl">4.8</span>, <span class="dv">150</span>)</span>
<span id="cb446-672"><a href="#cb446-672" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb446-673"><a href="#cb446-673" aria-hidden="true" tabindex="-1"></a>  <span class="at">has_mutation =</span> <span class="cn">FALSE</span></span>
<span id="cb446-674"><a href="#cb446-674" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-675"><a href="#cb446-675" aria-hidden="true" tabindex="-1"></a><span class="co"># Access specific patient</span></span>
<span id="cb446-676"><a href="#cb446-676" aria-hidden="true" tabindex="-1"></a>patient_data<span class="sc">$</span>P1001<span class="sc">$</span>test</span>
<span id="cb446-677"><a href="#cb446-677" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-678"><a href="#cb446-678" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-679"><a href="#cb446-679" aria-hidden="true" tabindex="-1"></a>For Batch Processing:</span>
<span id="cb446-680"><a href="#cb446-680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-683"><a href="#cb446-683" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-684"><a href="#cb446-684" aria-hidden="true" tabindex="-1"></a>patients <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb446-685"><a href="#cb446-685" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb446-686"><a href="#cb446-686" aria-hidden="true" tabindex="-1"></a>    <span class="at">id =</span> <span class="st">"P1001"</span>,</span>
<span id="cb446-687"><a href="#cb446-687" aria-hidden="true" tabindex="-1"></a>    <span class="at">tests =</span> <span class="fu">data.frame</span>(<span class="at">test =</span> <span class="fu">c</span>(<span class="st">"WBC"</span>, <span class="st">"RBC"</span>), <span class="at">value =</span> <span class="fu">c</span>(<span class="fl">4.5</span>, <span class="fl">5.1</span>)),</span>
<span id="cb446-688"><a href="#cb446-688" aria-hidden="true" tabindex="-1"></a>    <span class="at">has_mutation =</span> <span class="cn">TRUE</span></span>
<span id="cb446-689"><a href="#cb446-689" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb446-690"><a href="#cb446-690" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb446-691"><a href="#cb446-691" aria-hidden="true" tabindex="-1"></a>    <span class="at">id =</span> <span class="st">"P1002"</span>,</span>
<span id="cb446-692"><a href="#cb446-692" aria-hidden="true" tabindex="-1"></a>    <span class="at">tests =</span> <span class="fu">data.frame</span>(<span class="at">test =</span> <span class="fu">c</span>(<span class="st">"WBC"</span>, <span class="st">"RBC"</span>, <span class="st">"Platelets"</span>), <span class="at">value =</span> <span class="fu">c</span>(<span class="fl">6.2</span>, <span class="fl">4.8</span>, <span class="dv">150</span>)),</span>
<span id="cb446-693"><a href="#cb446-693" aria-hidden="true" tabindex="-1"></a>    <span class="at">has_mutation =</span> <span class="cn">FALSE</span></span>
<span id="cb446-694"><a href="#cb446-694" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb446-695"><a href="#cb446-695" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-696"><a href="#cb446-696" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-697"><a href="#cb446-697" aria-hidden="true" tabindex="-1"></a><span class="co"># Access 2nd patient's WBC value</span></span>
<span id="cb446-698"><a href="#cb446-698" aria-hidden="true" tabindex="-1"></a>patients[[<span class="dv">2</span>]]<span class="sc">$</span>tests<span class="sc">$</span>value[patients[[<span class="dv">2</span>]]<span class="sc">$</span>tests<span class="sc">$</span>test <span class="sc">==</span> <span class="st">"WBC"</span>]</span>
<span id="cb446-699"><a href="#cb446-699" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-700"><a href="#cb446-700" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-701"><a href="#cb446-701" aria-hidden="true" tabindex="-1"></a>Converting Between Structures</span>
<span id="cb446-702"><a href="#cb446-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-705"><a href="#cb446-705" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-706"><a href="#cb446-706" aria-hidden="true" tabindex="-1"></a><span class="co"># List ‚Üí Vector</span></span>
<span id="cb446-707"><a href="#cb446-707" aria-hidden="true" tabindex="-1"></a><span class="fu">unlist</span>(genomics_data[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span>
<span id="cb446-708"><a href="#cb446-708" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-709"><a href="#cb446-709" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-710"><a href="#cb446-710" aria-hidden="true" tabindex="-1"></a>Visualization</span>
<span id="cb446-711"><a href="#cb446-711" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-712"><a href="#cb446-712" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb446-713"><a href="#cb446-713" aria-hidden="true" tabindex="-1"></a><span class="co"># Base R plot from list data</span></span>
<span id="cb446-714"><a href="#cb446-714" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">unlist</span>(genomics_data[<span class="dv">2</span>]),</span>
<span id="cb446-715"><a href="#cb446-715" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> genomics_data[[<span class="dv">1</span>]])</span>
<span id="cb446-716"><a href="#cb446-716" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-717"><a href="#cb446-717" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-718"><a href="#cb446-718" aria-hidden="true" tabindex="-1"></a>This code won't work if you run. <span class="in">`unlist(genomics_data[2]`</span> creates a vector of length 6 from our 3<span class="sc">\*</span>2 matrix but <span class="in">`genomics_data[[1]]`</span> has 3 things inside the <span class="in">`gene_names`</span> vector. Debug like this:</span>
<span id="cb446-719"><a href="#cb446-719" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-722"><a href="#cb446-722" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-723"><a href="#cb446-723" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(genomics_data<span class="sc">$</span>expression)  <span class="co"># e.g., 2 rows x 2 cols</span></span>
<span id="cb446-724"><a href="#cb446-724" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(genomics_data<span class="sc">$</span>gene_names) <span class="co"># e.g., 3 genes</span></span>
<span id="cb446-725"><a href="#cb446-725" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-726"><a href="#cb446-726" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-727"><a href="#cb446-727" aria-hidden="true" tabindex="-1"></a>A. Gene-Centric (Mean Expression)</span>
<span id="cb446-728"><a href="#cb446-728" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-731"><a href="#cb446-731" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-732"><a href="#cb446-732" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">rowMeans</span>(genomics_data<span class="sc">$</span>expression),</span>
<span id="cb446-733"><a href="#cb446-733" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> genomics_data<span class="sc">$</span>gene_names,</span>
<span id="cb446-734"><a href="#cb446-734" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="st">"steelblue"</span>,</span>
<span id="cb446-735"><a href="#cb446-735" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Mean Expression"</span>,</span>
<span id="cb446-736"><a href="#cb446-736" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Average Gene Expression"</span>)</span>
<span id="cb446-737"><a href="#cb446-737" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-738"><a href="#cb446-738" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-739"><a href="#cb446-739" aria-hidden="true" tabindex="-1"></a>B. Sample-Centric (All Measurements)</span>
<span id="cb446-740"><a href="#cb446-740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-743"><a href="#cb446-743" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-744"><a href="#cb446-744" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(genomics_data<span class="sc">$</span>expression,</span>
<span id="cb446-745"><a href="#cb446-745" aria-hidden="true" tabindex="-1"></a>        <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb446-746"><a href="#cb446-746" aria-hidden="true" tabindex="-1"></a>        <span class="at">names.arg =</span> <span class="fu">paste0</span>(<span class="st">"Sample_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(genomics_data<span class="sc">$</span>expression)),</span>
<span id="cb446-747"><a href="#cb446-747" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> genomics_data<span class="sc">$</span>gene_names,</span>
<span id="cb446-748"><a href="#cb446-748" aria-hidden="true" tabindex="-1"></a>        <span class="at">args.legend =</span> <span class="fu">list</span>(<span class="at">x =</span> <span class="st">"topright"</span>, <span class="at">bty =</span> <span class="st">"n"</span>),</span>
<span id="cb446-749"><a href="#cb446-749" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"red"</span>, <span class="st">"green"</span>),</span>
<span id="cb446-750"><a href="#cb446-750" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Expression Across Samples"</span>)</span>
<span id="cb446-751"><a href="#cb446-751" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-752"><a href="#cb446-752" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-753"><a href="#cb446-753" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-754"><a href="#cb446-754" aria-hidden="true" tabindex="-1"></a>This matches real-world scenarios:</span>
<span id="cb446-755"><a href="#cb446-755" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-756"><a href="#cb446-756" aria-hidden="true" tabindex="-1"></a>RNA-seq: Rows=genes, cols=samples</span>
<span id="cb446-757"><a href="#cb446-757" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-758"><a href="#cb446-758" aria-hidden="true" tabindex="-1"></a>rowMeans() = average expression per gene</span>
<span id="cb446-759"><a href="#cb446-759" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-760"><a href="#cb446-760" aria-hidden="true" tabindex="-1"></a>beside=TRUE =<span class="sc">\&gt;</span> compare samples within genes</span>
<span id="cb446-761"><a href="#cb446-761" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-762"><a href="#cb446-762" aria-hidden="true" tabindex="-1"></a>Proteomics: Rows=proteins, cols=replicates</span>
<span id="cb446-763"><a href="#cb446-763" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-764"><a href="#cb446-764" aria-hidden="true" tabindex="-1"></a>Same principles apply</span>
<span id="cb446-765"><a href="#cb446-765" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-766"><a href="#cb446-766" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-769"><a href="#cb446-769" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-770"><a href="#cb446-770" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate stats</span></span>
<span id="cb446-771"><a href="#cb446-771" aria-hidden="true" tabindex="-1"></a>gene_means <span class="ot">&lt;-</span> <span class="fu">rowMeans</span>(genomics_data<span class="sc">$</span>expression)</span>
<span id="cb446-772"><a href="#cb446-772" aria-hidden="true" tabindex="-1"></a>gene_sds <span class="ot">&lt;-</span> <span class="fu">apply</span>(genomics_data<span class="sc">$</span>expression, <span class="dv">1</span>, sd)</span>
<span id="cb446-773"><a href="#cb446-773" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-774"><a href="#cb446-774" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot with error bars</span></span>
<span id="cb446-775"><a href="#cb446-775" aria-hidden="true" tabindex="-1"></a>bp <span class="ot">&lt;-</span> <span class="fu">barplot</span>(gene_means, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fu">max</span>(gene_means <span class="sc">+</span> gene_sds)))</span>
<span id="cb446-776"><a href="#cb446-776" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(bp, gene_means <span class="sc">-</span> gene_sds, bp, gene_means <span class="sc">+</span> gene_sds, </span>
<span id="cb446-777"><a href="#cb446-777" aria-hidden="true" tabindex="-1"></a>       <span class="at">angle =</span> <span class="dv">90</span>, <span class="at">code =</span> <span class="dv">3</span>)</span>
<span id="cb446-778"><a href="#cb446-778" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-779"><a href="#cb446-779" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-780"><a href="#cb446-780" aria-hidden="true" tabindex="-1"></a>**Task: Create a list containing:**</span>
<span id="cb446-781"><a href="#cb446-781" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-782"><a href="#cb446-782" aria-hidden="true" tabindex="-1"></a>i)  A character vector of 3 gene names</span>
<span id="cb446-783"><a href="#cb446-783" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-784"><a href="#cb446-784" aria-hidden="true" tabindex="-1"></a>ii) A numeric matrix of expression values</span>
<span id="cb446-785"><a href="#cb446-785" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-786"><a href="#cb446-786" aria-hidden="true" tabindex="-1"></a>iii) A logical vector indicating pathway membership</span>
<span id="cb446-787"><a href="#cb446-787" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-788"><a href="#cb446-788" aria-hidden="true" tabindex="-1"></a>iv) A nested list with lab metadata</span>
<span id="cb446-789"><a href="#cb446-789" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-790"><a href="#cb446-790" aria-hidden="true" tabindex="-1"></a><span class="fu">## üè° Homeworks: Matrix and List Operations</span></span>
<span id="cb446-791"><a href="#cb446-791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-792"><a href="#cb446-792" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß™ Protein Quantification in Biological Samples</span></span>
<span id="cb446-793"><a href="#cb446-793" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-794"><a href="#cb446-794" aria-hidden="true" tabindex="-1"></a>You are given the following protein concentration matrix:</span>
<span id="cb446-795"><a href="#cb446-795" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-796"><a href="#cb446-796" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-797"><a href="#cb446-797" aria-hidden="true" tabindex="-1"></a>\text{ProteinMatrix} =</span>
<span id="cb446-798"><a href="#cb446-798" aria-hidden="true" tabindex="-1"></a>\begin{bmatrix}</span>
<span id="cb446-799"><a href="#cb446-799" aria-hidden="true" tabindex="-1"></a>  5 &amp; 3 &amp; 2 <span class="sc">\\</span></span>
<span id="cb446-800"><a href="#cb446-800" aria-hidden="true" tabindex="-1"></a>  7 &amp; 6 &amp; 4 <span class="sc">\\</span></span>
<span id="cb446-801"><a href="#cb446-801" aria-hidden="true" tabindex="-1"></a>\end{bmatrix}</span>
<span id="cb446-802"><a href="#cb446-802" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-803"><a href="#cb446-803" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-804"><a href="#cb446-804" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Rows represent samples:</span>
<span id="cb446-805"><a href="#cb446-805" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Sample1</span>
<span id="cb446-806"><a href="#cb446-806" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Sample2</span>
<span id="cb446-807"><a href="#cb446-807" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Columns represent proteins:</span>
<span id="cb446-808"><a href="#cb446-808" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>ProteinX</span>
<span id="cb446-809"><a href="#cb446-809" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>ProteinY</span>
<span id="cb446-810"><a href="#cb446-810" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>ProteinZ</span>
<span id="cb446-811"><a href="#cb446-811" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-812"><a href="#cb446-812" aria-hidden="true" tabindex="-1"></a>You are also given a weight (importance) matrix for the proteins:</span>
<span id="cb446-813"><a href="#cb446-813" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-814"><a href="#cb446-814" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-815"><a href="#cb446-815" aria-hidden="true" tabindex="-1"></a>\text{WeightVector} =</span>
<span id="cb446-816"><a href="#cb446-816" aria-hidden="true" tabindex="-1"></a>\begin{bmatrix}</span>
<span id="cb446-817"><a href="#cb446-817" aria-hidden="true" tabindex="-1"></a>  0.5 <span class="sc">\\</span></span>
<span id="cb446-818"><a href="#cb446-818" aria-hidden="true" tabindex="-1"></a>  1.0 <span class="sc">\\</span></span>
<span id="cb446-819"><a href="#cb446-819" aria-hidden="true" tabindex="-1"></a>  1.5 <span class="sc">\\</span></span>
<span id="cb446-820"><a href="#cb446-820" aria-hidden="true" tabindex="-1"></a>\end{bmatrix}</span>
<span id="cb446-821"><a href="#cb446-821" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-822"><a href="#cb446-822" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-823"><a href="#cb446-823" aria-hidden="true" tabindex="-1"></a><span class="fu">### üîß Tasks</span></span>
<span id="cb446-824"><a href="#cb446-824" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-825"><a href="#cb446-825" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Make the matrices (with exact names) and multiply the ProteinMatrix by the WeightVector.</span>
<span id="cb446-826"><a href="#cb446-826" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-827"><a href="#cb446-827" aria-hidden="true" tabindex="-1"></a>That is:</span>
<span id="cb446-828"><a href="#cb446-828" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-829"><a href="#cb446-829" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-830"><a href="#cb446-830" aria-hidden="true" tabindex="-1"></a>  \text{ProteinMatrix} \times \text{WeightVector}</span>
<span id="cb446-831"><a href="#cb446-831" aria-hidden="true" tabindex="-1"></a>  $$</span>
<span id="cb446-832"><a href="#cb446-832" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-833"><a href="#cb446-833" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Transpose the ProteinMatrix and show what it looks like.</span>
<span id="cb446-834"><a href="#cb446-834" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-835"><a href="#cb446-835" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Create the Identity matrix of compatible size and show what happens when you multiply: $$</span>
<span id="cb446-836"><a href="#cb446-836" aria-hidden="true" tabindex="-1"></a>     \text{ProteinMatrix} \times I</span>
<span id="cb446-837"><a href="#cb446-837" aria-hidden="true" tabindex="-1"></a>     $$</span>
<span id="cb446-838"><a href="#cb446-838" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-839"><a href="#cb446-839" aria-hidden="true" tabindex="-1"></a>4.Do the calculations (rowSums, colSums, etc.) and visualization (barplot, heatmap) as shown in the class.</span>
<span id="cb446-840"><a href="#cb446-840" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-841"><a href="#cb446-841" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation Questions</span></span>
<span id="cb446-842"><a href="#cb446-842" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-843"><a href="#cb446-843" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What does multiplying the protein levels by the weight vector mean biologically?</span>
<span id="cb446-844"><a href="#cb446-844" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What does the result tell you about total protein burden (or total protein impact) for each sample?</span>
<span id="cb446-845"><a href="#cb446-845" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What do the identity matrix represent in the context of protein interactions or measurement biases?</span>
<span id="cb446-846"><a href="#cb446-846" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>If you changed the weight of ProteinZ to 3.0, how would the result change?</span>
<span id="cb446-847"><a href="#cb446-847" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-848"><a href="#cb446-848" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß¨ Gene-to-Protein Translation</span></span>
<span id="cb446-849"><a href="#cb446-849" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-850"><a href="#cb446-850" aria-hidden="true" tabindex="-1"></a>You are given the following matrix representing normalized gene expression levels (e.g., TPM): $$</span>
<span id="cb446-851"><a href="#cb446-851" aria-hidden="true" tabindex="-1"></a>\text{GeneExpression} =</span>
<span id="cb446-852"><a href="#cb446-852" aria-hidden="true" tabindex="-1"></a>\begin{bmatrix}</span>
<span id="cb446-853"><a href="#cb446-853" aria-hidden="true" tabindex="-1"></a>  10 &amp;  8 &amp;  5 <span class="sc">\\</span></span>
<span id="cb446-854"><a href="#cb446-854" aria-hidden="true" tabindex="-1"></a>  15 &amp; 12 &amp; 10 <span class="sc">\\</span></span>
<span id="cb446-855"><a href="#cb446-855" aria-hidden="true" tabindex="-1"></a>\end{bmatrix}</span>
<span id="cb446-856"><a href="#cb446-856" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-857"><a href="#cb446-857" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-858"><a href="#cb446-858" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Rows = Samples:</span>
<span id="cb446-859"><a href="#cb446-859" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Sample1</span>
<span id="cb446-860"><a href="#cb446-860" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Sample2</span>
<span id="cb446-861"><a href="#cb446-861" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Columns = Genes:</span>
<span id="cb446-862"><a href="#cb446-862" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>GeneA</span>
<span id="cb446-863"><a href="#cb446-863" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>GeneB</span>
<span id="cb446-864"><a href="#cb446-864" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>GeneC</span>
<span id="cb446-865"><a href="#cb446-865" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-866"><a href="#cb446-866" aria-hidden="true" tabindex="-1"></a>Each gene translates into proteins with a certain efficiency. The efficiency of translation from each gene to its corresponding protein is given by the following diagonal matrix:</span>
<span id="cb446-867"><a href="#cb446-867" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-868"><a href="#cb446-868" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-869"><a href="#cb446-869" aria-hidden="true" tabindex="-1"></a>\text{TranslationMatrix} =</span>
<span id="cb446-870"><a href="#cb446-870" aria-hidden="true" tabindex="-1"></a>\begin{bmatrix}</span>
<span id="cb446-871"><a href="#cb446-871" aria-hidden="true" tabindex="-1"></a>  1.5 &amp; 0   &amp; 0 <span class="sc">\\</span></span>
<span id="cb446-872"><a href="#cb446-872" aria-hidden="true" tabindex="-1"></a>  0   &amp; 1.2 &amp; 0 <span class="sc">\\</span></span>
<span id="cb446-873"><a href="#cb446-873" aria-hidden="true" tabindex="-1"></a>  0   &amp; 0   &amp; 1.8 <span class="sc">\\</span></span>
<span id="cb446-874"><a href="#cb446-874" aria-hidden="true" tabindex="-1"></a>\end{bmatrix}</span>
<span id="cb446-875"><a href="#cb446-875" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-876"><a href="#cb446-876" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-877"><a href="#cb446-877" aria-hidden="true" tabindex="-1"></a>This means:</span>
<span id="cb446-878"><a href="#cb446-878" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-879"><a href="#cb446-879" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>GeneA ‚Üí ProteinA with 1.5√ó efficiency\</span>
<span id="cb446-880"><a href="#cb446-880" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>GeneB ‚Üí ProteinB with 1.2√ó efficiency\</span>
<span id="cb446-881"><a href="#cb446-881" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>GeneC ‚Üí ProteinC with 1.8√ó efficiency</span>
<span id="cb446-882"><a href="#cb446-882" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-883"><a href="#cb446-883" aria-hidden="true" tabindex="-1"></a><span class="fu">### üîß Tasks</span></span>
<span id="cb446-884"><a href="#cb446-884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-885"><a href="#cb446-885" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Make the matrices and multiply GeneExpression √ó TranslationMatrix to compute the resulting ProteinMatrix.</span>
<span id="cb446-886"><a href="#cb446-886" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-887"><a href="#cb446-887" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Show the result step-by-step.</span>
<span id="cb446-888"><a href="#cb446-888" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-889"><a href="#cb446-889" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Transpose the GeneExpression matrix. What does this new matrix represent?</span>
<span id="cb446-890"><a href="#cb446-890" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-891"><a href="#cb446-891" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Create the Identity matrix I‚ÇÉ and multiply it with the TranslationMatrix. What happens?</span>
<span id="cb446-892"><a href="#cb446-892" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-893"><a href="#cb446-893" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Create a new matrix containing only the expression of GeneA and GeneB across both samples. Call this submatrix A.</span>
<span id="cb446-894"><a href="#cb446-894" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-895"><a href="#cb446-895" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Compute the inverse A‚Åª¬π using <span class="in">`solve()`</span> function.</span>
<span id="cb446-896"><a href="#cb446-896" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Then verify:\</span>
<span id="cb446-897"><a href="#cb446-897" aria-hidden="true" tabindex="-1"></a>    A √ó A‚Åª¬π = I‚ÇÇ</span>
<span id="cb446-898"><a href="#cb446-898" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-899"><a href="#cb446-899" aria-hidden="true" tabindex="-1"></a><span class="fu">### üìä Visualization Tasks</span></span>
<span id="cb446-900"><a href="#cb446-900" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-901"><a href="#cb446-901" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Plot a MARplot-style scatter plot:</span>
<span id="cb446-902"><a href="#cb446-902" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-903"><a href="#cb446-903" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>x-axis: Gene expression values (GeneExpression matrix, flattened)</span>
<span id="cb446-904"><a href="#cb446-904" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>y-axis: Corresponding Protein values (ProteinMatrix, flattened)</span>
<span id="cb446-905"><a href="#cb446-905" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Label each point as "Sample-Gene"</span>
<span id="cb446-906"><a href="#cb446-906" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-907"><a href="#cb446-907" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>Generate a heatmap of the ProteinMatrix using R‚Äôs <span class="in">`heatmap()`</span> function.</span>
<span id="cb446-908"><a href="#cb446-908" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-909"><a href="#cb446-909" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Add meaningful row and column labels.</span>
<span id="cb446-910"><a href="#cb446-910" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Enable clustering by rows and columns.</span>
<span id="cb446-911"><a href="#cb446-911" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-912"><a href="#cb446-912" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation Questions</span></span>
<span id="cb446-913"><a href="#cb446-913" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-914"><a href="#cb446-914" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What does matrix multiplication represent biologically in this case?</span>
<span id="cb446-915"><a href="#cb446-915" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Why does the diagonal TranslationMatrix make sense biologically?</span>
<span id="cb446-916"><a href="#cb446-916" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What does it mean if Sample2 has higher protein levels even with similar gene expression?</span>
<span id="cb446-917"><a href="#cb446-917" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How does the MARplot help interpret translation efficiency?</span>
<span id="cb446-918"><a href="#cb446-918" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How does clustering in the heatmap reveal relationships between samples and proteins?</span>
<span id="cb446-919"><a href="#cb446-919" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-920"><a href="#cb446-920" aria-hidden="true" tabindex="-1"></a><span class="fu">### üêÑ Animal Breeding ‚Äì Economic Ranking of Bulls by Traits</span></span>
<span id="cb446-921"><a href="#cb446-921" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-922"><a href="#cb446-922" aria-hidden="true" tabindex="-1"></a>You are evaluating two bulls for use in a dairy breeding program. Their Estimated Breeding Values (EBVs) are:</span>
<span id="cb446-923"><a href="#cb446-923" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-924"><a href="#cb446-924" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-925"><a href="#cb446-925" aria-hidden="true" tabindex="-1"></a>\text{BullEBVs} =</span>
<span id="cb446-926"><a href="#cb446-926" aria-hidden="true" tabindex="-1"></a>\begin{bmatrix}</span>
<span id="cb446-927"><a href="#cb446-927" aria-hidden="true" tabindex="-1"></a>  400 &amp; 1.2 &amp; 0.8 <span class="sc">\\\\</span></span>
<span id="cb446-928"><a href="#cb446-928" aria-hidden="true" tabindex="-1"></a>  500 &amp; 1.5 &amp; 0.6 <span class="sc">\\\\</span></span>
<span id="cb446-929"><a href="#cb446-929" aria-hidden="true" tabindex="-1"></a>\end{bmatrix}</span>
<span id="cb446-930"><a href="#cb446-930" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-931"><a href="#cb446-931" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-932"><a href="#cb446-932" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Rows:</span>
<span id="cb446-933"><a href="#cb446-933" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Bull1\</span>
<span id="cb446-934"><a href="#cb446-934" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Bull2</span>
<span id="cb446-935"><a href="#cb446-935" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Columns:</span>
<span id="cb446-936"><a href="#cb446-936" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Trait1 = Milk yield (liters/year)\</span>
<span id="cb446-937"><a href="#cb446-937" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Trait2 = Growth rate (kg/day)\</span>
<span id="cb446-938"><a href="#cb446-938" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>Trait3 = Fertility (calving interval adjustment)</span>
<span id="cb446-939"><a href="#cb446-939" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-940"><a href="#cb446-940" aria-hidden="true" tabindex="-1"></a>You assign economic weights to each trait:</span>
<span id="cb446-941"><a href="#cb446-941" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-942"><a href="#cb446-942" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-943"><a href="#cb446-943" aria-hidden="true" tabindex="-1"></a>\text{EconomicWeights} =</span>
<span id="cb446-944"><a href="#cb446-944" aria-hidden="true" tabindex="-1"></a>\begin{bmatrix}</span>
<span id="cb446-945"><a href="#cb446-945" aria-hidden="true" tabindex="-1"></a>  0.002 <span class="sc">\\\\</span></span>
<span id="cb446-946"><a href="#cb446-946" aria-hidden="true" tabindex="-1"></a>  50 <span class="sc">\\\\</span></span>
<span id="cb446-947"><a href="#cb446-947" aria-hidden="true" tabindex="-1"></a>  100 <span class="sc">\\\\</span></span>
<span id="cb446-948"><a href="#cb446-948" aria-hidden="true" tabindex="-1"></a>\end{bmatrix}</span>
<span id="cb446-949"><a href="#cb446-949" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-950"><a href="#cb446-950" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-951"><a href="#cb446-951" aria-hidden="true" tabindex="-1"></a><span class="fu">### üîß Tasks</span></span>
<span id="cb446-952"><a href="#cb446-952" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-953"><a href="#cb446-953" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Compute:</span>
<span id="cb446-954"><a href="#cb446-954" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-955"><a href="#cb446-955" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-956"><a href="#cb446-956" aria-hidden="true" tabindex="-1"></a>  \text{TotalValue} = \text{BullEBVs} \times \text{EconomicWeights}</span>
<span id="cb446-957"><a href="#cb446-957" aria-hidden="true" tabindex="-1"></a>  $$</span>
<span id="cb446-958"><a href="#cb446-958" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-959"><a href="#cb446-959" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What are the resulting values?</span>
<span id="cb446-960"><a href="#cb446-960" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Which bull is more valuable economically?</span>
<span id="cb446-961"><a href="#cb446-961" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-962"><a href="#cb446-962" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Interpret what multiplying by the economic weights means biologically.</span>
<span id="cb446-963"><a href="#cb446-963" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-964"><a href="#cb446-964" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Create the 3√ó3 identity matrix I‚ÇÉ and multiply it with BullEBVs.</span>
<span id="cb446-965"><a href="#cb446-965" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-966"><a href="#cb446-966" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What does it return?</span>
<span id="cb446-967"><a href="#cb446-967" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What does the identity matrix mean in this case?</span>
<span id="cb446-968"><a href="#cb446-968" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-969"><a href="#cb446-969" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Subset the BullEBVs matrix to remove Trait1 (milk yield) and recalculate TotalValue.</span>
<span id="cb446-970"><a href="#cb446-970" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-971"><a href="#cb446-971" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How does this change the ranking?</span>
<span id="cb446-972"><a href="#cb446-972" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-973"><a href="#cb446-973" aria-hidden="true" tabindex="-1"></a><span class="fu">### üìä Visualization Tasks</span></span>
<span id="cb446-974"><a href="#cb446-974" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-975"><a href="#cb446-975" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Create a bar plot comparing TotalValue for Bull1 and Bull2.</span>
<span id="cb446-976"><a href="#cb446-976" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-977"><a href="#cb446-977" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>Create a heatmap of the EBVs.</span>
<span id="cb446-978"><a href="#cb446-978" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-979"><a href="#cb446-979" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Label rows and columns.</span>
<span id="cb446-980"><a href="#cb446-980" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Enable clustering.</span>
<span id="cb446-981"><a href="#cb446-981" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-982"><a href="#cb446-982" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation Questions</span></span>
<span id="cb446-983"><a href="#cb446-983" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-984"><a href="#cb446-984" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How do economic weights affect trait importance?</span>
<span id="cb446-985"><a href="#cb446-985" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Why might you ignore milk yield in some breeding programs?</span>
<span id="cb446-986"><a href="#cb446-986" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What is the value of heatmaps in visualizing multivariate trait data?</span>
<span id="cb446-987"><a href="#cb446-987" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Can this method be extended to more bulls and more traits?</span>
<span id="cb446-988"><a href="#cb446-988" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-989"><a href="#cb446-989" aria-hidden="true" tabindex="-1"></a><span class="fu">### üåæ Plant Breeding ‚Äì Trait Contributions from Parental Lines</span></span>
<span id="cb446-990"><a href="#cb446-990" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-991"><a href="#cb446-991" aria-hidden="true" tabindex="-1"></a>You are breeding a new rice variety from three parental lines. The key traits are:</span>
<span id="cb446-992"><a href="#cb446-992" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-993"><a href="#cb446-993" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>T1 = Drought resistance</span>
<span id="cb446-994"><a href="#cb446-994" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>T2 = Yield</span>
<span id="cb446-995"><a href="#cb446-995" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>T3 = Maturation time</span>
<span id="cb446-996"><a href="#cb446-996" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-997"><a href="#cb446-997" aria-hidden="true" tabindex="-1"></a>The following trait values (normalized 1‚Äì10) have been measured:</span>
<span id="cb446-998"><a href="#cb446-998" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-999"><a href="#cb446-999" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-1000"><a href="#cb446-1000" aria-hidden="true" tabindex="-1"></a>\text{ParentTraits} =</span>
<span id="cb446-1001"><a href="#cb446-1001" aria-hidden="true" tabindex="-1"></a>\begin{bmatrix}</span>
<span id="cb446-1002"><a href="#cb446-1002" aria-hidden="true" tabindex="-1"></a>  7 &amp; 5 &amp; 3 <span class="sc">\\\\</span></span>
<span id="cb446-1003"><a href="#cb446-1003" aria-hidden="true" tabindex="-1"></a>  6 &amp; 8 &amp; 4 <span class="sc">\\\\</span></span>
<span id="cb446-1004"><a href="#cb446-1004" aria-hidden="true" tabindex="-1"></a>  5 &amp; 6 &amp; 6 <span class="sc">\\\\</span></span>
<span id="cb446-1005"><a href="#cb446-1005" aria-hidden="true" tabindex="-1"></a>\end{bmatrix}</span>
<span id="cb446-1006"><a href="#cb446-1006" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-1007"><a href="#cb446-1007" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1008"><a href="#cb446-1008" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Rows:</span>
<span id="cb446-1009"><a href="#cb446-1009" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>P1 (Parent 1)</span>
<span id="cb446-1010"><a href="#cb446-1010" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>P2 (Parent 2)</span>
<span id="cb446-1011"><a href="#cb446-1011" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>P3 (Parent 3)</span>
<span id="cb446-1012"><a href="#cb446-1012" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Columns:</span>
<span id="cb446-1013"><a href="#cb446-1013" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>T1 = Drought resistance</span>
<span id="cb446-1014"><a href="#cb446-1014" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>T2 = Yield</span>
<span id="cb446-1015"><a href="#cb446-1015" aria-hidden="true" tabindex="-1"></a><span class="ss">    -   </span>T3 = Maturation time</span>
<span id="cb446-1016"><a href="#cb446-1016" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1017"><a href="#cb446-1017" aria-hidden="true" tabindex="-1"></a>You design a hybrid with contributions from each parent as follows:</span>
<span id="cb446-1018"><a href="#cb446-1018" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1019"><a href="#cb446-1019" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-1020"><a href="#cb446-1020" aria-hidden="true" tabindex="-1"></a>\text{HybridWeights} =</span>
<span id="cb446-1021"><a href="#cb446-1021" aria-hidden="true" tabindex="-1"></a>\begin{bmatrix}</span>
<span id="cb446-1022"><a href="#cb446-1022" aria-hidden="true" tabindex="-1"></a>  0.5 <span class="sc">\\\\</span></span>
<span id="cb446-1023"><a href="#cb446-1023" aria-hidden="true" tabindex="-1"></a>  0.3 <span class="sc">\\\\</span></span>
<span id="cb446-1024"><a href="#cb446-1024" aria-hidden="true" tabindex="-1"></a>  0.2 <span class="sc">\\\\</span></span>
<span id="cb446-1025"><a href="#cb446-1025" aria-hidden="true" tabindex="-1"></a>\end{bmatrix}</span>
<span id="cb446-1026"><a href="#cb446-1026" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-1027"><a href="#cb446-1027" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1028"><a href="#cb446-1028" aria-hidden="true" tabindex="-1"></a><span class="fu">### üîß Tasks</span></span>
<span id="cb446-1029"><a href="#cb446-1029" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1030"><a href="#cb446-1030" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Compute the HybridTrait vector by:</span>
<span id="cb446-1031"><a href="#cb446-1031" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1032"><a href="#cb446-1032" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb446-1033"><a href="#cb446-1033" aria-hidden="true" tabindex="-1"></a>  \text{HybridTraits} = \text{HybridWeights}^T \times \text{ParentTraits}</span>
<span id="cb446-1034"><a href="#cb446-1034" aria-hidden="true" tabindex="-1"></a>  $$</span>
<span id="cb446-1035"><a href="#cb446-1035" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1036"><a href="#cb446-1036" aria-hidden="true" tabindex="-1"></a>Show the steps and result.</span>
<span id="cb446-1037"><a href="#cb446-1037" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1038"><a href="#cb446-1038" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Explain what it means biologically when one parent contributes more to a particular trait.</span>
<span id="cb446-1039"><a href="#cb446-1039" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1040"><a href="#cb446-1040" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Create an identity matrix I‚ÇÉ and multiply it with ParentTraits.</span>
<span id="cb446-1041"><a href="#cb446-1041" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1042"><a href="#cb446-1042" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What do you observe?</span>
<span id="cb446-1043"><a href="#cb446-1043" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What does I √ó ParentTraits represent?</span>
<span id="cb446-1044"><a href="#cb446-1044" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1045"><a href="#cb446-1045" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Subset the ParentTraits matrix to include only T1 and T2. Recalculate the hybrid traits.</span>
<span id="cb446-1046"><a href="#cb446-1046" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1047"><a href="#cb446-1047" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Discuss how removing a trait affects your outcome.</span>
<span id="cb446-1048"><a href="#cb446-1048" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1049"><a href="#cb446-1049" aria-hidden="true" tabindex="-1"></a><span class="fu">### üìä Visualization Tasks</span></span>
<span id="cb446-1050"><a href="#cb446-1050" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1051"><a href="#cb446-1051" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Generate a heatmap of the ParentTraits matrix.</span>
<span id="cb446-1052"><a href="#cb446-1052" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1053"><a href="#cb446-1053" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Label the rows with parent names and columns with trait names.</span>
<span id="cb446-1054"><a href="#cb446-1054" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Enable row/column clustering.</span>
<span id="cb446-1055"><a href="#cb446-1055" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1056"><a href="#cb446-1056" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>Create a bar plot showing the HybridTraits (T1, T2, T3).</span>
<span id="cb446-1057"><a href="#cb446-1057" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1058"><a href="#cb446-1058" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Color code each bar by trait.</span>
<span id="cb446-1059"><a href="#cb446-1059" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What trait contributes most?</span>
<span id="cb446-1060"><a href="#cb446-1060" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1061"><a href="#cb446-1061" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation Questions</span></span>
<span id="cb446-1062"><a href="#cb446-1062" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1063"><a href="#cb446-1063" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How does the weighting of parents affect the hybrid's performance?</span>
<span id="cb446-1064"><a href="#cb446-1064" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What does the identity matrix represent here?</span>
<span id="cb446-1065"><a href="#cb446-1065" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>If you used equal weights (‚Öì for each), how would the hybrid traits change?</span>
<span id="cb446-1066"><a href="#cb446-1066" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What real-world limitations does this simplified model ignore?</span>
<span id="cb446-1067"><a href="#cb446-1067" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1068"><a href="#cb446-1068" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Managing Matrices and Weight Vectors Using Lists in R</span></span>
<span id="cb446-1069"><a href="#cb446-1069" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1070"><a href="#cb446-1070" aria-hidden="true" tabindex="-1"></a>Now that you have completed four biological matrix problems ‚Äî Protein concentration, gene-to-protein mapping, bull breeding value ranking, and plant trait combinations ‚Äî it‚Äôs time to organize your data and weights using R‚Äôs list structure.</span>
<span id="cb446-1071"><a href="#cb446-1071" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1072"><a href="#cb446-1072" aria-hidden="true" tabindex="-1"></a>In this task, you will:</span>
<span id="cb446-1073"><a href="#cb446-1073" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1074"><a href="#cb446-1074" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Group each example‚Äôs matrix and its corresponding weight vector inside a named list.</span>
<span id="cb446-1075"><a href="#cb446-1075" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Combine these named lists into a larger list called bioList.</span>
<span id="cb446-1076"><a href="#cb446-1076" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Use list indexing to repeat your earlier calculations and visualizations.</span>
<span id="cb446-1077"><a href="#cb446-1077" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Reflect on the benefits and challenges of using structured data objects.</span>
<span id="cb446-1078"><a href="#cb446-1078" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1079"><a href="#cb446-1079" aria-hidden="true" tabindex="-1"></a><span class="fu">### üì¶ Step 1: Create a master list</span></span>
<span id="cb446-1080"><a href="#cb446-1080" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1081"><a href="#cb446-1081" aria-hidden="true" tabindex="-1"></a>You should now build a named list called bioList containing the following four elements:</span>
<span id="cb446-1082"><a href="#cb446-1082" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1083"><a href="#cb446-1083" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>ProteinConc = list(matrix = ProteinMatrix, weights = WeightVector)</span>
<span id="cb446-1084"><a href="#cb446-1084" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>ProteinMap = list(matrix = ProteinMapping, weights = TranslationWeights)</span>
<span id="cb446-1085"><a href="#cb446-1085" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Plant = list(matrix = ParentTraits, weights = HybridWeights)</span>
<span id="cb446-1086"><a href="#cb446-1086" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Animal = list(matrix = BullEBVs, weights = EconomicWeights)</span>
<span id="cb446-1087"><a href="#cb446-1087" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1088"><a href="#cb446-1088" aria-hidden="true" tabindex="-1"></a>Hint: Each inner list should contain both:</span>
<span id="cb446-1089"><a href="#cb446-1089" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1090"><a href="#cb446-1090" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>matrix = the main data matrix</span>
<span id="cb446-1091"><a href="#cb446-1091" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>weights = the vector used for multiplication</span>
<span id="cb446-1092"><a href="#cb446-1092" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1093"><a href="#cb446-1093" aria-hidden="true" tabindex="-1"></a>No R code is required here (You have them from previous part, use inside same rmd/notebook file) ‚Äî just structure your data like this in your workspace.</span>
<span id="cb446-1094"><a href="#cb446-1094" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1095"><a href="#cb446-1095" aria-hidden="true" tabindex="-1"></a><span class="fu">### üîß Tasks</span></span>
<span id="cb446-1096"><a href="#cb446-1096" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1097"><a href="#cb446-1097" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>List the full names of each component in bioList. What are the names of the top-level and nested components?</span>
<span id="cb446-1098"><a href="#cb446-1098" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1099"><a href="#cb446-1099" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Access each matrix and its corresponding weights using list indexing.</span>
<span id="cb446-1100"><a href="#cb446-1100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1101"><a href="#cb446-1101" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How would you extract only the matrix of the Plant entry?</span>
<span id="cb446-1102"><a href="#cb446-1102" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How would you extract the weights for the Protein concentration entry?</span>
<span id="cb446-1103"><a href="#cb446-1103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1104"><a href="#cb446-1104" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Use the correct matrix and weights to perform:</span>
<span id="cb446-1105"><a href="#cb446-1105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1106"><a href="#cb446-1106" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>ProteinConc: Weighted gene expression score</span>
<span id="cb446-1107"><a href="#cb446-1107" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>ProteinMap: Contribution of transcripts to each protein</span>
<span id="cb446-1108"><a href="#cb446-1108" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Plant: Hybrid trait values</span>
<span id="cb446-1109"><a href="#cb446-1109" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Animal: Bull total economic value</span>
<span id="cb446-1110"><a href="#cb446-1110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1111"><a href="#cb446-1111" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Subset one matrix in each sublist (e.g., drop a trait or feature) and repeat the weighted calculation.</span>
<span id="cb446-1112"><a href="#cb446-1112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1113"><a href="#cb446-1113" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What changes in the results?</span>
<span id="cb446-1114"><a href="#cb446-1114" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Which traits/genes have the strongest influence?</span>
<span id="cb446-1115"><a href="#cb446-1115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1116"><a href="#cb446-1116" aria-hidden="true" tabindex="-1"></a><span class="fu">### üìä Visualization Tasks</span></span>
<span id="cb446-1117"><a href="#cb446-1117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1118"><a href="#cb446-1118" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Generate one heatmap for any matrix stored in bioList.</span>
<span id="cb446-1119"><a href="#cb446-1119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1120"><a href="#cb446-1120" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Choose one (e.g., ProteinMap or Plant)</span>
<span id="cb446-1121"><a href="#cb446-1121" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Apply clustering to rows and/or columns</span>
<span id="cb446-1122"><a href="#cb446-1122" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Label appropriately</span>
<span id="cb446-1123"><a href="#cb446-1123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1124"><a href="#cb446-1124" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>Generate two bar plots:</span>
<span id="cb446-1125"><a href="#cb446-1125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1126"><a href="#cb446-1126" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>One showing the result of weighted trait aggregation for the Plant hybrid</span>
<span id="cb446-1127"><a href="#cb446-1127" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>One showing the total breeding values for each bull</span>
<span id="cb446-1128"><a href="#cb446-1128" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1129"><a href="#cb446-1129" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation Questions</span></span>
<span id="cb446-1130"><a href="#cb446-1130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1131"><a href="#cb446-1131" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How does structuring your data using a list help with clarity and reproducibility?</span>
<span id="cb446-1132"><a href="#cb446-1132" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What risks or challenges might occur when accessing elements from nested lists?</span>
<span id="cb446-1133"><a href="#cb446-1133" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Could this structure be scaled for real datasets with many samples or traits?</span>
<span id="cb446-1134"><a href="#cb446-1134" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How would you loop over all elements in bioList to apply the same function?</span>
<span id="cb446-1135"><a href="#cb446-1135" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How can this list structure be useful for building automated bioinformatics pipelines?</span>
<span id="cb446-1136"><a href="#cb446-1136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1137"><a href="#cb446-1137" aria-hidden="true" tabindex="-1"></a>üìù Your Rmarkdown file(s) should include:</span>
<span id="cb446-1138"><a href="#cb446-1138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1139"><a href="#cb446-1139" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>All matrix calculations (tasks). Also, name the rows and columns of each matrix accordingly.</span>
<span id="cb446-1140"><a href="#cb446-1140" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>All interpretation answers</span>
<span id="cb446-1141"><a href="#cb446-1141" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>All plots (output from embedded code)</span>
<span id="cb446-1142"><a href="#cb446-1142" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>And your commentary blocks for each code chunk</span>
<span id="cb446-1143"><a href="#cb446-1143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1144"><a href="#cb446-1144" aria-hidden="true" tabindex="-1"></a>**knit your rmd (or Notebook) file as html/pdf file and push both the rmd (or Notebook) and html/pdf files**</span>
<span id="cb446-1145"><a href="#cb446-1145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1146"><a href="#cb446-1146" aria-hidden="true" tabindex="-1"></a><span class="fu">## Factor Variables</span></span>
<span id="cb446-1147"><a href="#cb446-1147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1148"><a href="#cb446-1148" aria-hidden="true" tabindex="-1"></a>Important for categorical data</span>
<span id="cb446-1149"><a href="#cb446-1149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1150"><a href="#cb446-1150" aria-hidden="true" tabindex="-1"></a><span class="fu">### Creating Factors</span></span>
<span id="cb446-1151"><a href="#cb446-1151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1152"><a href="#cb446-1152" aria-hidden="true" tabindex="-1"></a>Factors are used to represent categorical data in R. They are particularly important for biological data like genotypes, phenotypes, and experimental conditions.</span>
<span id="cb446-1153"><a href="#cb446-1153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1154"><a href="#cb446-1154" aria-hidden="true" tabindex="-1"></a><span class="in">```{r creating-factors}</span></span>
<span id="cb446-1155"><a href="#cb446-1155" aria-hidden="true" tabindex="-1"></a><span class="co"># Simple factor: DNA sample origins</span></span>
<span id="cb446-1156"><a href="#cb446-1156" aria-hidden="true" tabindex="-1"></a>origins <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Human"</span>, <span class="st">"Mouse"</span>, <span class="st">"Human"</span>, <span class="st">"Zebrafish"</span>, <span class="st">"Mouse"</span>, <span class="st">"Human"</span>)</span>
<span id="cb446-1157"><a href="#cb446-1157" aria-hidden="true" tabindex="-1"></a>origins_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(origins, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Human"</span>, <span class="st">"Zebrafish"</span>, <span class="st">"Mouse"</span>))</span>
<span id="cb446-1158"><a href="#cb446-1158" aria-hidden="true" tabindex="-1"></a>origins_factor</span>
<span id="cb446-1159"><a href="#cb446-1159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1160"><a href="#cb446-1160" aria-hidden="true" tabindex="-1"></a><span class="co"># Check levels (categories)</span></span>
<span id="cb446-1161"><a href="#cb446-1161" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(origins_factor)</span>
<span id="cb446-1162"><a href="#cb446-1162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1163"><a href="#cb446-1163" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a factor with predefined levels</span></span>
<span id="cb446-1164"><a href="#cb446-1164" aria-hidden="true" tabindex="-1"></a>treatment_groups <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Control"</span>, <span class="st">"Low_dose"</span>, <span class="st">"High_dose"</span>, <span class="st">"Control"</span>, <span class="st">"Low_dose"</span>),</span>
<span id="cb446-1165"><a href="#cb446-1165" aria-hidden="true" tabindex="-1"></a>                         <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Control"</span>, <span class="st">"Low_dose"</span>, <span class="st">"High_dose"</span>))</span>
<span id="cb446-1166"><a href="#cb446-1166" aria-hidden="true" tabindex="-1"></a>treatment_groups</span>
<span id="cb446-1167"><a href="#cb446-1167" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1168"><a href="#cb446-1168" aria-hidden="true" tabindex="-1"></a><span class="co"># Ordered factors (important for severity, stages, etc.)</span></span>
<span id="cb446-1169"><a href="#cb446-1169" aria-hidden="true" tabindex="-1"></a>disease_severity <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Mild"</span>, <span class="st">"Severe"</span>, <span class="st">"Moderate"</span>, <span class="st">"Mild"</span>, <span class="st">"Critical"</span>),</span>
<span id="cb446-1170"><a href="#cb446-1170" aria-hidden="true" tabindex="-1"></a>                         <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Mild"</span>, <span class="st">"Moderate"</span>, <span class="st">"Severe"</span>, <span class="st">"Critical"</span>),</span>
<span id="cb446-1171"><a href="#cb446-1171" aria-hidden="true" tabindex="-1"></a>                         <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-1172"><a href="#cb446-1172" aria-hidden="true" tabindex="-1"></a>disease_severity</span>
<span id="cb446-1173"><a href="#cb446-1173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1174"><a href="#cb446-1174" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare with ordered factors</span></span>
<span id="cb446-1175"><a href="#cb446-1175" aria-hidden="true" tabindex="-1"></a>disease_severity[<span class="dv">1</span>] <span class="sc">&gt;</span> disease_severity[<span class="dv">2</span>]  <span class="co"># Is Mild less severe than Severe?</span></span>
<span id="cb446-1176"><a href="#cb446-1176" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1177"><a href="#cb446-1177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1178"><a href="#cb446-1178" aria-hidden="true" tabindex="-1"></a><span class="fu">### Factor Operations</span></span>
<span id="cb446-1179"><a href="#cb446-1179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1180"><a href="#cb446-1180" aria-hidden="true" tabindex="-1"></a><span class="in">```{r factor-operations}</span></span>
<span id="cb446-1181"><a href="#cb446-1181" aria-hidden="true" tabindex="-1"></a><span class="co"># Count frequencies</span></span>
<span id="cb446-1182"><a href="#cb446-1182" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">table</span>(origins_factor)</span>
<span id="cb446-1183"><a href="#cb446-1183" aria-hidden="true" tabindex="-1"></a>out</span>
<span id="cb446-1184"><a href="#cb446-1184" aria-hidden="true" tabindex="-1"></a>out[<span class="st">"Human"</span>]</span>
<span id="cb446-1185"><a href="#cb446-1185" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate proportions</span></span>
<span id="cb446-1186"><a href="#cb446-1186" aria-hidden="true" tabindex="-1"></a><span class="fu">prop.table</span>(out)</span>
<span id="cb446-1187"><a href="#cb446-1187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1188"><a href="#cb446-1188" aria-hidden="true" tabindex="-1"></a><span class="co"># Change reference level (important for statistical models)</span></span>
<span id="cb446-1189"><a href="#cb446-1189" aria-hidden="true" tabindex="-1"></a>origins_factor_relevel <span class="ot">&lt;-</span> <span class="fu">relevel</span>(origins_factor, <span class="at">ref =</span> <span class="st">"Mouse"</span>)</span>
<span id="cb446-1190"><a href="#cb446-1190" aria-hidden="true" tabindex="-1"></a>origins_factor_relevel</span>
<span id="cb446-1191"><a href="#cb446-1191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1192"><a href="#cb446-1192" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert to character</span></span>
<span id="cb446-1193"><a href="#cb446-1193" aria-hidden="true" tabindex="-1"></a>origins_char <span class="ot">&lt;-</span> <span class="fu">as.character</span>(origins_factor)</span>
<span id="cb446-1194"><a href="#cb446-1194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1195"><a href="#cb446-1195" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot factors - Basic barplot</span></span>
<span id="cb446-1196"><a href="#cb446-1196" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(origins_factor), </span>
<span id="cb446-1197"><a href="#cb446-1197" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"blue"</span>, <span class="st">"green"</span>, <span class="st">"red"</span>),</span>
<span id="cb446-1198"><a href="#cb446-1198" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Sample Origins"</span>,</span>
<span id="cb446-1199"><a href="#cb446-1199" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Count"</span>)</span>
<span id="cb446-1200"><a href="#cb446-1200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1201"><a href="#cb446-1201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1202"><a href="#cb446-1202" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-1203"><a href="#cb446-1203" aria-hidden="true" tabindex="-1"></a>Factor <span class="in">`level`</span>-ing and <span class="in">`relevel`</span>-ing are different. <span class="in">`relevel`</span> redefines what the reference should be. For example, in an experiment, you have <span class="in">`control`</span>, <span class="in">`treatment1`</span>, <span class="in">`treatment2`</span> groups. Your reference might be <span class="in">`control`</span>. So, all of your comparisons/statistics are on the basis of <span class="in">`control`</span>. But you might change the reference (by <span class="in">`relevel`</span> to <span class="in">`treatment1`</span> and all of your comparison will be on the basis of <span class="in">`treatment1`</span> group. Got it?</span>
<span id="cb446-1204"><a href="#cb446-1204" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1205"><a href="#cb446-1205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1206"><a href="#cb446-1206" aria-hidden="true" tabindex="-1"></a>More advanced plot with factors:</span>
<span id="cb446-1207"><a href="#cb446-1207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1210"><a href="#cb446-1210" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1211"><a href="#cb446-1211" aria-hidden="true" tabindex="-1"></a>gene_expr <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">5.2</span>, <span class="fl">7.8</span>, <span class="fl">4.5</span>, <span class="fl">12.3</span>, <span class="fl">8.1</span>, <span class="fl">3.7</span>)</span>
<span id="cb446-1212"><a href="#cb446-1212" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(gene_expr) <span class="ot">&lt;-</span> <span class="fu">as.character</span>(origins)</span>
<span id="cb446-1213"><a href="#cb446-1213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1214"><a href="#cb446-1214" aria-hidden="true" tabindex="-1"></a><span class="co"># Boxplot by factor</span></span>
<span id="cb446-1215"><a href="#cb446-1215" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(gene_expr <span class="sc">~</span> origins, </span>
<span id="cb446-1216"><a href="#cb446-1216" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="st">"lightblue"</span>,</span>
<span id="cb446-1217"><a href="#cb446-1217" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Gene Expression by Sample Origin"</span>, </span>
<span id="cb446-1218"><a href="#cb446-1218" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"Origin"</span>, </span>
<span id="cb446-1219"><a href="#cb446-1219" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Expression Level"</span>)</span>
<span id="cb446-1220"><a href="#cb446-1220" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1221"><a href="#cb446-1221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1222"><a href="#cb446-1222" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-1223"><a href="#cb446-1223" aria-hidden="true" tabindex="-1"></a>Did you notice how factor <span class="in">`level`</span>-ing changes the appearance of the categories in the plots? See the <span class="in">`barplot`</span> and the <span class="in">`boxplot`</span> again. Where are Zebrafish and Mouse now in the plots? Why are their positions on the x-axis changed?</span>
<span id="cb446-1224"><a href="#cb446-1224" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1225"><a href="#cb446-1225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1226"><a href="#cb446-1226" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-1227"><a href="#cb446-1227" aria-hidden="true" tabindex="-1"></a>Keep noticing the output formats. Sometimes the output is just a number, sometimes a vector or table or list, etc. Check <span class="in">`prop.table(table(origins_factor))`</span>. How is it?</span>
<span id="cb446-1228"><a href="#cb446-1228" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1229"><a href="#cb446-1229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1230"><a href="#cb446-1230" aria-hidden="true" tabindex="-1"></a>::: {.callout-note collapse="true"}</span>
<span id="cb446-1231"><a href="#cb446-1231" aria-hidden="true" tabindex="-1"></a><span class="fu">## Got it?</span></span>
<span id="cb446-1232"><a href="#cb446-1232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1233"><a href="#cb446-1233" aria-hidden="true" tabindex="-1"></a><span class="in">`prop &lt;- prop.table(table(origins_factor))`</span> -- is a named numeric vector (atomic vector). <span class="in">`prop$Human`</span> or similar won't work. Check this way: <span class="in">`prop`</span> <span class="in">`prop["Human"]; prop["Mouse"]; prop["Zebrafish"]`</span></span>
<span id="cb446-1234"><a href="#cb446-1234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1235"><a href="#cb446-1235" aria-hidden="true" tabindex="-1"></a>Or make it a data frame (df) first, then try to use normal way of handling df.</span>
<span id="cb446-1236"><a href="#cb446-1236" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1237"><a href="#cb446-1237" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1238"><a href="#cb446-1238" aria-hidden="true" tabindex="-1"></a>Accessing the Output:</span>
<span id="cb446-1239"><a href="#cb446-1239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1242"><a href="#cb446-1242" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1243"><a href="#cb446-1243" aria-hidden="true" tabindex="-1"></a>prop <span class="ot">&lt;-</span> <span class="fu">prop.table</span>(<span class="fu">table</span>(origins_factor))</span>
<span id="cb446-1244"><a href="#cb446-1244" aria-hidden="true" tabindex="-1"></a>prop <span class="co">#What do you see? A data frame? No difference?</span></span>
<span id="cb446-1245"><a href="#cb446-1245" aria-hidden="true" tabindex="-1"></a>prop[<span class="st">"Human"</span>]; prop[<span class="st">"Mouse"</span>]; prop[<span class="st">"Zebrafish"</span>]</span>
<span id="cb446-1246"><a href="#cb446-1246" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1247"><a href="#cb446-1247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1248"><a href="#cb446-1248" aria-hidden="true" tabindex="-1"></a><span class="fu">## Subsetting Data</span></span>
<span id="cb446-1249"><a href="#cb446-1249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1250"><a href="#cb446-1250" aria-hidden="true" tabindex="-1"></a><span class="fu">### Vectors</span></span>
<span id="cb446-1251"><a href="#cb446-1251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1252"><a href="#cb446-1252" aria-hidden="true" tabindex="-1"></a><span class="in">```{r subsetting-vectors}</span></span>
<span id="cb446-1253"><a href="#cb446-1253" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a vector</span></span>
<span id="cb446-1254"><a href="#cb446-1254" aria-hidden="true" tabindex="-1"></a>expression_data <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">3.2</span>, <span class="fl">4.5</span>, <span class="fl">2.1</span>, <span class="fl">6.7</span>, <span class="fl">5.9</span>, <span class="fl">3.3</span>, <span class="fl">7.8</span>, <span class="fl">2.9</span>)</span>
<span id="cb446-1255"><a href="#cb446-1255" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(expression_data) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Sample_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>)</span>
<span id="cb446-1256"><a href="#cb446-1256" aria-hidden="true" tabindex="-1"></a>expression_data</span>
<span id="cb446-1257"><a href="#cb446-1257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1258"><a href="#cb446-1258" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset by position</span></span>
<span id="cb446-1259"><a href="#cb446-1259" aria-hidden="true" tabindex="-1"></a>expression_data[<span class="dv">3</span>]             <span class="co"># Single element</span></span>
<span id="cb446-1260"><a href="#cb446-1260" aria-hidden="true" tabindex="-1"></a>expression_data[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">5</span>)]    <span class="co"># Multiple elements</span></span>
<span id="cb446-1261"><a href="#cb446-1261" aria-hidden="true" tabindex="-1"></a>expression_data[<span class="dv">2</span><span class="sc">:</span><span class="dv">5</span>]           <span class="co"># Range</span></span>
<span id="cb446-1262"><a href="#cb446-1262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1263"><a href="#cb446-1263" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset by name</span></span>
<span id="cb446-1264"><a href="#cb446-1264" aria-hidden="true" tabindex="-1"></a>expression_data[<span class="st">"Sample_6"</span>]</span>
<span id="cb446-1265"><a href="#cb446-1265" aria-hidden="true" tabindex="-1"></a>expression_data[<span class="fu">c</span>(<span class="st">"Sample_1"</span>, <span class="st">"Sample_8"</span>)]</span>
<span id="cb446-1266"><a href="#cb446-1266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1267"><a href="#cb446-1267" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset by condition</span></span>
<span id="cb446-1268"><a href="#cb446-1268" aria-hidden="true" tabindex="-1"></a>expression_data[expression_data <span class="sc">&gt;</span> <span class="dv">5</span>]              <span class="co"># Values &gt; 5</span></span>
<span id="cb446-1269"><a href="#cb446-1269" aria-hidden="true" tabindex="-1"></a>expression_data[expression_data <span class="sc">&gt;=</span> <span class="dv">3</span> <span class="sc">&amp;</span> expression_data <span class="sc">&lt;=</span> <span class="dv">6</span>]  <span class="co"># Values between 3 and 6</span></span>
<span id="cb446-1270"><a href="#cb446-1270" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1271"><a href="#cb446-1271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1272"><a href="#cb446-1272" aria-hidden="true" tabindex="-1"></a><span class="fu">### Data Frames</span></span>
<span id="cb446-1273"><a href="#cb446-1273" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1274"><a href="#cb446-1274" aria-hidden="true" tabindex="-1"></a><span class="in">```{r subsetting-dataframes}</span></span>
<span id="cb446-1275"><a href="#cb446-1275" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a data frame</span></span>
<span id="cb446-1276"><a href="#cb446-1276" aria-hidden="true" tabindex="-1"></a>gene_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb446-1277"><a href="#cb446-1277" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_id =</span> <span class="fu">c</span>(<span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"EGFR"</span>, <span class="st">"GAPDH"</span>),</span>
<span id="cb446-1278"><a href="#cb446-1278" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression =</span> <span class="fu">c</span>(<span class="fl">8.2</span>, <span class="fl">6.1</span>, <span class="fl">9.5</span>, <span class="fl">7.0</span>, <span class="fl">10.0</span>),</span>
<span id="cb446-1279"><a href="#cb446-1279" aria-hidden="true" tabindex="-1"></a>  <span class="at">mutation =</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"No"</span>)),</span>
<span id="cb446-1280"><a href="#cb446-1280" aria-hidden="true" tabindex="-1"></a>  <span class="at">pathway =</span> <span class="fu">c</span>(<span class="st">"DNA Repair"</span>, <span class="st">"Apoptosis"</span>, <span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>, <span class="st">"Metabolism"</span>)</span>
<span id="cb446-1281"><a href="#cb446-1281" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-1282"><a href="#cb446-1282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1283"><a href="#cb446-1283" aria-hidden="true" tabindex="-1"></a>gene_df</span>
<span id="cb446-1284"><a href="#cb446-1284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1285"><a href="#cb446-1285" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsetting by row index</span></span>
<span id="cb446-1286"><a href="#cb446-1286" aria-hidden="true" tabindex="-1"></a>gene_df[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, ]         <span class="co"># First three rows, all columns</span></span>
<span id="cb446-1287"><a href="#cb446-1287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1288"><a href="#cb446-1288" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsetting by column index</span></span>
<span id="cb446-1289"><a href="#cb446-1289" aria-hidden="true" tabindex="-1"></a>gene_df[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]     <span class="co"># All rows, first two columns</span></span>
<span id="cb446-1290"><a href="#cb446-1290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1291"><a href="#cb446-1291" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsetting by column name</span></span>
<span id="cb446-1292"><a href="#cb446-1292" aria-hidden="true" tabindex="-1"></a>gene_df[, <span class="fu">c</span>(<span class="st">"gene_id"</span>, <span class="st">"mutation"</span>)]</span>
<span id="cb446-1293"><a href="#cb446-1293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1294"><a href="#cb446-1294" aria-hidden="true" tabindex="-1"></a><span class="co"># Using the $ operator</span></span>
<span id="cb446-1295"><a href="#cb446-1295" aria-hidden="true" tabindex="-1"></a>gene_df<span class="sc">$</span>expression</span>
<span id="cb446-1296"><a href="#cb446-1296" aria-hidden="true" tabindex="-1"></a>gene_df<span class="sc">$</span>mutation</span>
<span id="cb446-1297"><a href="#cb446-1297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1298"><a href="#cb446-1298" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsetting by condition</span></span>
<span id="cb446-1299"><a href="#cb446-1299" aria-hidden="true" tabindex="-1"></a>gene_df[gene_df<span class="sc">$</span>expression <span class="sc">&gt;</span> <span class="dv">8</span>, ]</span>
<span id="cb446-1300"><a href="#cb446-1300" aria-hidden="true" tabindex="-1"></a>gene_df[gene_df<span class="sc">$</span>mutation <span class="sc">==</span> <span class="st">"Yes"</span>, ]</span>
<span id="cb446-1301"><a href="#cb446-1301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1302"><a href="#cb446-1302" aria-hidden="true" tabindex="-1"></a><span class="co"># Multiple conditions</span></span>
<span id="cb446-1303"><a href="#cb446-1303" aria-hidden="true" tabindex="-1"></a>gene_df[gene_df<span class="sc">$</span>expression <span class="sc">&gt;</span> <span class="dv">7</span> <span class="sc">&amp;</span> gene_df<span class="sc">$</span>mutation <span class="sc">==</span> <span class="st">"No"</span>, ]</span>
<span id="cb446-1304"><a href="#cb446-1304" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1305"><a href="#cb446-1305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1306"><a href="#cb446-1306" aria-hidden="true" tabindex="-1"></a>**Logical Operators**</span>
<span id="cb446-1307"><a href="#cb446-1307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1308"><a href="#cb446-1308" aria-hidden="true" tabindex="-1"></a>| Operator | Meaning          | Example          |</span>
<span id="cb446-1309"><a href="#cb446-1309" aria-hidden="true" tabindex="-1"></a>|----------|------------------|------------------|</span>
<span id="cb446-1310"><a href="#cb446-1310" aria-hidden="true" tabindex="-1"></a>| <span class="in">`==`</span>     | Equal to         | <span class="in">`x == 5`</span>         |</span>
<span id="cb446-1311"><a href="#cb446-1311" aria-hidden="true" tabindex="-1"></a>| <span class="in">`!=`</span>     | Not equal        | <span class="in">`x != 5`</span>         |</span>
<span id="cb446-1312"><a href="#cb446-1312" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&lt;`</span>      | Less than        | <span class="in">`x &lt; 5`</span>          |</span>
<span id="cb446-1313"><a href="#cb446-1313" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&gt;`</span>      | Greater than     | <span class="in">`x &gt; 5`</span>          |</span>
<span id="cb446-1314"><a href="#cb446-1314" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&lt;=`</span>     | Less or equal    | <span class="in">`x &lt;= 5`</span>         |</span>
<span id="cb446-1315"><a href="#cb446-1315" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&gt;=`</span>     | Greater or equal | <span class="in">`x &gt;= 5`</span>         |</span>
<span id="cb446-1316"><a href="#cb446-1316" aria-hidden="true" tabindex="-1"></a>| <span class="in">`!`</span>      | Not              | <span class="in">`!(x &lt; 5)`</span>       |</span>
<span id="cb446-1317"><a href="#cb446-1317" aria-hidden="true" tabindex="-1"></a>| <span class="in">`|`</span>      | OR               | <span class="in">`x &lt; 5 | x &gt; 10`</span> |</span>
<span id="cb446-1318"><a href="#cb446-1318" aria-hidden="true" tabindex="-1"></a>| <span class="in">`&amp;`</span>      | AND              | <span class="in">`x &gt; 5 &amp; x &lt; 10`</span> |</span>
<span id="cb446-1319"><a href="#cb446-1319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1320"><a href="#cb446-1320" aria-hidden="true" tabindex="-1"></a><span class="fu">### Row Names in Data Frames</span></span>
<span id="cb446-1321"><a href="#cb446-1321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1322"><a href="#cb446-1322" aria-hidden="true" tabindex="-1"></a>Row names are particularly important in bioinformatics where genes, proteins, or samples are often used as identifiers.</span>
<span id="cb446-1323"><a href="#cb446-1323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1324"><a href="#cb446-1324" aria-hidden="true" tabindex="-1"></a><span class="in">```{r rownames}</span></span>
<span id="cb446-1325"><a href="#cb446-1325" aria-hidden="true" tabindex="-1"></a><span class="co"># Setting row names for gene_df</span></span>
<span id="cb446-1326"><a href="#cb446-1326" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(gene_df) <span class="ot">&lt;-</span> gene_df<span class="sc">$</span>gene_id</span>
<span id="cb446-1327"><a href="#cb446-1327" aria-hidden="true" tabindex="-1"></a>gene_df</span>
<span id="cb446-1328"><a href="#cb446-1328" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1329"><a href="#cb446-1329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1330"><a href="#cb446-1330" aria-hidden="true" tabindex="-1"></a>We can now drop the gene_id column, if required.</span>
<span id="cb446-1331"><a href="#cb446-1331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1334"><a href="#cb446-1334" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1335"><a href="#cb446-1335" aria-hidden="true" tabindex="-1"></a>gene_df_clean <span class="ot">&lt;-</span> gene_df[, <span class="sc">-</span><span class="dv">1</span>]  <span class="co"># Remove the first column</span></span>
<span id="cb446-1336"><a href="#cb446-1336" aria-hidden="true" tabindex="-1"></a>gene_df_clean</span>
<span id="cb446-1337"><a href="#cb446-1337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1338"><a href="#cb446-1338" aria-hidden="true" tabindex="-1"></a><span class="co"># Access rows by name</span></span>
<span id="cb446-1339"><a href="#cb446-1339" aria-hidden="true" tabindex="-1"></a>gene_df_clean[<span class="st">"TP53"</span>, ]</span>
<span id="cb446-1340"><a href="#cb446-1340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1341"><a href="#cb446-1341" aria-hidden="true" tabindex="-1"></a><span class="co"># Check if row names are unique</span></span>
<span id="cb446-1342"><a href="#cb446-1342" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(<span class="fu">duplicated</span>(<span class="fu">rownames</span>(gene_df_clean)))</span>
<span id="cb446-1343"><a href="#cb446-1343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1344"><a href="#cb446-1344" aria-hidden="true" tabindex="-1"></a><span class="co"># Handle potential duplicated row names</span></span>
<span id="cb446-1345"><a href="#cb446-1345" aria-hidden="true" tabindex="-1"></a><span class="co"># </span><span class="al">NOTE</span><span class="co">: R doesn't allow duplicate row names by default</span></span>
<span id="cb446-1346"><a href="#cb446-1346" aria-hidden="true" tabindex="-1"></a>dup_genes <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb446-1347"><a href="#cb446-1347" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression =</span> <span class="fu">c</span>(<span class="fl">5.2</span>, <span class="fl">6.3</span>, <span class="fl">5.2</span>, <span class="fl">8.1</span>),</span>
<span id="cb446-1348"><a href="#cb446-1348" aria-hidden="true" tabindex="-1"></a>  <span class="at">mutation =</span> <span class="fu">c</span>(<span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span>)</span>
<span id="cb446-1349"><a href="#cb446-1349" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-1350"><a href="#cb446-1350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1351"><a href="#cb446-1351" aria-hidden="true" tabindex="-1"></a><span class="co"># This would cause an error:</span></span>
<span id="cb446-1352"><a href="#cb446-1352" aria-hidden="true" tabindex="-1"></a><span class="co">#rownames(dup_genes) &lt;- c("BRCA1", "BRCA1", "TP53", "EGFR")</span></span>
<span id="cb446-1353"><a href="#cb446-1353" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1354"><a href="#cb446-1354" aria-hidden="true" tabindex="-1"></a><span class="co"># Instead, we can preemptively make them unique:</span></span>
<span id="cb446-1355"><a href="#cb446-1355" aria-hidden="true" tabindex="-1"></a>proposed_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"BRCA1"</span>, <span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"EGFR"</span>)</span>
<span id="cb446-1356"><a href="#cb446-1356" aria-hidden="true" tabindex="-1"></a>unique_names <span class="ot">&lt;-</span> <span class="fu">make.unique</span>(proposed_names)</span>
<span id="cb446-1357"><a href="#cb446-1357" aria-hidden="true" tabindex="-1"></a>unique_names  <span class="co"># Show the generated unique names</span></span>
<span id="cb446-1358"><a href="#cb446-1358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1359"><a href="#cb446-1359" aria-hidden="true" tabindex="-1"></a><span class="co"># Now we can safely assign them</span></span>
<span id="cb446-1360"><a href="#cb446-1360" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(dup_genes) <span class="ot">&lt;-</span> unique_names</span>
<span id="cb446-1361"><a href="#cb446-1361" aria-hidden="true" tabindex="-1"></a>dup_genes</span>
<span id="cb446-1362"><a href="#cb446-1362" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1363"><a href="#cb446-1363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1364"><a href="#cb446-1364" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-1365"><a href="#cb446-1365" aria-hidden="true" tabindex="-1"></a>Why is unique name important for us? Imagine this meaningful biological scenario: one gene might transcribed into many transcript isoforms and hence many protein isoforms. From RNAseq data, we might get alignment count for each gene. But then we can separate the count for each transcript. One gene has one name or ID, but the transcripts are many for the same gene! So, we can denote, for example, 21 isoform of geneA like genA.1, geneA.2, geneA.3,......., geneA.21. See this link for <span class="co">[</span><span class="ot">MBP gene</span><span class="co">](https://www.ncbi.nlm.nih.gov/gene?Db=gene&amp;Cmd=DetailsSearch&amp;Term=4155)</span>. How many transcript isoforms does it have?</span>
<span id="cb446-1366"><a href="#cb446-1366" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1367"><a href="#cb446-1367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1368"><a href="#cb446-1368" aria-hidden="true" tabindex="-1"></a><span class="fu">## üè° Homeworks: Factors, Subsetting, and Biological Insight</span></span>
<span id="cb446-1369"><a href="#cb446-1369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1370"><a href="#cb446-1370" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>**(Factor vs Character)** Explain the difference between a character vector and a factor in R. Why would <span class="in">`mutation_status`</span> be a factor and not just a character vector?</span>
<span id="cb446-1371"><a href="#cb446-1371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1372"><a href="#cb446-1372" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>**(Factor Level Order)** You observed the following bacterial species in gut microbiome samples:</span>
<span id="cb446-1373"><a href="#cb446-1373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1376"><a href="#cb446-1376" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1377"><a href="#cb446-1377" aria-hidden="true" tabindex="-1"></a>species <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Lactobacillus"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Lactobacillus"</span>)</span>
<span id="cb446-1378"><a href="#cb446-1378" aria-hidden="true" tabindex="-1"></a>species_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(species, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Lactobacillus"</span>))</span>
<span id="cb446-1379"><a href="#cb446-1379" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1380"><a href="#cb446-1380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1381"><a href="#cb446-1381" aria-hidden="true" tabindex="-1"></a>What will <span class="in">`levels(species_factor)`</span> return? Why?</span>
<span id="cb446-1382"><a href="#cb446-1382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1383"><a href="#cb446-1383" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Given the factor:</span>
<span id="cb446-1384"><a href="#cb446-1384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1387"><a href="#cb446-1387" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1388"><a href="#cb446-1388" aria-hidden="true" tabindex="-1"></a>disease_severity <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Mild"</span>, <span class="st">"Severe"</span>, <span class="st">"Moderate"</span>), <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Mild"</span>, <span class="st">"Moderate"</span>, <span class="st">"Severe"</span>, <span class="st">"Critical"</span>), <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-1389"><a href="#cb446-1389" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1390"><a href="#cb446-1390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1391"><a href="#cb446-1391" aria-hidden="true" tabindex="-1"></a>What will be the result of <span class="in">`disease_severity[1] &lt; disease_severity[2]`</span> and why?</span>
<span id="cb446-1392"><a href="#cb446-1392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1393"><a href="#cb446-1393" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>You computed:</span>
<span id="cb446-1394"><a href="#cb446-1394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1397"><a href="#cb446-1397" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1398"><a href="#cb446-1398" aria-hidden="true" tabindex="-1"></a>prop <span class="ot">&lt;-</span> <span class="fu">prop.table</span>(<span class="fu">table</span>(species_factor))</span>
<span id="cb446-1399"><a href="#cb446-1399" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1400"><a href="#cb446-1400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1401"><a href="#cb446-1401" aria-hidden="true" tabindex="-1"></a>How do you extract the proportion of "Escherichia" samples from prop? Is <span class="in">`prop$Escherichia`</span> valid?</span>
<span id="cb446-1402"><a href="#cb446-1402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1403"><a href="#cb446-1403" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Interpret what this query returns:</span>
<span id="cb446-1404"><a href="#cb446-1404" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1405"><a href="#cb446-1405" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb446-1406"><a href="#cb446-1406" aria-hidden="true" tabindex="-1"></a>gene_df[gene_df<span class="sc">$</span>expression <span class="sc">&gt;</span> <span class="dv">7</span> <span class="sc">&amp;</span> gene_df<span class="sc">$</span>mutation <span class="sc">==</span> <span class="st">"No"</span>, ]</span>
<span id="cb446-1407"><a href="#cb446-1407" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1408"><a href="#cb446-1408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1409"><a href="#cb446-1409" aria-hidden="true" tabindex="-1"></a>What type of genes does it select?</span>
<span id="cb446-1410"><a href="#cb446-1410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1411"><a href="#cb446-1411" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>You have:</span>
<span id="cb446-1412"><a href="#cb446-1412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1415"><a href="#cb446-1415" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1416"><a href="#cb446-1416" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span>)</span>
<span id="cb446-1417"><a href="#cb446-1417" aria-hidden="true" tabindex="-1"></a>expression <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">5.2</span>, <span class="fl">8.1</span>, <span class="fl">4.3</span>, <span class="fl">9.0</span>, <span class="fl">5.7</span>)</span>
<span id="cb446-1418"><a href="#cb446-1418" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1419"><a href="#cb446-1419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1420"><a href="#cb446-1420" aria-hidden="true" tabindex="-1"></a>Make a dataframe using these 2 vectors first. Then,</span>
<span id="cb446-1421"><a href="#cb446-1421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1422"><a href="#cb446-1422" aria-hidden="true" tabindex="-1"></a>(a) Create a factor <span class="in">`group_factor`</span> for the samples.</span>
<span id="cb446-1423"><a href="#cb446-1423" aria-hidden="true" tabindex="-1"></a>(b) Use <span class="in">`tapply()`</span> to calculate mean expression per group.</span>
<span id="cb446-1424"><a href="#cb446-1424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1425"><a href="#cb446-1425" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-1426"><a href="#cb446-1426" aria-hidden="true" tabindex="-1"></a>Use <span class="in">`?tapply()`</span> to see how to use it.</span>
<span id="cb446-1427"><a href="#cb446-1427" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1428"><a href="#cb446-1428" aria-hidden="true" tabindex="-1"></a>**Hint:** You need to provide things for <span class="in">`X`</span>, <span class="in">`INDEX`</span>, <span class="in">`FUN`</span>. You have <span class="in">`X`</span>, <span class="in">`INDEX`</span> in this small dataframe. The <span class="in">`FUN`</span> should be applied thinking of what you are trying to do. You are trying to get the mean or average, right?</span>
<span id="cb446-1429"><a href="#cb446-1429" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1430"><a href="#cb446-1430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1431"><a href="#cb446-1431" aria-hidden="true" tabindex="-1"></a>(c) Plot a <span class="in">`barplot`</span> of average expression for each group.</span>
<span id="cb446-1432"><a href="#cb446-1432" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1433"><a href="#cb446-1433" aria-hidden="true" tabindex="-1"></a><span class="in">```{=html}</span></span>
<span id="cb446-1434"><a href="#cb446-1434" aria-hidden="true" tabindex="-1"></a><span class="in">&lt;!-- --&gt;</span></span>
<span id="cb446-1435"><a href="#cb446-1435" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1436"><a href="#cb446-1436" aria-hidden="true" tabindex="-1"></a><span class="ss">7.  </span>Use the <span class="in">`gene_df`</span> example. Subset the data to find genes with:</span>
<span id="cb446-1437"><a href="#cb446-1437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1438"><a href="#cb446-1438" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>expression <span class="sc">\&gt;</span> 8</span>
<span id="cb446-1439"><a href="#cb446-1439" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>pathway is either "Cell Cycle" or "Signaling"</span>
<span id="cb446-1440"><a href="#cb446-1440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1441"><a href="#cb446-1441" aria-hidden="true" tabindex="-1"></a><span class="ss">8.  </span>Create an ordered factor for the disease stages: <span class="in">`c("Stage I", "Stage III", "Stage II", "Stage IV", "Stage I")`</span>. Then plot the number of patients per stage using <span class="in">`barplot()`</span>. Confirm that <span class="in">`"Stage III" &gt; "Stage I"`</span> is logical in your factor.</span>
<span id="cb446-1442"><a href="#cb446-1442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1443"><a href="#cb446-1443" aria-hidden="true" tabindex="-1"></a><span class="ss">9.  </span>Suppose <span class="in">`gene_data`</span> has a column type with values "Oncogene", "Tumor Suppressor", and "Housekeeping".</span>
<span id="cb446-1444"><a href="#cb446-1444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1445"><a href="#cb446-1445" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Subset all "Oncogene" rows where expression <span class="sc">\&gt;</span> 8.</span>
<span id="cb446-1446"><a href="#cb446-1446" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Change the reference level of the factor type to "Housekeeping"</span>
<span id="cb446-1447"><a href="#cb446-1447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1448"><a href="#cb446-1448" aria-hidden="true" tabindex="-1"></a><span class="ss">10. </span>Simulate expression data for 3 tissues (see the code chunk below): We are going to use <span class="in">`rnorm()`</span> function to generate random values from a normal distribution for this purpose. The example values inside the <span class="in">`rnorm()`</span> function means we want:</span>
<span id="cb446-1449"><a href="#cb446-1449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1450"><a href="#cb446-1450" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>30 values in total,</span>
<span id="cb446-1451"><a href="#cb446-1451" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>average or mean value = 8,</span>
<span id="cb446-1452"><a href="#cb446-1452" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>standard deviation of expression is 2.</span>
<span id="cb446-1453"><a href="#cb446-1453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1454"><a href="#cb446-1454" aria-hidden="true" tabindex="-1"></a>You can play with the numbers to make your own values.</span>
<span id="cb446-1455"><a href="#cb446-1455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1456"><a href="#cb446-1456" aria-hidden="true" tabindex="-1"></a><span class="in">`rep()`</span> function is to replicate things (many times). In this example, we have <span class="in">`rep(c("brain", "liver", "kidney"), each = 10)`</span>. We will be having 10x "brains", followed by 10x "liver", followed by 10x "kidney". So, if you have changed your values inside the <span class="in">`rnorm()`</span> function, make this value meaningful for you. Now we have 3 things, each=10. So, 3<span class="sc">\*</span>10=30 is matching with the total value inside <span class="in">`rnorm()`</span> function. Got it?</span>
<span id="cb446-1457"><a href="#cb446-1457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1460"><a href="#cb446-1460" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1461"><a href="#cb446-1461" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>) <span class="co">#just for reproducibility. Not completely needed</span></span>
<span id="cb446-1462"><a href="#cb446-1462" aria-hidden="true" tabindex="-1"></a>gene_expr <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">30</span>, <span class="at">mean =</span> <span class="dv">8</span>, <span class="at">sd =</span> <span class="dv">2</span>)</span>
<span id="cb446-1463"><a href="#cb446-1463" aria-hidden="true" tabindex="-1"></a>tissue <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"brain"</span>, <span class="st">"liver"</span>, <span class="st">"kidney"</span>), <span class="at">each =</span> <span class="dv">10</span>)</span>
<span id="cb446-1464"><a href="#cb446-1464" aria-hidden="true" tabindex="-1"></a>tissue_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(tissue, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"liver"</span>, <span class="st">"brain"</span>, <span class="st">"kidney"</span>))</span>
<span id="cb446-1465"><a href="#cb446-1465" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1466"><a href="#cb446-1466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1467"><a href="#cb446-1467" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Make a boxplot showing expression per tissue.</span>
<span id="cb446-1468"><a href="#cb446-1468" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Which tissue shows the most variable gene expression? (Use <span class="in">`tapply()`</span> + <span class="in">`sd()`</span>)</span>
<span id="cb446-1469"><a href="#cb446-1469" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1470"><a href="#cb446-1470" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-1471"><a href="#cb446-1471" aria-hidden="true" tabindex="-1"></a>**Hint:** Variability is an expression of measuring standard deviation (<span class="in">`sd`</span>) just by squaring it. So, <span class="in">`var`</span> = <span class="in">`sd^2`</span>. Well, do you see how to use <span class="in">`sd`</span> inside <span class="in">`tapply()`</span> function? Use <span class="in">`?tapply()`</span> to know how to use it.</span>
<span id="cb446-1472"><a href="#cb446-1472" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1473"><a href="#cb446-1473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1474"><a href="#cb446-1474" aria-hidden="true" tabindex="-1"></a>Use these questions as a self-check -- reflect on why each step works before moving on to the next level (question).</span>
<span id="cb446-1475"><a href="#cb446-1475" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1476"><a href="#cb446-1476" aria-hidden="true" tabindex="-1"></a>**Push your `.Rmd` file and share by Friday 10PM BD Time.**</span>
<span id="cb446-1477"><a href="#cb446-1477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1478"><a href="#cb446-1478" aria-hidden="true" tabindex="-1"></a><span class="fu">## Conditionals</span></span>
<span id="cb446-1479"><a href="#cb446-1479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1480"><a href="#cb446-1480" aria-hidden="true" tabindex="-1"></a><span class="fu">### `if-else` statement</span></span>
<span id="cb446-1481"><a href="#cb446-1481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1482"><a href="#cb446-1482" aria-hidden="true" tabindex="-1"></a>General structure of <span class="in">`if-else`</span> statement:</span>
<span id="cb446-1483"><a href="#cb446-1483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1484"><a href="#cb446-1484" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb446-1485"><a href="#cb446-1485" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (condition1) {</span>
<span id="cb446-1486"><a href="#cb446-1486" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Code executed when condition1 is TRUE</span></span>
<span id="cb446-1487"><a href="#cb446-1487" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span> (condition2) {</span>
<span id="cb446-1488"><a href="#cb446-1488" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Code executed when condition1 is FALSE but condition2 is TRUE</span></span>
<span id="cb446-1489"><a href="#cb446-1489" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb446-1490"><a href="#cb446-1490" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Code executed when all conditions above are FALSE</span></span>
<span id="cb446-1491"><a href="#cb446-1491" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1492"><a href="#cb446-1492" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1493"><a href="#cb446-1493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1494"><a href="#cb446-1494" aria-hidden="true" tabindex="-1"></a>Let's use it now.</span>
<span id="cb446-1495"><a href="#cb446-1495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1496"><a href="#cb446-1496" aria-hidden="true" tabindex="-1"></a><span class="in">```{r conditionals}</span></span>
<span id="cb446-1497"><a href="#cb446-1497" aria-hidden="true" tabindex="-1"></a><span class="co"># if-else statement</span></span>
<span id="cb446-1498"><a href="#cb446-1498" aria-hidden="true" tabindex="-1"></a>gene_value <span class="ot">&lt;-</span> <span class="fl">6.8</span></span>
<span id="cb446-1499"><a href="#cb446-1499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1500"><a href="#cb446-1500" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(gene_value <span class="sc">&gt;</span> <span class="dv">10</span>) {</span>
<span id="cb446-1501"><a href="#cb446-1501" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"High expression"</span>)</span>
<span id="cb446-1502"><a href="#cb446-1502" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> <span class="cf">if</span>(gene_value <span class="sc">&gt;</span> <span class="dv">5</span>) {</span>
<span id="cb446-1503"><a href="#cb446-1503" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Medium expression"</span>)</span>
<span id="cb446-1504"><a href="#cb446-1504" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb446-1505"><a href="#cb446-1505" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Low expression"</span>)</span>
<span id="cb446-1506"><a href="#cb446-1506" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1507"><a href="#cb446-1507" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1508"><a href="#cb446-1508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1509"><a href="#cb446-1509" aria-hidden="true" tabindex="-1"></a>Visualize it using the image below.</span>
<span id="cb446-1510"><a href="#cb446-1510" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1511"><a href="#cb446-1511" aria-hidden="true" tabindex="-1"></a>::: text-center</span>
<span id="cb446-1512"><a href="#cb446-1512" aria-hidden="true" tabindex="-1"></a><span class="al">![](img/if_else.png)</span>{width="80%"}</span>
<span id="cb446-1513"><a href="#cb446-1513" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1514"><a href="#cb446-1514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1515"><a href="#cb446-1515" aria-hidden="true" tabindex="-1"></a><span class="fu">### `ifelse` statement for vectors</span></span>
<span id="cb446-1516"><a href="#cb446-1516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1517"><a href="#cb446-1517" aria-hidden="true" tabindex="-1"></a><span class="in">`ifelse`</span> is binary in nature. So, we can categorize only 2 things using <span class="in">`ifelse`</span>. The structure is:</span>
<span id="cb446-1518"><a href="#cb446-1518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1519"><a href="#cb446-1519" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb446-1520"><a href="#cb446-1520" aria-hidden="true" tabindex="-1"></a><span class="fu">ifelse</span>(test_or_condition, <span class="st">"value_if_condition_is_TRUE"</span>, <span class="st">"value_if_condition_is_FALSE"</span>)</span>
<span id="cb446-1521"><a href="#cb446-1521" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1522"><a href="#cb446-1522" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1523"><a href="#cb446-1523" aria-hidden="true" tabindex="-1"></a>See this example:</span>
<span id="cb446-1524"><a href="#cb446-1524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1527"><a href="#cb446-1527" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1528"><a href="#cb446-1528" aria-hidden="true" tabindex="-1"></a>expression_values <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">12.5</span>, <span class="fl">4.3</span>, <span class="fl">8.1</span>, <span class="fl">2.2</span>)</span>
<span id="cb446-1529"><a href="#cb446-1529" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(expression_values <span class="sc">&gt;</span> <span class="dv">5</span>, <span class="st">"Upregulated"</span>, <span class="st">"Downregulated"</span>)</span>
<span id="cb446-1530"><a href="#cb446-1530" aria-hidden="true" tabindex="-1"></a>labels</span>
<span id="cb446-1531"><a href="#cb446-1531" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1532"><a href="#cb446-1532" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1533"><a href="#cb446-1533" aria-hidden="true" tabindex="-1"></a>::: text-center</span>
<span id="cb446-1534"><a href="#cb446-1534" aria-hidden="true" tabindex="-1"></a><span class="al">![](img/elseif.png)</span>{width="80%"}</span>
<span id="cb446-1535"><a href="#cb446-1535" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1536"><a href="#cb446-1536" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1537"><a href="#cb446-1537" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-1538"><a href="#cb446-1538" aria-hidden="true" tabindex="-1"></a><span class="in">`ifelse`</span> has 3 things inside the parentheses, right? The first one is the condition, the second one is the category we define if the condition is met, and the third thing is the other remaining category we want to assign if the condition is not met. So, it's usage is perfect to say if a gene/transcript is upregulated or downregulated (binary classification).</span>
<span id="cb446-1539"><a href="#cb446-1539" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1540"><a href="#cb446-1540" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1541"><a href="#cb446-1541" aria-hidden="true" tabindex="-1"></a>If we still want to categorize more than 2 categories using <span class="in">`ifelse`</span>, we need to use it in a nested way. The structure will be like this:</span>
<span id="cb446-1542"><a href="#cb446-1542" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1543"><a href="#cb446-1543" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb446-1544"><a href="#cb446-1544" aria-hidden="true" tabindex="-1"></a><span class="fu">ifelse</span>(test1, value1,</span>
<span id="cb446-1545"><a href="#cb446-1545" aria-hidden="true" tabindex="-1"></a>       <span class="fu">ifelse</span>(test2, value2, </span>
<span id="cb446-1546"><a href="#cb446-1546" aria-hidden="true" tabindex="-1"></a>              value3))</span>
<span id="cb446-1547"><a href="#cb446-1547" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1548"><a href="#cb446-1548" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1549"><a href="#cb446-1549" aria-hidden="true" tabindex="-1"></a>See this example:</span>
<span id="cb446-1550"><a href="#cb446-1550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1553"><a href="#cb446-1553" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1554"><a href="#cb446-1554" aria-hidden="true" tabindex="-1"></a><span class="co"># ifelse() for vectors</span></span>
<span id="cb446-1555"><a href="#cb446-1555" aria-hidden="true" tabindex="-1"></a>expression_levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">2.5</span>, <span class="fl">5.8</span>, <span class="fl">7.2</span>, <span class="fl">3.1</span>, <span class="fl">6.9</span>)</span>
<span id="cb446-1556"><a href="#cb446-1556" aria-hidden="true" tabindex="-1"></a>expression_category <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(expression_levels <span class="sc">&gt;</span> <span class="dv">6</span>, </span>
<span id="cb446-1557"><a href="#cb446-1557" aria-hidden="true" tabindex="-1"></a>                             <span class="st">"High"</span>, </span>
<span id="cb446-1558"><a href="#cb446-1558" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">ifelse</span>(expression_levels <span class="sc">&gt;</span> <span class="dv">4</span>, <span class="st">"Medium"</span>, <span class="st">"Low"</span>))</span>
<span id="cb446-1559"><a href="#cb446-1559" aria-hidden="true" tabindex="-1"></a>expression_category</span>
<span id="cb446-1560"><a href="#cb446-1560" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1561"><a href="#cb446-1561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1562"><a href="#cb446-1562" aria-hidden="true" tabindex="-1"></a>::: text-center</span>
<span id="cb446-1563"><a href="#cb446-1563" aria-hidden="true" tabindex="-1"></a><span class="al">![](img/nested_elseif.png)</span>{width="80%"}</span>
<span id="cb446-1564"><a href="#cb446-1564" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1565"><a href="#cb446-1565" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1566"><a href="#cb446-1566" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-1567"><a href="#cb446-1567" aria-hidden="true" tabindex="-1"></a>You remember the general structure of <span class="in">`ifelse`</span> loop, right? the second thing after the first <span class="in">`,`</span> is the assigned category if the condition is met. So, we assigned it as <span class="in">`High`</span> here in this example. But then after the second <span class="in">`,`</span> there is a second <span class="in">`ifelse`</span> loop instead of a category. The second loop makes 2 more binary categories <span class="in">`Medium`</span> and <span class="in">`Low`</span>, and our task of assigning 3 categories is achieved.</span>
<span id="cb446-1568"><a href="#cb446-1568" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1569"><a href="#cb446-1569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1570"><a href="#cb446-1570" aria-hidden="true" tabindex="-1"></a><span class="in">`dplyr`</span> package has a function named <span class="in">`case-when()`</span> to help us use as many categories we want. The same task would be achieved like this:</span>
<span id="cb446-1571"><a href="#cb446-1571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1574"><a href="#cb446-1574" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1575"><a href="#cb446-1575" aria-hidden="true" tabindex="-1"></a><span class="co"># Requires dplyr package</span></span>
<span id="cb446-1576"><a href="#cb446-1576" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("dplyr") #decomment if you need to install the package</span></span>
<span id="cb446-1577"><a href="#cb446-1577" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb446-1578"><a href="#cb446-1578" aria-hidden="true" tabindex="-1"></a>expression_levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">2.5</span>, <span class="fl">5.8</span>, <span class="fl">7.2</span>, <span class="fl">3.1</span>, <span class="fl">6.9</span>)</span>
<span id="cb446-1579"><a href="#cb446-1579" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="fu">case_when</span>(</span>
<span id="cb446-1580"><a href="#cb446-1580" aria-hidden="true" tabindex="-1"></a>  expression_levels <span class="sc">&gt;</span> <span class="dv">6</span> <span class="sc">~</span> <span class="st">"High"</span>,</span>
<span id="cb446-1581"><a href="#cb446-1581" aria-hidden="true" tabindex="-1"></a>  expression_levels <span class="sc">&gt;</span> <span class="dv">4</span> <span class="sc">~</span> <span class="st">"Medium"</span>,</span>
<span id="cb446-1582"><a href="#cb446-1582" aria-hidden="true" tabindex="-1"></a>  <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Low"</span>  <span class="co"># Default case</span></span>
<span id="cb446-1583"><a href="#cb446-1583" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-1584"><a href="#cb446-1584" aria-hidden="true" tabindex="-1"></a>labels</span>
<span id="cb446-1585"><a href="#cb446-1585" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1586"><a href="#cb446-1586" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1587"><a href="#cb446-1587" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb446-1588"><a href="#cb446-1588" aria-hidden="true" tabindex="-1"></a>Do you see the point how you would use the <span class="in">`ifelse`</span> loop if you wanted to write a function to make 4 or 5 categories? If not, pause and re-think. You need to see the point. But anyway, categorizing more than 2 is better using <span class="in">`if-else`</span> statement</span>
<span id="cb446-1589"><a href="#cb446-1589" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb446-1590"><a href="#cb446-1590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1591"><a href="#cb446-1591" aria-hidden="true" tabindex="-1"></a><span class="fu">### `for` loop</span></span>
<span id="cb446-1592"><a href="#cb446-1592" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1595"><a href="#cb446-1595" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1596"><a href="#cb446-1596" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"CDC2"</span>, <span class="st">"MBP"</span>)</span>
<span id="cb446-1597"><a href="#cb446-1597" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">8.2</span>, <span class="fl">5.4</span>, <span class="fl">11.0</span>, <span class="fl">5.4</span>, <span class="fl">13.0</span>)</span>
<span id="cb446-1598"><a href="#cb446-1598" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1599"><a href="#cb446-1599" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(genes)) {</span>
<span id="cb446-1600"><a href="#cb446-1600" aria-hidden="true" tabindex="-1"></a>  status <span class="ot">&lt;-</span> <span class="cf">if</span> (expr[i] <span class="sc">&gt;</span> <span class="dv">10</span>) <span class="st">"High"</span> </span>
<span id="cb446-1601"><a href="#cb446-1601" aria-hidden="true" tabindex="-1"></a>  <span class="cf">else</span> <span class="cf">if</span> (expr[i] <span class="sc">&gt;</span> <span class="dv">6</span>) <span class="st">"Moderate"</span> </span>
<span id="cb446-1602"><a href="#cb446-1602" aria-hidden="true" tabindex="-1"></a>  <span class="cf">else</span> <span class="st">"Low"</span></span>
<span id="cb446-1603"><a href="#cb446-1603" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(genes[i], <span class="st">"has"</span>, status, <span class="st">"expression</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb446-1604"><a href="#cb446-1604" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1605"><a href="#cb446-1605" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1606"><a href="#cb446-1606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1607"><a href="#cb446-1607" aria-hidden="true" tabindex="-1"></a>In-class Task:</span>
<span id="cb446-1608"><a href="#cb446-1608" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1609"><a href="#cb446-1609" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Make a data frame using <span class="in">`genes`</span> and <span class="in">`expr`</span>.</span>
<span id="cb446-1610"><a href="#cb446-1610" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Add/flag the categories <span class="in">`High`</span>, <span class="in">`Moderate`</span> and <span class="in">`Low`</span> you get using the <span class="in">`for`</span> loop in a new column named <span class="in">`expression_level`</span> or similar.</span>
<span id="cb446-1611"><a href="#cb446-1611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1614"><a href="#cb446-1614" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1615"><a href="#cb446-1615" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 1: Vectors</span></span>
<span id="cb446-1616"><a href="#cb446-1616" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"CDC2"</span>, <span class="st">"MBP"</span>)</span>
<span id="cb446-1617"><a href="#cb446-1617" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">8.2</span>, <span class="fl">5.4</span>, <span class="fl">11.0</span>, <span class="fl">5.4</span>, <span class="fl">13.0</span>)</span>
<span id="cb446-1618"><a href="#cb446-1618" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1619"><a href="#cb446-1619" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Create a data frame</span></span>
<span id="cb446-1620"><a href="#cb446-1620" aria-hidden="true" tabindex="-1"></a>gene_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">gene =</span> genes, <span class="at">expression =</span> expr)</span>
<span id="cb446-1621"><a href="#cb446-1621" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1622"><a href="#cb446-1622" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 3: Add an empty column for expression level</span></span>
<span id="cb446-1623"><a href="#cb446-1623" aria-hidden="true" tabindex="-1"></a>gene_df<span class="sc">$</span>expression_level <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb446-1624"><a href="#cb446-1624" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1625"><a href="#cb446-1625" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 4: Use for loop to fill in the expression_level column</span></span>
<span id="cb446-1626"><a href="#cb446-1626" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(gene_df)) {</span>
<span id="cb446-1627"><a href="#cb446-1627" aria-hidden="true" tabindex="-1"></a>  gene_df<span class="sc">$</span>expression_level[i] <span class="ot">&lt;-</span> <span class="cf">if</span> (gene_df<span class="sc">$</span>expression[i] <span class="sc">&gt;</span> <span class="dv">10</span>) {</span>
<span id="cb446-1628"><a href="#cb446-1628" aria-hidden="true" tabindex="-1"></a>    <span class="st">"High"</span></span>
<span id="cb446-1629"><a href="#cb446-1629" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (gene_df<span class="sc">$</span>expression[i] <span class="sc">&gt;</span> <span class="dv">6</span>) {</span>
<span id="cb446-1630"><a href="#cb446-1630" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Moderate"</span></span>
<span id="cb446-1631"><a href="#cb446-1631" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb446-1632"><a href="#cb446-1632" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Low"</span></span>
<span id="cb446-1633"><a href="#cb446-1633" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb446-1634"><a href="#cb446-1634" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1635"><a href="#cb446-1635" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1636"><a href="#cb446-1636" aria-hidden="true" tabindex="-1"></a><span class="co"># View the final data frame</span></span>
<span id="cb446-1637"><a href="#cb446-1637" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(gene_df)</span>
<span id="cb446-1638"><a href="#cb446-1638" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1639"><a href="#cb446-1639" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1640"><a href="#cb446-1640" aria-hidden="true" tabindex="-1"></a><span class="fu">### `while` loop</span></span>
<span id="cb446-1641"><a href="#cb446-1641" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1642"><a href="#cb446-1642" aria-hidden="true" tabindex="-1"></a>Context:</span>
<span id="cb446-1643"><a href="#cb446-1643" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1644"><a href="#cb446-1644" aria-hidden="true" tabindex="-1"></a>You are preparing biological samples (e.g., blood, DNA extracts) for analysis. You have a set of samples labeled <span class="in">`Sample 1`</span> to <span class="in">`Sample 5`</span>. You want to check each one in order and confirm that it's ready for analysis. Use a while loop to process the samples sequentially.</span>
<span id="cb446-1645"><a href="#cb446-1645" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1648"><a href="#cb446-1648" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1649"><a href="#cb446-1649" aria-hidden="true" tabindex="-1"></a>i <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb446-1650"><a href="#cb446-1650" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> (i <span class="sc">&lt;=</span> <span class="dv">5</span>) {</span>
<span id="cb446-1651"><a href="#cb446-1651" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Sample"</span>, i, <span class="st">"is ready for analysis</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb446-1652"><a href="#cb446-1652" aria-hidden="true" tabindex="-1"></a>  i <span class="ot">&lt;-</span> i <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb446-1653"><a href="#cb446-1653" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1654"><a href="#cb446-1654" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1655"><a href="#cb446-1655" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1656"><a href="#cb446-1656" aria-hidden="true" tabindex="-1"></a><span class="fu">### `next` and `break`</span></span>
<span id="cb446-1657"><a href="#cb446-1657" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1658"><a href="#cb446-1658" aria-hidden="true" tabindex="-1"></a>Context:</span>
<span id="cb446-1659"><a href="#cb446-1659" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1660"><a href="#cb446-1660" aria-hidden="true" tabindex="-1"></a>You are screening biological samples (e.g., tissue or blood) in a quality control process. Some samples are good, some are suboptimal (not contaminated but poor quality), and some are contaminated (must be flagged and stop further processing). Use <span class="in">`next`</span> to skip suboptimal samples and <span class="in">`break`</span> to immediately stop when a contaminated sample is found.</span>
<span id="cb446-1661"><a href="#cb446-1661" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1664"><a href="#cb446-1664" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1665"><a href="#cb446-1665" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"good"</span>, <span class="st">"bad"</span>, <span class="st">"good"</span>, <span class="st">"contaminated"</span>)</span>
<span id="cb446-1666"><a href="#cb446-1666" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1667"><a href="#cb446-1667" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> samples) {</span>
<span id="cb446-1668"><a href="#cb446-1668" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (i <span class="sc">==</span> <span class="st">"contaminated"</span>) {</span>
<span id="cb446-1669"><a href="#cb446-1669" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="st">"Stop! Contaminated sample."</span>)</span>
<span id="cb446-1670"><a href="#cb446-1670" aria-hidden="true" tabindex="-1"></a>    <span class="cf">break</span></span>
<span id="cb446-1671"><a href="#cb446-1671" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb446-1672"><a href="#cb446-1672" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (i <span class="sc">==</span> <span class="st">"bad"</span>) <span class="cf">next</span></span>
<span id="cb446-1673"><a href="#cb446-1673" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Processing"</span>, i))</span>
<span id="cb446-1674"><a href="#cb446-1674" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1675"><a href="#cb446-1675" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1676"><a href="#cb446-1676" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1677"><a href="#cb446-1677" aria-hidden="true" tabindex="-1"></a><span class="fu">## Writing Functions in R</span></span>
<span id="cb446-1678"><a href="#cb446-1678" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1679"><a href="#cb446-1679" aria-hidden="true" tabindex="-1"></a>The syntax to write an R function is:</span>
<span id="cb446-1680"><a href="#cb446-1680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1681"><a href="#cb446-1681" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb446-1682"><a href="#cb446-1682" aria-hidden="true" tabindex="-1"></a>function_name <span class="ot">&lt;-</span> <span class="cf">function</span>() {</span>
<span id="cb446-1683"><a href="#cb446-1683" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1684"><a href="#cb446-1684" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1685"><a href="#cb446-1685" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1686"><a href="#cb446-1686" aria-hidden="true" tabindex="-1"></a>Let's use it.</span>
<span id="cb446-1687"><a href="#cb446-1687" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1688"><a href="#cb446-1688" aria-hidden="true" tabindex="-1"></a><span class="fu">### Flag gene expression</span></span>
<span id="cb446-1689"><a href="#cb446-1689" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1692"><a href="#cb446-1692" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1693"><a href="#cb446-1693" aria-hidden="true" tabindex="-1"></a>flag_expression <span class="ot">&lt;-</span> <span class="cf">function</span>(value) {</span>
<span id="cb446-1694"><a href="#cb446-1694" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (value <span class="sc">&gt;</span> <span class="dv">10</span>) {</span>
<span id="cb446-1695"><a href="#cb446-1695" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="st">"High"</span>)</span>
<span id="cb446-1696"><a href="#cb446-1696" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> <span class="cf">if</span> (value <span class="sc">&gt;</span> <span class="dv">5</span>) {</span>
<span id="cb446-1697"><a href="#cb446-1697" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="st">"Moderate"</span>)</span>
<span id="cb446-1698"><a href="#cb446-1698" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb446-1699"><a href="#cb446-1699" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="st">"Low"</span>)</span>
<span id="cb446-1700"><a href="#cb446-1700" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb446-1701"><a href="#cb446-1701" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1702"><a href="#cb446-1702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1703"><a href="#cb446-1703" aria-hidden="true" tabindex="-1"></a><span class="fu">flag_expression</span>(<span class="fl">8.3</span>)</span>
<span id="cb446-1704"><a href="#cb446-1704" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1705"><a href="#cb446-1705" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1706"><a href="#cb446-1706" aria-hidden="true" tabindex="-1"></a>**Apply to a vector**</span>
<span id="cb446-1707"><a href="#cb446-1707" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1710"><a href="#cb446-1710" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1711"><a href="#cb446-1711" aria-hidden="true" tabindex="-1"></a>expr_values <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">12.2</span>, <span class="fl">4.4</span>, <span class="fl">7.5</span>)</span>
<span id="cb446-1712"><a href="#cb446-1712" aria-hidden="true" tabindex="-1"></a><span class="fu">sapply</span>(<span class="at">X=</span>expr_values, <span class="at">FUN=</span>flag_expression)</span>
<span id="cb446-1713"><a href="#cb446-1713" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1714"><a href="#cb446-1714" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1715"><a href="#cb446-1715" aria-hidden="true" tabindex="-1"></a><span class="fu">### Function with multiple arguments</span></span>
<span id="cb446-1716"><a href="#cb446-1716" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1719"><a href="#cb446-1719" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1720"><a href="#cb446-1720" aria-hidden="true" tabindex="-1"></a>gene_status <span class="ot">&lt;-</span> <span class="cf">function</span>(gene, expression, <span class="at">threshold =</span> <span class="dv">6</span>) { <span class="co">#gene = "string", expression = value</span></span>
<span id="cb446-1721"><a href="#cb446-1721" aria-hidden="true" tabindex="-1"></a>  label <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(expression <span class="sc">&gt;</span> threshold, <span class="st">"Up"</span>, <span class="st">"Down"</span>)</span>
<span id="cb446-1722"><a href="#cb446-1722" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">paste</span>(gene, <span class="st">"is"</span>, label, <span class="st">"regulated"</span>))</span>
<span id="cb446-1723"><a href="#cb446-1723" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1724"><a href="#cb446-1724" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1725"><a href="#cb446-1725" aria-hidden="true" tabindex="-1"></a><span class="fu">gene_status</span>(<span class="st">"TP53"</span>, <span class="fl">8.1</span>, <span class="at">threshold =</span> <span class="dv">9</span>)</span>
<span id="cb446-1726"><a href="#cb446-1726" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1727"><a href="#cb446-1727" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1728"><a href="#cb446-1728" aria-hidden="true" tabindex="-1"></a><span class="fu">### Return a list</span></span>
<span id="cb446-1729"><a href="#cb446-1729" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1732"><a href="#cb446-1732" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1733"><a href="#cb446-1733" aria-hidden="true" tabindex="-1"></a>calc_stats <span class="ot">&lt;-</span> <span class="cf">function</span>(values) {</span>
<span id="cb446-1734"><a href="#cb446-1734" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">mean =</span> <span class="fu">mean</span>(values), <span class="at">sd =</span> <span class="fu">sd</span>(values)))</span>
<span id="cb446-1735"><a href="#cb446-1735" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb446-1736"><a href="#cb446-1736" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1737"><a href="#cb446-1737" aria-hidden="true" tabindex="-1"></a><span class="fu">calc_stats</span>(<span class="fu">c</span>(<span class="fl">4.2</span>, <span class="fl">5.5</span>, <span class="fl">7.8</span>))</span>
<span id="cb446-1738"><a href="#cb446-1738" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1739"><a href="#cb446-1739" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1740"><a href="#cb446-1740" aria-hidden="true" tabindex="-1"></a><span class="fu">## üè° Homeworks: Control Structures and Functions</span></span>
<span id="cb446-1741"><a href="#cb446-1741" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1742"><a href="#cb446-1742" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span><span class="in">`for`</span> Loop + Conditional: Gene Expression Classifier</span>
<span id="cb446-1743"><a href="#cb446-1743" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1744"><a href="#cb446-1744" aria-hidden="true" tabindex="-1"></a>You have:</span>
<span id="cb446-1745"><a href="#cb446-1745" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1748"><a href="#cb446-1748" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1749"><a href="#cb446-1749" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"TP53"</span>, <span class="st">"EGFR"</span>, <span class="st">"MYC"</span>, <span class="st">"BRCA2"</span>)</span>
<span id="cb446-1750"><a href="#cb446-1750" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">4.8</span>, <span class="fl">6.3</span>, <span class="fl">11.7</span>, <span class="fl">9.5</span>)</span>
<span id="cb446-1751"><a href="#cb446-1751" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1752"><a href="#cb446-1752" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1753"><a href="#cb446-1753" aria-hidden="true" tabindex="-1"></a>**Task:**</span>
<span id="cb446-1754"><a href="#cb446-1754" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1755"><a href="#cb446-1755" aria-hidden="true" tabindex="-1"></a>Loop over the gene names and print: <span class="in">`&lt;Gene&gt; has &lt;Low/Moderate/High&gt; expression`</span></span>
<span id="cb446-1756"><a href="#cb446-1756" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1757"><a href="#cb446-1757" aria-hidden="true" tabindex="-1"></a>Based on:</span>
<span id="cb446-1758"><a href="#cb446-1758" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1759"><a href="#cb446-1759" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>High: <span class="sc">\&gt;</span>10</span>
<span id="cb446-1760"><a href="#cb446-1760" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Moderate: 6‚Äì10</span>
<span id="cb446-1761"><a href="#cb446-1761" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Low: ‚â§6</span>
<span id="cb446-1762"><a href="#cb446-1762" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1763"><a href="#cb446-1763" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span><span class="in">`next`</span>: Sample Screening Skips Bad Samples</span>
<span id="cb446-1764"><a href="#cb446-1764" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1767"><a href="#cb446-1767" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1768"><a href="#cb446-1768" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"good"</span>, <span class="st">"bad"</span>, <span class="st">"good"</span>, <span class="st">"contaminated"</span>, <span class="st">"good"</span>, <span class="st">"bad"</span>)</span>
<span id="cb446-1769"><a href="#cb446-1769" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1770"><a href="#cb446-1770" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1771"><a href="#cb446-1771" aria-hidden="true" tabindex="-1"></a>**Context:**</span>
<span id="cb446-1772"><a href="#cb446-1772" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1773"><a href="#cb446-1773" aria-hidden="true" tabindex="-1"></a>Bad samples should be skipped, only good and contaminated should be reported.</span>
<span id="cb446-1774"><a href="#cb446-1774" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1775"><a href="#cb446-1775" aria-hidden="true" tabindex="-1"></a>**Task:**</span>
<span id="cb446-1776"><a href="#cb446-1776" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1777"><a href="#cb446-1777" aria-hidden="true" tabindex="-1"></a>Use a <span class="in">`for`</span> loop with <span class="in">`next`</span> to skip <span class="in">`bad`</span> samples and print a message for others: <span class="in">`Sample X is flagged for analysis`</span></span>
<span id="cb446-1778"><a href="#cb446-1778" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1779"><a href="#cb446-1779" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span><span class="in">`break`</span>: Stop at Contaminated Sample Use the same samples vector above.</span>
<span id="cb446-1780"><a href="#cb446-1780" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1781"><a href="#cb446-1781" aria-hidden="true" tabindex="-1"></a>**Task:**</span>
<span id="cb446-1782"><a href="#cb446-1782" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1783"><a href="#cb446-1783" aria-hidden="true" tabindex="-1"></a>Print <span class="in">`Analyzing sample X`</span> for each sample. But if a <span class="in">`contaminated`</span> sample is found, stop processing immediately and print <span class="in">`Contamination detected! Halting...`</span></span>
<span id="cb446-1784"><a href="#cb446-1784" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1785"><a href="#cb446-1785" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span><span class="in">`while`</span> Loop: Sample Prep Countdown</span>
<span id="cb446-1786"><a href="#cb446-1786" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1787"><a href="#cb446-1787" aria-hidden="true" tabindex="-1"></a>**Context:**</span>
<span id="cb446-1788"><a href="#cb446-1788" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1789"><a href="#cb446-1789" aria-hidden="true" tabindex="-1"></a>You‚Äôre preparing 5 samples.</span>
<span id="cb446-1790"><a href="#cb446-1790" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1791"><a href="#cb446-1791" aria-hidden="true" tabindex="-1"></a>**Task:**</span>
<span id="cb446-1792"><a href="#cb446-1792" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1793"><a href="#cb446-1793" aria-hidden="true" tabindex="-1"></a>Use a while loop to print: <span class="in">`Sample &lt;i&gt; is ready for analysis`</span> from 1 to 5.</span>
<span id="cb446-1794"><a href="#cb446-1794" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1795"><a href="#cb446-1795" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span><span class="in">`while`</span> + Condition: Until Threshold You‚Äôre measuring a protein level that starts at 2. With each measurement, it increases randomly between 0.5 and 1.5 units.</span>
<span id="cb446-1796"><a href="#cb446-1796" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1797"><a href="#cb446-1797" aria-hidden="true" tabindex="-1"></a>**Task:** Use a while loop to simulate the increase until the protein level exceeds 10. Print each step.</span>
<span id="cb446-1798"><a href="#cb446-1798" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1801"><a href="#cb446-1801" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1802"><a href="#cb446-1802" aria-hidden="true" tabindex="-1"></a>level <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb446-1803"><a href="#cb446-1803" aria-hidden="true" tabindex="-1"></a><span class="co">#your code goes here</span></span>
<span id="cb446-1804"><a href="#cb446-1804" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1805"><a href="#cb446-1805" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1806"><a href="#cb446-1806" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>Custom Function: Classify BMI</span>
<span id="cb446-1807"><a href="#cb446-1807" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1808"><a href="#cb446-1808" aria-hidden="true" tabindex="-1"></a>Write a function <span class="in">`bmi_category(weight, height)`</span> that:</span>
<span id="cb446-1809"><a href="#cb446-1809" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1810"><a href="#cb446-1810" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Takes weight (kg) and height (m)</span>
<span id="cb446-1811"><a href="#cb446-1811" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1812"><a href="#cb446-1812" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Calculates BMI: weight/height¬≤</span>
<span id="cb446-1813"><a href="#cb446-1813" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1814"><a href="#cb446-1814" aria-hidden="true" tabindex="-1"></a>Returns <span class="in">`Underweight`</span> if <span class="in">`&lt;18.5`</span>, <span class="in">`Normal`</span> if <span class="in">`18.5‚Äì24.9`</span>, <span class="in">`Overweight`</span> if <span class="in">`25‚Äì29.9`</span>, <span class="in">`Obese`</span> if <span class="in">`‚â•30`</span>.</span>
<span id="cb446-1815"><a href="#cb446-1815" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1816"><a href="#cb446-1816" aria-hidden="true" tabindex="-1"></a>Test it:</span>
<span id="cb446-1817"><a href="#cb446-1817" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1818"><a href="#cb446-1818" aria-hidden="true" tabindex="-1"></a><span class="in">`bmi_category(65, 1.70)`</span></span>
<span id="cb446-1819"><a href="#cb446-1819" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1820"><a href="#cb446-1820" aria-hidden="true" tabindex="-1"></a><span class="ss">7.  </span>Function: Gene Risk Calculator</span>
<span id="cb446-1821"><a href="#cb446-1821" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1822"><a href="#cb446-1822" aria-hidden="true" tabindex="-1"></a>Write a function <span class="in">`gene_risk(expr, mut_status)`</span>:</span>
<span id="cb446-1823"><a href="#cb446-1823" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1824"><a href="#cb446-1824" aria-hidden="true" tabindex="-1"></a><span class="in">`High`</span> if <span class="in">`expression &gt; 9`</span> and <span class="in">`mutation = Yes`</span>.</span>
<span id="cb446-1825"><a href="#cb446-1825" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1826"><a href="#cb446-1826" aria-hidden="true" tabindex="-1"></a><span class="in">`Moderate`</span> if <span class="in">`expression &gt; 6`</span>.</span>
<span id="cb446-1827"><a href="#cb446-1827" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1828"><a href="#cb446-1828" aria-hidden="true" tabindex="-1"></a><span class="in">`Low`</span> otherwise.</span>
<span id="cb446-1829"><a href="#cb446-1829" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1830"><a href="#cb446-1830" aria-hidden="true" tabindex="-1"></a>Test on vectors:</span>
<span id="cb446-1831"><a href="#cb446-1831" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1834"><a href="#cb446-1834" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-1835"><a href="#cb446-1835" aria-hidden="true" tabindex="-1"></a>expr <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">5.5</span>, <span class="fl">10.1</span>, <span class="fl">7.3</span>)</span>
<span id="cb446-1836"><a href="#cb446-1836" aria-hidden="true" tabindex="-1"></a>mut_status <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span>)</span>
<span id="cb446-1837"><a href="#cb446-1837" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1838"><a href="#cb446-1838" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1839"><a href="#cb446-1839" aria-hidden="true" tabindex="-1"></a>Hint: Use <span class="in">`mapply()`</span>.</span>
<span id="cb446-1840"><a href="#cb446-1840" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1841"><a href="#cb446-1841" aria-hidden="true" tabindex="-1"></a><span class="ss">8.  </span><span class="in">`for`</span> Loop + Data Frame Make a data frame of 4 genes and expression values. Add a new column category using a for loop that assigns <span class="in">`High`</span>, <span class="in">`Moderate`</span>, <span class="in">`Low`</span> based on expression.</span>
<span id="cb446-1842"><a href="#cb446-1842" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1843"><a href="#cb446-1843" aria-hidden="true" tabindex="-1"></a><span class="ss">9.  </span>Function + Error Checking Write a function <span class="in">`get_expression_level(gene, df)`</span> that:</span>
<span id="cb446-1844"><a href="#cb446-1844" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1845"><a href="#cb446-1845" aria-hidden="true" tabindex="-1"></a>    Takes a gene name and a data frame with <span class="in">`gene`</span> and <span class="in">`expr`</span> columns</span>
<span id="cb446-1846"><a href="#cb446-1846" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1847"><a href="#cb446-1847" aria-hidden="true" tabindex="-1"></a>    If the gene is present, returns <span class="in">`Low`</span>, <span class="in">`Moderate`</span> or <span class="in">`High`</span>.</span>
<span id="cb446-1848"><a href="#cb446-1848" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1849"><a href="#cb446-1849" aria-hidden="true" tabindex="-1"></a>    If missing, returns <span class="in">`Gene not found`</span></span>
<span id="cb446-1850"><a href="#cb446-1850" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1851"><a href="#cb446-1851" aria-hidden="true" tabindex="-1"></a>    Test it with:</span>
<span id="cb446-1852"><a href="#cb446-1852" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1853"><a href="#cb446-1853" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE}</span></span>
<span id="cb446-1854"><a href="#cb446-1854" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">gene =</span> <span class="fu">c</span>(<span class="st">"TP53"</span>, <span class="st">"BRCA1"</span>), <span class="at">expr =</span> <span class="fu">c</span>(<span class="fl">5.5</span>, <span class="fl">10.5</span>))</span>
<span id="cb446-1855"><a href="#cb446-1855" aria-hidden="true" tabindex="-1"></a><span class="fu">get_expression_level</span>(<span class="st">"BRCA1"</span>, df)</span>
<span id="cb446-1856"><a href="#cb446-1856" aria-hidden="true" tabindex="-1"></a><span class="fu">get_expression_level</span>(<span class="st">"EGFR"</span>, df)</span>
<span id="cb446-1857"><a href="#cb446-1857" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1858"><a href="#cb446-1858" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1859"><a href="#cb446-1859" aria-hidden="true" tabindex="-1"></a><span class="fu">## Handling Missing/Wrong Values</span></span>
<span id="cb446-1860"><a href="#cb446-1860" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1861"><a href="#cb446-1861" aria-hidden="true" tabindex="-1"></a><span class="fu">### Identifying Issues</span></span>
<span id="cb446-1862"><a href="#cb446-1862" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1863"><a href="#cb446-1863" aria-hidden="true" tabindex="-1"></a><span class="in">```{r missing-values}</span></span>
<span id="cb446-1864"><a href="#cb446-1864" aria-hidden="true" tabindex="-1"></a><span class="co"># Create data with missing values</span></span>
<span id="cb446-1865"><a href="#cb446-1865" aria-hidden="true" tabindex="-1"></a>clinical_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb446-1866"><a href="#cb446-1866" aria-hidden="true" tabindex="-1"></a>  <span class="at">patient_id =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,</span>
<span id="cb446-1867"><a href="#cb446-1867" aria-hidden="true" tabindex="-1"></a>  <span class="at">age =</span> <span class="fu">c</span>(<span class="dv">25</span>, <span class="dv">99</span>, <span class="dv">30</span>, <span class="sc">-</span><span class="dv">5</span>, <span class="dv">40</span>),    <span class="co"># -5 is wrong, 99 is suspect</span></span>
<span id="cb446-1868"><a href="#cb446-1868" aria-hidden="true" tabindex="-1"></a>  <span class="at">bp =</span> <span class="fu">c</span>(<span class="dv">120</span>, <span class="cn">NA</span>, <span class="dv">115</span>, <span class="dv">125</span>, <span class="dv">118</span>),  <span class="co"># NA is missing</span></span>
<span id="cb446-1869"><a href="#cb446-1869" aria-hidden="true" tabindex="-1"></a>  <span class="at">weight =</span> <span class="fu">c</span>(<span class="dv">65</span>, <span class="dv">70</span>, <span class="cn">NA</span>, <span class="dv">68</span>, <span class="sc">-</span><span class="dv">1</span>)   <span class="co"># -1 is wrong</span></span>
<span id="cb446-1870"><a href="#cb446-1870" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-1871"><a href="#cb446-1871" aria-hidden="true" tabindex="-1"></a>clinical_data</span>
<span id="cb446-1872"><a href="#cb446-1872" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1873"><a href="#cb446-1873" aria-hidden="true" tabindex="-1"></a><span class="co"># Check for missing values</span></span>
<span id="cb446-1874"><a href="#cb446-1874" aria-hidden="true" tabindex="-1"></a><span class="fu">is.na</span>(clinical_data)</span>
<span id="cb446-1875"><a href="#cb446-1875" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">is.na</span>(clinical_data))  <span class="co"># Count NAs by column</span></span>
<span id="cb446-1876"><a href="#cb446-1876" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1877"><a href="#cb446-1877" aria-hidden="true" tabindex="-1"></a><span class="co"># Check for impossible values</span></span>
<span id="cb446-1878"><a href="#cb446-1878" aria-hidden="true" tabindex="-1"></a>clinical_data<span class="sc">$</span>age <span class="sc">&lt;</span> <span class="dv">0</span></span>
<span id="cb446-1879"><a href="#cb446-1879" aria-hidden="true" tabindex="-1"></a>clinical_data<span class="sc">$</span>weight <span class="sc">&lt;</span> <span class="dv">0</span></span>
<span id="cb446-1880"><a href="#cb446-1880" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1881"><a href="#cb446-1881" aria-hidden="true" tabindex="-1"></a><span class="co"># Find indices of problematic values</span></span>
<span id="cb446-1882"><a href="#cb446-1882" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(clinical_data<span class="sc">$</span>age <span class="sc">&lt;</span> <span class="dv">0</span> <span class="sc">|</span> clinical_data<span class="sc">$</span>age <span class="sc">&gt;</span> <span class="dv">90</span>)</span>
<span id="cb446-1883"><a href="#cb446-1883" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1884"><a href="#cb446-1884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1885"><a href="#cb446-1885" aria-hidden="true" tabindex="-1"></a><span class="fu">### Fixing Data</span></span>
<span id="cb446-1886"><a href="#cb446-1886" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1887"><a href="#cb446-1887" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fixing-data}</span></span>
<span id="cb446-1888"><a href="#cb446-1888" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace impossible values with NA</span></span>
<span id="cb446-1889"><a href="#cb446-1889" aria-hidden="true" tabindex="-1"></a>clinical_data<span class="sc">$</span>age[clinical_data<span class="sc">$</span>age <span class="sc">&lt;</span> <span class="dv">0</span> <span class="sc">|</span> clinical_data<span class="sc">$</span>age <span class="sc">&gt;</span> <span class="dv">90</span>] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb446-1890"><a href="#cb446-1890" aria-hidden="true" tabindex="-1"></a>clinical_data<span class="sc">$</span>weight[clinical_data<span class="sc">$</span>weight <span class="sc">&lt;</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb446-1891"><a href="#cb446-1891" aria-hidden="true" tabindex="-1"></a>clinical_data</span>
<span id="cb446-1892"><a href="#cb446-1892" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1893"><a href="#cb446-1893" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace NAs with mean (common in biological data)</span></span>
<span id="cb446-1894"><a href="#cb446-1894" aria-hidden="true" tabindex="-1"></a>clinical_data<span class="sc">$</span>bp[<span class="fu">is.na</span>(clinical_data<span class="sc">$</span>bp)] <span class="ot">&lt;-</span> <span class="fu">mean</span>(clinical_data<span class="sc">$</span>bp, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-1895"><a href="#cb446-1895" aria-hidden="true" tabindex="-1"></a>clinical_data<span class="sc">$</span>weight[<span class="fu">is.na</span>(clinical_data<span class="sc">$</span>weight)] <span class="ot">&lt;-</span> <span class="fu">mean</span>(clinical_data<span class="sc">$</span>weight, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-1896"><a href="#cb446-1896" aria-hidden="true" tabindex="-1"></a>clinical_data</span>
<span id="cb446-1897"><a href="#cb446-1897" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1898"><a href="#cb446-1898" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace NAs with median (better for skewed data)</span></span>
<span id="cb446-1899"><a href="#cb446-1899" aria-hidden="true" tabindex="-1"></a>clinical_data<span class="sc">$</span>age[<span class="fu">is.na</span>(clinical_data<span class="sc">$</span>age)] <span class="ot">&lt;-</span> <span class="fu">median</span>(clinical_data<span class="sc">$</span>age, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-1900"><a href="#cb446-1900" aria-hidden="true" tabindex="-1"></a>clinical_data</span>
<span id="cb446-1901"><a href="#cb446-1901" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1902"><a href="#cb446-1902" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1903"><a href="#cb446-1903" aria-hidden="true" tabindex="-1"></a><span class="fu">## Data Transformation</span></span>
<span id="cb446-1904"><a href="#cb446-1904" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1905"><a href="#cb446-1905" aria-hidden="true" tabindex="-1"></a><span class="fu">### Introduction to Outliers</span></span>
<span id="cb446-1906"><a href="#cb446-1906" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1907"><a href="#cb446-1907" aria-hidden="true" tabindex="-1"></a>Outliers can significantly affect statistical analyses, especially in biological data where sample variation can be high.</span>
<span id="cb446-1908"><a href="#cb446-1908" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1909"><a href="#cb446-1909" aria-hidden="true" tabindex="-1"></a><span class="in">```{r outliers}</span></span>
<span id="cb446-1910"><a href="#cb446-1910" aria-hidden="true" tabindex="-1"></a><span class="co"># Create data with outliers</span></span>
<span id="cb446-1911"><a href="#cb446-1911" aria-hidden="true" tabindex="-1"></a>expression_levels <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">2.3</span>, <span class="fl">2.7</span>, <span class="fl">3.1</span>, <span class="fl">2.9</span>, <span class="fl">2.5</span>, <span class="fl">3.0</span>, <span class="fl">15.2</span>, <span class="fl">2.8</span>)</span>
<span id="cb446-1912"><a href="#cb446-1912" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(expression_levels, </span>
<span id="cb446-1913"><a href="#cb446-1913" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Expression Levels with Outlier"</span>,</span>
<span id="cb446-1914"><a href="#cb446-1914" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Expression"</span>)</span>
<span id="cb446-1915"><a href="#cb446-1915" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1916"><a href="#cb446-1916" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1917"><a href="#cb446-1917" aria-hidden="true" tabindex="-1"></a><span class="fu">### Identifying Outliers</span></span>
<span id="cb446-1918"><a href="#cb446-1918" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1919"><a href="#cb446-1919" aria-hidden="true" tabindex="-1"></a><span class="in">```{r identify-outliers}</span></span>
<span id="cb446-1920"><a href="#cb446-1920" aria-hidden="true" tabindex="-1"></a><span class="co"># Statistical approach: Values beyond 1.5*IQR</span></span>
<span id="cb446-1921"><a href="#cb446-1921" aria-hidden="true" tabindex="-1"></a>data_summary <span class="ot">&lt;-</span> <span class="fu">summary</span>(expression_levels)</span>
<span id="cb446-1922"><a href="#cb446-1922" aria-hidden="true" tabindex="-1"></a>data_summary</span>
<span id="cb446-1923"><a href="#cb446-1923" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1924"><a href="#cb446-1924" aria-hidden="true" tabindex="-1"></a>IQR_value <span class="ot">&lt;-</span> <span class="fu">IQR</span>(expression_levels)</span>
<span id="cb446-1925"><a href="#cb446-1925" aria-hidden="true" tabindex="-1"></a>upper_bound <span class="ot">&lt;-</span> data_summary[<span class="st">"3rd Qu."</span>] <span class="sc">+</span> <span class="fl">1.5</span> <span class="sc">*</span> IQR_value</span>
<span id="cb446-1926"><a href="#cb446-1926" aria-hidden="true" tabindex="-1"></a>lower_bound <span class="ot">&lt;-</span> data_summary[<span class="st">"1st Qu."</span>] <span class="sc">-</span> <span class="fl">1.5</span> <span class="sc">*</span> IQR_value</span>
<span id="cb446-1927"><a href="#cb446-1927" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1928"><a href="#cb446-1928" aria-hidden="true" tabindex="-1"></a><span class="co"># Find outliers</span></span>
<span id="cb446-1929"><a href="#cb446-1929" aria-hidden="true" tabindex="-1"></a>outliers <span class="ot">&lt;-</span> expression_levels[expression_levels <span class="sc">&gt;</span> upper_bound <span class="sc">|</span> </span>
<span id="cb446-1930"><a href="#cb446-1930" aria-hidden="true" tabindex="-1"></a>                             expression_levels <span class="sc">&lt;</span> lower_bound]</span>
<span id="cb446-1931"><a href="#cb446-1931" aria-hidden="true" tabindex="-1"></a>outliers</span>
<span id="cb446-1932"><a href="#cb446-1932" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1933"><a href="#cb446-1933" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1934"><a href="#cb446-1934" aria-hidden="true" tabindex="-1"></a><span class="fu">### Transforming Vectors</span></span>
<span id="cb446-1935"><a href="#cb446-1935" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1936"><a href="#cb446-1936" aria-hidden="true" tabindex="-1"></a>Mathematical transformations can normalize data, reduce outlier effects, and make data more suitable for statistical analyses.</span>
<span id="cb446-1937"><a href="#cb446-1937" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1938"><a href="#cb446-1938" aria-hidden="true" tabindex="-1"></a><span class="in">```{r transformations}</span></span>
<span id="cb446-1939"><a href="#cb446-1939" aria-hidden="true" tabindex="-1"></a><span class="co"># Original data</span></span>
<span id="cb446-1940"><a href="#cb446-1940" aria-hidden="true" tabindex="-1"></a>gene_exp <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">15</span>, <span class="dv">42</span>, <span class="dv">87</span>, <span class="dv">115</span>, <span class="dv">320</span>, <span class="dv">560</span>, <span class="dv">1120</span>)</span>
<span id="cb446-1941"><a href="#cb446-1941" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(gene_exp, <span class="at">main =</span> <span class="st">"Original Expression Values"</span>, <span class="at">xlab =</span> <span class="st">"Expression"</span>)</span>
<span id="cb446-1942"><a href="#cb446-1942" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1943"><a href="#cb446-1943" aria-hidden="true" tabindex="-1"></a><span class="co"># Log transformation (common in gene expression analysis)</span></span>
<span id="cb446-1944"><a href="#cb446-1944" aria-hidden="true" tabindex="-1"></a>log_exp <span class="ot">&lt;-</span> <span class="fu">log2</span>(gene_exp)</span>
<span id="cb446-1945"><a href="#cb446-1945" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(log_exp, <span class="at">main =</span> <span class="st">"Log2 Transformed Expression"</span>, <span class="at">xlab =</span> <span class="st">"Log2 Expression"</span>)</span>
<span id="cb446-1946"><a href="#cb446-1946" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1947"><a href="#cb446-1947" aria-hidden="true" tabindex="-1"></a><span class="co"># Square root transformation (less aggressive than log)</span></span>
<span id="cb446-1948"><a href="#cb446-1948" aria-hidden="true" tabindex="-1"></a>sqrt_exp <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(gene_exp)</span>
<span id="cb446-1949"><a href="#cb446-1949" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(sqrt_exp, <span class="at">main =</span> <span class="st">"Square Root Transformed Expression"</span>, <span class="at">xlab =</span> <span class="st">"Sqrt Expression"</span>)</span>
<span id="cb446-1950"><a href="#cb446-1950" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1951"><a href="#cb446-1951" aria-hidden="true" tabindex="-1"></a><span class="co"># Z-score normalization (standardization)</span></span>
<span id="cb446-1952"><a href="#cb446-1952" aria-hidden="true" tabindex="-1"></a>z_exp <span class="ot">&lt;-</span> <span class="fu">scale</span>(gene_exp)</span>
<span id="cb446-1953"><a href="#cb446-1953" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(z_exp, <span class="at">main =</span> <span class="st">"Z-score Normalized Expression"</span>, <span class="at">xlab =</span> <span class="st">"Z-score"</span>)</span>
<span id="cb446-1954"><a href="#cb446-1954" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1955"><a href="#cb446-1955" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare transformations</span></span>
<span id="cb446-1956"><a href="#cb446-1956" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb446-1957"><a href="#cb446-1957" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(gene_exp, <span class="at">main =</span> <span class="st">"Original"</span>)</span>
<span id="cb446-1958"><a href="#cb446-1958" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(log_exp, <span class="at">main =</span> <span class="st">"Log2"</span>)</span>
<span id="cb446-1959"><a href="#cb446-1959" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(sqrt_exp, <span class="at">main =</span> <span class="st">"Square Root"</span>)</span>
<span id="cb446-1960"><a href="#cb446-1960" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(z_exp, <span class="at">main =</span> <span class="st">"Z-score"</span>)</span>
<span id="cb446-1961"><a href="#cb446-1961" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>))  <span class="co"># Reset plotting layout</span></span>
<span id="cb446-1962"><a href="#cb446-1962" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1963"><a href="#cb446-1963" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1964"><a href="#cb446-1964" aria-hidden="true" tabindex="-1"></a><span class="fu">### Logical Expressions</span></span>
<span id="cb446-1965"><a href="#cb446-1965" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1966"><a href="#cb446-1966" aria-hidden="true" tabindex="-1"></a><span class="in">```{r logical-expressions}</span></span>
<span id="cb446-1967"><a href="#cb446-1967" aria-hidden="true" tabindex="-1"></a><span class="co"># Create gene expression vector</span></span>
<span id="cb446-1968"><a href="#cb446-1968" aria-hidden="true" tabindex="-1"></a>exp_data <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">5.2</span>, <span class="fl">3.8</span>, <span class="fl">7.1</span>, <span class="fl">2.9</span>, <span class="fl">6.5</span>, <span class="fl">8.0</span>, <span class="fl">4.3</span>)</span>
<span id="cb446-1969"><a href="#cb446-1969" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(exp_data) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Gene_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>)</span>
<span id="cb446-1970"><a href="#cb446-1970" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1971"><a href="#cb446-1971" aria-hidden="true" tabindex="-1"></a><span class="co"># Basic comparisons</span></span>
<span id="cb446-1972"><a href="#cb446-1972" aria-hidden="true" tabindex="-1"></a>exp_data <span class="sc">&gt;</span> <span class="dv">5</span>    <span class="co"># Which genes have expression &gt; 5?</span></span>
<span id="cb446-1973"><a href="#cb446-1973" aria-hidden="true" tabindex="-1"></a>exp_data <span class="sc">&lt;=</span> <span class="dv">4</span>   <span class="co"># Which genes have expression &lt;= 4?</span></span>
<span id="cb446-1974"><a href="#cb446-1974" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1975"><a href="#cb446-1975" aria-hidden="true" tabindex="-1"></a><span class="co"># Store results in logical vector</span></span>
<span id="cb446-1976"><a href="#cb446-1976" aria-hidden="true" tabindex="-1"></a>high_exp <span class="ot">&lt;-</span> exp_data <span class="sc">&gt;</span> <span class="dv">6</span></span>
<span id="cb446-1977"><a href="#cb446-1977" aria-hidden="true" tabindex="-1"></a>high_exp</span>
<span id="cb446-1978"><a href="#cb446-1978" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1979"><a href="#cb446-1979" aria-hidden="true" tabindex="-1"></a><span class="co"># Use logical vectors for subsetting</span></span>
<span id="cb446-1980"><a href="#cb446-1980" aria-hidden="true" tabindex="-1"></a>exp_data[high_exp]  <span class="co"># Get high expression values</span></span>
<span id="cb446-1981"><a href="#cb446-1981" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1982"><a href="#cb446-1982" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1983"><a href="#cb446-1983" aria-hidden="true" tabindex="-1"></a><span class="fu">### Logical Operators</span></span>
<span id="cb446-1984"><a href="#cb446-1984" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1985"><a href="#cb446-1985" aria-hidden="true" tabindex="-1"></a><span class="in">```{r logical-operators}</span></span>
<span id="cb446-1986"><a href="#cb446-1986" aria-hidden="true" tabindex="-1"></a><span class="co"># Combining conditions with AND (&amp;)</span></span>
<span id="cb446-1987"><a href="#cb446-1987" aria-hidden="true" tabindex="-1"></a>exp_data <span class="sc">&gt;</span> <span class="dv">4</span> <span class="sc">&amp;</span> exp_data <span class="sc">&lt;</span> <span class="dv">7</span>  <span class="co"># Expression between 4 and 7</span></span>
<span id="cb446-1988"><a href="#cb446-1988" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1989"><a href="#cb446-1989" aria-hidden="true" tabindex="-1"></a><span class="co"># Combining conditions with OR (|)</span></span>
<span id="cb446-1990"><a href="#cb446-1990" aria-hidden="true" tabindex="-1"></a>exp_data <span class="sc">&lt;</span> <span class="dv">4</span> <span class="sc">|</span> exp_data <span class="sc">&gt;</span> <span class="dv">7</span>  <span class="co"># Expression less than 4 OR greater than 7</span></span>
<span id="cb446-1991"><a href="#cb446-1991" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1992"><a href="#cb446-1992" aria-hidden="true" tabindex="-1"></a><span class="co"># Using NOT (!)</span></span>
<span id="cb446-1993"><a href="#cb446-1993" aria-hidden="true" tabindex="-1"></a><span class="sc">!</span>high_exp  <span class="co"># Not high expression</span></span>
<span id="cb446-1994"><a href="#cb446-1994" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1995"><a href="#cb446-1995" aria-hidden="true" tabindex="-1"></a><span class="co"># Subsetting with combined conditions</span></span>
<span id="cb446-1996"><a href="#cb446-1996" aria-hidden="true" tabindex="-1"></a>exp_data[exp_data <span class="sc">&gt;</span> <span class="dv">4</span> <span class="sc">&amp;</span> exp_data <span class="sc">&lt;</span> <span class="dv">7</span>]  <span class="co"># Get values between 4 and 7</span></span>
<span id="cb446-1997"><a href="#cb446-1997" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-1998"><a href="#cb446-1998" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-1999"><a href="#cb446-1999" aria-hidden="true" tabindex="-1"></a><span class="fu">### Logical Functions</span></span>
<span id="cb446-2000"><a href="#cb446-2000" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2001"><a href="#cb446-2001" aria-hidden="true" tabindex="-1"></a><span class="in">```{r logical-functions}</span></span>
<span id="cb446-2002"><a href="#cb446-2002" aria-hidden="true" tabindex="-1"></a><span class="co"># all() - Are all values TRUE?</span></span>
<span id="cb446-2003"><a href="#cb446-2003" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(exp_data <span class="sc">&gt;</span> <span class="dv">0</span>)  <span class="co"># Are all expressions positive?</span></span>
<span id="cb446-2004"><a href="#cb446-2004" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2005"><a href="#cb446-2005" aria-hidden="true" tabindex="-1"></a><span class="co"># any() - Is at least one value TRUE?</span></span>
<span id="cb446-2006"><a href="#cb446-2006" aria-hidden="true" tabindex="-1"></a><span class="fu">any</span>(exp_data <span class="sc">&gt;</span> <span class="dv">7</span>)  <span class="co"># Is any expression greater than 7?</span></span>
<span id="cb446-2007"><a href="#cb446-2007" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2008"><a href="#cb446-2008" aria-hidden="true" tabindex="-1"></a><span class="co"># which() - Get indices of TRUE values</span></span>
<span id="cb446-2009"><a href="#cb446-2009" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(exp_data <span class="sc">&gt;</span> <span class="dv">6</span>)  <span class="co"># Which elements have expressions &gt; 6?</span></span>
<span id="cb446-2010"><a href="#cb446-2010" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2011"><a href="#cb446-2011" aria-hidden="true" tabindex="-1"></a><span class="co"># %in% operator - Test for membership</span></span>
<span id="cb446-2012"><a href="#cb446-2012" aria-hidden="true" tabindex="-1"></a>test_genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Gene_1"</span>, <span class="st">"Gene_5"</span>, <span class="st">"Gene_9"</span>)</span>
<span id="cb446-2013"><a href="#cb446-2013" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(exp_data) <span class="sc">%in%</span> test_genes  <span class="co"># Which names match test_genes?</span></span>
<span id="cb446-2014"><a href="#cb446-2014" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-2015"><a href="#cb446-2015" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2016"><a href="#cb446-2016" aria-hidden="true" tabindex="-1"></a><span class="fu">### Practical Session</span></span>
<span id="cb446-2017"><a href="#cb446-2017" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2018"><a href="#cb446-2018" aria-hidden="true" tabindex="-1"></a>Check out this repo: <span class="in">`https://github.com/genomicsclass/dagdata/`</span></span>
<span id="cb446-2019"><a href="#cb446-2019" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2020"><a href="#cb446-2020" aria-hidden="true" tabindex="-1"></a><span class="in">```{r practical, eval=FALSE}</span></span>
<span id="cb446-2021"><a href="#cb446-2021" aria-hidden="true" tabindex="-1"></a><span class="co"># Download small example dataset</span></span>
<span id="cb446-2022"><a href="#cb446-2022" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://github.com/genomicsclass/dagdata/raw/master/inst/extdata/msleep_ggplot2.csv"</span>,</span>
<span id="cb446-2023"><a href="#cb446-2023" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="st">"msleep_data.csv"</span>)</span>
<span id="cb446-2024"><a href="#cb446-2024" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2025"><a href="#cb446-2025" aria-hidden="true" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb446-2026"><a href="#cb446-2026" aria-hidden="true" tabindex="-1"></a>msleep <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">"msleep_data.csv"</span>)</span>
<span id="cb446-2027"><a href="#cb446-2027" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-2028"><a href="#cb446-2028" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2029"><a href="#cb446-2029" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Convert 'vore' column to factor and plot its distribution.</span>
<span id="cb446-2030"><a href="#cb446-2030" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>Create a matrix of sleep data columns and add row names.</span>
<span id="cb446-2031"><a href="#cb446-2031" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Find and handle any missing values.</span>
<span id="cb446-2032"><a href="#cb446-2032" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Calculate mean sleep time by diet category (vore).</span>
<span id="cb446-2033"><a href="#cb446-2033" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Identify outliers in sleep_total.</span>
<span id="cb446-2034"><a href="#cb446-2034" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2035"><a href="#cb446-2035" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary of the Lesson</span></span>
<span id="cb446-2036"><a href="#cb446-2036" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2037"><a href="#cb446-2037" aria-hidden="true" tabindex="-1"></a>In this lesson, we covered:</span>
<span id="cb446-2038"><a href="#cb446-2038" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2039"><a href="#cb446-2039" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>**Factor Variables**: Essential for categorical data in biology (genotypes, treatments, etc.)</span>
<span id="cb446-2040"><a href="#cb446-2040" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2041"><a href="#cb446-2041" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Creation, levels, ordering, and visualization</span>
<span id="cb446-2042"><a href="#cb446-2042" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2043"><a href="#cb446-2043" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>**Subsetting Techniques**: Critical for data extraction and analysis</span>
<span id="cb446-2044"><a href="#cb446-2044" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2045"><a href="#cb446-2045" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Vector and data frame subsetting with various methods</span>
<span id="cb446-2046"><a href="#cb446-2046" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Using row names effectively for biological identifiers</span>
<span id="cb446-2047"><a href="#cb446-2047" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2048"><a href="#cb446-2048" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>**Matrix Operations**: Fundamental for expression data</span>
<span id="cb446-2049"><a href="#cb446-2049" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2050"><a href="#cb446-2050" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Creation, manipulation, and biological applications</span>
<span id="cb446-2051"><a href="#cb446-2051" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Calculating fold changes and other common operations</span>
<span id="cb446-2052"><a href="#cb446-2052" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2053"><a href="#cb446-2053" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>**Missing Values**: Practical approaches for real-world biological data</span>
<span id="cb446-2054"><a href="#cb446-2054" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2055"><a href="#cb446-2055" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Identification and appropriate replacement methods</span>
<span id="cb446-2056"><a href="#cb446-2056" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2057"><a href="#cb446-2057" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>**Data Transformation**: Making data suitable for statistical analysis</span>
<span id="cb446-2058"><a href="#cb446-2058" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2059"><a href="#cb446-2059" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Log, square root, and z-score transformations</span>
<span id="cb446-2060"><a href="#cb446-2060" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Outlier identification and handling</span>
<span id="cb446-2061"><a href="#cb446-2061" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2062"><a href="#cb446-2062" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>**Logical Operations**: For data filtering and decision making</span>
<span id="cb446-2063"><a href="#cb446-2063" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2064"><a href="#cb446-2064" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Conditions, combinations, and applications</span>
<span id="cb446-2065"><a href="#cb446-2065" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2066"><a href="#cb446-2066" aria-hidden="true" tabindex="-1"></a>    These skills form the foundation for the more advanced visualization techniques we'll cover in future lessons.</span>
<span id="cb446-2067"><a href="#cb446-2067" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2068"><a href="#cb446-2068" aria-hidden="true" tabindex="-1"></a><span class="ss">7.  </span>**List**: Fundamental for many biological data and packages' output.</span>
<span id="cb446-2069"><a href="#cb446-2069" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2070"><a href="#cb446-2070" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Properties, accessing, and applications</span>
<span id="cb446-2071"><a href="#cb446-2071" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2072"><a href="#cb446-2072" aria-hidden="true" tabindex="-1"></a><span class="ss">8.  </span>We will know more about conditionals, R packages to handle data and visualization in a better and efficient way.</span>
<span id="cb446-2073"><a href="#cb446-2073" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2074"><a href="#cb446-2074" aria-hidden="true" tabindex="-1"></a><span class="fu">## Homework</span></span>
<span id="cb446-2075"><a href="#cb446-2075" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2076"><a href="#cb446-2076" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>**Matrix Operations**:</span>
<span id="cb446-2077"><a href="#cb446-2077" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2078"><a href="#cb446-2078" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Create a gene expression matrix with 8 genes and 4 conditions</span>
<span id="cb446-2079"><a href="#cb446-2079" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Calculate the mean expression for each gene</span>
<span id="cb446-2080"><a href="#cb446-2080" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Calculate fold change between condition 4 and condition 1</span>
<span id="cb446-2081"><a href="#cb446-2081" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Create a heatmap of your matrix</span>
<span id="cb446-2082"><a href="#cb446-2082" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2083"><a href="#cb446-2083" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>**Factor Analysis**:</span>
<span id="cb446-2084"><a href="#cb446-2084" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2085"><a href="#cb446-2085" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Using the <span class="in">`iris`</span> dataset, convert Species to an ordered factor</span>
<span id="cb446-2086"><a href="#cb446-2086" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Create boxplots showing Sepal.Length by Species</span>
<span id="cb446-2087"><a href="#cb446-2087" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Calculate mean petal length for each species level</span>
<span id="cb446-2088"><a href="#cb446-2088" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2089"><a href="#cb446-2089" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>**Data Cleaning Challenge**:</span>
<span id="cb446-2090"><a href="#cb446-2090" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2091"><a href="#cb446-2091" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>In the downloaded <span class="in">`msleep_data.csv`</span>:</span>
<span id="cb446-2092"><a href="#cb446-2092" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Identify all columns with missing values</span>
<span id="cb446-2093"><a href="#cb446-2093" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Replace missing values appropriately</span>
<span id="cb446-2094"><a href="#cb446-2094" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Create a new categorical variable "sleep_duration" with levels "Short", "Medium", "Long"</span>
<span id="cb446-2095"><a href="#cb446-2095" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2096"><a href="#cb446-2096" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>**List challenge**:</span>
<span id="cb446-2097"><a href="#cb446-2097" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2098"><a href="#cb446-2098" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Make your own lists</span>
<span id="cb446-2099"><a href="#cb446-2099" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Replicate all the tasks we did</span>
<span id="cb446-2100"><a href="#cb446-2100" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>You may ask AI to give you beginner-level questions but don't ask to solve the questions programmatically. Tell AI not to provide answers.</span>
<span id="cb446-2101"><a href="#cb446-2101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2102"><a href="#cb446-2102" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>**Complete Documentation**:</span>
<span id="cb446-2103"><a href="#cb446-2103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2104"><a href="#cb446-2104" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Write all code in R Markdown</span>
<span id="cb446-2105"><a href="#cb446-2105" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Include comments explaining your approach</span>
<span id="cb446-2106"><a href="#cb446-2106" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Push to GitHub</span>
<span id="cb446-2107"><a href="#cb446-2107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2108"><a href="#cb446-2108" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Due date: Friday 10pm BD Time</span></span>
<span id="cb446-2109"><a href="#cb446-2109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2110"><a href="#cb446-2110" aria-hidden="true" tabindex="-1"></a><span class="in">```{r save-examples, include=FALSE}</span></span>
<span id="cb446-2111"><a href="#cb446-2111" aria-hidden="true" tabindex="-1"></a><span class="co"># Save example data for students who missed class</span></span>
<span id="cb446-2112"><a href="#cb446-2112" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(expr_matrix, <span class="st">"example_expression_matrix.csv"</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-2113"><a href="#cb446-2113" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(gene_df, <span class="st">"example_gene_data.csv"</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb446-2114"><a href="#cb446-2114" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-2115"><a href="#cb446-2115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2118"><a href="#cb446-2118" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-2119"><a href="#cb446-2119" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>) <span class="co"># For reproducibility</span></span>
<span id="cb446-2120"><a href="#cb446-2120" aria-hidden="true" tabindex="-1"></a>gene_expr <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">30</span>, <span class="at">mean =</span> <span class="dv">8</span>, <span class="at">sd =</span> <span class="dv">2</span>)</span>
<span id="cb446-2121"><a href="#cb446-2121" aria-hidden="true" tabindex="-1"></a>tissue <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"brain"</span>, <span class="st">"liver"</span>, <span class="st">"kidney"</span>), <span class="at">each =</span> <span class="dv">10</span>)</span>
<span id="cb446-2122"><a href="#cb446-2122" aria-hidden="true" tabindex="-1"></a>tissue_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(tissue, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"liver"</span>, <span class="st">"brain"</span>, <span class="st">"kidney"</span>))</span>
<span id="cb446-2123"><a href="#cb446-2123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2124"><a href="#cb446-2124" aria-hidden="true" tabindex="-1"></a>tissue_factor</span>
<span id="cb446-2125"><a href="#cb446-2125" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-2126"><a href="#cb446-2126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2129"><a href="#cb446-2129" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-2130"><a href="#cb446-2130" aria-hidden="true" tabindex="-1"></a>expr_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb446-2131"><a href="#cb446-2131" aria-hidden="true" tabindex="-1"></a>  <span class="at">tissue =</span> tissue_factor,</span>
<span id="cb446-2132"><a href="#cb446-2132" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression =</span> gene_expr</span>
<span id="cb446-2133"><a href="#cb446-2133" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb446-2134"><a href="#cb446-2134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2135"><a href="#cb446-2135" aria-hidden="true" tabindex="-1"></a><span class="co">#expr_data</span></span>
<span id="cb446-2136"><a href="#cb446-2136" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(expr_data)</span>
<span id="cb446-2137"><a href="#cb446-2137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2138"><a href="#cb446-2138" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-2139"><a href="#cb446-2139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2142"><a href="#cb446-2142" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-2143"><a href="#cb446-2143" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(expression <span class="sc">~</span> tissue, </span>
<span id="cb446-2144"><a href="#cb446-2144" aria-hidden="true" tabindex="-1"></a>        <span class="at">data =</span> expr_data,</span>
<span id="cb446-2145"><a href="#cb446-2145" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Gene Expression by Tissue"</span>,</span>
<span id="cb446-2146"><a href="#cb446-2146" aria-hidden="true" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">"Tissue"</span>,</span>
<span id="cb446-2147"><a href="#cb446-2147" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Expression"</span>,</span>
<span id="cb446-2148"><a href="#cb446-2148" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"lightblue"</span>, <span class="st">"lightgreen"</span>, <span class="st">"lightpink"</span>))</span>
<span id="cb446-2149"><a href="#cb446-2149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2150"><a href="#cb446-2150" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-2151"><a href="#cb446-2151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2154"><a href="#cb446-2154" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-2155"><a href="#cb446-2155" aria-hidden="true" tabindex="-1"></a>sd_expression <span class="ot">&lt;-</span> <span class="fu">tapply</span>(expr_data<span class="sc">$</span>expression, expr_data<span class="sc">$</span>tissue, sd)</span>
<span id="cb446-2156"><a href="#cb446-2156" aria-hidden="true" tabindex="-1"></a>sd_expression</span>
<span id="cb446-2157"><a href="#cb446-2157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2158"><a href="#cb446-2158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb446-2159"><a href="#cb446-2159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb446-2162"><a href="#cb446-2162" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb446-2163"><a href="#cb446-2163" aria-hidden="true" tabindex="-1"></a>max_sd_tissue <span class="ot">&lt;-</span> <span class="fu">names</span>(sd_expression)[<span class="fu">which.max</span>(sd_expression)]</span>
<span id="cb446-2164"><a href="#cb446-2164" aria-hidden="true" tabindex="-1"></a>max_sd_value <span class="ot">&lt;-</span> sd_expression[<span class="fu">which.max</span>(sd_expression)]</span>
<span id="cb446-2165"><a href="#cb446-2165" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"The tissue with the highest sd is:"</span>, max_sd_tissue,<span class="st">", with SD ="</span>, <span class="fu">round</span>(max_sd_value, <span class="dv">2</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb446-2166"><a href="#cb446-2166" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>