<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="dcterms.date" content="2025-05-10">
<title>pythrash - HW solutions</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="../../_include/styles.css">
</head>
<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">PythRaSh</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-r" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">R</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-r">
<li>
    <a class="dropdown-item" href="../../ch/rbasics/git.html" rel="" target="">
 <span class="dropdown-text">git</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/firststeps.html" rel="" target="">
 <span class="dropdown-text">Basic R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/tidyverse.html" rel="" target="">
 <span class="dropdown-text">The tidyverse</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/correlation_regression.html" rel="" target="">
 <span class="dropdown-text">Advanced R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/baddata.html" rel="" target="">
 <span class="dropdown-text">Data Wrangling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/solutions.html" rel="" target="">
 <span class="dropdown-text">Problem Solutions</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-python" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Python</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-python">
<li>
    <a class="dropdown-item" href="../../ch/python/basics.html" rel="" target="">
 <span class="dropdown-text">Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/panda.html" rel="" target="">
 <span class="dropdown-text">Advanced Python</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/MLA.html" rel="" target="">
 <span class="dropdown-text">Machine Learning Algorithms</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/AI.html" rel="" target="">
 <span class="dropdown-text">AI</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-linux-and-clt" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Linux and CLT</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-linux-and-clt">
<li>
    <a class="dropdown-item" href="../../ch/linux-and-ctl/basics.html" rel="" target="">
 <span class="dropdown-text">Linux Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/linux-and-ctl/advanced.html" rel="" target="">
 <span class="dropdown-text">Advanced Command Line</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-probability-and-statistics" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Probability and Statistics</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-probability-and-statistics">
<li>
    <a class="dropdown-item" href="../../ch/prob_stat/prob_basic.html" rel="" target="">
 <span class="dropdown-text">Probability Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/prob_stat/stat.html" rel="" target="">
 <span class="dropdown-text">Statistics</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-genomics" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Genomics</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-genomics">
<li>
    <a class="dropdown-item" href="../../ch/genomics/basics.html" rel="" target="">
 <span class="dropdown-text">Genomics Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/genomics/advanced.html" rel="" target="">
 <span class="dropdown-text">Advanced Genomics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/genomics/pipeline.html" rel="" target="">
 <span class="dropdown-text">Pipelines</span></a>
  </li>  
    </ul>
</li>
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-blogs" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Blogs</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-blogs">
<li>
    <a class="dropdown-item" href="../../ch/blogs/en/index.html" rel="" target="">
 <span class="dropdown-text">English</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/blogs/bn/index.html" rel="" target="">
 <span class="dropdown-text">Bangla</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/MdRasheduz-zaman" rel="" target="">
 <span class="menu-text">Contact</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/MdRasheduz-zaman/PythRaSh" rel="" target=""><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#l3-matrices-and-lists" id="toc-l3-matrices-and-lists" class="nav-link active" data-scroll-target="#l3-matrices-and-lists">L3: Matrices and Lists</a>
  <ul class="collapse">
<li>
<a href="#task-1-protein-quantification-in-biological-samples" id="toc-task-1-protein-quantification-in-biological-samples" class="nav-link" data-scroll-target="#task-1-protein-quantification-in-biological-samples">Task 1: Protein Quantification in Biological Samples</a>
  <ul class="collapse">
<li><a href="#matrices" id="toc-matrices" class="nav-link" data-scroll-target="#matrices">1. Matrices</a></li>
  <li><a href="#making-transpose-of-proteinmatrix" id="toc-making-transpose-of-proteinmatrix" class="nav-link" data-scroll-target="#making-transpose-of-proteinmatrix">2. Making transpose of ProteinMatrix</a></li>
  <li><a href="#identity-matrix" id="toc-identity-matrix" class="nav-link" data-scroll-target="#identity-matrix">3. Identity Matrix</a></li>
  <li><a href="#total-protein-per-sample" id="toc-total-protein-per-sample" class="nav-link" data-scroll-target="#total-protein-per-sample">4.1. Total Protein per Sample</a></li>
  <li><a href="#total-protein-per-protein-type" id="toc-total-protein-per-protein-type" class="nav-link" data-scroll-target="#total-protein-per-protein-type">4.2. Total Protein per Protein Type</a></li>
  <li><a href="#heatmap-of-protein-concentrations" id="toc-heatmap-of-protein-concentrations" class="nav-link" data-scroll-target="#heatmap-of-protein-concentrations">4.3. Heatmap of Protein Concentrations</a></li>
  <li><a href="#interpretation" id="toc-interpretation" class="nav-link" data-scroll-target="#interpretation">üß† Interpretation</a></li>
  </ul>
</li>
  <li>
<a href="#task-2-gene-to-protein-translation" id="toc-task-2-gene-to-protein-translation" class="nav-link" data-scroll-target="#task-2-gene-to-protein-translation">Task 2: Gene-to-Protein Translation</a>
  <ul class="collapse">
<li><a href="#protein-output" id="toc-protein-output" class="nav-link" data-scroll-target="#protein-output">Protein Output</a></li>
  <li><a href="#transpose" id="toc-transpose" class="nav-link" data-scroll-target="#transpose">2. Transpose</a></li>
  <li><a href="#identity-matrix-and-multipliccation" id="toc-identity-matrix-and-multipliccation" class="nav-link" data-scroll-target="#identity-matrix-and-multipliccation">3. Identity matrix and multipliccation</a></li>
  <li><a href="#sub-matrix" id="toc-sub-matrix" class="nav-link" data-scroll-target="#sub-matrix">4. Sub-matrix:</a></li>
  <li><a href="#visualization-tasks" id="toc-visualization-tasks" class="nav-link" data-scroll-target="#visualization-tasks">Visualization tasks:</a></li>
  <li><a href="#interpretation-1" id="toc-interpretation-1" class="nav-link" data-scroll-target="#interpretation-1">üß† Interpretation</a></li>
  </ul>
</li>
  <li>
<a href="#task-3-animal-breeding-bull-ranking-by-economic-traits" id="toc-task-3-animal-breeding-bull-ranking-by-economic-traits" class="nav-link" data-scroll-target="#task-3-animal-breeding-bull-ranking-by-economic-traits">Task 3: Animal Breeding ‚Äì Bull Ranking by Economic Traits</a>
  <ul class="collapse">
<li><a href="#define-data" id="toc-define-data" class="nav-link" data-scroll-target="#define-data">Define Data</a></li>
  <li><a href="#compute-total-economic-value" id="toc-compute-total-economic-value" class="nav-link" data-scroll-target="#compute-total-economic-value">Compute Total Economic Value</a></li>
  <li><a href="#multiply-with-identity-matrix" id="toc-multiply-with-identity-matrix" class="nav-link" data-scroll-target="#multiply-with-identity-matrix">Multiply with Identity Matrix</a></li>
  <li><a href="#remove-milk-yield-and-recalculate-total-value" id="toc-remove-milk-yield-and-recalculate-total-value" class="nav-link" data-scroll-target="#remove-milk-yield-and-recalculate-total-value">Remove Milk Yield and Recalculate Total Value</a></li>
  <li><a href="#bar-plot-total-economic-value" id="toc-bar-plot-total-economic-value" class="nav-link" data-scroll-target="#bar-plot-total-economic-value">Bar Plot: Total Economic Value</a></li>
  <li><a href="#heatmap-of-ebvs" id="toc-heatmap-of-ebvs" class="nav-link" data-scroll-target="#heatmap-of-ebvs">Heatmap of EBVs</a></li>
  <li><a href="#interpretation-questions" id="toc-interpretation-questions" class="nav-link" data-scroll-target="#interpretation-questions">üß† Interpretation Questions</a></li>
  </ul>
</li>
  <li>
<a href="#task-4-plant-breeding-trait-contributions-from-parental-lines" id="toc-task-4-plant-breeding-trait-contributions-from-parental-lines" class="nav-link" data-scroll-target="#task-4-plant-breeding-trait-contributions-from-parental-lines">Task 4: Plant Breeding ‚Äì Trait Contributions from Parental Lines</a>
  <ul class="collapse">
<li><a href="#define-data-1" id="toc-define-data-1" class="nav-link" data-scroll-target="#define-data-1">Define Data</a></li>
  <li><a href="#compute-hybridtraits-vector" id="toc-compute-hybridtraits-vector" class="nav-link" data-scroll-target="#compute-hybridtraits-vector">Compute HybridTraits Vector</a></li>
  <li><a href="#multiply-with-identity-matrix-1" id="toc-multiply-with-identity-matrix-1" class="nav-link" data-scroll-target="#multiply-with-identity-matrix-1">Multiply with Identity Matrix</a></li>
  <li><a href="#remove-t3-maturation-time-and-recalculate" id="toc-remove-t3-maturation-time-and-recalculate" class="nav-link" data-scroll-target="#remove-t3-maturation-time-and-recalculate">Remove T3 (Maturation Time) and Recalculate</a></li>
  <li><a href="#heatmap-of-parenttraits" id="toc-heatmap-of-parenttraits" class="nav-link" data-scroll-target="#heatmap-of-parenttraits">Heatmap of ParentTraits</a></li>
  <li><a href="#bar-plot-of-hybrid-traits" id="toc-bar-plot-of-hybrid-traits" class="nav-link" data-scroll-target="#bar-plot-of-hybrid-traits">Bar Plot of Hybrid Traits</a></li>
  <li><a href="#interpretation-questions-1" id="toc-interpretation-questions-1" class="nav-link" data-scroll-target="#interpretation-questions-1">üß† Interpretation Questions</a></li>
  </ul>
</li>
  <li>
<a href="#task-5-managing-matrices-and-weight-vectors-using-lists-in-r" id="toc-task-5-managing-matrices-and-weight-vectors-using-lists-in-r" class="nav-link" data-scroll-target="#task-5-managing-matrices-and-weight-vectors-using-lists-in-r">Task 5: Managing Matrices and Weight Vectors Using Lists in R</a>
  <ul class="collapse">
<li><a href="#create-a-master-list" id="toc-create-a-master-list" class="nav-link" data-scroll-target="#create-a-master-list">Create a Master List</a></li>
  <li><a href="#list-the-structure" id="toc-list-the-structure" class="nav-link" data-scroll-target="#list-the-structure">List the Structure</a></li>
  <li><a href="#indexing-elements-from-lists" id="toc-indexing-elements-from-lists" class="nav-link" data-scroll-target="#indexing-elements-from-lists">Indexing Elements from Lists</a></li>
  <li><a href="#perform-weighted-calculations" id="toc-perform-weighted-calculations" class="nav-link" data-scroll-target="#perform-weighted-calculations">Perform Weighted Calculations</a></li>
  <li><a href="#subset-and-recalculate" id="toc-subset-and-recalculate" class="nav-link" data-scroll-target="#subset-and-recalculate">Subset and Recalculate</a></li>
  <li><a href="#visualization-tasks-1" id="toc-visualization-tasks-1" class="nav-link" data-scroll-target="#visualization-tasks-1">Visualization Tasks</a></li>
  <li><a href="#interpretation-questions-2" id="toc-interpretation-questions-2" class="nav-link" data-scroll-target="#interpretation-questions-2">üß† Interpretation Questions</a></li>
  </ul>
</li>
  <li><a href="#homework-solutions-factors-subsetting-and-biological-insight" id="toc-homework-solutions-factors-subsetting-and-biological-insight" class="nav-link" data-scroll-target="#homework-solutions-factors-subsetting-and-biological-insight">üß© Homework solutions: Factors, Subsetting, and Biological Insight</a></li>
  </ul>
</li>
  </ul><div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="solutions.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">HW solutions</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Md Rasheduzzaman </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://www.fli.de/de/institute/institut-fuer-infektionsmedizin-imed/">
            FLI Insel Riems
            </a>
          </p>
        <p class="affiliation">
            <a href="https://github.com/MdRasheduz-zaman">
            &amp; Freelancer
            </a>
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Last updated</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 10, 2025</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="abstract-title">Content summary</div>
    dataframe, matrices, list, factor, vector, etc.
  </div>
</div>

</header><style>
  .content {
    text-align: justify !important;
  }
  p {
    text-align: justify !important;
  }
  .quarto-body {
    text-align: justify !important;
  }
</style>
<section id="l3-matrices-and-lists" class="level1"><h1>L3: Matrices and Lists</h1>
<section id="task-1-protein-quantification-in-biological-samples" class="level2"><h2 class="anchored" data-anchor-id="task-1-protein-quantification-in-biological-samples">Task 1: Protein Quantification in Biological Samples</h2>
<p>We measured the concentration (in ¬µg/¬µL) of three proteins (P1, P2, P3) in four samples (S1‚ÄìS4):</p>
<section id="matrices" class="level3"><h3 class="anchored" data-anchor-id="matrices">1. Matrices</h3>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Making Protein Matrix</span></span>
<span><span class="va">ProteinMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">3</span>, <span class="fl">2</span>,</span>
<span>    <span class="fl">7</span>, <span class="fl">6</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample1"</span>, <span class="st">"Sample2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="va">ProteinMatrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        ProteinX ProteinY ProteinZ
Sample1        5        3        2
Sample2        7        6        4</code></pre>
</div>
</div>
<p>Now goes the weight matrix</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Making weight matrix</span></span>
<span><span class="va">WeightVector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>  nrow<span class="op">=</span><span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">WeightVector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5</code></pre>
</div>
</div>
<p>Now, multiply them.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Multiplying Matrices</span></span>
<span><span class="va">TotalConc</span> <span class="op">=</span> <span class="va">ProteinMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">WeightVector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TotalConc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">TotalConc</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Total_Protein_Conc
Sample1                8.5
Sample2               15.5</code></pre>
</div>
</div>
</section><section id="making-transpose-of-proteinmatrix" class="level3"><h3 class="anchored" data-anchor-id="making-transpose-of-proteinmatrix">2. Making transpose of ProteinMatrix</h3>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ProteinMatTranspose</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span></span>
<span><span class="va">ProteinMatTranspose</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Sample1 Sample2
ProteinX       5       7
ProteinY       3       6
ProteinZ       2       4</code></pre>
</div>
</div>
</section><section id="identity-matrix" class="level3"><h3 class="anchored" data-anchor-id="identity-matrix">3. Identity Matrix</h3>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">I</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">Identitycheck</span> <span class="op">=</span> <span class="va">ProteinMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Identitycheck</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="va">Identitycheck</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        ProteinX ProteinY ProteinZ
Sample1        5        3        2
Sample2        7        6        4</code></pre>
</div>
</div>
</section><section id="total-protein-per-sample" class="level3"><h3 class="anchored" data-anchor-id="total-protein-per-sample">4.1. Total Protein per Sample</h3>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sample1 Sample2 
     10      17 </code></pre>
</div>
</div>
</section><section id="total-protein-per-protein-type" class="level3"><h3 class="anchored" data-anchor-id="total-protein-per-protein-type">4.2. Total Protein per Protein Type</h3>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ProteinX ProteinY ProteinZ 
      12        9        6 </code></pre>
</div>
</div>
</section><section id="heatmap-of-protein-concentrations" class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-protein-concentrations">4.3. Heatmap of Protein Concentrations</h3>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">ProteinMatrix</span>, scale <span class="op">=</span> <span class="st">"none"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation" class="level3"><h3 class="anchored" data-anchor-id="interpretation">üß† Interpretation</h3>
<ul>
<li>Multiplying the protein levels by the weight vector shows how much each protein contributes in a sample. The result shows total protein concentration per sample.</li>
<li>The result shows that sample S2 has the highest protein burden.</li>
<li>The identity matrix represents no protein interactions or measurement biases. It is a simple matrix calculation.</li>
<li>New calculation:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># changing the weight of ProteinZ to 3.0</span></span>
<span><span class="va">newweightvector</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">3.0</span><span class="op">)</span>,</span>
<span>  nrow<span class="op">=</span><span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">newTotalconc</span> <span class="op">=</span> <span class="va">ProteinMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">newweightvector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">newTotalconc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span><span class="va">newTotalconc</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Total_Protein_Conc
Sample1               11.5
Sample2               21.5</code></pre>
</div>
</div>
<p>Still, S2 has more protein burden.</p>
<p>Bonus:</p>
<ul>
<li>Heatmap reveals PX is most abundant across all samples.</li>
</ul>
<hr></section></section><section id="task-2-gene-to-protein-translation" class="level2"><h2 class="anchored" data-anchor-id="task-2-gene-to-protein-translation">Task 2: Gene-to-Protein Translation</h2>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># making  Gene Expression matrix</span></span>
<span><span class="va">GeneExpression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">8</span>, <span class="fl">5</span>,</span>
<span>    <span class="fl">15</span>, <span class="fl">12</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">GeneExpression</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample1"</span>, <span class="st">"Sample2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">GeneExpression</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span><span class="va">GeneExpression</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        GeneA GeneB GeneC
Sample1    10     8     5
Sample2    15    12    10</code></pre>
</div>
</div>
<p>Translation efficiency:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># making Translation Matrix</span></span>
<span><span class="va">TranslationMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">0</span> , <span class="fl">0</span>,</span>
<span>  <span class="fl">0</span>, <span class="fl">1.2</span>, <span class="fl">0</span>,</span>
<span>  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1.8</span><span class="op">)</span>,</span>
<span>nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">TranslationMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TranslationMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"protA"</span>, <span class="st">"protB"</span>, <span class="st">"protC"</span><span class="op">)</span></span>
<span><span class="va">TranslationMatrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      protA protB protC
GeneA   1.5   0.0   0.0
GeneB   0.0   1.2   0.0
GeneC   0.0   0.0   1.8</code></pre>
</div>
</div>
<section id="protein-output" class="level3"><h3 class="anchored" data-anchor-id="protein-output">Protein Output</h3>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># computing Protein matrix</span></span>
<span><span class="va">Protein_matrix</span> <span class="op">&lt;-</span> <span class="va">GeneExpression</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">TranslationMatrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Protein_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"total_protA"</span>, <span class="st">"total_protB"</span>, <span class="st">"total_protC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Protein_matrix</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        total_protA total_protB total_protC
Sample1        15.0         9.6           9
Sample2        22.5        14.4          18</code></pre>
</div>
</div>
</section><section id="transpose" class="level3"><h3 class="anchored" data-anchor-id="transpose">2. Transpose</h3>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Transpose of GeneExpression matrix</span></span>
<span><span class="va">GeneExpression_Transpose</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">GeneExpression</span><span class="op">)</span></span>
<span><span class="va">GeneExpression_Transpose</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Sample1 Sample2
GeneA      10      15
GeneB       8      12
GeneC       5      10</code></pre>
</div>
</div>
<p>The new matrix represnts a matrix where the rows and columns of GeneExpression matrix have been interchanged.</p>
</section><section id="identity-matrix-and-multipliccation" class="level3"><h3 class="anchored" data-anchor-id="identity-matrix-and-multipliccation">3. Identity matrix and multipliccation</h3>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Creating Identity matrix</span></span>
<span><span class="va">I</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">I</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3]
[1,]    1    0    0
[2,]    0    1    0
[3,]    0    0    1</code></pre>
</div>
</div>
<p>Now, multiply:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">Product_matrix</span> <span class="op">=</span> <span class="va">TranslationMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span>  <span class="va">I</span> </span>
<span><span class="va">Product_matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1] [,2] [,3]
GeneA  1.5  0.0  0.0
GeneB  0.0  1.2  0.0
GeneC  0.0  0.0  1.8</code></pre>
</div>
</div>
<p>The product is identical to TranslationMatrix</p>
</section><section id="sub-matrix" class="level3"><h3 class="anchored" data-anchor-id="sub-matrix">4. Sub-matrix:</h3>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># making submatrix A</span></span>
<span><span class="va">A</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">8</span>,</span>
<span>    <span class="fl">15</span>, <span class="fl">12</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">A</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        GeneA GeneB
sample1    10     8
sample2    15    12</code></pre>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># finding inverse of A</span></span>
<span><span class="co">#inv_A &lt;- solve(A)</span></span>
<span><span class="co">#inv_A</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The inverse matrix could not be calculated since A is a singular matrix. So, A * A^-1 is also not possible.</p>
</section><section id="visualization-tasks" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks">Visualization tasks:</h3>
<ul>
<li><ol type="1">
<li>MARplot</li>
</ol></li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># generating MARplot-style scatter plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">GeneExpression</span>, <span class="va">Protein_matrix</span>, type<span class="op">=</span><span class="st">"p"</span>, main<span class="op">=</span><span class="st">"Protein level vs. Gene Expression level"</span><span class="op">)</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="st">"Sample-Gene"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">GeneExpression</span>, <span class="va">Protein_matrix</span>, labels <span class="op">=</span> <span class="va">labels</span>, pos<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># generating a heatmap</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">Protein_matrix</span>, main<span class="op">=</span> <span class="st">"Heatmap of Protein Level"</span>, Rowv <span class="op">=</span> <span class="cn">TRUE</span>, Colv <span class="op">=</span> <span class="cn">TRUE</span>, labRow<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Protein_matrix</span><span class="op">)</span>, labCol<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinA"</span>, <span class="st">"ProteinB"</span>, <span class="st">"ProteinC"</span><span class="op">)</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">topo.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span> <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid" width="1650"></p>
</div>
</div>
<ul>
<li><ol start="2" type="1">
<li>Heatmap of Expression:</li>
</ol></li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">GeneExpression</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">terrain.colors</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>, scale <span class="op">=</span> <span class="st">"column"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation-1" class="level3"><h3 class="anchored" data-anchor-id="interpretation-1">üß† Interpretation</h3>
<ol type="1">
<li><p>Matrix multiplication allows each gene in both samples to be multiplied to their respective translation efficiency. So, the product shows how successfully each gene is translated‚Äù)</p></li>
<li><p>The diagonal TranslationMatrix make sense biologically because they show translation efficiency of each gene and there is no other interaction between them. Although there could be interaction in real-world scenarios.</p></li>
<li><p>If Sample2 has higher protein levels even with similar gene expression, it means that more mRNAs are translated to proteins compared to Sample1‚Äù</p></li>
<li><p>The upward trend in MARplot may indicate an increase in translation efficacy and downward trend may indicate a decline in translation efficacy‚Äù</p></li>
<li><p>Clustering in the heatmap may suggest which samples are most similar to each other based on their prot.</p></li>
</ol></section></section><section id="task-3-animal-breeding-bull-ranking-by-economic-traits" class="level2"><h2 class="anchored" data-anchor-id="task-3-animal-breeding-bull-ranking-by-economic-traits">Task 3: Animal Breeding ‚Äì Bull Ranking by Economic Traits</h2>
<section id="define-data" class="level3"><h3 class="anchored" data-anchor-id="define-data">Define Data</h3>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define Bull EBVs</span></span>
<span><span class="va">BullEBVs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="fl">400</span>, <span class="fl">1.2</span>, <span class="fl">0.8</span>,</span>
<span>  <span class="fl">500</span>, <span class="fl">1.5</span>, <span class="fl">0.6</span></span>
<span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Bull1"</span>, <span class="st">"Bull2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Milk_yield"</span>, <span class="st">"Growth_rate"</span>, <span class="st">"Fertility"</span><span class="op">)</span></span>
<span><span class="va">BullEBVs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Milk_yield Growth_rate Fertility
Bull1        400         1.2       0.8
Bull2        500         1.5       0.6</code></pre>
</div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define Economic Weights</span></span>
<span><span class="va">EconomicWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.002</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">EconomicWeights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">EconomicWeights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">EconomicWeights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Weight
Milk_yield   2e-03
Growth_rate  5e+01
Fertility    1e+02</code></pre>
</div>
</div>
</section><section id="compute-total-economic-value" class="level3"><h3 class="anchored" data-anchor-id="compute-total-economic-value">Compute Total Economic Value</h3>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">TotalValue</span> <span class="op">&lt;-</span> <span class="va">BullEBVs</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">EconomicWeights</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TotalValue</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Merit"</span><span class="op">)</span></span>
<span><span class="va">TotalValue</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Merit
Bull1 140.8
Bull2 136.0</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Bull1: (400 √ó 0.002) + (1.2 √ó 50) + (0.8 √ó 100) = 140.8</p>
<p>Bull2: (500 √ó 0.002) + (1.5 √ó 50) + (0.6 √ó 100) = 136.0</p>
<p>Bull1 is more valuable economically.</p>
<p><strong>Biological Interpretation</strong></p>
<p>Economic weights convert genetic merit (EBVs, Estimated Breeding Values) into economic merit. Traits with higher financial importance have a larger impact, regardless of absolute EBV values.</p>
</section><section id="multiply-with-identity-matrix" class="level3"><h3 class="anchored" data-anchor-id="multiply-with-identity-matrix">Multiply with Identity Matrix</h3>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">I3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">I3</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">I3</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span></span>
<span><span class="va">I3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Milk_yield Growth_rate Fertility
Milk_yield           1           0         0
Growth_rate          0           1         0
Fertility            0           0         1</code></pre>
</div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">BullEBVs_identity</span> <span class="op">&lt;-</span> <span class="va">BullEBVs</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I3</span></span>
<span><span class="va">BullEBVs_identity</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Milk_yield Growth_rate Fertility
Bull1        400         1.2       0.8
Bull2        500         1.5       0.6</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Multiplying by identity matrix returns the original matrix. It confirms that EBV structure is preserved.</p>
</section><section id="remove-milk-yield-and-recalculate-total-value" class="level3"><h3 class="anchored" data-anchor-id="remove-milk-yield-and-recalculate-total-value">Remove Milk Yield and Recalculate Total Value</h3>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">BullEBVs_noMilk</span> <span class="op">&lt;-</span> <span class="va">BullEBVs</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">EconomicWeights_noMilk</span> <span class="op">&lt;-</span> <span class="va">EconomicWeights</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="va">TotalValue_noMilk</span> <span class="op">&lt;-</span> <span class="va">BullEBVs_noMilk</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">EconomicWeights_noMilk</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TotalValue_noMilk</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"New_Merit"</span><span class="op">)</span></span>
<span><span class="va">TotalValue_noMilk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      New_Merit
Bull1       140
Bull2       135</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Bull1: (1.2 √ó 50) + (0.8 √ó 100) = 140</p>
<p>Bull2: (1.5 √ó 50) + (0.6 √ó 100) = 135</p>
<p>Bull1 still ranks higher, but by a smaller margin.</p>
</section><section id="bar-plot-total-economic-value" class="level3"><h3 class="anchored" data-anchor-id="bar-plot-total-economic-value">Bar Plot: Total Economic Value</h3>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">TotalValue</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  names.arg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"skyblue"</span>, <span class="st">"orange"</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Total Economic Value of Bulls"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Total Value"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="heatmap-of-ebvs" class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-ebvs">Heatmap of EBVs</h3>
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">BullEBVs</span>,</span>
<span>  Rowv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  Colv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Heatmap of Bull EBVs"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation-questions" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions">üß† Interpretation Questions</h3>
<ul>
<li>How do economic weights affect trait importance?</li>
</ul>
<p>Traits with higher weights contribute more to the total economic value. This makes them more influential in ranking and selection.</p>
<ul>
<li>Why might you ignore milk yield?</li>
</ul>
<p>Milk yield may be excluded in systems focusing on fertility, growth, or when it is no longer a limiting factor. Environmental or economic contexts may also shift trait priorities.</p>
<ul>
<li>What is the value of heatmaps?</li>
</ul>
<p>Heatmaps visually compare EBVs across bulls and traits.They help detect patterns, outliers, and clusters easily in multivariate data.</p>
<ul>
<li>Can this method be extended to more bulls and traits?</li>
</ul>
<p>Yes. This method scales to any number of bulls or traits. Just ensure the EBVs matrix and economic weights are dimensionally compatible.</p>
</section></section><section id="task-4-plant-breeding-trait-contributions-from-parental-lines" class="level2"><h2 class="anchored" data-anchor-id="task-4-plant-breeding-trait-contributions-from-parental-lines">Task 4: Plant Breeding ‚Äì Trait Contributions from Parental Lines</h2>
<section id="define-data-1" class="level3"><h3 class="anchored" data-anchor-id="define-data-1">Define Data</h3>
<p><strong>Parent trait values (normalized 1‚Äì10)</strong></p>
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ParentTraits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="fl">7</span>, <span class="fl">5</span>, <span class="fl">3</span>,</span>
<span>  <span class="fl">6</span>, <span class="fl">8</span>, <span class="fl">4</span>,</span>
<span>  <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">6</span></span>
<span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"P1"</span>, <span class="st">"P2"</span>, <span class="st">"P3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Drought_resistance"</span>, <span class="st">"Yield"</span>, <span class="st">"Maturation_time"</span><span class="op">)</span></span>
<span><span class="va">ParentTraits</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define Hybrid Weights</span></span>
<span><span class="va">HybridWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">HybridWeights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">HybridWeights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">HybridWeights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Drought_resistance Yield Maturation_time
Weight                0.5   0.3             0.2</code></pre>
</div>
</div>
</section><section id="compute-hybridtraits-vector" class="level3"><h3 class="anchored" data-anchor-id="compute-hybridtraits-vector">Compute HybridTraits Vector</h3>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">HybridTraits</span> <span class="op">&lt;-</span> <span class="va">HybridWeights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">ParentTraits</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">HybridTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Contribution"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">HybridTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span></span>
<span><span class="va">HybridTraits</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              P1  P2  P3
Contribution 6.3 6.1 3.9</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>HybridTraits = (0.5 √ó P1) + (0.3 √ó P2) + (0.2 √ó P3)</p>
<p>Drought_resistance = (0.5 √ó 7) + (0.3 √ó 6) + (0.2 √ó 5) = 6.3</p>
<p>Yield = (0.5 √ó 5) + (0.3 √ó 8) + (0.2 √ó 6) = 6.1</p>
<p>Maturation_time = (0.5 √ó 3) + (0.3 √ó 4) + (0.2 √ó 6) = 3.9</p>
<p>The hybrid is moderately strong in drought resistance and yield, and has a relatively shorter maturation time.</p>
<p><strong>Biological Meaning of Unequal Contribution</strong></p>
<p>When one parent contributes more to a trait, it suggests that the trait‚Äôs heritable strength comes disproportionately from that parent. Breeders can use this knowledge to amplify desirable traits using the best parent.</p>
</section><section id="multiply-with-identity-matrix-1" class="level3"><h3 class="anchored" data-anchor-id="multiply-with-identity-matrix-1">Multiply with Identity Matrix</h3>
<div class="cell">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">I3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">ParentTraits_identity</span> <span class="op">&lt;-</span> <span class="va">ParentTraits</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits_identity</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span></span>
<span><span class="va">ParentTraits_identity</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Multiplying by identity matrix returns the original matrix. This operation verifies structural consistency and dimensionality.</p>
</section><section id="remove-t3-maturation-time-and-recalculate" class="level3"><h3 class="anchored" data-anchor-id="remove-t3-maturation-time-and-recalculate">Remove T3 (Maturation Time) and Recalculate</h3>
<div class="cell">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ParentTraits_T1T2</span> <span class="op">&lt;-</span> <span class="va">ParentTraits</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">ParentTraits_T1T2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield
P1                  7     5
P2                  6     8
P3                  5     6</code></pre>
</div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">HybridTraits_T1T2</span> <span class="op">&lt;-</span> <span class="va">HybridWeights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">ParentTraits_T1T2</span></span>
<span><span class="va">HybridTraits_T1T2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Drought_resistance Yield
Weight                6.3   6.1</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Drought_resistance = (0.5 √ó 7) + (0.3 √ó 6) + (0.2 √ó 5) = 6.3</p>
<p>Yield = (0.5 √ó 5) + (0.3 √ó 8) + (0.2 √ó 6) = 6.1</p>
<p>Removing a trait (T3) changes the trait profile. Hybrid selection may now favor traits that remain.</p>
</section><section id="heatmap-of-parenttraits" class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-parenttraits">Heatmap of ParentTraits</h3>
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">ParentTraits</span>,</span>
<span>  Rowv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  Colv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Heatmap of Parent Traits"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="bar-plot-of-hybrid-traits" class="level3"><h3 class="anchored" data-anchor-id="bar-plot-of-hybrid-traits">Bar Plot of Hybrid Traits</h3>
<div class="cell">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">HybridTraits</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  names.arg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#66c2a5"</span>, <span class="st">"#fc8d62"</span>, <span class="st">"#8da0cb"</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Hybrid Trait Profile"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Trait Value"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation-questions-1" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-1">üß† Interpretation Questions</h3>
<ul>
<li>
<p>How does the weighting of parents affect the hybrid‚Äôs performance?</p>
<p>Stronger weights mean more genetic contribution. Traits from highly weighted parents dominate the hybrid profile.</p>
</li>
<li>
<p>What does the identity matrix represent here?</p>
<p>It represents a neutral transformation. It confirms data integrity when used in matrix multiplication.</p>
</li>
<li>
<p>If you used equal weights (‚Öì for each), how would the hybrid traits change?</p>
<p>Traits would reflect an even mix, potentially leading to balanced but less specialized performance.</p>
</li>
<li>
<p>What real-world limitations does this simplified model ignore?</p>
<ol type="i">
<li><p>Non-additive genetic effects (dominance, epistasis)</p></li>
<li><p>Environmental interactions</p></li>
<li><p>Trait heritability and correlations</p></li>
<li><p>Breeding feasibility and cost</p></li>
</ol>
</li>
</ul></section></section><section id="task-5-managing-matrices-and-weight-vectors-using-lists-in-r" class="level2"><h2 class="anchored" data-anchor-id="task-5-managing-matrices-and-weight-vectors-using-lists-in-r">Task 5: Managing Matrices and Weight Vectors Using Lists in R</h2>
<p>Now that we‚Äôve explored trait-based decisions using matrices, it‚Äôs time to organize our work using R‚Äôs list structure. Lists help bundle related objects like matrices and weight vectors, keeping the analysis modular and scalable.</p>
<section id="create-a-master-list" class="level3"><h3 class="anchored" data-anchor-id="create-a-master-list">Create a Master List</h3>
<div class="cell">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Assuming previous matrices and weights are already defined:</span></span>
<span></span>
<span><span class="co"># Making a MasterList</span></span>
<span><span class="va">bioList</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  ProteinConc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">ProteinMatrix</span>, weights <span class="op">=</span> <span class="va">WeightVector</span><span class="op">)</span>,</span>
<span>  ProteinMap  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">GeneExpression</span>, weights <span class="op">=</span> <span class="va">TranslationMatrix</span><span class="op">)</span>,</span>
<span>  Animal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">BullEBVs</span>, weights <span class="op">=</span> <span class="va">EconomicWeights</span><span class="op">)</span>,</span>
<span>  Plant <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">ParentTraits</span>, weights <span class="op">=</span> <span class="va">HybridWeights</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">bioList</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$ProteinConc
$ProteinConc$matrix
        ProteinX ProteinY ProteinZ
Sample1        5        3        2
Sample2        7        6        4

$ProteinConc$weights
         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5


$ProteinMap
$ProteinMap$matrix
        GeneA GeneB GeneC
Sample1    10     8     5
Sample2    15    12    10

$ProteinMap$weights
      protA protB protC
GeneA   1.5   0.0   0.0
GeneB   0.0   1.2   0.0
GeneC   0.0   0.0   1.8


$Animal
$Animal$matrix
      Milk_yield Growth_rate Fertility
Bull1        400         1.2       0.8
Bull2        500         1.5       0.6

$Animal$weights
            Weight
Milk_yield   2e-03
Growth_rate  5e+01
Fertility    1e+02


$Plant
$Plant$matrix
   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6

$Plant$weights
       Drought_resistance Yield Maturation_time
Weight                0.5   0.3             0.2</code></pre>
</div>
</div>
</section><section id="list-the-structure" class="level3"><h3 class="anchored" data-anchor-id="list-the-structure">List the Structure</h3>
<div class="cell">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">bioList</span><span class="op">)</span>               <span class="co"># Top-level list names</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ProteinConc" "ProteinMap"  "Animal"      "Plant"      </code></pre>
</div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span><span class="op">(</span><span class="va">bioList</span><span class="op">)</span>             <span class="co"># Number of components in each sublist</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ProteinConc  ProteinMap      Animal       Plant 
          2           2           2           2 </code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Each top-level entry (e.g., ProteinConc, Plant) contains two components:</p>
<ul>
<li>A matrix (e.g., ProteinMatrix)</li>
<li>A corresponding weight vector or matrix</li>
</ul></section><section id="indexing-elements-from-lists" class="level3"><h3 class="anchored" data-anchor-id="indexing-elements-from-lists">Indexing Elements from Lists</h3>
<div class="cell">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Access the trait matrix for Plant</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#or</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Access the weight vector for ProteinConc</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5</code></pre>
</div>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#or</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Use double brackets [[ ]] to extract unnamed list elements by position. But we named our list, so they are easily extractable using the <code>$</code> notation.</p>
</section><section id="perform-weighted-calculations" class="level3"><h3 class="anchored" data-anchor-id="perform-weighted-calculations">Perform Weighted Calculations</h3>
<div class="cell">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Protein concentration score</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Weight
Sample1    8.5
Sample2   15.5</code></pre>
</div>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Gene ‚Üí Protein contribution</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinMap</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">ProteinMap</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        protA protB protC
Sample1  15.0   9.6     9
Sample2  22.5  14.4    18</code></pre>
</div>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Bull economic value</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Weight
Bull1  140.8
Bull2  136.0</code></pre>
</div>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Hybrid trait value</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">weights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Drought_resistance Yield Maturation_time
Weight                6.3   6.1             3.9</code></pre>
</div>
</div>
</section><section id="subset-and-recalculate" class="level3"><h3 class="anchored" data-anchor-id="subset-and-recalculate">Subset and Recalculate</h3>
<div class="cell">
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Remove last trait from ParentTraits</span></span>
<span><span class="va">ParentSubset</span> <span class="op">&lt;-</span> <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">NewWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.4</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Recalculated hybrid score</span></span>
<span><span class="va">SubsetHybridScore</span> <span class="op">&lt;-</span> <span class="va">ParentSubset</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">NewWeights</span></span>
<span><span class="va">SubsetHybridScore</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   [,1]
P1  6.2
P2  6.8
P3  5.4</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Dropping a trait and reweighting highlights its influence in trait aggregation and selection.</p>
</section><section id="visualization-tasks-1" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks-1">Visualization Tasks</h3>
<section id="heatmap-gene-expression" class="level4"><h4 class="anchored" data-anchor-id="heatmap-gene-expression">Heatmap: Gene Expression</h4>
<div class="cell">
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">bioList</span><span class="op">$</span><span class="va">ProteinMap</span><span class="op">$</span><span class="va">matrix</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Gene Expression Heatmap"</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"Proteins"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Genes"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="bar-plots" class="level4"><h4 class="anchored" data-anchor-id="bar-plots">Bar Plots</h4>
<section id="hybrid-traits" class="level5"><h5 class="anchored" data-anchor-id="hybrid-traits">Hybrid traits</h5>
<div class="cell">
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">weights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Hybrid Trait Contributions"</span>,</span>
<span>  col <span class="op">=</span> <span class="st">"#66c2a5"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Score"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="bull-ebv-economic-breeding-values" class="level5"><h5 class="anchored" data-anchor-id="bull-ebv-economic-breeding-values">Bull EBV (Economic Breeding Values)</h5>
<div class="cell">
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">weights</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Bull EBVs (Economic Values)"</span>,</span>
<span>  col <span class="op">=</span> <span class="st">"#fc8d62"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Score"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section></section></section><section id="interpretation-questions-2" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-2">üß† Interpretation Questions</h3>
<ul>
<li>Why use a list structure?</li>
</ul>
<p>Keeps each dataset and its weights together. Facilitates automated workflows and reuse.</p>
<ul>
<li>What‚Äôs tricky about <code>[[ ]]</code> access?</li>
</ul>
<p>You must remember the order (<code>[[1]]</code> = <code>matrix</code>, <code>[[2]]</code> = <code>weights</code>). No names means you can‚Äôt use $matrix, only positional access.</p>
<p>Loop across all list entries</p>
<p>Weighted scores for all entries <code>lapply(bioList, function(x) x[[2]] %*% x[[1]])</code></p>
<ul>
<li>How does this help in large-scale pipelines?</li>
</ul>
<p>You can use this format with <code><a href="https://rdrr.io/r/base/lapply.html">lapply()</a></code>, <code><a href="https://purrr.tidyverse.org/reference/map.html">purrr::map()</a></code>, or in targets pipelines for reproducibility and modular processing.</p>
</section></section><section id="homework-solutions-factors-subsetting-and-biological-insight" class="level2"><h2 class="anchored" data-anchor-id="homework-solutions-factors-subsetting-and-biological-insight">üß© Homework solutions: Factors, Subsetting, and Biological Insight</h2>
<ol type="1">
<li><p><strong>Character vs Factor</strong><br>
A character vector simply holds string values, but a factor is a categorical variable with fixed levels, used especially in modeling.<br>
For <code>mutation_status</code>, a factor ensures consistent categories (e.g., <code>"Yes"</code> or <code>"No"</code>) and helps control level order and statistical reference groups.</p></li>
<li><p><strong>Factor Levels</strong></p></li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Lactobacillus"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Lactobacillus"</span><span class="op">)</span></span>
<span><span class="va">species_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">species</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Lactobacillus"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">species_factor</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Bacteroides"   "Escherichia"   "Lactobacillus"</code></pre>
</div>
</div>
<p>Because we defined the level order explicitly, R maintains that order regardless of data input.</p>
<ol start="3" type="1">
<li>Ordered Factor Comparison</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">disease_severity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mild"</span>, <span class="st">"Severe"</span>, <span class="st">"Moderate"</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mild"</span>, <span class="st">"Moderate"</span>, <span class="st">"Severe"</span>, <span class="st">"Critical"</span><span class="op">)</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">disease_severity</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="va">disease_severity</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>‚ÄúMild‚Äù is less severe than ‚ÄúSevere‚Äù based on the defined order.</p>
<ol start="4" type="1">
<li>Proportion Extraction</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">species_factor</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">prop</span><span class="op">[</span><span class="st">"Escherichia"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Escherichia 
        0.2 </code></pre>
</div>
</div>
<p><code>prop$Escherichia</code> won‚Äôt work ‚Äî named numeric vectors require bracket-based access.</p>
<ol start="5" type="1">
<li>Subsetting by Conditions</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"EGFR"</span>, <span class="st">"GAPDH"</span><span class="op">)</span>,</span>
<span>  expression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8.2</span>, <span class="fl">6.1</span>, <span class="fl">9.5</span>, <span class="fl">7.0</span>, <span class="fl">10.0</span><span class="op">)</span>,</span>
<span>  mutation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"No"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  pathway <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DNA Repair"</span>, <span class="st">"Apoptosis"</span>, <span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>, <span class="st">"Metabolism"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene_df</span><span class="op">$</span><span class="va">gene_id</span> <span class="co">#name the rows by the gene IDs</span></span>
<span><span class="va">gene_df</span> <span class="op">&lt;-</span> <span class="va">gene_df</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="co">#remove the first column which is not needed anymore</span></span>
<span><span class="co">#gene_df</span></span>
<span><span class="va">gene_df</span><span class="op">[</span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">7</span> <span class="op">&amp;</span> <span class="va">gene_df</span><span class="op">$</span><span class="va">mutation</span> <span class="op">==</span> <span class="st">"No"</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      expression mutation    pathway
GAPDH         10       No Metabolism</code></pre>
</div>
</div>
<p>Returns genes with <strong>high expression (&gt;7)</strong> <code>and</code> <strong>no mutation</strong> ‚Äî potentially highly active but wild-type genes.</p>
<ol start="6" type="1">
<li>Group-wise Expression Summary</li>
</ol>
<p>The given vectors are:</p>
<div class="cell">
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span><span class="op">)</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5.2</span>, <span class="fl">8.1</span>, <span class="fl">4.3</span>, <span class="fl">9.0</span>, <span class="fl">5.7</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The solution would be:</p>
<div class="cell">
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">group_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Mean expression</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span><span class="op">(</span><span class="va">expression</span>, <span class="va">group_factor</span>, <span class="va">mean</span><span class="op">)</span> <span class="co">## KO: 8.55, WT: 5.07</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      KO       WT 
8.550000 5.066667 </code></pre>
</div>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span><span class="op">(</span><span class="va">expression</span>, <span class="va">group_factor</span>, <span class="va">mean</span><span class="op">)</span>, </span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"skyblue"</span>, <span class="st">"salmon"</span><span class="op">)</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Mean Expression"</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Group-wise Expression"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<ol start="7" type="1">
<li>Gene Subsetting</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_df</span><span class="op">[</span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">8</span> <span class="op">&amp;</span> </span>
<span>        <span class="va">gene_df</span><span class="op">$</span><span class="va">pathway</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span><span class="op">)</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    expression mutation    pathway
MYC        9.5      Yes Cell Cycle</code></pre>
</div>
</div>
<p>It filters for genes highly expressed and involved in key biological pathways.</p>
<ol start="8" type="1">
<li>Disease Stage Visualization</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">stages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Stage I"</span>, <span class="st">"Stage III"</span>, <span class="st">"Stage II"</span>, <span class="st">"Stage IV"</span>, <span class="st">"Stage I"</span><span class="op">)</span></span>
<span><span class="va">disease_stage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">stages</span>, </span>
<span>                        levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Stage I"</span>, <span class="st">"Stage II"</span>, <span class="st">"Stage III"</span>, <span class="st">"Stage IV"</span><span class="op">)</span>, </span>
<span>                        ordered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">disease_stage</span><span class="op">)</span>, </span>
<span>        col <span class="op">=</span> <span class="st">"lightgreen"</span>, </span>
<span>        main <span class="op">=</span> <span class="st">"Patient Count by Disease Stage"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<p>Let‚Äôs do the severity order check:</p>
<div class="cell">
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Comparison</span></span>
<span><span class="va">disease_stage</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">disease_stage</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>  <span class="co"># TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>So, ‚ÄúStage III‚Äù is more sever than ‚ÄúStage I‚Äù.</p>
<ol start="9" type="1">
<li>Oncogene Subsetting and Releveling</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define a small gene dataset</span></span>
<span><span class="va">gene_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"MYC"</span>, <span class="st">"GAPDH"</span>, <span class="st">"EGFR"</span><span class="op">)</span>,</span>
<span>  expression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">9.1</span>, <span class="fl">7.3</span>, <span class="fl">10.5</span>, <span class="fl">5.2</span>, <span class="fl">8.6</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Tumor Suppressor"</span>, <span class="st">"Oncogene"</span>, <span class="st">"Oncogene"</span>, <span class="st">"Housekeeping"</span>, <span class="st">"Oncogene"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset: Oncogene rows with expression &gt; 8</span></span>
<span><span class="va">gene_data</span><span class="op">[</span><span class="va">gene_data</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"Oncogene"</span> <span class="op">&amp;</span> <span class="va">gene_data</span><span class="op">$</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">8</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  gene expression     type
3  MYC       10.5 Oncogene
5 EGFR        8.6 Oncogene</code></pre>
</div>
</div>
<p>Let‚Äôs relevel now, ‚ÄúHousekeeping‚Äù is the reference:</p>
<div class="cell">
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Relevel: make "Housekeeping" the reference level</span></span>
<span><span class="va">gene_data</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">gene_data</span><span class="op">$</span><span class="va">type</span>, ref <span class="op">=</span> <span class="st">"Housekeeping"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check the new levels</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">gene_data</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Housekeeping"     "Oncogene"         "Tumor Suppressor"</code></pre>
</div>
</div>
<ol start="10" type="1">
<li>Simulated Expression by Tissue</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">gene_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">45</span>, mean <span class="op">=</span> <span class="fl">8</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"brain"</span>, <span class="st">"liver"</span>, <span class="st">"kidney"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="va">tissue_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">tissue</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"liver"</span>, <span class="st">"brain"</span>, <span class="st">"kidney"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">gene_expr</span> <span class="op">~</span> <span class="va">tissue_factor</span>, </span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"orange"</span>, <span class="st">"skyblue"</span>, <span class="st">"lightgreen"</span><span class="op">)</span>, </span>
<span>        main <span class="op">=</span> <span class="st">"Expression by Tissue"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Expression Level"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<p>Let‚Äôs calculate variability per tissue type now:</p>
<div class="cell">
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Variability</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span><span class="op">(</span><span class="va">gene_expr</span>, <span class="va">tissue_factor</span>, <span class="va">sd</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   liver    brain   kidney 
2.713940 2.050487 1.993668 </code></pre>
</div>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Returns standard deviation per tissue group</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


<!-- -->

</section></section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents"><h2 class="anchored quarto-appendix-heading">Citation</h2><div><div class="quarto-appendix-secondary-label">BibTeX citation:</div><pre class="sourceCode code-with-copy quarto-appendix-bibtex"><code class="sourceCode bibtex">@online{rasheduzzaman2025,
  author = {Md Rasheduzzaman},
  title = {HW Solutions},
  date = {2025-05-10},
  langid = {en},
  abstract = {dataframe, matrices, list, factor, vector, etc.}
}
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre><div class="quarto-appendix-secondary-label">For attribution, please cite this work as:</div><div id="ref-rasheduzzaman2025" class="csl-entry quarto-appendix-citeas" role="listitem">
Md Rasheduzzaman. 2025. <span>‚ÄúHW Solutions.‚Äù</span> May 10, 2025.
</div></div></section></div></main><!-- /main --><hr>
<div style="margin-top: 2em;">
  <p style="font-weight: 500; font-size: 1.1em;">
    üí¨ Have thoughts or questions? Join the discussion below using your GitHub account!
  </p>
  <p style="font-size: 0.95em; color: #666;">
    You can <strong>edit or delete</strong> your own comments. Reactions like üëç ‚ù§Ô∏è üöÄ are also supported.
  </p>
</div>

<div id="giscus_thread" style="margin-top: 1.5em;"></div>

<script src="https://giscus.app/client.js" data-repo="MdRasheduz-zaman/PythRaSh" data-repo-id="R_kgDOOZN9Xw" data-category-id="DIC_kwDOOZN9X84CpE6N" data-mapping="pathname" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="preferred_color_scheme" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "Óßã";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb117" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "HW solutions"</span></span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a><span class="an">abstract:</span><span class="co"> "dataframe, matrices, list, factor, vector, etc."</span></span>
<span id="cb117-4"><a href="#cb117-4" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb117-5"><a href="#cb117-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb117-6"><a href="#cb117-6" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;style&gt;</span></span>
<span id="cb117-7"><a href="#cb117-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">.content</span> {</span>
<span id="cb117-8"><a href="#cb117-8" aria-hidden="true" tabindex="-1"></a>    <span class="kw">text-align</span>: <span class="dv">justify</span> <span class="at">!important</span><span class="op">;</span></span>
<span id="cb117-9"><a href="#cb117-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb117-10"><a href="#cb117-10" aria-hidden="true" tabindex="-1"></a>  p {</span>
<span id="cb117-11"><a href="#cb117-11" aria-hidden="true" tabindex="-1"></a>    <span class="kw">text-align</span>: <span class="dv">justify</span> <span class="at">!important</span><span class="op">;</span></span>
<span id="cb117-12"><a href="#cb117-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb117-13"><a href="#cb117-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">.quarto-body</span> {</span>
<span id="cb117-14"><a href="#cb117-14" aria-hidden="true" tabindex="-1"></a>    <span class="kw">text-align</span>: <span class="dv">justify</span> <span class="at">!important</span><span class="op">;</span></span>
<span id="cb117-15"><a href="#cb117-15" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb117-16"><a href="#cb117-16" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/style&gt;</span></span>
<span id="cb117-17"><a href="#cb117-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-18"><a href="#cb117-18" aria-hidden="true" tabindex="-1"></a><span class="fu"># L3: Matrices and Lists</span></span>
<span id="cb117-19"><a href="#cb117-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-20"><a href="#cb117-20" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 1: Protein Quantification in Biological Samples</span></span>
<span id="cb117-21"><a href="#cb117-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-22"><a href="#cb117-22" aria-hidden="true" tabindex="-1"></a>We measured the concentration (in ¬µg/¬µL) of three proteins (P1, P2, P3) in four samples (S1‚ÄìS4):</span>
<span id="cb117-23"><a href="#cb117-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-24"><a href="#cb117-24" aria-hidden="true" tabindex="-1"></a><span class="fu">### 1. Matrices</span></span>
<span id="cb117-25"><a href="#cb117-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-28"><a href="#cb117-28" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-29"><a href="#cb117-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Making Protein Matrix</span></span>
<span id="cb117-30"><a href="#cb117-30" aria-hidden="true" tabindex="-1"></a>ProteinMatrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb117-31"><a href="#cb117-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">2</span>,</span>
<span id="cb117-32"><a href="#cb117-32" aria-hidden="true" tabindex="-1"></a>    <span class="dv">7</span>, <span class="dv">6</span>, <span class="dv">4</span>),</span>
<span id="cb117-33"><a href="#cb117-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb117-34"><a href="#cb117-34" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-35"><a href="#cb117-35" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ProteinMatrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Sample1"</span>, <span class="st">"Sample2"</span>)</span>
<span id="cb117-36"><a href="#cb117-36" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ProteinMatrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb117-37"><a href="#cb117-37" aria-hidden="true" tabindex="-1"></a>ProteinMatrix</span>
<span id="cb117-38"><a href="#cb117-38" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-39"><a href="#cb117-39" aria-hidden="true" tabindex="-1"></a>Now goes the weight matrix</span>
<span id="cb117-40"><a href="#cb117-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-43"><a href="#cb117-43" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-44"><a href="#cb117-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Making weight matrix</span></span>
<span id="cb117-45"><a href="#cb117-45" aria-hidden="true" tabindex="-1"></a>WeightVector <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb117-46"><a href="#cb117-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">1.5</span>),</span>
<span id="cb117-47"><a href="#cb117-47" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow=</span><span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb117-48"><a href="#cb117-48" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-49"><a href="#cb117-49" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb117-50"><a href="#cb117-50" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb117-51"><a href="#cb117-51" aria-hidden="true" tabindex="-1"></a>WeightVector</span>
<span id="cb117-52"><a href="#cb117-52" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-53"><a href="#cb117-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-54"><a href="#cb117-54" aria-hidden="true" tabindex="-1"></a>Now, multiply them.</span>
<span id="cb117-55"><a href="#cb117-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-58"><a href="#cb117-58" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-59"><a href="#cb117-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Multiplying Matrices</span></span>
<span id="cb117-60"><a href="#cb117-60" aria-hidden="true" tabindex="-1"></a>TotalConc <span class="ot">=</span> ProteinMatrix <span class="sc">%*%</span> WeightVector</span>
<span id="cb117-61"><a href="#cb117-61" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TotalConc) <span class="ot">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span id="cb117-62"><a href="#cb117-62" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(TotalConc)</span>
<span id="cb117-63"><a href="#cb117-63" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-64"><a href="#cb117-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-65"><a href="#cb117-65" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2. Making transpose of ProteinMatrix</span></span>
<span id="cb117-66"><a href="#cb117-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-69"><a href="#cb117-69" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-70"><a href="#cb117-70" aria-hidden="true" tabindex="-1"></a>ProteinMatTranspose <span class="ot">=</span> <span class="fu">t</span>(ProteinMatrix)</span>
<span id="cb117-71"><a href="#cb117-71" aria-hidden="true" tabindex="-1"></a>ProteinMatTranspose</span>
<span id="cb117-72"><a href="#cb117-72" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-73"><a href="#cb117-73" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3. Identity Matrix</span></span>
<span id="cb117-74"><a href="#cb117-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-77"><a href="#cb117-77" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-78"><a href="#cb117-78" aria-hidden="true" tabindex="-1"></a>I <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb117-79"><a href="#cb117-79" aria-hidden="true" tabindex="-1"></a>Identitycheck <span class="ot">=</span> ProteinMatrix <span class="sc">%*%</span> I</span>
<span id="cb117-80"><a href="#cb117-80" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Identitycheck) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb117-81"><a href="#cb117-81" aria-hidden="true" tabindex="-1"></a>Identitycheck</span>
<span id="cb117-82"><a href="#cb117-82" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-83"><a href="#cb117-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-84"><a href="#cb117-84" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4.1. Total Protein per Sample</span></span>
<span id="cb117-85"><a href="#cb117-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-88"><a href="#cb117-88" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-89"><a href="#cb117-89" aria-hidden="true" tabindex="-1"></a><span class="fu">rowSums</span>(ProteinMatrix)</span>
<span id="cb117-90"><a href="#cb117-90" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-91"><a href="#cb117-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-92"><a href="#cb117-92" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4.2. Total Protein per Protein Type</span></span>
<span id="cb117-93"><a href="#cb117-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-96"><a href="#cb117-96" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-97"><a href="#cb117-97" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(ProteinMatrix)</span>
<span id="cb117-98"><a href="#cb117-98" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-99"><a href="#cb117-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-100"><a href="#cb117-100" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4.3. Heatmap of Protein Concentrations</span></span>
<span id="cb117-101"><a href="#cb117-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-104"><a href="#cb117-104" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-105"><a href="#cb117-105" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(ProteinMatrix, <span class="at">scale =</span> <span class="st">"none"</span>, <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">10</span>))</span>
<span id="cb117-106"><a href="#cb117-106" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-107"><a href="#cb117-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-108"><a href="#cb117-108" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation</span></span>
<span id="cb117-109"><a href="#cb117-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-110"><a href="#cb117-110" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Multiplying the protein levels by the weight vector shows how much each protein contributes in a sample. The result shows total protein concentration per sample.</span>
<span id="cb117-111"><a href="#cb117-111" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The result shows that sample S2 has the highest protein burden.</span>
<span id="cb117-112"><a href="#cb117-112" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The identity matrix represents no protein interactions or measurement biases. It is a simple matrix calculation.</span>
<span id="cb117-113"><a href="#cb117-113" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>New calculation:</span>
<span id="cb117-114"><a href="#cb117-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-117"><a href="#cb117-117" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-118"><a href="#cb117-118" aria-hidden="true" tabindex="-1"></a><span class="co"># changing the weight of ProteinZ to 3.0</span></span>
<span id="cb117-119"><a href="#cb117-119" aria-hidden="true" tabindex="-1"></a>newweightvector <span class="ot">=</span> <span class="fu">matrix</span>(</span>
<span id="cb117-120"><a href="#cb117-120" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">3.0</span>),</span>
<span id="cb117-121"><a href="#cb117-121" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow=</span><span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb117-122"><a href="#cb117-122" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-123"><a href="#cb117-123" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb117-124"><a href="#cb117-124" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb117-125"><a href="#cb117-125" aria-hidden="true" tabindex="-1"></a>newTotalconc <span class="ot">=</span> ProteinMatrix <span class="sc">%*%</span> newweightvector</span>
<span id="cb117-126"><a href="#cb117-126" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(newTotalconc) <span class="ot">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span id="cb117-127"><a href="#cb117-127" aria-hidden="true" tabindex="-1"></a>newTotalconc</span>
<span id="cb117-128"><a href="#cb117-128" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-129"><a href="#cb117-129" aria-hidden="true" tabindex="-1"></a>Still, S2 has more protein burden.</span>
<span id="cb117-130"><a href="#cb117-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-131"><a href="#cb117-131" aria-hidden="true" tabindex="-1"></a>Bonus:</span>
<span id="cb117-132"><a href="#cb117-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-133"><a href="#cb117-133" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Heatmap reveals PX is most abundant across all samples.</span>
<span id="cb117-134"><a href="#cb117-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-135"><a href="#cb117-135" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb117-136"><a href="#cb117-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-137"><a href="#cb117-137" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 2: Gene-to-Protein Translation</span></span>
<span id="cb117-138"><a href="#cb117-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-141"><a href="#cb117-141" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-142"><a href="#cb117-142" aria-hidden="true" tabindex="-1"></a><span class="co"># making  Gene Expression matrix</span></span>
<span id="cb117-143"><a href="#cb117-143" aria-hidden="true" tabindex="-1"></a>GeneExpression <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb117-144"><a href="#cb117-144" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">8</span>, <span class="dv">5</span>,</span>
<span id="cb117-145"><a href="#cb117-145" aria-hidden="true" tabindex="-1"></a>    <span class="dv">15</span>, <span class="dv">12</span>, <span class="dv">10</span>),</span>
<span id="cb117-146"><a href="#cb117-146" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb117-147"><a href="#cb117-147" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-148"><a href="#cb117-148" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(GeneExpression) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Sample1"</span>, <span class="st">"Sample2"</span>)</span>
<span id="cb117-149"><a href="#cb117-149" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(GeneExpression) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb117-150"><a href="#cb117-150" aria-hidden="true" tabindex="-1"></a>GeneExpression</span>
<span id="cb117-151"><a href="#cb117-151" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-152"><a href="#cb117-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-153"><a href="#cb117-153" aria-hidden="true" tabindex="-1"></a>Translation efficiency:</span>
<span id="cb117-154"><a href="#cb117-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-157"><a href="#cb117-157" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-158"><a href="#cb117-158" aria-hidden="true" tabindex="-1"></a><span class="co"># making Translation Matrix</span></span>
<span id="cb117-159"><a href="#cb117-159" aria-hidden="true" tabindex="-1"></a>TranslationMatrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb117-160"><a href="#cb117-160" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="dv">0</span> , <span class="dv">0</span>,</span>
<span id="cb117-161"><a href="#cb117-161" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span>, <span class="fl">1.2</span>, <span class="dv">0</span>,</span>
<span id="cb117-162"><a href="#cb117-162" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">1.8</span>),</span>
<span id="cb117-163"><a href="#cb117-163" aria-hidden="true" tabindex="-1"></a><span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb117-164"><a href="#cb117-164" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-165"><a href="#cb117-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-166"><a href="#cb117-166" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(TranslationMatrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb117-167"><a href="#cb117-167" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TranslationMatrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"protA"</span>, <span class="st">"protB"</span>, <span class="st">"protC"</span>)</span>
<span id="cb117-168"><a href="#cb117-168" aria-hidden="true" tabindex="-1"></a>TranslationMatrix</span>
<span id="cb117-169"><a href="#cb117-169" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-170"><a href="#cb117-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-171"><a href="#cb117-171" aria-hidden="true" tabindex="-1"></a><span class="fu">### Protein Output</span></span>
<span id="cb117-172"><a href="#cb117-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-175"><a href="#cb117-175" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-176"><a href="#cb117-176" aria-hidden="true" tabindex="-1"></a><span class="co"># computing Protein matrix</span></span>
<span id="cb117-177"><a href="#cb117-177" aria-hidden="true" tabindex="-1"></a>Protein_matrix <span class="ot">&lt;-</span> GeneExpression <span class="sc">%*%</span> TranslationMatrix</span>
<span id="cb117-178"><a href="#cb117-178" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Protein_matrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"total_protA"</span>, <span class="st">"total_protB"</span>, <span class="st">"total_protC"</span>)</span>
<span id="cb117-179"><a href="#cb117-179" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Protein_matrix)</span>
<span id="cb117-180"><a href="#cb117-180" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-181"><a href="#cb117-181" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2. Transpose</span></span>
<span id="cb117-184"><a href="#cb117-184" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-185"><a href="#cb117-185" aria-hidden="true" tabindex="-1"></a><span class="co"># Transpose of GeneExpression matrix</span></span>
<span id="cb117-186"><a href="#cb117-186" aria-hidden="true" tabindex="-1"></a>GeneExpression_Transpose <span class="ot">&lt;-</span> <span class="fu">t</span>(GeneExpression)</span>
<span id="cb117-187"><a href="#cb117-187" aria-hidden="true" tabindex="-1"></a>GeneExpression_Transpose</span>
<span id="cb117-188"><a href="#cb117-188" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-189"><a href="#cb117-189" aria-hidden="true" tabindex="-1"></a>The new matrix represnts a matrix where the rows and columns of GeneExpression matrix have been interchanged.</span>
<span id="cb117-190"><a href="#cb117-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-191"><a href="#cb117-191" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3. Identity matrix and multipliccation</span></span>
<span id="cb117-194"><a href="#cb117-194" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-195"><a href="#cb117-195" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating Identity matrix</span></span>
<span id="cb117-196"><a href="#cb117-196" aria-hidden="true" tabindex="-1"></a>I <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb117-197"><a href="#cb117-197" aria-hidden="true" tabindex="-1"></a>I</span>
<span id="cb117-198"><a href="#cb117-198" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-199"><a href="#cb117-199" aria-hidden="true" tabindex="-1"></a>Now, multiply:</span>
<span id="cb117-202"><a href="#cb117-202" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-203"><a href="#cb117-203" aria-hidden="true" tabindex="-1"></a>Product_matrix <span class="ot">=</span> TranslationMatrix <span class="sc">%*%</span>  I </span>
<span id="cb117-204"><a href="#cb117-204" aria-hidden="true" tabindex="-1"></a>Product_matrix</span>
<span id="cb117-205"><a href="#cb117-205" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-206"><a href="#cb117-206" aria-hidden="true" tabindex="-1"></a>The product is identical to TranslationMatrix</span>
<span id="cb117-207"><a href="#cb117-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-208"><a href="#cb117-208" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4. Sub-matrix:</span></span>
<span id="cb117-211"><a href="#cb117-211" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-212"><a href="#cb117-212" aria-hidden="true" tabindex="-1"></a><span class="co"># making submatrix A</span></span>
<span id="cb117-213"><a href="#cb117-213" aria-hidden="true" tabindex="-1"></a>A <span class="ot">=</span> <span class="fu">matrix</span>(</span>
<span id="cb117-214"><a href="#cb117-214" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">8</span>,</span>
<span id="cb117-215"><a href="#cb117-215" aria-hidden="true" tabindex="-1"></a>    <span class="dv">15</span>, <span class="dv">12</span>), <span class="at">nrow=</span><span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb117-216"><a href="#cb117-216" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-217"><a href="#cb117-217" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(A) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"sample1"</span>, <span class="st">"sample2"</span>)</span>
<span id="cb117-218"><a href="#cb117-218" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(A) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>)</span>
<span id="cb117-219"><a href="#cb117-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-220"><a href="#cb117-220" aria-hidden="true" tabindex="-1"></a>A</span>
<span id="cb117-221"><a href="#cb117-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-222"><a href="#cb117-222" aria-hidden="true" tabindex="-1"></a><span class="co"># finding inverse of A</span></span>
<span id="cb117-223"><a href="#cb117-223" aria-hidden="true" tabindex="-1"></a><span class="co">#inv_A &lt;- solve(A)</span></span>
<span id="cb117-224"><a href="#cb117-224" aria-hidden="true" tabindex="-1"></a><span class="co">#inv_A</span></span>
<span id="cb117-225"><a href="#cb117-225" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-226"><a href="#cb117-226" aria-hidden="true" tabindex="-1"></a>The inverse matrix could not be calculated since A is a singular matrix.</span>
<span id="cb117-227"><a href="#cb117-227" aria-hidden="true" tabindex="-1"></a>So, A * A^-1 is also not possible.</span>
<span id="cb117-228"><a href="#cb117-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-229"><a href="#cb117-229" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualization tasks:</span></span>
<span id="cb117-230"><a href="#cb117-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-231"><a href="#cb117-231" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>1. MARplot</span>
<span id="cb117-234"><a href="#cb117-234" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-235"><a href="#cb117-235" aria-hidden="true" tabindex="-1"></a><span class="co"># generating MARplot-style scatter plot</span></span>
<span id="cb117-236"><a href="#cb117-236" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(GeneExpression, Protein_matrix, <span class="at">type=</span><span class="st">"p"</span>, <span class="at">main=</span><span class="st">"Protein level vs. Gene Expression level"</span>)</span>
<span id="cb117-237"><a href="#cb117-237" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="st">"Sample-Gene"</span></span>
<span id="cb117-238"><a href="#cb117-238" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(GeneExpression, Protein_matrix, <span class="at">labels =</span> labels, <span class="at">pos=</span><span class="dv">3</span>)</span>
<span id="cb117-239"><a href="#cb117-239" aria-hidden="true" tabindex="-1"></a><span class="co"># generating a heatmap</span></span>
<span id="cb117-240"><a href="#cb117-240" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(Protein_matrix, <span class="at">main=</span> <span class="st">"Heatmap of Protein Level"</span>, <span class="at">Rowv =</span> <span class="cn">TRUE</span>, <span class="at">Colv =</span> <span class="cn">TRUE</span>, <span class="at">labRow=</span> <span class="fu">rownames</span>(Protein_matrix), <span class="at">labCol=</span> <span class="fu">c</span>(<span class="st">"ProteinA"</span>, <span class="st">"ProteinB"</span>, <span class="st">"ProteinC"</span>), <span class="at">col=</span><span class="fu">topo.colors</span>(<span class="dv">256</span>) )</span>
<span id="cb117-241"><a href="#cb117-241" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-242"><a href="#cb117-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-243"><a href="#cb117-243" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>2. Heatmap of Expression:</span>
<span id="cb117-244"><a href="#cb117-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-247"><a href="#cb117-247" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-248"><a href="#cb117-248" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(GeneExpression, <span class="at">col =</span> <span class="fu">terrain.colors</span>(<span class="dv">10</span>), <span class="at">scale =</span> <span class="st">"column"</span>)</span>
<span id="cb117-249"><a href="#cb117-249" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-250"><a href="#cb117-250" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation</span></span>
<span id="cb117-251"><a href="#cb117-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-252"><a href="#cb117-252" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Matrix multiplication allows each gene in both samples to be multiplied to their respective translation efficiency. So, the product shows how successfully each gene is translated")</span>
<span id="cb117-253"><a href="#cb117-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-254"><a href="#cb117-254" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>The diagonal TranslationMatrix make sense biologically because they show translation efficiency of each gene and there is no other interaction between them. Although there could be interaction in real-world scenarios.</span>
<span id="cb117-255"><a href="#cb117-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-256"><a href="#cb117-256" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>If Sample2 has higher protein levels even with similar gene expression, it means that more mRNAs are translated to proteins compared to Sample1"</span>
<span id="cb117-257"><a href="#cb117-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-258"><a href="#cb117-258" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>The upward trend in MARplot may indicate an increase in translation efficacy and downward trend may indicate a decline in translation efficacy"</span>
<span id="cb117-259"><a href="#cb117-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-260"><a href="#cb117-260" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>Clustering in the heatmap may suggest which samples are most similar to each other based on their prot.</span>
<span id="cb117-261"><a href="#cb117-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-262"><a href="#cb117-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-263"><a href="#cb117-263" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 3: Animal Breeding ‚Äì Bull Ranking by Economic Traits</span></span>
<span id="cb117-264"><a href="#cb117-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-265"><a href="#cb117-265" aria-hidden="true" tabindex="-1"></a><span class="fu">### Define Data</span></span>
<span id="cb117-266"><a href="#cb117-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-269"><a href="#cb117-269" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-270"><a href="#cb117-270" aria-hidden="true" tabindex="-1"></a><span class="co"># Define Bull EBVs</span></span>
<span id="cb117-271"><a href="#cb117-271" aria-hidden="true" tabindex="-1"></a>BullEBVs <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb117-272"><a href="#cb117-272" aria-hidden="true" tabindex="-1"></a>  <span class="dv">400</span>, <span class="fl">1.2</span>, <span class="fl">0.8</span>,</span>
<span id="cb117-273"><a href="#cb117-273" aria-hidden="true" tabindex="-1"></a>  <span class="dv">500</span>, <span class="fl">1.5</span>, <span class="fl">0.6</span></span>
<span id="cb117-274"><a href="#cb117-274" aria-hidden="true" tabindex="-1"></a>), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb117-275"><a href="#cb117-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-276"><a href="#cb117-276" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(BullEBVs) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Bull1"</span>, <span class="st">"Bull2"</span>)</span>
<span id="cb117-277"><a href="#cb117-277" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(BullEBVs) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Milk_yield"</span>, <span class="st">"Growth_rate"</span>, <span class="st">"Fertility"</span>)</span>
<span id="cb117-278"><a href="#cb117-278" aria-hidden="true" tabindex="-1"></a>BullEBVs</span>
<span id="cb117-279"><a href="#cb117-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-280"><a href="#cb117-280" aria-hidden="true" tabindex="-1"></a><span class="co"># Define Economic Weights</span></span>
<span id="cb117-281"><a href="#cb117-281" aria-hidden="true" tabindex="-1"></a>EconomicWeights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.002</span>, <span class="dv">50</span>, <span class="dv">100</span>), <span class="at">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb117-282"><a href="#cb117-282" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(EconomicWeights) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(BullEBVs)</span>
<span id="cb117-283"><a href="#cb117-283" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(EconomicWeights) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb117-284"><a href="#cb117-284" aria-hidden="true" tabindex="-1"></a>EconomicWeights</span>
<span id="cb117-285"><a href="#cb117-285" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-286"><a href="#cb117-286" aria-hidden="true" tabindex="-1"></a><span class="fu">### Compute Total Economic Value</span></span>
<span id="cb117-287"><a href="#cb117-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-290"><a href="#cb117-290" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-291"><a href="#cb117-291" aria-hidden="true" tabindex="-1"></a>TotalValue <span class="ot">&lt;-</span> BullEBVs <span class="sc">%*%</span> EconomicWeights</span>
<span id="cb117-292"><a href="#cb117-292" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TotalValue) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Merit"</span>)</span>
<span id="cb117-293"><a href="#cb117-293" aria-hidden="true" tabindex="-1"></a>TotalValue</span>
<span id="cb117-294"><a href="#cb117-294" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-295"><a href="#cb117-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-296"><a href="#cb117-296" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb117-297"><a href="#cb117-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-298"><a href="#cb117-298" aria-hidden="true" tabindex="-1"></a>Bull1: (400 √ó 0.002) + (1.2 √ó 50) + (0.8 √ó 100) = 140.8</span>
<span id="cb117-299"><a href="#cb117-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-300"><a href="#cb117-300" aria-hidden="true" tabindex="-1"></a>Bull2: (500 √ó 0.002) + (1.5 √ó 50) + (0.6 √ó 100) = 136.0</span>
<span id="cb117-301"><a href="#cb117-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-302"><a href="#cb117-302" aria-hidden="true" tabindex="-1"></a>Bull1 is more valuable economically.</span>
<span id="cb117-303"><a href="#cb117-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-304"><a href="#cb117-304" aria-hidden="true" tabindex="-1"></a>**Biological Interpretation**</span>
<span id="cb117-305"><a href="#cb117-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-306"><a href="#cb117-306" aria-hidden="true" tabindex="-1"></a>Economic weights convert genetic merit (EBVs, Estimated Breeding Values) into economic merit. Traits with higher financial importance have a larger impact, regardless of absolute EBV values.</span>
<span id="cb117-307"><a href="#cb117-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-308"><a href="#cb117-308" aria-hidden="true" tabindex="-1"></a><span class="fu">### Multiply with Identity Matrix</span></span>
<span id="cb117-309"><a href="#cb117-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-312"><a href="#cb117-312" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-313"><a href="#cb117-313" aria-hidden="true" tabindex="-1"></a>I3 <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb117-314"><a href="#cb117-314" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(I3) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(BullEBVs)</span>
<span id="cb117-315"><a href="#cb117-315" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(I3) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(BullEBVs)</span>
<span id="cb117-316"><a href="#cb117-316" aria-hidden="true" tabindex="-1"></a>I3</span>
<span id="cb117-317"><a href="#cb117-317" aria-hidden="true" tabindex="-1"></a>BullEBVs_identity <span class="ot">&lt;-</span> BullEBVs <span class="sc">%*%</span> I3</span>
<span id="cb117-318"><a href="#cb117-318" aria-hidden="true" tabindex="-1"></a>BullEBVs_identity</span>
<span id="cb117-319"><a href="#cb117-319" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-320"><a href="#cb117-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-321"><a href="#cb117-321" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb117-322"><a href="#cb117-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-323"><a href="#cb117-323" aria-hidden="true" tabindex="-1"></a>Multiplying by identity matrix returns the original matrix. It confirms that EBV structure is preserved.</span>
<span id="cb117-324"><a href="#cb117-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-325"><a href="#cb117-325" aria-hidden="true" tabindex="-1"></a><span class="fu">### Remove Milk Yield and Recalculate Total Value</span></span>
<span id="cb117-326"><a href="#cb117-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-329"><a href="#cb117-329" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-330"><a href="#cb117-330" aria-hidden="true" tabindex="-1"></a>BullEBVs_noMilk <span class="ot">&lt;-</span> BullEBVs[, <span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb117-331"><a href="#cb117-331" aria-hidden="true" tabindex="-1"></a>EconomicWeights_noMilk <span class="ot">&lt;-</span> EconomicWeights[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb117-332"><a href="#cb117-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-333"><a href="#cb117-333" aria-hidden="true" tabindex="-1"></a>TotalValue_noMilk <span class="ot">&lt;-</span> BullEBVs_noMilk <span class="sc">%*%</span> EconomicWeights_noMilk</span>
<span id="cb117-334"><a href="#cb117-334" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TotalValue_noMilk) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"New_Merit"</span>)</span>
<span id="cb117-335"><a href="#cb117-335" aria-hidden="true" tabindex="-1"></a>TotalValue_noMilk</span>
<span id="cb117-336"><a href="#cb117-336" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-337"><a href="#cb117-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-338"><a href="#cb117-338" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb117-339"><a href="#cb117-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-340"><a href="#cb117-340" aria-hidden="true" tabindex="-1"></a>Bull1: (1.2 √ó 50) + (0.8 √ó 100) = 140</span>
<span id="cb117-341"><a href="#cb117-341" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-342"><a href="#cb117-342" aria-hidden="true" tabindex="-1"></a>Bull2: (1.5 √ó 50) + (0.6 √ó 100) = 135</span>
<span id="cb117-343"><a href="#cb117-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-344"><a href="#cb117-344" aria-hidden="true" tabindex="-1"></a>Bull1 still ranks higher, but by a smaller margin.</span>
<span id="cb117-345"><a href="#cb117-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-346"><a href="#cb117-346" aria-hidden="true" tabindex="-1"></a><span class="fu">### Bar Plot: Total Economic Value</span></span>
<span id="cb117-347"><a href="#cb117-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-350"><a href="#cb117-350" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-351"><a href="#cb117-351" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb117-352"><a href="#cb117-352" aria-hidden="true" tabindex="-1"></a>  TotalValue,</span>
<span id="cb117-353"><a href="#cb117-353" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb117-354"><a href="#cb117-354" aria-hidden="true" tabindex="-1"></a>  <span class="at">names.arg =</span> <span class="fu">rownames</span>(BullEBVs),</span>
<span id="cb117-355"><a href="#cb117-355" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"orange"</span>),</span>
<span id="cb117-356"><a href="#cb117-356" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Total Economic Value of Bulls"</span>,</span>
<span id="cb117-357"><a href="#cb117-357" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Total Value"</span></span>
<span id="cb117-358"><a href="#cb117-358" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-359"><a href="#cb117-359" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-360"><a href="#cb117-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-361"><a href="#cb117-361" aria-hidden="true" tabindex="-1"></a><span class="fu">### Heatmap of EBVs</span></span>
<span id="cb117-362"><a href="#cb117-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-365"><a href="#cb117-365" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-366"><a href="#cb117-366" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb117-367"><a href="#cb117-367" aria-hidden="true" tabindex="-1"></a>  BullEBVs,</span>
<span id="cb117-368"><a href="#cb117-368" aria-hidden="true" tabindex="-1"></a>  <span class="at">Rowv =</span> <span class="cn">NA</span>,</span>
<span id="cb117-369"><a href="#cb117-369" aria-hidden="true" tabindex="-1"></a>  <span class="at">Colv =</span> <span class="cn">NA</span>,</span>
<span id="cb117-370"><a href="#cb117-370" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb117-371"><a href="#cb117-371" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">256</span>),</span>
<span id="cb117-372"><a href="#cb117-372" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Heatmap of Bull EBVs"</span></span>
<span id="cb117-373"><a href="#cb117-373" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-374"><a href="#cb117-374" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-375"><a href="#cb117-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-376"><a href="#cb117-376" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation Questions</span></span>
<span id="cb117-377"><a href="#cb117-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-378"><a href="#cb117-378" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How do economic weights affect trait importance?</span>
<span id="cb117-379"><a href="#cb117-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-380"><a href="#cb117-380" aria-hidden="true" tabindex="-1"></a>Traits with higher weights contribute more to the total economic value. This makes them more influential in ranking and selection.</span>
<span id="cb117-381"><a href="#cb117-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-382"><a href="#cb117-382" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Why might you ignore milk yield?</span>
<span id="cb117-383"><a href="#cb117-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-384"><a href="#cb117-384" aria-hidden="true" tabindex="-1"></a>Milk yield may be excluded in systems focusing on fertility, growth, or when it is no longer a limiting factor. Environmental or economic contexts may also shift trait priorities.</span>
<span id="cb117-385"><a href="#cb117-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-386"><a href="#cb117-386" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What is the value of heatmaps?</span>
<span id="cb117-387"><a href="#cb117-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-388"><a href="#cb117-388" aria-hidden="true" tabindex="-1"></a>Heatmaps visually compare EBVs across bulls and traits.They help detect patterns, outliers, and clusters easily in multivariate data.</span>
<span id="cb117-389"><a href="#cb117-389" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-390"><a href="#cb117-390" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Can this method be extended to more bulls and traits?</span>
<span id="cb117-391"><a href="#cb117-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-392"><a href="#cb117-392" aria-hidden="true" tabindex="-1"></a>Yes. This method scales to any number of bulls or traits. Just ensure the EBVs matrix and economic weights are dimensionally compatible.</span>
<span id="cb117-393"><a href="#cb117-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-394"><a href="#cb117-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-395"><a href="#cb117-395" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 4: Plant Breeding ‚Äì Trait Contributions from Parental Lines</span></span>
<span id="cb117-396"><a href="#cb117-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-397"><a href="#cb117-397" aria-hidden="true" tabindex="-1"></a><span class="fu">### Define Data</span></span>
<span id="cb117-398"><a href="#cb117-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-399"><a href="#cb117-399" aria-hidden="true" tabindex="-1"></a>**Parent trait values (normalized 1‚Äì10)**</span>
<span id="cb117-402"><a href="#cb117-402" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-403"><a href="#cb117-403" aria-hidden="true" tabindex="-1"></a>ParentTraits <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb117-404"><a href="#cb117-404" aria-hidden="true" tabindex="-1"></a>  <span class="dv">7</span>, <span class="dv">5</span>, <span class="dv">3</span>,</span>
<span id="cb117-405"><a href="#cb117-405" aria-hidden="true" tabindex="-1"></a>  <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">4</span>,</span>
<span id="cb117-406"><a href="#cb117-406" aria-hidden="true" tabindex="-1"></a>  <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">6</span></span>
<span id="cb117-407"><a href="#cb117-407" aria-hidden="true" tabindex="-1"></a>), <span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb117-408"><a href="#cb117-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-409"><a href="#cb117-409" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ParentTraits) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"P1"</span>, <span class="st">"P2"</span>, <span class="st">"P3"</span>)</span>
<span id="cb117-410"><a href="#cb117-410" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ParentTraits) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Drought_resistance"</span>, <span class="st">"Yield"</span>, <span class="st">"Maturation_time"</span>)</span>
<span id="cb117-411"><a href="#cb117-411" aria-hidden="true" tabindex="-1"></a>ParentTraits</span>
<span id="cb117-412"><a href="#cb117-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-413"><a href="#cb117-413" aria-hidden="true" tabindex="-1"></a><span class="co"># Define Hybrid Weights</span></span>
<span id="cb117-414"><a href="#cb117-414" aria-hidden="true" tabindex="-1"></a>HybridWeights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span>), <span class="at">nrow =</span> <span class="dv">1</span>)</span>
<span id="cb117-415"><a href="#cb117-415" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(HybridWeights) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ParentTraits)</span>
<span id="cb117-416"><a href="#cb117-416" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(HybridWeights) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb117-417"><a href="#cb117-417" aria-hidden="true" tabindex="-1"></a>HybridWeights</span>
<span id="cb117-418"><a href="#cb117-418" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-419"><a href="#cb117-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-420"><a href="#cb117-420" aria-hidden="true" tabindex="-1"></a><span class="fu">### Compute HybridTraits Vector</span></span>
<span id="cb117-421"><a href="#cb117-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-424"><a href="#cb117-424" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-425"><a href="#cb117-425" aria-hidden="true" tabindex="-1"></a>HybridTraits <span class="ot">&lt;-</span> HybridWeights <span class="sc">%*%</span> ParentTraits</span>
<span id="cb117-426"><a href="#cb117-426" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(HybridTraits) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Contribution"</span>)</span>
<span id="cb117-427"><a href="#cb117-427" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(HybridTraits) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(ParentTraits)</span>
<span id="cb117-428"><a href="#cb117-428" aria-hidden="true" tabindex="-1"></a>HybridTraits</span>
<span id="cb117-429"><a href="#cb117-429" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-430"><a href="#cb117-430" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb117-431"><a href="#cb117-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-432"><a href="#cb117-432" aria-hidden="true" tabindex="-1"></a>HybridTraits = (0.5 √ó P1) + (0.3 √ó P2) + (0.2 √ó P3)</span>
<span id="cb117-433"><a href="#cb117-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-434"><a href="#cb117-434" aria-hidden="true" tabindex="-1"></a>Drought_resistance = (0.5 √ó 7) + (0.3 √ó 6) + (0.2 √ó 5) = 6.3</span>
<span id="cb117-435"><a href="#cb117-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-436"><a href="#cb117-436" aria-hidden="true" tabindex="-1"></a>Yield = (0.5 √ó 5) + (0.3 √ó 8) + (0.2 √ó 6) = 6.1</span>
<span id="cb117-437"><a href="#cb117-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-438"><a href="#cb117-438" aria-hidden="true" tabindex="-1"></a>Maturation_time = (0.5 √ó 3) + (0.3 √ó 4) + (0.2 √ó 6) = 3.9</span>
<span id="cb117-439"><a href="#cb117-439" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-440"><a href="#cb117-440" aria-hidden="true" tabindex="-1"></a>The hybrid is moderately strong in drought resistance and yield, and has a relatively shorter maturation time.</span>
<span id="cb117-441"><a href="#cb117-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-442"><a href="#cb117-442" aria-hidden="true" tabindex="-1"></a>**Biological Meaning of Unequal Contribution**</span>
<span id="cb117-443"><a href="#cb117-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-444"><a href="#cb117-444" aria-hidden="true" tabindex="-1"></a>When one parent contributes more to a trait, it suggests that the trait's heritable strength comes disproportionately from that parent. Breeders can use this knowledge to amplify desirable traits using the best parent.</span>
<span id="cb117-445"><a href="#cb117-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-446"><a href="#cb117-446" aria-hidden="true" tabindex="-1"></a><span class="fu">### Multiply with Identity Matrix</span></span>
<span id="cb117-447"><a href="#cb117-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-450"><a href="#cb117-450" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-451"><a href="#cb117-451" aria-hidden="true" tabindex="-1"></a>I3 <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb117-452"><a href="#cb117-452" aria-hidden="true" tabindex="-1"></a>ParentTraits_identity <span class="ot">&lt;-</span> ParentTraits <span class="sc">%*%</span> I3</span>
<span id="cb117-453"><a href="#cb117-453" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ParentTraits_identity) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ParentTraits)</span>
<span id="cb117-454"><a href="#cb117-454" aria-hidden="true" tabindex="-1"></a>ParentTraits_identity</span>
<span id="cb117-455"><a href="#cb117-455" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-456"><a href="#cb117-456" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-457"><a href="#cb117-457" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb117-458"><a href="#cb117-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-459"><a href="#cb117-459" aria-hidden="true" tabindex="-1"></a>Multiplying by identity matrix returns the original matrix. This operation verifies structural consistency and dimensionality.</span>
<span id="cb117-460"><a href="#cb117-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-461"><a href="#cb117-461" aria-hidden="true" tabindex="-1"></a><span class="fu">### Remove T3 (Maturation Time) and Recalculate</span></span>
<span id="cb117-462"><a href="#cb117-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-465"><a href="#cb117-465" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-466"><a href="#cb117-466" aria-hidden="true" tabindex="-1"></a>ParentTraits_T1T2 <span class="ot">&lt;-</span> ParentTraits[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb117-467"><a href="#cb117-467" aria-hidden="true" tabindex="-1"></a>ParentTraits_T1T2</span>
<span id="cb117-468"><a href="#cb117-468" aria-hidden="true" tabindex="-1"></a>HybridTraits_T1T2 <span class="ot">&lt;-</span> HybridWeights <span class="sc">%*%</span> ParentTraits_T1T2</span>
<span id="cb117-469"><a href="#cb117-469" aria-hidden="true" tabindex="-1"></a>HybridTraits_T1T2</span>
<span id="cb117-470"><a href="#cb117-470" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-471"><a href="#cb117-471" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-472"><a href="#cb117-472" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb117-473"><a href="#cb117-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-474"><a href="#cb117-474" aria-hidden="true" tabindex="-1"></a>Drought_resistance = (0.5 √ó 7) + (0.3 √ó 6) + (0.2 √ó 5) = 6.3</span>
<span id="cb117-475"><a href="#cb117-475" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-476"><a href="#cb117-476" aria-hidden="true" tabindex="-1"></a>Yield = (0.5 √ó 5) + (0.3 √ó 8) + (0.2 √ó 6) = 6.1</span>
<span id="cb117-477"><a href="#cb117-477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-478"><a href="#cb117-478" aria-hidden="true" tabindex="-1"></a>Removing a trait (T3) changes the trait profile. Hybrid selection may now favor traits that remain.</span>
<span id="cb117-479"><a href="#cb117-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-480"><a href="#cb117-480" aria-hidden="true" tabindex="-1"></a><span class="fu">### Heatmap of ParentTraits</span></span>
<span id="cb117-481"><a href="#cb117-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-484"><a href="#cb117-484" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-485"><a href="#cb117-485" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb117-486"><a href="#cb117-486" aria-hidden="true" tabindex="-1"></a>  ParentTraits,</span>
<span id="cb117-487"><a href="#cb117-487" aria-hidden="true" tabindex="-1"></a>  <span class="at">Rowv =</span> <span class="cn">NA</span>,</span>
<span id="cb117-488"><a href="#cb117-488" aria-hidden="true" tabindex="-1"></a>  <span class="at">Colv =</span> <span class="cn">NA</span>,</span>
<span id="cb117-489"><a href="#cb117-489" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb117-490"><a href="#cb117-490" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">256</span>),</span>
<span id="cb117-491"><a href="#cb117-491" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Heatmap of Parent Traits"</span></span>
<span id="cb117-492"><a href="#cb117-492" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-493"><a href="#cb117-493" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-494"><a href="#cb117-494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-495"><a href="#cb117-495" aria-hidden="true" tabindex="-1"></a><span class="fu">### Bar Plot of Hybrid Traits</span></span>
<span id="cb117-496"><a href="#cb117-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-499"><a href="#cb117-499" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-500"><a href="#cb117-500" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb117-501"><a href="#cb117-501" aria-hidden="true" tabindex="-1"></a>  HybridTraits,</span>
<span id="cb117-502"><a href="#cb117-502" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb117-503"><a href="#cb117-503" aria-hidden="true" tabindex="-1"></a>  <span class="at">names.arg =</span> <span class="fu">colnames</span>(ParentTraits),</span>
<span id="cb117-504"><a href="#cb117-504" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#66c2a5"</span>, <span class="st">"#fc8d62"</span>, <span class="st">"#8da0cb"</span>),</span>
<span id="cb117-505"><a href="#cb117-505" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Hybrid Trait Profile"</span>,</span>
<span id="cb117-506"><a href="#cb117-506" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Trait Value"</span></span>
<span id="cb117-507"><a href="#cb117-507" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-508"><a href="#cb117-508" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-509"><a href="#cb117-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-510"><a href="#cb117-510" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation Questions</span></span>
<span id="cb117-511"><a href="#cb117-511" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-512"><a href="#cb117-512" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How does the weighting of parents affect the hybrid's performance?</span>
<span id="cb117-513"><a href="#cb117-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-514"><a href="#cb117-514" aria-hidden="true" tabindex="-1"></a>  Stronger weights mean more genetic contribution. Traits from highly weighted parents dominate the hybrid profile.</span>
<span id="cb117-515"><a href="#cb117-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-516"><a href="#cb117-516" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What does the identity matrix represent here?</span>
<span id="cb117-517"><a href="#cb117-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-518"><a href="#cb117-518" aria-hidden="true" tabindex="-1"></a>  It represents a neutral transformation. It confirms data integrity when used in matrix multiplication.</span>
<span id="cb117-519"><a href="#cb117-519" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-520"><a href="#cb117-520" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>If you used equal weights (‚Öì for each), how would the hybrid traits change?</span>
<span id="cb117-521"><a href="#cb117-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-522"><a href="#cb117-522" aria-hidden="true" tabindex="-1"></a>  Traits would reflect an even mix, potentially leading to balanced but less specialized performance.</span>
<span id="cb117-523"><a href="#cb117-523" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-524"><a href="#cb117-524" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What real-world limitations does this simplified model ignore?</span>
<span id="cb117-525"><a href="#cb117-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-526"><a href="#cb117-526" aria-hidden="true" tabindex="-1"></a>  i. Non-additive genetic effects (dominance, epistasis)</span>
<span id="cb117-527"><a href="#cb117-527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-528"><a href="#cb117-528" aria-hidden="true" tabindex="-1"></a>  ii. Environmental interactions</span>
<span id="cb117-529"><a href="#cb117-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-530"><a href="#cb117-530" aria-hidden="true" tabindex="-1"></a>  iii. Trait heritability and correlations</span>
<span id="cb117-531"><a href="#cb117-531" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-532"><a href="#cb117-532" aria-hidden="true" tabindex="-1"></a>  iv. Breeding feasibility and cost</span>
<span id="cb117-533"><a href="#cb117-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-534"><a href="#cb117-534" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 5: Managing Matrices and Weight Vectors Using Lists in R</span></span>
<span id="cb117-535"><a href="#cb117-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-536"><a href="#cb117-536" aria-hidden="true" tabindex="-1"></a>Now that we've explored trait-based decisions using matrices, it's time to organize our work using R's list structure. Lists help bundle related objects like matrices and weight vectors, keeping the analysis modular and scalable.</span>
<span id="cb117-537"><a href="#cb117-537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-538"><a href="#cb117-538" aria-hidden="true" tabindex="-1"></a><span class="fu">### Create a Master List</span></span>
<span id="cb117-539"><a href="#cb117-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-542"><a href="#cb117-542" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-543"><a href="#cb117-543" aria-hidden="true" tabindex="-1"></a><span class="co"># Assuming previous matrices and weights are already defined:</span></span>
<span id="cb117-544"><a href="#cb117-544" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-545"><a href="#cb117-545" aria-hidden="true" tabindex="-1"></a><span class="co"># Making a MasterList</span></span>
<span id="cb117-546"><a href="#cb117-546" aria-hidden="true" tabindex="-1"></a>bioList <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb117-547"><a href="#cb117-547" aria-hidden="true" tabindex="-1"></a>  <span class="at">ProteinConc =</span> <span class="fu">list</span>(<span class="at">matrix =</span> ProteinMatrix, <span class="at">weights =</span> WeightVector),</span>
<span id="cb117-548"><a href="#cb117-548" aria-hidden="true" tabindex="-1"></a>  <span class="at">ProteinMap  =</span> <span class="fu">list</span>(<span class="at">matrix =</span> GeneExpression, <span class="at">weights =</span> TranslationMatrix),</span>
<span id="cb117-549"><a href="#cb117-549" aria-hidden="true" tabindex="-1"></a>  <span class="at">Animal =</span> <span class="fu">list</span>(<span class="at">matrix =</span> BullEBVs, <span class="at">weights =</span> EconomicWeights),</span>
<span id="cb117-550"><a href="#cb117-550" aria-hidden="true" tabindex="-1"></a>  <span class="at">Plant =</span> <span class="fu">list</span>(<span class="at">matrix =</span> ParentTraits, <span class="at">weights =</span> HybridWeights)</span>
<span id="cb117-551"><a href="#cb117-551" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-552"><a href="#cb117-552" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-553"><a href="#cb117-553" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(bioList)</span>
<span id="cb117-554"><a href="#cb117-554" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-555"><a href="#cb117-555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-556"><a href="#cb117-556" aria-hidden="true" tabindex="-1"></a><span class="fu">### List the Structure</span></span>
<span id="cb117-557"><a href="#cb117-557" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-560"><a href="#cb117-560" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-561"><a href="#cb117-561" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bioList)               <span class="co"># Top-level list names</span></span>
<span id="cb117-562"><a href="#cb117-562" aria-hidden="true" tabindex="-1"></a><span class="fu">lengths</span>(bioList)             <span class="co"># Number of components in each sublist</span></span>
<span id="cb117-563"><a href="#cb117-563" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-564"><a href="#cb117-564" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-565"><a href="#cb117-565" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb117-566"><a href="#cb117-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-567"><a href="#cb117-567" aria-hidden="true" tabindex="-1"></a>Each top-level entry (e.g., ProteinConc, Plant) contains two components:</span>
<span id="cb117-568"><a href="#cb117-568" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-569"><a href="#cb117-569" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>A matrix (e.g., ProteinMatrix)</span>
<span id="cb117-570"><a href="#cb117-570" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>A corresponding weight vector or matrix</span>
<span id="cb117-571"><a href="#cb117-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-572"><a href="#cb117-572" aria-hidden="true" tabindex="-1"></a><span class="fu">### Indexing Elements from Lists</span></span>
<span id="cb117-573"><a href="#cb117-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-576"><a href="#cb117-576" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-577"><a href="#cb117-577" aria-hidden="true" tabindex="-1"></a><span class="co"># Access the trait matrix for Plant</span></span>
<span id="cb117-578"><a href="#cb117-578" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Plant[[<span class="dv">1</span>]]</span>
<span id="cb117-579"><a href="#cb117-579" aria-hidden="true" tabindex="-1"></a><span class="co">#or</span></span>
<span id="cb117-580"><a href="#cb117-580" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix</span>
<span id="cb117-581"><a href="#cb117-581" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-582"><a href="#cb117-582" aria-hidden="true" tabindex="-1"></a><span class="co"># Access the weight vector for ProteinConc</span></span>
<span id="cb117-583"><a href="#cb117-583" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinConc[[<span class="dv">2</span>]]</span>
<span id="cb117-584"><a href="#cb117-584" aria-hidden="true" tabindex="-1"></a><span class="co">#or</span></span>
<span id="cb117-585"><a href="#cb117-585" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinConc<span class="sc">$</span>weights</span>
<span id="cb117-586"><a href="#cb117-586" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-587"><a href="#cb117-587" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-588"><a href="#cb117-588" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb117-589"><a href="#cb117-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-590"><a href="#cb117-590" aria-hidden="true" tabindex="-1"></a>Use double brackets [<span class="co">[ ]</span>] to extract unnamed list elements by position. But we named our list, so they are easily extractable using the <span class="in">`$`</span> notation.</span>
<span id="cb117-591"><a href="#cb117-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-592"><a href="#cb117-592" aria-hidden="true" tabindex="-1"></a><span class="fu">### Perform Weighted Calculations</span></span>
<span id="cb117-593"><a href="#cb117-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-596"><a href="#cb117-596" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-597"><a href="#cb117-597" aria-hidden="true" tabindex="-1"></a><span class="co"># Protein concentration score</span></span>
<span id="cb117-598"><a href="#cb117-598" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinConc<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>ProteinConc<span class="sc">$</span>weights</span>
<span id="cb117-599"><a href="#cb117-599" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-600"><a href="#cb117-600" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene ‚Üí Protein contribution</span></span>
<span id="cb117-601"><a href="#cb117-601" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinMap<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>ProteinMap<span class="sc">$</span>weights</span>
<span id="cb117-602"><a href="#cb117-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-603"><a href="#cb117-603" aria-hidden="true" tabindex="-1"></a><span class="co"># Bull economic value</span></span>
<span id="cb117-604"><a href="#cb117-604" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Animal<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>Animal<span class="sc">$</span>weights</span>
<span id="cb117-605"><a href="#cb117-605" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-606"><a href="#cb117-606" aria-hidden="true" tabindex="-1"></a><span class="co"># Hybrid trait value</span></span>
<span id="cb117-607"><a href="#cb117-607" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Plant<span class="sc">$</span>weights <span class="sc">%*%</span> bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix</span>
<span id="cb117-608"><a href="#cb117-608" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-609"><a href="#cb117-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-610"><a href="#cb117-610" aria-hidden="true" tabindex="-1"></a><span class="fu">### Subset and Recalculate</span></span>
<span id="cb117-611"><a href="#cb117-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-614"><a href="#cb117-614" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-615"><a href="#cb117-615" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove last trait from ParentTraits</span></span>
<span id="cb117-616"><a href="#cb117-616" aria-hidden="true" tabindex="-1"></a>ParentSubset <span class="ot">&lt;-</span> bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb117-617"><a href="#cb117-617" aria-hidden="true" tabindex="-1"></a>NewWeights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.6</span>, <span class="fl">0.4</span>), <span class="at">nrow =</span> <span class="dv">2</span>)</span>
<span id="cb117-618"><a href="#cb117-618" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-619"><a href="#cb117-619" aria-hidden="true" tabindex="-1"></a><span class="co"># Recalculated hybrid score</span></span>
<span id="cb117-620"><a href="#cb117-620" aria-hidden="true" tabindex="-1"></a>SubsetHybridScore <span class="ot">&lt;-</span> ParentSubset <span class="sc">%*%</span> NewWeights</span>
<span id="cb117-621"><a href="#cb117-621" aria-hidden="true" tabindex="-1"></a>SubsetHybridScore</span>
<span id="cb117-622"><a href="#cb117-622" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-623"><a href="#cb117-623" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-624"><a href="#cb117-624" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb117-625"><a href="#cb117-625" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-626"><a href="#cb117-626" aria-hidden="true" tabindex="-1"></a>Dropping a trait and reweighting highlights its influence in trait aggregation and selection.</span>
<span id="cb117-627"><a href="#cb117-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-628"><a href="#cb117-628" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualization Tasks</span></span>
<span id="cb117-629"><a href="#cb117-629" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-630"><a href="#cb117-630" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Heatmap: Gene Expression</span></span>
<span id="cb117-631"><a href="#cb117-631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-634"><a href="#cb117-634" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-635"><a href="#cb117-635" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb117-636"><a href="#cb117-636" aria-hidden="true" tabindex="-1"></a>  bioList<span class="sc">$</span>ProteinMap<span class="sc">$</span>matrix,</span>
<span id="cb117-637"><a href="#cb117-637" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb117-638"><a href="#cb117-638" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">256</span>),</span>
<span id="cb117-639"><a href="#cb117-639" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Gene Expression Heatmap"</span>,</span>
<span id="cb117-640"><a href="#cb117-640" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Proteins"</span>,</span>
<span id="cb117-641"><a href="#cb117-641" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Genes"</span></span>
<span id="cb117-642"><a href="#cb117-642" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-643"><a href="#cb117-643" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-644"><a href="#cb117-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-645"><a href="#cb117-645" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Bar Plots</span></span>
<span id="cb117-646"><a href="#cb117-646" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-647"><a href="#cb117-647" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Hybrid traits</span></span>
<span id="cb117-648"><a href="#cb117-648" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-651"><a href="#cb117-651" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-652"><a href="#cb117-652" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb117-653"><a href="#cb117-653" aria-hidden="true" tabindex="-1"></a>  bioList<span class="sc">$</span>Plant<span class="sc">$</span>weights <span class="sc">%*%</span> bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix,</span>
<span id="cb117-654"><a href="#cb117-654" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb117-655"><a href="#cb117-655" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Hybrid Trait Contributions"</span>,</span>
<span id="cb117-656"><a href="#cb117-656" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="st">"#66c2a5"</span>,</span>
<span id="cb117-657"><a href="#cb117-657" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Score"</span></span>
<span id="cb117-658"><a href="#cb117-658" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-659"><a href="#cb117-659" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-660"><a href="#cb117-660" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-661"><a href="#cb117-661" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Bull EBV (Economic Breeding Values)</span></span>
<span id="cb117-662"><a href="#cb117-662" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-665"><a href="#cb117-665" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-666"><a href="#cb117-666" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb117-667"><a href="#cb117-667" aria-hidden="true" tabindex="-1"></a>  bioList<span class="sc">$</span>Animal<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>Animal<span class="sc">$</span>weights,</span>
<span id="cb117-668"><a href="#cb117-668" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb117-669"><a href="#cb117-669" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Bull EBVs (Economic Values)"</span>,</span>
<span id="cb117-670"><a href="#cb117-670" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="st">"#fc8d62"</span>,</span>
<span id="cb117-671"><a href="#cb117-671" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Score"</span></span>
<span id="cb117-672"><a href="#cb117-672" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-673"><a href="#cb117-673" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-674"><a href="#cb117-674" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-675"><a href="#cb117-675" aria-hidden="true" tabindex="-1"></a><span class="fu">### üß† Interpretation Questions</span></span>
<span id="cb117-676"><a href="#cb117-676" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-677"><a href="#cb117-677" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Why use a list structure?</span>
<span id="cb117-678"><a href="#cb117-678" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-679"><a href="#cb117-679" aria-hidden="true" tabindex="-1"></a>Keeps each dataset and its weights together.</span>
<span id="cb117-680"><a href="#cb117-680" aria-hidden="true" tabindex="-1"></a>Facilitates automated workflows and reuse.</span>
<span id="cb117-681"><a href="#cb117-681" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-682"><a href="#cb117-682" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What‚Äôs tricky about <span class="in">`[[ ]]`</span> access?</span>
<span id="cb117-683"><a href="#cb117-683" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-684"><a href="#cb117-684" aria-hidden="true" tabindex="-1"></a>You must remember the order (<span class="in">`[[1]]`</span> = <span class="in">`matrix`</span>, <span class="in">`[[2]]`</span> = <span class="in">`weights`</span>).</span>
<span id="cb117-685"><a href="#cb117-685" aria-hidden="true" tabindex="-1"></a>No names means you can't use $matrix, only positional access.</span>
<span id="cb117-686"><a href="#cb117-686" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-687"><a href="#cb117-687" aria-hidden="true" tabindex="-1"></a>Loop across all list entries</span>
<span id="cb117-688"><a href="#cb117-688" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-689"><a href="#cb117-689" aria-hidden="true" tabindex="-1"></a>Weighted scores for all entries</span>
<span id="cb117-690"><a href="#cb117-690" aria-hidden="true" tabindex="-1"></a><span class="in">`lapply(bioList, function(x) x[[2]] %*% x[[1]])`</span></span>
<span id="cb117-691"><a href="#cb117-691" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-692"><a href="#cb117-692" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How does this help in large-scale pipelines?</span>
<span id="cb117-693"><a href="#cb117-693" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-694"><a href="#cb117-694" aria-hidden="true" tabindex="-1"></a>You can use this format with <span class="in">`lapply()`</span>, <span class="in">`purrr::map()`</span>, or in targets pipelines for reproducibility and modular processing.</span>
<span id="cb117-695"><a href="#cb117-695" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-696"><a href="#cb117-696" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-697"><a href="#cb117-697" aria-hidden="true" tabindex="-1"></a><span class="fu">## üß© Homework solutions: Factors, Subsetting, and Biological Insight</span></span>
<span id="cb117-698"><a href="#cb117-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-699"><a href="#cb117-699" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Character vs Factor**  </span>
<span id="cb117-700"><a href="#cb117-700" aria-hidden="true" tabindex="-1"></a>   A character vector simply holds string values, but a factor is a categorical variable with fixed levels, used especially in modeling.  </span>
<span id="cb117-701"><a href="#cb117-701" aria-hidden="true" tabindex="-1"></a>   For <span class="in">`mutation_status`</span>, a factor ensures consistent categories (e.g., <span class="in">`"Yes"`</span> or <span class="in">`"No"`</span>) and helps control level order and statistical reference groups.</span>
<span id="cb117-702"><a href="#cb117-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-703"><a href="#cb117-703" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Factor Levels**  </span>
<span id="cb117-704"><a href="#cb117-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-707"><a href="#cb117-707" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-708"><a href="#cb117-708" aria-hidden="true" tabindex="-1"></a>species <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Lactobacillus"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Lactobacillus"</span>)</span>
<span id="cb117-709"><a href="#cb117-709" aria-hidden="true" tabindex="-1"></a>species_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(species, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Lactobacillus"</span>))</span>
<span id="cb117-710"><a href="#cb117-710" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(species_factor)</span>
<span id="cb117-711"><a href="#cb117-711" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-712"><a href="#cb117-712" aria-hidden="true" tabindex="-1"></a>Because we defined the level order explicitly, R maintains that order regardless of data input.</span>
<span id="cb117-713"><a href="#cb117-713" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-714"><a href="#cb117-714" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Ordered Factor Comparison</span>
<span id="cb117-717"><a href="#cb117-717" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-718"><a href="#cb117-718" aria-hidden="true" tabindex="-1"></a>disease_severity <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Mild"</span>, <span class="st">"Severe"</span>, <span class="st">"Moderate"</span>), <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Mild"</span>, <span class="st">"Moderate"</span>, <span class="st">"Severe"</span>, <span class="st">"Critical"</span>), <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb117-719"><a href="#cb117-719" aria-hidden="true" tabindex="-1"></a>disease_severity[<span class="dv">1</span>] <span class="sc">&lt;</span> disease_severity[<span class="dv">2</span>]</span>
<span id="cb117-720"><a href="#cb117-720" aria-hidden="true" tabindex="-1"></a><span class="co"># TRUE</span></span>
<span id="cb117-721"><a href="#cb117-721" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-722"><a href="#cb117-722" aria-hidden="true" tabindex="-1"></a>"Mild" is less severe than "Severe" based on the defined order.</span>
<span id="cb117-723"><a href="#cb117-723" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-724"><a href="#cb117-724" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Proportion Extraction</span>
<span id="cb117-727"><a href="#cb117-727" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-728"><a href="#cb117-728" aria-hidden="true" tabindex="-1"></a>prop <span class="ot">&lt;-</span> <span class="fu">prop.table</span>(<span class="fu">table</span>(species_factor))</span>
<span id="cb117-729"><a href="#cb117-729" aria-hidden="true" tabindex="-1"></a>prop[<span class="st">"Escherichia"</span>]</span>
<span id="cb117-730"><a href="#cb117-730" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-731"><a href="#cb117-731" aria-hidden="true" tabindex="-1"></a><span class="in">`prop$Escherichia`</span> won't work ‚Äî named numeric vectors require bracket-based access.</span>
<span id="cb117-732"><a href="#cb117-732" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-733"><a href="#cb117-733" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>Subsetting by Conditions</span>
<span id="cb117-736"><a href="#cb117-736" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-737"><a href="#cb117-737" aria-hidden="true" tabindex="-1"></a>gene_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb117-738"><a href="#cb117-738" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_id =</span> <span class="fu">c</span>(<span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"EGFR"</span>, <span class="st">"GAPDH"</span>),</span>
<span id="cb117-739"><a href="#cb117-739" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression =</span> <span class="fu">c</span>(<span class="fl">8.2</span>, <span class="fl">6.1</span>, <span class="fl">9.5</span>, <span class="fl">7.0</span>, <span class="fl">10.0</span>),</span>
<span id="cb117-740"><a href="#cb117-740" aria-hidden="true" tabindex="-1"></a>  <span class="at">mutation =</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"No"</span>)),</span>
<span id="cb117-741"><a href="#cb117-741" aria-hidden="true" tabindex="-1"></a>  <span class="at">pathway =</span> <span class="fu">c</span>(<span class="st">"DNA Repair"</span>, <span class="st">"Apoptosis"</span>, <span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>, <span class="st">"Metabolism"</span>)</span>
<span id="cb117-742"><a href="#cb117-742" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-743"><a href="#cb117-743" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(gene_df) <span class="ot">&lt;-</span> gene_df<span class="sc">$</span>gene_id <span class="co">#name the rows by the gene IDs</span></span>
<span id="cb117-744"><a href="#cb117-744" aria-hidden="true" tabindex="-1"></a>gene_df <span class="ot">&lt;-</span> gene_df[, <span class="sc">-</span><span class="dv">1</span>] <span class="co">#remove the first column which is not needed anymore</span></span>
<span id="cb117-745"><a href="#cb117-745" aria-hidden="true" tabindex="-1"></a><span class="co">#gene_df</span></span>
<span id="cb117-746"><a href="#cb117-746" aria-hidden="true" tabindex="-1"></a>gene_df[gene_df<span class="sc">$</span>expression <span class="sc">&gt;</span> <span class="dv">7</span> <span class="sc">&amp;</span> gene_df<span class="sc">$</span>mutation <span class="sc">==</span> <span class="st">"No"</span>, ]</span>
<span id="cb117-747"><a href="#cb117-747" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-748"><a href="#cb117-748" aria-hidden="true" tabindex="-1"></a>Returns genes with **high expression (&gt;7)** `and` **no mutation** ‚Äî potentially highly active but wild-type genes.</span>
<span id="cb117-749"><a href="#cb117-749" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-750"><a href="#cb117-750" aria-hidden="true" tabindex="-1"></a><span class="ss">6. </span>Group-wise Expression Summary</span>
<span id="cb117-751"><a href="#cb117-751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-752"><a href="#cb117-752" aria-hidden="true" tabindex="-1"></a>The given vectors are:</span>
<span id="cb117-755"><a href="#cb117-755" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-756"><a href="#cb117-756" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span>)</span>
<span id="cb117-757"><a href="#cb117-757" aria-hidden="true" tabindex="-1"></a>expression <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">5.2</span>, <span class="fl">8.1</span>, <span class="fl">4.3</span>, <span class="fl">9.0</span>, <span class="fl">5.7</span>)</span>
<span id="cb117-758"><a href="#cb117-758" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-759"><a href="#cb117-759" aria-hidden="true" tabindex="-1"></a>The solution would be:</span>
<span id="cb117-760"><a href="#cb117-760" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-763"><a href="#cb117-763" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-764"><a href="#cb117-764" aria-hidden="true" tabindex="-1"></a>group_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(samples)</span>
<span id="cb117-765"><a href="#cb117-765" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-766"><a href="#cb117-766" aria-hidden="true" tabindex="-1"></a><span class="co"># Mean expression</span></span>
<span id="cb117-767"><a href="#cb117-767" aria-hidden="true" tabindex="-1"></a><span class="fu">tapply</span>(expression, group_factor, mean) <span class="do">## KO: 8.55, WT: 5.07</span></span>
<span id="cb117-768"><a href="#cb117-768" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-769"><a href="#cb117-769" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb117-770"><a href="#cb117-770" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">tapply</span>(expression, group_factor, mean), </span>
<span id="cb117-771"><a href="#cb117-771" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"salmon"</span>),</span>
<span id="cb117-772"><a href="#cb117-772" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Mean Expression"</span>,</span>
<span id="cb117-773"><a href="#cb117-773" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Group-wise Expression"</span>)</span>
<span id="cb117-774"><a href="#cb117-774" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-775"><a href="#cb117-775" aria-hidden="true" tabindex="-1"></a><span class="ss">7. </span>Gene Subsetting</span>
<span id="cb117-776"><a href="#cb117-776" aria-hidden="true" tabindex="-1"></a><span class="in">```{r gene_df_subset}</span></span>
<span id="cb117-777"><a href="#cb117-777" aria-hidden="true" tabindex="-1"></a>gene_df[gene_df<span class="sc">$</span>expression <span class="sc">&gt;</span> <span class="dv">8</span> <span class="sc">&amp;</span> </span>
<span id="cb117-778"><a href="#cb117-778" aria-hidden="true" tabindex="-1"></a>        gene_df<span class="sc">$</span>pathway <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>), ]</span>
<span id="cb117-779"><a href="#cb117-779" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-780"><a href="#cb117-780" aria-hidden="true" tabindex="-1"></a>It filters for genes highly expressed and involved in key biological pathways.</span>
<span id="cb117-781"><a href="#cb117-781" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-782"><a href="#cb117-782" aria-hidden="true" tabindex="-1"></a><span class="ss">8. </span>Disease Stage Visualization</span>
<span id="cb117-785"><a href="#cb117-785" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-786"><a href="#cb117-786" aria-hidden="true" tabindex="-1"></a>stages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Stage I"</span>, <span class="st">"Stage III"</span>, <span class="st">"Stage II"</span>, <span class="st">"Stage IV"</span>, <span class="st">"Stage I"</span>)</span>
<span id="cb117-787"><a href="#cb117-787" aria-hidden="true" tabindex="-1"></a>disease_stage <span class="ot">&lt;-</span> <span class="fu">factor</span>(stages, </span>
<span id="cb117-788"><a href="#cb117-788" aria-hidden="true" tabindex="-1"></a>                        <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Stage I"</span>, <span class="st">"Stage II"</span>, <span class="st">"Stage III"</span>, <span class="st">"Stage IV"</span>), </span>
<span id="cb117-789"><a href="#cb117-789" aria-hidden="true" tabindex="-1"></a>                        <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb117-790"><a href="#cb117-790" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-791"><a href="#cb117-791" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(disease_stage), </span>
<span id="cb117-792"><a href="#cb117-792" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="st">"lightgreen"</span>, </span>
<span id="cb117-793"><a href="#cb117-793" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Patient Count by Disease Stage"</span>,</span>
<span id="cb117-794"><a href="#cb117-794" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Count"</span>)</span>
<span id="cb117-795"><a href="#cb117-795" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-796"><a href="#cb117-796" aria-hidden="true" tabindex="-1"></a>Let's do the severity order check:</span>
<span id="cb117-799"><a href="#cb117-799" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-800"><a href="#cb117-800" aria-hidden="true" tabindex="-1"></a><span class="co"># Comparison</span></span>
<span id="cb117-801"><a href="#cb117-801" aria-hidden="true" tabindex="-1"></a>disease_stage[<span class="dv">2</span>] <span class="sc">&gt;</span> disease_stage[<span class="dv">1</span>]  <span class="co"># TRUE</span></span>
<span id="cb117-802"><a href="#cb117-802" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-803"><a href="#cb117-803" aria-hidden="true" tabindex="-1"></a>So, "Stage III" is more sever than "Stage I".</span>
<span id="cb117-804"><a href="#cb117-804" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-805"><a href="#cb117-805" aria-hidden="true" tabindex="-1"></a><span class="ss">9. </span>Oncogene Subsetting and Releveling</span>
<span id="cb117-808"><a href="#cb117-808" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-809"><a href="#cb117-809" aria-hidden="true" tabindex="-1"></a><span class="co"># Define a small gene dataset</span></span>
<span id="cb117-810"><a href="#cb117-810" aria-hidden="true" tabindex="-1"></a>gene_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb117-811"><a href="#cb117-811" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene =</span> <span class="fu">c</span>(<span class="st">"TP53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"MYC"</span>, <span class="st">"GAPDH"</span>, <span class="st">"EGFR"</span>),</span>
<span id="cb117-812"><a href="#cb117-812" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression =</span> <span class="fu">c</span>(<span class="fl">9.1</span>, <span class="fl">7.3</span>, <span class="fl">10.5</span>, <span class="fl">5.2</span>, <span class="fl">8.6</span>),</span>
<span id="cb117-813"><a href="#cb117-813" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Tumor Suppressor"</span>, <span class="st">"Oncogene"</span>, <span class="st">"Oncogene"</span>, <span class="st">"Housekeeping"</span>, <span class="st">"Oncogene"</span>))</span>
<span id="cb117-814"><a href="#cb117-814" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb117-815"><a href="#cb117-815" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-816"><a href="#cb117-816" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset: Oncogene rows with expression &gt; 8</span></span>
<span id="cb117-817"><a href="#cb117-817" aria-hidden="true" tabindex="-1"></a>gene_data[gene_data<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"Oncogene"</span> <span class="sc">&amp;</span> gene_data<span class="sc">$</span>expression <span class="sc">&gt;</span> <span class="dv">8</span>, ]</span>
<span id="cb117-818"><a href="#cb117-818" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-819"><a href="#cb117-819" aria-hidden="true" tabindex="-1"></a>Let's relevel now, "Housekeeping" is the reference:</span>
<span id="cb117-822"><a href="#cb117-822" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-823"><a href="#cb117-823" aria-hidden="true" tabindex="-1"></a><span class="co"># Relevel: make "Housekeeping" the reference level</span></span>
<span id="cb117-824"><a href="#cb117-824" aria-hidden="true" tabindex="-1"></a>gene_data<span class="sc">$</span>type <span class="ot">&lt;-</span> <span class="fu">relevel</span>(gene_data<span class="sc">$</span>type, <span class="at">ref =</span> <span class="st">"Housekeeping"</span>)</span>
<span id="cb117-825"><a href="#cb117-825" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-826"><a href="#cb117-826" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the new levels</span></span>
<span id="cb117-827"><a href="#cb117-827" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(gene_data<span class="sc">$</span>type)</span>
<span id="cb117-828"><a href="#cb117-828" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-829"><a href="#cb117-829" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-830"><a href="#cb117-830" aria-hidden="true" tabindex="-1"></a><span class="ss">10. </span>Simulated Expression by Tissue</span>
<span id="cb117-833"><a href="#cb117-833" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-834"><a href="#cb117-834" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb117-835"><a href="#cb117-835" aria-hidden="true" tabindex="-1"></a>gene_expr <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">45</span>, <span class="at">mean =</span> <span class="dv">8</span>, <span class="at">sd =</span> <span class="dv">2</span>)</span>
<span id="cb117-836"><a href="#cb117-836" aria-hidden="true" tabindex="-1"></a>tissue <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"brain"</span>, <span class="st">"liver"</span>, <span class="st">"kidney"</span>), <span class="at">each =</span> <span class="dv">15</span>)</span>
<span id="cb117-837"><a href="#cb117-837" aria-hidden="true" tabindex="-1"></a>tissue_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(tissue, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"liver"</span>, <span class="st">"brain"</span>, <span class="st">"kidney"</span>))</span>
<span id="cb117-838"><a href="#cb117-838" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-839"><a href="#cb117-839" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(gene_expr <span class="sc">~</span> tissue_factor, </span>
<span id="cb117-840"><a href="#cb117-840" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"orange"</span>, <span class="st">"skyblue"</span>, <span class="st">"lightgreen"</span>), </span>
<span id="cb117-841"><a href="#cb117-841" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Expression by Tissue"</span>,</span>
<span id="cb117-842"><a href="#cb117-842" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Expression Level"</span>)</span>
<span id="cb117-843"><a href="#cb117-843" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-844"><a href="#cb117-844" aria-hidden="true" tabindex="-1"></a>Let's calculate variability per tissue type now:</span>
<span id="cb117-847"><a href="#cb117-847" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb117-848"><a href="#cb117-848" aria-hidden="true" tabindex="-1"></a><span class="co"># Variability</span></span>
<span id="cb117-849"><a href="#cb117-849" aria-hidden="true" tabindex="-1"></a><span class="fu">tapply</span>(gene_expr, tissue_factor, sd)</span>
<span id="cb117-850"><a href="#cb117-850" aria-hidden="true" tabindex="-1"></a><span class="co"># Returns standard deviation per tissue group</span></span>
<span id="cb117-851"><a href="#cb117-851" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb117-852"><a href="#cb117-852" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-853"><a href="#cb117-853" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb117-854"><a href="#cb117-854" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>