<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="dcterms.date" content="2025-09-09">
<title>pythrash - HW solutions</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="../../_include/styles.css">
</head>
<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">PythRaSh</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-r" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">R</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-r">
<li>
    <a class="dropdown-item" href="../../ch/rbasics/git.html" rel="" target="">
 <span class="dropdown-text">git</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/firststeps.html" rel="" target="">
 <span class="dropdown-text">Basic R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/tidyverse.html" rel="" target="">
 <span class="dropdown-text">The tidyverse</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/correlation_regression.html" rel="" target="">
 <span class="dropdown-text">Advanced R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/baddata.html" rel="" target="">
 <span class="dropdown-text">Data Wrangling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/solutions.html" rel="" target="">
 <span class="dropdown-text">Problem Solutions</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-python" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Python</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-python">
<li>
    <a class="dropdown-item" href="../../ch/python/basics.html" rel="" target="">
 <span class="dropdown-text">Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/panda.html" rel="" target="">
 <span class="dropdown-text">Advanced Python</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/MLA.html" rel="" target="">
 <span class="dropdown-text">Machine Learning Algorithms</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/AI.html" rel="" target="">
 <span class="dropdown-text">AI</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-linux-and-clt" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Linux and CLT</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-linux-and-clt">
<li>
    <a class="dropdown-item" href="../../ch/linux-and-ctl/basics.html" rel="" target="">
 <span class="dropdown-text">Linux Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/linux-and-ctl/advanced.html" rel="" target="">
 <span class="dropdown-text">Advanced Command Line</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-probability-and-statistics" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Probability and Statistics</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-probability-and-statistics">
<li>
    <a class="dropdown-item" href="../../ch/prob_stat/prob_basic.html" rel="" target="">
 <span class="dropdown-text">Probability Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/prob_stat/stat.html" rel="" target="">
 <span class="dropdown-text">Statistics</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-genomics" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Genomics</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-genomics">
<li>
    <a class="dropdown-item" href="../../ch/genomics/basics.html" rel="" target="">
 <span class="dropdown-text">Genomics Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/genomics/advanced.html" rel="" target="">
 <span class="dropdown-text">Advanced Genomics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/genomics/pipeline.html" rel="" target="">
 <span class="dropdown-text">Pipelines</span></a>
  </li>  
    </ul>
</li>
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-blogs" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Blogs</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-blogs">
<li>
    <a class="dropdown-item" href="../../ch/blogs/en/index.html" rel="" target="">
 <span class="dropdown-text">English</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/blogs/bn/index.html" rel="" target="">
 <span class="dropdown-text">Bangla</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/MdRasheduz-zaman" rel="" target="">
 <span class="menu-text">Contact</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/MdRasheduz-zaman/PythRaSh" rel="" target=""><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#l3-matrices-and-lists" id="toc-l3-matrices-and-lists" class="nav-link active" data-scroll-target="#l3-matrices-and-lists">L3: Matrices and Lists</a>
  <ul class="collapse">
<li>
<a href="#task-1-protein-quantification-in-biological-samples" id="toc-task-1-protein-quantification-in-biological-samples" class="nav-link" data-scroll-target="#task-1-protein-quantification-in-biological-samples">Task 1: Protein Quantification in Biological Samples</a>
  <ul class="collapse">
<li><a href="#matrices" id="toc-matrices" class="nav-link" data-scroll-target="#matrices">1. Matrices</a></li>
  <li><a href="#making-transpose-of-proteinmatrix" id="toc-making-transpose-of-proteinmatrix" class="nav-link" data-scroll-target="#making-transpose-of-proteinmatrix">2. Making transpose of ProteinMatrix</a></li>
  <li><a href="#identity-matrix" id="toc-identity-matrix" class="nav-link" data-scroll-target="#identity-matrix">3. Identity Matrix</a></li>
  <li><a href="#total-protein-per-sample" id="toc-total-protein-per-sample" class="nav-link" data-scroll-target="#total-protein-per-sample">4.1. Total Protein per Sample</a></li>
  <li><a href="#total-protein-per-protein-type" id="toc-total-protein-per-protein-type" class="nav-link" data-scroll-target="#total-protein-per-protein-type">4.2. Total Protein per Protein Type</a></li>
  <li><a href="#heatmap-of-protein-concentrations" id="toc-heatmap-of-protein-concentrations" class="nav-link" data-scroll-target="#heatmap-of-protein-concentrations">4.3. Heatmap of Protein Concentrations</a></li>
  <li><a href="#interpretation" id="toc-interpretation" class="nav-link" data-scroll-target="#interpretation">🧠 Interpretation</a></li>
  </ul>
</li>
  <li>
<a href="#task-2-gene-to-protein-translation" id="toc-task-2-gene-to-protein-translation" class="nav-link" data-scroll-target="#task-2-gene-to-protein-translation">Task 2: Gene-to-Protein Translation</a>
  <ul class="collapse">
<li><a href="#protein-output" id="toc-protein-output" class="nav-link" data-scroll-target="#protein-output">Protein Output</a></li>
  <li><a href="#transpose" id="toc-transpose" class="nav-link" data-scroll-target="#transpose">2. Transpose</a></li>
  <li><a href="#identity-matrix-and-multipliccation" id="toc-identity-matrix-and-multipliccation" class="nav-link" data-scroll-target="#identity-matrix-and-multipliccation">3. Identity matrix and multipliccation</a></li>
  <li><a href="#sub-matrix" id="toc-sub-matrix" class="nav-link" data-scroll-target="#sub-matrix">4. Sub-matrix:</a></li>
  <li><a href="#visualization-tasks" id="toc-visualization-tasks" class="nav-link" data-scroll-target="#visualization-tasks">Visualization tasks:</a></li>
  <li><a href="#interpretation-1" id="toc-interpretation-1" class="nav-link" data-scroll-target="#interpretation-1">🧠 Interpretation</a></li>
  </ul>
</li>
  <li>
<a href="#task-3-animal-breeding-bull-ranking-by-economic-traits" id="toc-task-3-animal-breeding-bull-ranking-by-economic-traits" class="nav-link" data-scroll-target="#task-3-animal-breeding-bull-ranking-by-economic-traits">Task 3: Animal Breeding – Bull Ranking by Economic Traits</a>
  <ul class="collapse">
<li><a href="#define-data" id="toc-define-data" class="nav-link" data-scroll-target="#define-data">Define Data</a></li>
  <li><a href="#compute-total-economic-value" id="toc-compute-total-economic-value" class="nav-link" data-scroll-target="#compute-total-economic-value">Compute Total Economic Value</a></li>
  <li><a href="#multiply-with-identity-matrix" id="toc-multiply-with-identity-matrix" class="nav-link" data-scroll-target="#multiply-with-identity-matrix">Multiply with Identity Matrix</a></li>
  <li><a href="#remove-milk-yield-and-recalculate-total-value" id="toc-remove-milk-yield-and-recalculate-total-value" class="nav-link" data-scroll-target="#remove-milk-yield-and-recalculate-total-value">Remove Milk Yield and Recalculate Total Value</a></li>
  <li><a href="#bar-plot-total-economic-value" id="toc-bar-plot-total-economic-value" class="nav-link" data-scroll-target="#bar-plot-total-economic-value">Bar Plot: Total Economic Value</a></li>
  <li><a href="#heatmap-of-ebvs" id="toc-heatmap-of-ebvs" class="nav-link" data-scroll-target="#heatmap-of-ebvs">Heatmap of EBVs</a></li>
  <li><a href="#interpretation-questions" id="toc-interpretation-questions" class="nav-link" data-scroll-target="#interpretation-questions">🧠 Interpretation Questions</a></li>
  </ul>
</li>
  <li>
<a href="#task-4-plant-breeding-trait-contributions-from-parental-lines" id="toc-task-4-plant-breeding-trait-contributions-from-parental-lines" class="nav-link" data-scroll-target="#task-4-plant-breeding-trait-contributions-from-parental-lines">Task 4: Plant Breeding – Trait Contributions from Parental Lines</a>
  <ul class="collapse">
<li><a href="#define-data-1" id="toc-define-data-1" class="nav-link" data-scroll-target="#define-data-1">Define Data</a></li>
  <li><a href="#compute-hybridtraits-vector" id="toc-compute-hybridtraits-vector" class="nav-link" data-scroll-target="#compute-hybridtraits-vector">Compute HybridTraits Vector</a></li>
  <li><a href="#multiply-with-identity-matrix-1" id="toc-multiply-with-identity-matrix-1" class="nav-link" data-scroll-target="#multiply-with-identity-matrix-1">Multiply with Identity Matrix</a></li>
  <li><a href="#remove-t3-maturation-time-and-recalculate" id="toc-remove-t3-maturation-time-and-recalculate" class="nav-link" data-scroll-target="#remove-t3-maturation-time-and-recalculate">Remove T3 (Maturation Time) and Recalculate</a></li>
  <li><a href="#heatmap-of-parenttraits" id="toc-heatmap-of-parenttraits" class="nav-link" data-scroll-target="#heatmap-of-parenttraits">Heatmap of ParentTraits</a></li>
  <li><a href="#bar-plot-of-hybrid-traits" id="toc-bar-plot-of-hybrid-traits" class="nav-link" data-scroll-target="#bar-plot-of-hybrid-traits">Bar Plot of Hybrid Traits</a></li>
  <li><a href="#interpretation-questions-1" id="toc-interpretation-questions-1" class="nav-link" data-scroll-target="#interpretation-questions-1">🧠 Interpretation Questions</a></li>
  </ul>
</li>
  <li>
<a href="#task-5-managing-matrices-and-weight-vectors-using-lists-in-r" id="toc-task-5-managing-matrices-and-weight-vectors-using-lists-in-r" class="nav-link" data-scroll-target="#task-5-managing-matrices-and-weight-vectors-using-lists-in-r">Task 5: Managing Matrices and Weight Vectors Using Lists in R</a>
  <ul class="collapse">
<li><a href="#create-a-master-list" id="toc-create-a-master-list" class="nav-link" data-scroll-target="#create-a-master-list">Create a Master List</a></li>
  <li><a href="#list-the-structure" id="toc-list-the-structure" class="nav-link" data-scroll-target="#list-the-structure">List the Structure</a></li>
  <li><a href="#indexing-elements-from-lists" id="toc-indexing-elements-from-lists" class="nav-link" data-scroll-target="#indexing-elements-from-lists">Indexing Elements from Lists</a></li>
  <li><a href="#perform-weighted-calculations" id="toc-perform-weighted-calculations" class="nav-link" data-scroll-target="#perform-weighted-calculations">Perform Weighted Calculations</a></li>
  <li><a href="#subset-and-recalculate" id="toc-subset-and-recalculate" class="nav-link" data-scroll-target="#subset-and-recalculate">Subset and Recalculate</a></li>
  <li><a href="#visualization-tasks-1" id="toc-visualization-tasks-1" class="nav-link" data-scroll-target="#visualization-tasks-1">Visualization Tasks</a></li>
  <li><a href="#interpretation-questions-2" id="toc-interpretation-questions-2" class="nav-link" data-scroll-target="#interpretation-questions-2">🧠 Interpretation Questions</a></li>
  </ul>
</li>
  <li><a href="#homework-solutions-factors-subsetting-and-biological-insight" id="toc-homework-solutions-factors-subsetting-and-biological-insight" class="nav-link" data-scroll-target="#homework-solutions-factors-subsetting-and-biological-insight">🧩 Homework solutions: Factors, Subsetting, and Biological Insight</a></li>
  <li><a href="#homework-solution-tidyverse" id="toc-homework-solution-tidyverse" class="nav-link" data-scroll-target="#homework-solution-tidyverse">🧩 Homework Solution tidyverse</a></li>
  </ul>
</li>
  </ul><div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="solutions.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">HW solutions</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Md Rasheduzzaman </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://www.fli.de/de/institute/institut-fuer-infektionsmedizin-imed/">
            FLI Insel Riems
            </a>
          </p>
        <p class="affiliation">
            <a href="https://github.com/MdRasheduz-zaman">
            &amp; Freelancer
            </a>
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Last updated</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 9, 2025</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="abstract-title">Content summary</div>
    dataframe, matrices, list, factor, vector, tidyverse, etc.
  </div>
</div>

</header><section id="l3-matrices-and-lists" class="level1"><h1>L3: Matrices and Lists</h1>
<section id="task-1-protein-quantification-in-biological-samples" class="level2"><h2 class="anchored" data-anchor-id="task-1-protein-quantification-in-biological-samples">Task 1: Protein Quantification in Biological Samples</h2>
<p>We measured the concentration (in µg/µL) of three proteins (P1, P2, P3) in four samples (S1–S4):</p>
<section id="matrices" class="level3"><h3 class="anchored" data-anchor-id="matrices">1. Matrices</h3>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Making Protein Matrix</span></span>
<span><span class="va">ProteinMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">3</span>, <span class="fl">2</span>,</span>
<span>    <span class="fl">7</span>, <span class="fl">6</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample1"</span>, <span class="st">"Sample2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="va">ProteinMatrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        ProteinX ProteinY ProteinZ
Sample1        5        3        2
Sample2        7        6        4</code></pre>
</div>
</div>
<p>Now goes the weight matrix</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Making weight matrix</span></span>
<span><span class="va">WeightVector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>  nrow<span class="op">=</span><span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">WeightVector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5</code></pre>
</div>
</div>
<p>Now, multiply them.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Multiplying Matrices</span></span>
<span><span class="va">TotalConc</span> <span class="op">=</span> <span class="va">ProteinMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">WeightVector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TotalConc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">TotalConc</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Total_Protein_Conc
Sample1                8.5
Sample2               15.5</code></pre>
</div>
</div>
</section><section id="making-transpose-of-proteinmatrix" class="level3"><h3 class="anchored" data-anchor-id="making-transpose-of-proteinmatrix">2. Making transpose of ProteinMatrix</h3>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ProteinMatTranspose</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span></span>
<span><span class="va">ProteinMatTranspose</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Sample1 Sample2
ProteinX       5       7
ProteinY       3       6
ProteinZ       2       4</code></pre>
</div>
</div>
</section><section id="identity-matrix" class="level3"><h3 class="anchored" data-anchor-id="identity-matrix">3. Identity Matrix</h3>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">I</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">Identitycheck</span> <span class="op">=</span> <span class="va">ProteinMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Identitycheck</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="va">Identitycheck</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        ProteinX ProteinY ProteinZ
Sample1        5        3        2
Sample2        7        6        4</code></pre>
</div>
</div>
</section><section id="total-protein-per-sample" class="level3"><h3 class="anchored" data-anchor-id="total-protein-per-sample">4.1. Total Protein per Sample</h3>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sample1 Sample2 
     10      17 </code></pre>
</div>
</div>
</section><section id="total-protein-per-protein-type" class="level3"><h3 class="anchored" data-anchor-id="total-protein-per-protein-type">4.2. Total Protein per Protein Type</h3>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ProteinX ProteinY ProteinZ 
      12        9        6 </code></pre>
</div>
</div>
</section><section id="heatmap-of-protein-concentrations" class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-protein-concentrations">4.3. Heatmap of Protein Concentrations</h3>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">ProteinMatrix</span>, scale <span class="op">=</span> <span class="st">"none"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation" class="level3"><h3 class="anchored" data-anchor-id="interpretation">🧠 Interpretation</h3>
<ul>
<li>Multiplying the protein levels by the weight vector shows how much each protein contributes in a sample. The result shows total protein concentration per sample.</li>
<li>The result shows that sample S2 has the highest protein burden.</li>
<li>The identity matrix represents no protein interactions or measurement biases. It is a simple matrix calculation.</li>
<li>New calculation:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># changing the weight of ProteinZ to 3.0</span></span>
<span><span class="va">newweightvector</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">3.0</span><span class="op">)</span>,</span>
<span>  nrow<span class="op">=</span><span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">newTotalconc</span> <span class="op">=</span> <span class="va">ProteinMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">newweightvector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">newTotalconc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span><span class="va">newTotalconc</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Total_Protein_Conc
Sample1               11.5
Sample2               21.5</code></pre>
</div>
</div>
<p>Still, S2 has more protein burden.</p>
<p>Bonus:</p>
<ul>
<li>Heatmap reveals PX is most abundant across all samples.</li>
</ul></section></section><section id="task-2-gene-to-protein-translation" class="level2"><h2 class="anchored" data-anchor-id="task-2-gene-to-protein-translation">Task 2: Gene-to-Protein Translation</h2>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># making  Gene Expression matrix</span></span>
<span><span class="va">GeneExpression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">8</span>, <span class="fl">5</span>,</span>
<span>    <span class="fl">15</span>, <span class="fl">12</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">GeneExpression</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample1"</span>, <span class="st">"Sample2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">GeneExpression</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span><span class="va">GeneExpression</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        GeneA GeneB GeneC
Sample1    10     8     5
Sample2    15    12    10</code></pre>
</div>
</div>
<p>Translation efficiency:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># making Translation Matrix</span></span>
<span><span class="va">TranslationMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">0</span> , <span class="fl">0</span>,</span>
<span>  <span class="fl">0</span>, <span class="fl">1.2</span>, <span class="fl">0</span>,</span>
<span>  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1.8</span><span class="op">)</span>,</span>
<span>nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">TranslationMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TranslationMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"protA"</span>, <span class="st">"protB"</span>, <span class="st">"protC"</span><span class="op">)</span></span>
<span><span class="va">TranslationMatrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      protA protB protC
GeneA   1.5   0.0   0.0
GeneB   0.0   1.2   0.0
GeneC   0.0   0.0   1.8</code></pre>
</div>
</div>
<section id="protein-output" class="level3"><h3 class="anchored" data-anchor-id="protein-output">Protein Output</h3>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># computing Protein matrix</span></span>
<span><span class="va">Protein_matrix</span> <span class="op">&lt;-</span> <span class="va">GeneExpression</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">TranslationMatrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Protein_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"total_protA"</span>, <span class="st">"total_protB"</span>, <span class="st">"total_protC"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Protein_matrix</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        total_protA total_protB total_protC
Sample1        15.0         9.6           9
Sample2        22.5        14.4          18</code></pre>
</div>
</div>
</section><section id="transpose" class="level3"><h3 class="anchored" data-anchor-id="transpose">2. Transpose</h3>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Transpose of GeneExpression matrix</span></span>
<span><span class="va">GeneExpression_Transpose</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">GeneExpression</span><span class="op">)</span></span>
<span><span class="va">GeneExpression_Transpose</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Sample1 Sample2
GeneA      10      15
GeneB       8      12
GeneC       5      10</code></pre>
</div>
</div>
<p>The new matrix represnts a matrix where the rows and columns of GeneExpression matrix have been interchanged.</p>
</section><section id="identity-matrix-and-multipliccation" class="level3"><h3 class="anchored" data-anchor-id="identity-matrix-and-multipliccation">3. Identity matrix and multipliccation</h3>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Creating Identity matrix</span></span>
<span><span class="va">I</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">I</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3]
[1,]    1    0    0
[2,]    0    1    0
[3,]    0    0    1</code></pre>
</div>
</div>
<p>Now, multiply:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">Product_matrix</span> <span class="op">=</span> <span class="va">TranslationMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span>  <span class="va">I</span> </span>
<span><span class="va">Product_matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1] [,2] [,3]
GeneA  1.5  0.0  0.0
GeneB  0.0  1.2  0.0
GeneC  0.0  0.0  1.8</code></pre>
</div>
</div>
<p>The product is identical to TranslationMatrix</p>
</section><section id="sub-matrix" class="level3"><h3 class="anchored" data-anchor-id="sub-matrix">4. Sub-matrix:</h3>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># making submatrix A</span></span>
<span><span class="va">A</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">8</span>,</span>
<span>    <span class="fl">15</span>, <span class="fl">12</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">A</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        GeneA GeneB
sample1    10     8
sample2    15    12</code></pre>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># finding inverse of A</span></span>
<span><span class="co">#inv_A &lt;- solve(A)</span></span>
<span><span class="co">#inv_A</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The inverse matrix could not be calculated since A is a singular matrix. So, A * A^-1 is also not possible.</p>
</section><section id="visualization-tasks" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks">Visualization tasks:</h3>
<ul>
<li><ol type="1">
<li>MARplot</li>
</ol></li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># generating MARplot-style scatter plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">GeneExpression</span>, <span class="va">Protein_matrix</span>, type<span class="op">=</span><span class="st">"p"</span>, main<span class="op">=</span><span class="st">"Protein level vs. Gene Expression level"</span><span class="op">)</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="st">"Sample-Gene"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">GeneExpression</span>, <span class="va">Protein_matrix</span>, labels <span class="op">=</span> <span class="va">labels</span>, pos<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># generating a heatmap</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">Protein_matrix</span>, main<span class="op">=</span> <span class="st">"Heatmap of Protein Level"</span>, Rowv <span class="op">=</span> <span class="cn">TRUE</span>, Colv <span class="op">=</span> <span class="cn">TRUE</span>, labRow<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Protein_matrix</span><span class="op">)</span>, labCol<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinA"</span>, <span class="st">"ProteinB"</span>, <span class="st">"ProteinC"</span><span class="op">)</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">topo.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span> <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid" width="1650"></p>
</div>
</div>
<ul>
<li><ol start="2" type="1">
<li>Heatmap of Expression:</li>
</ol></li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">GeneExpression</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">terrain.colors</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>, scale <span class="op">=</span> <span class="st">"column"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation-1" class="level3"><h3 class="anchored" data-anchor-id="interpretation-1">🧠 Interpretation</h3>
<ol type="1">
<li><p>Matrix multiplication allows each gene in both samples to be multiplied to their respective translation efficiency. So, the product shows how successfully each gene is translated”)</p></li>
<li><p>The diagonal TranslationMatrix make sense biologically because they show translation efficiency of each gene and there is no other interaction between them. Although there could be interaction in real-world scenarios.</p></li>
<li><p>If Sample2 has higher protein levels even with similar gene expression, it means that more mRNAs are translated to proteins compared to Sample1”</p></li>
<li><p>The upward trend in MARplot may indicate an increase in translation efficacy and downward trend may indicate a decline in translation efficacy”</p></li>
<li><p>Clustering in the heatmap may suggest which samples are most similar to each other based on their prot.</p></li>
</ol></section></section><section id="task-3-animal-breeding-bull-ranking-by-economic-traits" class="level2"><h2 class="anchored" data-anchor-id="task-3-animal-breeding-bull-ranking-by-economic-traits">Task 3: Animal Breeding – Bull Ranking by Economic Traits</h2>
<section id="define-data" class="level3"><h3 class="anchored" data-anchor-id="define-data">Define Data</h3>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define Bull EBVs</span></span>
<span><span class="va">BullEBVs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="fl">400</span>, <span class="fl">1.2</span>, <span class="fl">0.8</span>,</span>
<span>  <span class="fl">500</span>, <span class="fl">1.5</span>, <span class="fl">0.6</span></span>
<span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Bull1"</span>, <span class="st">"Bull2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Milk_yield"</span>, <span class="st">"Growth_rate"</span>, <span class="st">"Fertility"</span><span class="op">)</span></span>
<span><span class="va">BullEBVs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Milk_yield Growth_rate Fertility
Bull1        400         1.2       0.8
Bull2        500         1.5       0.6</code></pre>
</div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define Economic Weights</span></span>
<span><span class="va">EconomicWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.002</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">EconomicWeights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">EconomicWeights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">EconomicWeights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Weight
Milk_yield   2e-03
Growth_rate  5e+01
Fertility    1e+02</code></pre>
</div>
</div>
</section><section id="compute-total-economic-value" class="level3"><h3 class="anchored" data-anchor-id="compute-total-economic-value">Compute Total Economic Value</h3>
<div class="cell">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">TotalValue</span> <span class="op">&lt;-</span> <span class="va">BullEBVs</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">EconomicWeights</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TotalValue</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Merit"</span><span class="op">)</span></span>
<span><span class="va">TotalValue</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Merit
Bull1 140.8
Bull2 136.0</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Bull1: (400 × 0.002) + (1.2 × 50) + (0.8 × 100) = 140.8</p>
<p>Bull2: (500 × 0.002) + (1.5 × 50) + (0.6 × 100) = 136.0</p>
<p>Bull1 is more valuable economically.</p>
<p><strong>Biological Interpretation</strong></p>
<p>Economic weights convert genetic merit (EBVs, Estimated Breeding Values) into economic merit. Traits with higher financial importance have a larger impact, regardless of absolute EBV values.</p>
</section><section id="multiply-with-identity-matrix" class="level3"><h3 class="anchored" data-anchor-id="multiply-with-identity-matrix">Multiply with Identity Matrix</h3>
<div class="cell">
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">I3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">I3</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">I3</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span></span>
<span><span class="va">I3</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            Milk_yield Growth_rate Fertility
Milk_yield           1           0         0
Growth_rate          0           1         0
Fertility            0           0         1</code></pre>
</div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">BullEBVs_identity</span> <span class="op">&lt;-</span> <span class="va">BullEBVs</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I3</span></span>
<span><span class="va">BullEBVs_identity</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Milk_yield Growth_rate Fertility
Bull1        400         1.2       0.8
Bull2        500         1.5       0.6</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Multiplying by identity matrix returns the original matrix. It confirms that EBV structure is preserved.</p>
</section><section id="remove-milk-yield-and-recalculate-total-value" class="level3"><h3 class="anchored" data-anchor-id="remove-milk-yield-and-recalculate-total-value">Remove Milk Yield and Recalculate Total Value</h3>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">BullEBVs_noMilk</span> <span class="op">&lt;-</span> <span class="va">BullEBVs</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">EconomicWeights_noMilk</span> <span class="op">&lt;-</span> <span class="va">EconomicWeights</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="va">TotalValue_noMilk</span> <span class="op">&lt;-</span> <span class="va">BullEBVs_noMilk</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">EconomicWeights_noMilk</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TotalValue_noMilk</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"New_Merit"</span><span class="op">)</span></span>
<span><span class="va">TotalValue_noMilk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      New_Merit
Bull1       140
Bull2       135</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Bull1: (1.2 × 50) + (0.8 × 100) = 140</p>
<p>Bull2: (1.5 × 50) + (0.6 × 100) = 135</p>
<p>Bull1 still ranks higher, but by a smaller margin.</p>
</section><section id="bar-plot-total-economic-value" class="level3"><h3 class="anchored" data-anchor-id="bar-plot-total-economic-value">Bar Plot: Total Economic Value</h3>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">TotalValue</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  names.arg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"skyblue"</span>, <span class="st">"orange"</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Total Economic Value of Bulls"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Total Value"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="heatmap-of-ebvs" class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-ebvs">Heatmap of EBVs</h3>
<div class="cell">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">BullEBVs</span>,</span>
<span>  Rowv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  Colv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Heatmap of Bull EBVs"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation-questions" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions">🧠 Interpretation Questions</h3>
<ul>
<li>How do economic weights affect trait importance?</li>
</ul>
<p>Traits with higher weights contribute more to the total economic value. This makes them more influential in ranking and selection.</p>
<ul>
<li>Why might you ignore milk yield?</li>
</ul>
<p>Milk yield may be excluded in systems focusing on fertility, growth, or when it is no longer a limiting factor. Environmental or economic contexts may also shift trait priorities.</p>
<ul>
<li>What is the value of heatmaps?</li>
</ul>
<p>Heatmaps visually compare EBVs across bulls and traits.They help detect patterns, outliers, and clusters easily in multivariate data.</p>
<ul>
<li>Can this method be extended to more bulls and traits?</li>
</ul>
<p>Yes. This method scales to any number of bulls or traits. Just ensure the EBVs matrix and economic weights are dimensionally compatible.</p>
</section></section><section id="task-4-plant-breeding-trait-contributions-from-parental-lines" class="level2"><h2 class="anchored" data-anchor-id="task-4-plant-breeding-trait-contributions-from-parental-lines">Task 4: Plant Breeding – Trait Contributions from Parental Lines</h2>
<section id="define-data-1" class="level3"><h3 class="anchored" data-anchor-id="define-data-1">Define Data</h3>
<p><strong>Parent trait values (normalized 1–10)</strong></p>
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ParentTraits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="fl">7</span>, <span class="fl">5</span>, <span class="fl">3</span>,</span>
<span>  <span class="fl">6</span>, <span class="fl">8</span>, <span class="fl">4</span>,</span>
<span>  <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">6</span></span>
<span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"P1"</span>, <span class="st">"P2"</span>, <span class="st">"P3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Drought_resistance"</span>, <span class="st">"Yield"</span>, <span class="st">"Maturation_time"</span><span class="op">)</span></span>
<span><span class="va">ParentTraits</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define Hybrid Weights</span></span>
<span><span class="va">HybridWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">HybridWeights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">HybridWeights</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">HybridWeights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Drought_resistance Yield Maturation_time
Weight                0.5   0.3             0.2</code></pre>
</div>
</div>
</section><section id="compute-hybridtraits-vector" class="level3"><h3 class="anchored" data-anchor-id="compute-hybridtraits-vector">Compute HybridTraits Vector</h3>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">HybridTraits</span> <span class="op">&lt;-</span> <span class="va">HybridWeights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">ParentTraits</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">HybridTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Contribution"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">HybridTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span></span>
<span><span class="va">HybridTraits</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              P1  P2  P3
Contribution 6.3 6.1 3.9</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>HybridTraits = (0.5 × P1) + (0.3 × P2) + (0.2 × P3)</p>
<p>Drought_resistance = (0.5 × 7) + (0.3 × 6) + (0.2 × 5) = 6.3</p>
<p>Yield = (0.5 × 5) + (0.3 × 8) + (0.2 × 6) = 6.1</p>
<p>Maturation_time = (0.5 × 3) + (0.3 × 4) + (0.2 × 6) = 3.9</p>
<p>The hybrid is moderately strong in drought resistance and yield, and has a relatively shorter maturation time.</p>
<p><strong>Biological Meaning of Unequal Contribution</strong></p>
<p>When one parent contributes more to a trait, it suggests that the trait’s heritable strength comes disproportionately from that parent. Breeders can use this knowledge to amplify desirable traits using the best parent.</p>
</section><section id="multiply-with-identity-matrix-1" class="level3"><h3 class="anchored" data-anchor-id="multiply-with-identity-matrix-1">Multiply with Identity Matrix</h3>
<div class="cell">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">I3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">ParentTraits_identity</span> <span class="op">&lt;-</span> <span class="va">ParentTraits</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits_identity</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span></span>
<span><span class="va">ParentTraits_identity</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Multiplying by identity matrix returns the original matrix. This operation verifies structural consistency and dimensionality.</p>
</section><section id="remove-t3-maturation-time-and-recalculate" class="level3"><h3 class="anchored" data-anchor-id="remove-t3-maturation-time-and-recalculate">Remove T3 (Maturation Time) and Recalculate</h3>
<div class="cell">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ParentTraits_T1T2</span> <span class="op">&lt;-</span> <span class="va">ParentTraits</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">ParentTraits_T1T2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield
P1                  7     5
P2                  6     8
P3                  5     6</code></pre>
</div>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">HybridTraits_T1T2</span> <span class="op">&lt;-</span> <span class="va">HybridWeights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">ParentTraits_T1T2</span></span>
<span><span class="va">HybridTraits_T1T2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Drought_resistance Yield
Weight                6.3   6.1</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Drought_resistance = (0.5 × 7) + (0.3 × 6) + (0.2 × 5) = 6.3</p>
<p>Yield = (0.5 × 5) + (0.3 × 8) + (0.2 × 6) = 6.1</p>
<p>Removing a trait (T3) changes the trait profile. Hybrid selection may now favor traits that remain.</p>
</section><section id="heatmap-of-parenttraits" class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-parenttraits">Heatmap of ParentTraits</h3>
<div class="cell">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">ParentTraits</span>,</span>
<span>  Rowv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  Colv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Heatmap of Parent Traits"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="bar-plot-of-hybrid-traits" class="level3"><h3 class="anchored" data-anchor-id="bar-plot-of-hybrid-traits">Bar Plot of Hybrid Traits</h3>
<div class="cell">
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">HybridTraits</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  names.arg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#66c2a5"</span>, <span class="st">"#fc8d62"</span>, <span class="st">"#8da0cb"</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Hybrid Trait Profile"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Trait Value"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation-questions-1" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-1">🧠 Interpretation Questions</h3>
<ul>
<li>
<p>How does the weighting of parents affect the hybrid’s performance?</p>
<p>Stronger weights mean more genetic contribution. Traits from highly weighted parents dominate the hybrid profile.</p>
</li>
<li>
<p>What does the identity matrix represent here?</p>
<p>It represents a neutral transformation. It confirms data integrity when used in matrix multiplication.</p>
</li>
<li>
<p>If you used equal weights (⅓ for each), how would the hybrid traits change?</p>
<p>Traits would reflect an even mix, potentially leading to balanced but less specialized performance.</p>
</li>
<li>
<p>What real-world limitations does this simplified model ignore?</p>
<ol type="i">
<li><p>Non-additive genetic effects (dominance, epistasis)</p></li>
<li><p>Environmental interactions</p></li>
<li><p>Trait heritability and correlations</p></li>
<li><p>Breeding feasibility and cost</p></li>
</ol>
</li>
</ul></section></section><section id="task-5-managing-matrices-and-weight-vectors-using-lists-in-r" class="level2"><h2 class="anchored" data-anchor-id="task-5-managing-matrices-and-weight-vectors-using-lists-in-r">Task 5: Managing Matrices and Weight Vectors Using Lists in R</h2>
<p>Now that we’ve explored trait-based decisions using matrices, it’s time to organize our work using R’s list structure. Lists help bundle related objects like matrices and weight vectors, keeping the analysis modular and scalable.</p>
<section id="create-a-master-list" class="level3"><h3 class="anchored" data-anchor-id="create-a-master-list">Create a Master List</h3>
<div class="cell">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Assuming previous matrices and weights are already defined:</span></span>
<span></span>
<span><span class="co"># Making a MasterList</span></span>
<span><span class="va">bioList</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  ProteinConc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">ProteinMatrix</span>, weights <span class="op">=</span> <span class="va">WeightVector</span><span class="op">)</span>,</span>
<span>  ProteinMap  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">GeneExpression</span>, weights <span class="op">=</span> <span class="va">TranslationMatrix</span><span class="op">)</span>,</span>
<span>  Animal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">BullEBVs</span>, weights <span class="op">=</span> <span class="va">EconomicWeights</span><span class="op">)</span>,</span>
<span>  Plant <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">ParentTraits</span>, weights <span class="op">=</span> <span class="va">HybridWeights</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">bioList</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$ProteinConc
$ProteinConc$matrix
        ProteinX ProteinY ProteinZ
Sample1        5        3        2
Sample2        7        6        4

$ProteinConc$weights
         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5


$ProteinMap
$ProteinMap$matrix
        GeneA GeneB GeneC
Sample1    10     8     5
Sample2    15    12    10

$ProteinMap$weights
      protA protB protC
GeneA   1.5   0.0   0.0
GeneB   0.0   1.2   0.0
GeneC   0.0   0.0   1.8


$Animal
$Animal$matrix
      Milk_yield Growth_rate Fertility
Bull1        400         1.2       0.8
Bull2        500         1.5       0.6

$Animal$weights
            Weight
Milk_yield   2e-03
Growth_rate  5e+01
Fertility    1e+02


$Plant
$Plant$matrix
   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6

$Plant$weights
       Drought_resistance Yield Maturation_time
Weight                0.5   0.3             0.2</code></pre>
</div>
</div>
</section><section id="list-the-structure" class="level3"><h3 class="anchored" data-anchor-id="list-the-structure">List the Structure</h3>
<div class="cell">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">bioList</span><span class="op">)</span>               <span class="co"># Top-level list names</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ProteinConc" "ProteinMap"  "Animal"      "Plant"      </code></pre>
</div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span><span class="op">(</span><span class="va">bioList</span><span class="op">)</span>             <span class="co"># Number of components in each sublist</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ProteinConc  ProteinMap      Animal       Plant 
          2           2           2           2 </code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Each top-level entry (e.g., ProteinConc, Plant) contains two components:</p>
<ul>
<li>A matrix (e.g., ProteinMatrix)</li>
<li>A corresponding weight vector or matrix</li>
</ul></section><section id="indexing-elements-from-lists" class="level3"><h3 class="anchored" data-anchor-id="indexing-elements-from-lists">Indexing Elements from Lists</h3>
<div class="cell">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Access the trait matrix for Plant</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#or</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Access the weight vector for ProteinConc</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5</code></pre>
</div>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#or</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Use double brackets <code>[[ ]]</code> to extract unnamed list elements by position. But we named our list, so they are easily extractable using the <code>$</code> notation.</p>
</section><section id="perform-weighted-calculations" class="level3"><h3 class="anchored" data-anchor-id="perform-weighted-calculations">Perform Weighted Calculations</h3>
<div class="cell">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Protein concentration score</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Weight
Sample1    8.5
Sample2   15.5</code></pre>
</div>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Gene → Protein contribution</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinMap</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">ProteinMap</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        protA protB protC
Sample1  15.0   9.6     9
Sample2  22.5  14.4    18</code></pre>
</div>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Bull economic value</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Weight
Bull1  140.8
Bull2  136.0</code></pre>
</div>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Hybrid trait value</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">weights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Drought_resistance Yield Maturation_time
Weight                6.3   6.1             3.9</code></pre>
</div>
</div>
</section><section id="subset-and-recalculate" class="level3"><h3 class="anchored" data-anchor-id="subset-and-recalculate">Subset and Recalculate</h3>
<div class="cell">
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Remove last trait from ParentTraits</span></span>
<span><span class="va">ParentSubset</span> <span class="op">&lt;-</span> <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">NewWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.4</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Recalculated hybrid score</span></span>
<span><span class="va">SubsetHybridScore</span> <span class="op">&lt;-</span> <span class="va">ParentSubset</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">NewWeights</span></span>
<span><span class="va">SubsetHybridScore</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   [,1]
P1  6.2
P2  6.8
P3  5.4</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Dropping a trait and reweighting highlights its influence in trait aggregation and selection.</p>
</section><section id="visualization-tasks-1" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks-1">Visualization Tasks</h3>
<section id="heatmap-gene-expression" class="level4"><h4 class="anchored" data-anchor-id="heatmap-gene-expression">Heatmap: Gene Expression</h4>
<div class="cell">
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">bioList</span><span class="op">$</span><span class="va">ProteinMap</span><span class="op">$</span><span class="va">matrix</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Gene Expression Heatmap"</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"Proteins"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Genes"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="bar-plots" class="level4"><h4 class="anchored" data-anchor-id="bar-plots">Bar Plots</h4>
<section id="hybrid-traits" class="level5"><h5 class="anchored" data-anchor-id="hybrid-traits">Hybrid traits</h5>
<div class="cell">
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">weights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Hybrid Trait Contributions"</span>,</span>
<span>  col <span class="op">=</span> <span class="st">"#66c2a5"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Score"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="bull-ebv-economic-breeding-values" class="level5"><h5 class="anchored" data-anchor-id="bull-ebv-economic-breeding-values">Bull EBV (Economic Breeding Values)</h5>
<div class="cell">
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">weights</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Bull EBVs (Economic Values)"</span>,</span>
<span>  col <span class="op">=</span> <span class="st">"#fc8d62"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Score"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section></section></section><section id="interpretation-questions-2" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-2">🧠 Interpretation Questions</h3>
<ul>
<li>Why use a list structure?</li>
</ul>
<p>Keeps each dataset and its weights together. Facilitates automated workflows and reuse.</p>
<ul>
<li>What’s tricky about <code>[[ ]]</code> access?</li>
</ul>
<p>You must remember the order (<code>[[1]]</code> = <code>matrix</code>, <code>[[2]]</code> = <code>weights</code>). No names means you can’t use <code>$matrix</code>, only positional access.</p>
<p>Loop across all list entries</p>
<p>Weighted scores for all entries <code>lapply(bioList, function(x) x[[2]] %*% x[[1]])</code></p>
<ul>
<li>How does this help in large-scale pipelines?</li>
</ul>
<p>You can use this format with <code><a href="https://rdrr.io/r/base/lapply.html">lapply()</a></code>, <code><a href="https://purrr.tidyverse.org/reference/map.html">purrr::map()</a></code>, or in targets pipelines for reproducibility and modular processing.</p>
</section></section><section id="homework-solutions-factors-subsetting-and-biological-insight" class="level2"><h2 class="anchored" data-anchor-id="homework-solutions-factors-subsetting-and-biological-insight">🧩 Homework solutions: Factors, Subsetting, and Biological Insight</h2>
<ol type="1">
<li><p><strong>Character vs Factor</strong> A character vector simply holds string values, but a factor is a categorical variable with fixed levels, used especially in modeling.<br>
For <code>mutation_status</code>, a factor ensures consistent categories (e.g., <code>"Yes"</code> or <code>"No"</code>) and helps control level order and statistical reference groups.</p></li>
<li><p><strong>Factor Levels</strong></p></li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Lactobacillus"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Lactobacillus"</span><span class="op">)</span></span>
<span><span class="va">species_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">species</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Lactobacillus"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">species_factor</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Bacteroides"   "Escherichia"   "Lactobacillus"</code></pre>
</div>
</div>
<p>Because we defined the level order explicitly, R maintains that order regardless of data input.</p>
<ol start="3" type="1">
<li>Ordered Factor Comparison</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">disease_severity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mild"</span>, <span class="st">"Severe"</span>, <span class="st">"Moderate"</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mild"</span>, <span class="st">"Moderate"</span>, <span class="st">"Severe"</span>, <span class="st">"Critical"</span><span class="op">)</span>, ordered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">disease_severity</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;</span> <span class="va">disease_severity</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>“Mild” is less severe than “Severe” based on the defined order.</p>
<ol start="4" type="1">
<li>Proportion Extraction</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">prop</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">species_factor</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">prop</span><span class="op">[</span><span class="st">"Escherichia"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Escherichia 
        0.2 </code></pre>
</div>
</div>
<p><code>prop$Escherichia</code> won’t work — named numeric vectors require bracket-based access.</p>
<ol start="5" type="1">
<li>Subsetting by Conditions</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  gene_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"EGFR"</span>, <span class="st">"GAPDH"</span><span class="op">)</span>,</span>
<span>  expression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8.2</span>, <span class="fl">6.1</span>, <span class="fl">9.5</span>, <span class="fl">7.0</span>, <span class="fl">10.0</span><span class="op">)</span>,</span>
<span>  mutation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"No"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  pathway <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"DNA Repair"</span>, <span class="st">"Apoptosis"</span>, <span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>, <span class="st">"Metabolism"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">gene_df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">gene_df</span><span class="op">$</span><span class="va">gene_id</span> <span class="co">#name the rows by the gene IDs</span></span>
<span><span class="va">gene_df</span> <span class="op">&lt;-</span> <span class="va">gene_df</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="co">#remove the first column which is not needed anymore</span></span>
<span><span class="co">#gene_df</span></span>
<span><span class="va">gene_df</span><span class="op">[</span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">7</span> <span class="op">&amp;</span> <span class="va">gene_df</span><span class="op">$</span><span class="va">mutation</span> <span class="op">==</span> <span class="st">"No"</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      expression mutation    pathway
GAPDH         10       No Metabolism</code></pre>
</div>
</div>
<p>Returns genes with <strong>high expression (&gt;7)</strong> <code>and</code> <strong>no mutation</strong> — potentially highly active but wild-type genes.</p>
<ol start="6" type="1">
<li>Group-wise Expression Summary</li>
</ol>
<p>The given vectors are:</p>
<div class="cell">
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span><span class="op">)</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5.2</span>, <span class="fl">8.1</span>, <span class="fl">4.3</span>, <span class="fl">9.0</span>, <span class="fl">5.7</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The solution would be:</p>
<div class="cell">
<div class="sourceCode" id="cb101"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">group_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Mean expression</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span><span class="op">(</span><span class="va">expression</span>, <span class="va">group_factor</span>, <span class="va">mean</span><span class="op">)</span> <span class="co">## KO: 8.55, WT: 5.07</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      KO       WT 
8.550000 5.066667 </code></pre>
</div>
<div class="sourceCode" id="cb103"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span><span class="op">(</span><span class="va">expression</span>, <span class="va">group_factor</span>, <span class="va">mean</span><span class="op">)</span>, </span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"skyblue"</span>, <span class="st">"salmon"</span><span class="op">)</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Mean Expression"</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Group-wise Expression"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<ol start="7" type="1">
<li>Gene Subsetting</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene_df</span><span class="op">[</span><span class="va">gene_df</span><span class="op">$</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">8</span> <span class="op">&amp;</span> </span>
<span>        <span class="va">gene_df</span><span class="op">$</span><span class="va">pathway</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span><span class="op">)</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    expression mutation    pathway
MYC        9.5      Yes Cell Cycle</code></pre>
</div>
</div>
<p>It filters for genes highly expressed and involved in key biological pathways.</p>
<ol start="8" type="1">
<li>Disease Stage Visualization</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">stages</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Stage I"</span>, <span class="st">"Stage III"</span>, <span class="st">"Stage II"</span>, <span class="st">"Stage IV"</span>, <span class="st">"Stage I"</span><span class="op">)</span></span>
<span><span class="va">disease_stage</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">stages</span>, </span>
<span>                        levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Stage I"</span>, <span class="st">"Stage II"</span>, <span class="st">"Stage III"</span>, <span class="st">"Stage IV"</span><span class="op">)</span>, </span>
<span>                        ordered <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">disease_stage</span><span class="op">)</span>, </span>
<span>        col <span class="op">=</span> <span class="st">"lightgreen"</span>, </span>
<span>        main <span class="op">=</span> <span class="st">"Patient Count by Disease Stage"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<p>Let’s do the severity order check:</p>
<div class="cell">
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Comparison</span></span>
<span><span class="va">disease_stage</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">disease_stage</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>  <span class="co"># TRUE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>So, “Stage III” is more sever than “Stage I”.</p>
<ol start="9" type="1">
<li>Oncogene Subsetting and Releveling</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define a small gene dataset</span></span>
<span><span class="va">gene_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  gene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"MYC"</span>, <span class="st">"GAPDH"</span>, <span class="st">"EGFR"</span><span class="op">)</span>,</span>
<span>  expression <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">9.1</span>, <span class="fl">7.3</span>, <span class="fl">10.5</span>, <span class="fl">5.2</span>, <span class="fl">8.6</span><span class="op">)</span>,</span>
<span>  type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Tumor Suppressor"</span>, <span class="st">"Oncogene"</span>, <span class="st">"Oncogene"</span>, <span class="st">"Housekeeping"</span>, <span class="st">"Oncogene"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Subset: Oncogene rows with expression &gt; 8</span></span>
<span><span class="va">gene_data</span><span class="op">[</span><span class="va">gene_data</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"Oncogene"</span> <span class="op">&amp;</span> <span class="va">gene_data</span><span class="op">$</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">8</span>, <span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  gene expression     type
3  MYC       10.5 Oncogene
5 EGFR        8.6 Oncogene</code></pre>
</div>
</div>
<p>Let’s relevel now, “Housekeeping” is the reference:</p>
<div class="cell">
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Relevel: make "Housekeeping" the reference level</span></span>
<span><span class="va">gene_data</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">gene_data</span><span class="op">$</span><span class="va">type</span>, ref <span class="op">=</span> <span class="st">"Housekeeping"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check the new levels</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">gene_data</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Housekeeping"     "Oncogene"         "Tumor Suppressor"</code></pre>
</div>
</div>
<p>But the “Housekeeping” was already a reference by default (using alphabetic ordering by R). Making something else the reference would make more sense. Our code above work, but nothing new is done.</p>
<ol start="10" type="1">
<li>Simulated Expression by Tissue</li>
</ol>
<div class="cell">
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">gene_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="fl">45</span>, mean <span class="op">=</span> <span class="fl">8</span>, sd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">tissue</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"brain"</span>, <span class="st">"liver"</span>, <span class="st">"kidney"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="va">tissue_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">tissue</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"liver"</span>, <span class="st">"brain"</span>, <span class="st">"kidney"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">gene_expr</span> <span class="op">~</span> <span class="va">tissue_factor</span>, </span>
<span>        col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"orange"</span>, <span class="st">"skyblue"</span>, <span class="st">"lightgreen"</span><span class="op">)</span>, </span>
<span>        main <span class="op">=</span> <span class="st">"Expression by Tissue"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"Expression Level"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<p>Let’s calculate variability per tissue type now:</p>
<div class="cell">
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Variability</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/tapply.html">tapply</a></span><span class="op">(</span><span class="va">gene_expr</span>, <span class="va">tissue_factor</span>, <span class="va">sd</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   liver    brain   kidney 
2.713940 2.050487 1.993668 </code></pre>
</div>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Returns standard deviation per tissue group</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="homework-solution-tidyverse" class="level2"><h2 class="anchored" data-anchor-id="homework-solution-tidyverse">🧩 Homework Solution tidyverse</h2>
<div class="cell">
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Load necessary package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use the built-in iris dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">iris</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create the plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">iris</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Petal.Length</span>, y <span class="op">=</span> <span class="va">Petal.Width</span>, color <span class="op">=</span> <span class="va">Species</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>  <span class="co"># scatter plot points</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span>, se <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span>  <span class="co"># linear regression lines</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Relationship between Petal Length and Width by Species"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Petal Length (cm)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Petal Width (cm)"</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"Iris Species"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span>  <span class="co"># apply a clean minimal theme</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">iris</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Petal.Length</span>, y <span class="op">=</span> <span class="va">Petal.Width</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">Species</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span>  <span class="co"># color points by Species</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"lm"</span>, se <span class="op">=</span> <span class="cn">FALSE</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span>  <span class="co"># single regression line</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">labs</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Overall Regression: Petal Length vs Width (Iris Dataset)"</span>,</span>
<span>    x <span class="op">=</span> <span class="st">"Petal Length (cm)"</span>,</span>
<span>    y <span class="op">=</span> <span class="st">"Petal Width (cm)"</span>,</span>
<span>    color <span class="op">=</span> <span class="st">"Iris Species"</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid" width="1650"></p>
</div>
</div>


<!-- -->

</section></section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents"><h2 class="anchored quarto-appendix-heading">Citation</h2><div><div class="quarto-appendix-secondary-label">BibTeX citation:</div><pre class="sourceCode code-with-copy quarto-appendix-bibtex"><code class="sourceCode bibtex">@online{rasheduzzaman2025,
  author = {Md Rasheduzzaman},
  title = {HW Solutions},
  date = {2025-09-09},
  langid = {en},
  abstract = {dataframe, matrices, list, factor, vector, tidyverse,
    etc.}
}
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre><div class="quarto-appendix-secondary-label">For attribution, please cite this work as:</div><div id="ref-rasheduzzaman2025" class="csl-entry quarto-appendix-citeas" role="listitem">
Md Rasheduzzaman. 2025. <span>“HW Solutions.”</span> September 9, 2025.
</div></div></section></div></main><!-- /main --><hr>
<div style="margin-top: 2em;">
  <p style="font-weight: 500; font-size: 1.1em;">
    💬 Have thoughts or questions? Join the discussion below using your GitHub account!
  </p>
  <p style="font-size: 0.95em; color: #666;">
    You can <strong>edit or delete</strong> your own comments. Reactions like 👍 ❤️ 🚀 are also supported.
  </p>
</div>

<div id="giscus_thread" style="margin-top: 1.5em;"></div>

<script src="https://giscus.app/client.js" data-repo="MdRasheduz-zaman/PythRaSh" data-repo-id="R_kgDOOZN9Xw" data-category-id="DIC_kwDOOZN9X84CpE6N" data-mapping="pathname" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="preferred_color_scheme" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb119" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "HW solutions"</span></span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a><span class="an">abstract:</span><span class="co"> "dataframe, matrices, list, factor, vector, tidyverse, etc."</span></span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a><span class="fu"># L3: Matrices and Lists</span></span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-8"><a href="#cb119-8" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 1: Protein Quantification in Biological Samples</span></span>
<span id="cb119-9"><a href="#cb119-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-10"><a href="#cb119-10" aria-hidden="true" tabindex="-1"></a>We measured the concentration (in µg/µL) of three proteins (P1, P2, P3) in four samples (S1–S4):</span>
<span id="cb119-11"><a href="#cb119-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-12"><a href="#cb119-12" aria-hidden="true" tabindex="-1"></a><span class="fu">### 1. Matrices</span></span>
<span id="cb119-13"><a href="#cb119-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-16"><a href="#cb119-16" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-17"><a href="#cb119-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Making Protein Matrix</span></span>
<span id="cb119-18"><a href="#cb119-18" aria-hidden="true" tabindex="-1"></a>ProteinMatrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb119-19"><a href="#cb119-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">2</span>,</span>
<span id="cb119-20"><a href="#cb119-20" aria-hidden="true" tabindex="-1"></a>    <span class="dv">7</span>, <span class="dv">6</span>, <span class="dv">4</span>),</span>
<span id="cb119-21"><a href="#cb119-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb119-22"><a href="#cb119-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-23"><a href="#cb119-23" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ProteinMatrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Sample1"</span>, <span class="st">"Sample2"</span>)</span>
<span id="cb119-24"><a href="#cb119-24" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ProteinMatrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb119-25"><a href="#cb119-25" aria-hidden="true" tabindex="-1"></a>ProteinMatrix</span>
<span id="cb119-26"><a href="#cb119-26" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-27"><a href="#cb119-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-28"><a href="#cb119-28" aria-hidden="true" tabindex="-1"></a>Now goes the weight matrix</span>
<span id="cb119-29"><a href="#cb119-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-32"><a href="#cb119-32" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-33"><a href="#cb119-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Making weight matrix</span></span>
<span id="cb119-34"><a href="#cb119-34" aria-hidden="true" tabindex="-1"></a>WeightVector <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb119-35"><a href="#cb119-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">1.5</span>),</span>
<span id="cb119-36"><a href="#cb119-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow=</span><span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb119-37"><a href="#cb119-37" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-38"><a href="#cb119-38" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb119-39"><a href="#cb119-39" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb119-40"><a href="#cb119-40" aria-hidden="true" tabindex="-1"></a>WeightVector</span>
<span id="cb119-41"><a href="#cb119-41" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-42"><a href="#cb119-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-43"><a href="#cb119-43" aria-hidden="true" tabindex="-1"></a>Now, multiply them.</span>
<span id="cb119-44"><a href="#cb119-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-47"><a href="#cb119-47" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-48"><a href="#cb119-48" aria-hidden="true" tabindex="-1"></a><span class="co"># Multiplying Matrices</span></span>
<span id="cb119-49"><a href="#cb119-49" aria-hidden="true" tabindex="-1"></a>TotalConc <span class="ot">=</span> ProteinMatrix <span class="sc">%*%</span> WeightVector</span>
<span id="cb119-50"><a href="#cb119-50" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TotalConc) <span class="ot">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span id="cb119-51"><a href="#cb119-51" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(TotalConc)</span>
<span id="cb119-52"><a href="#cb119-52" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-53"><a href="#cb119-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-54"><a href="#cb119-54" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2. Making transpose of ProteinMatrix</span></span>
<span id="cb119-55"><a href="#cb119-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-58"><a href="#cb119-58" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-59"><a href="#cb119-59" aria-hidden="true" tabindex="-1"></a>ProteinMatTranspose <span class="ot">=</span> <span class="fu">t</span>(ProteinMatrix)</span>
<span id="cb119-60"><a href="#cb119-60" aria-hidden="true" tabindex="-1"></a>ProteinMatTranspose</span>
<span id="cb119-61"><a href="#cb119-61" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-62"><a href="#cb119-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-63"><a href="#cb119-63" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3. Identity Matrix</span></span>
<span id="cb119-64"><a href="#cb119-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-67"><a href="#cb119-67" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-68"><a href="#cb119-68" aria-hidden="true" tabindex="-1"></a>I <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb119-69"><a href="#cb119-69" aria-hidden="true" tabindex="-1"></a>Identitycheck <span class="ot">=</span> ProteinMatrix <span class="sc">%*%</span> I</span>
<span id="cb119-70"><a href="#cb119-70" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Identitycheck) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb119-71"><a href="#cb119-71" aria-hidden="true" tabindex="-1"></a>Identitycheck</span>
<span id="cb119-72"><a href="#cb119-72" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-73"><a href="#cb119-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-74"><a href="#cb119-74" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4.1. Total Protein per Sample</span></span>
<span id="cb119-75"><a href="#cb119-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-78"><a href="#cb119-78" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-79"><a href="#cb119-79" aria-hidden="true" tabindex="-1"></a><span class="fu">rowSums</span>(ProteinMatrix)</span>
<span id="cb119-80"><a href="#cb119-80" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-81"><a href="#cb119-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-82"><a href="#cb119-82" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4.2. Total Protein per Protein Type</span></span>
<span id="cb119-83"><a href="#cb119-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-86"><a href="#cb119-86" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-87"><a href="#cb119-87" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(ProteinMatrix)</span>
<span id="cb119-88"><a href="#cb119-88" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-89"><a href="#cb119-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-90"><a href="#cb119-90" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4.3. Heatmap of Protein Concentrations</span></span>
<span id="cb119-91"><a href="#cb119-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-94"><a href="#cb119-94" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-95"><a href="#cb119-95" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(ProteinMatrix, <span class="at">scale =</span> <span class="st">"none"</span>, <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">10</span>))</span>
<span id="cb119-96"><a href="#cb119-96" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-97"><a href="#cb119-97" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-98"><a href="#cb119-98" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation</span></span>
<span id="cb119-99"><a href="#cb119-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-100"><a href="#cb119-100" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Multiplying the protein levels by the weight vector shows how much each protein contributes in a sample. The result shows total protein concentration per sample.</span>
<span id="cb119-101"><a href="#cb119-101" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The result shows that sample S2 has the highest protein burden.</span>
<span id="cb119-102"><a href="#cb119-102" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>The identity matrix represents no protein interactions or measurement biases. It is a simple matrix calculation.</span>
<span id="cb119-103"><a href="#cb119-103" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>New calculation:</span>
<span id="cb119-104"><a href="#cb119-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-107"><a href="#cb119-107" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-108"><a href="#cb119-108" aria-hidden="true" tabindex="-1"></a><span class="co"># changing the weight of ProteinZ to 3.0</span></span>
<span id="cb119-109"><a href="#cb119-109" aria-hidden="true" tabindex="-1"></a>newweightvector <span class="ot">=</span> <span class="fu">matrix</span>(</span>
<span id="cb119-110"><a href="#cb119-110" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">3.0</span>),</span>
<span id="cb119-111"><a href="#cb119-111" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow=</span><span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb119-112"><a href="#cb119-112" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-113"><a href="#cb119-113" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb119-114"><a href="#cb119-114" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb119-115"><a href="#cb119-115" aria-hidden="true" tabindex="-1"></a>newTotalconc <span class="ot">=</span> ProteinMatrix <span class="sc">%*%</span> newweightvector</span>
<span id="cb119-116"><a href="#cb119-116" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(newTotalconc) <span class="ot">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span id="cb119-117"><a href="#cb119-117" aria-hidden="true" tabindex="-1"></a>newTotalconc</span>
<span id="cb119-118"><a href="#cb119-118" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-119"><a href="#cb119-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-120"><a href="#cb119-120" aria-hidden="true" tabindex="-1"></a>Still, S2 has more protein burden.</span>
<span id="cb119-121"><a href="#cb119-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-122"><a href="#cb119-122" aria-hidden="true" tabindex="-1"></a>Bonus:</span>
<span id="cb119-123"><a href="#cb119-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-124"><a href="#cb119-124" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Heatmap reveals PX is most abundant across all samples.  </span>
<span id="cb119-125"><a href="#cb119-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-126"><a href="#cb119-126" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 2: Gene-to-Protein Translation</span></span>
<span id="cb119-127"><a href="#cb119-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-130"><a href="#cb119-130" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-131"><a href="#cb119-131" aria-hidden="true" tabindex="-1"></a><span class="co"># making  Gene Expression matrix</span></span>
<span id="cb119-132"><a href="#cb119-132" aria-hidden="true" tabindex="-1"></a>GeneExpression <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb119-133"><a href="#cb119-133" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">8</span>, <span class="dv">5</span>,</span>
<span id="cb119-134"><a href="#cb119-134" aria-hidden="true" tabindex="-1"></a>    <span class="dv">15</span>, <span class="dv">12</span>, <span class="dv">10</span>),</span>
<span id="cb119-135"><a href="#cb119-135" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb119-136"><a href="#cb119-136" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-137"><a href="#cb119-137" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(GeneExpression) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Sample1"</span>, <span class="st">"Sample2"</span>)</span>
<span id="cb119-138"><a href="#cb119-138" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(GeneExpression) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb119-139"><a href="#cb119-139" aria-hidden="true" tabindex="-1"></a>GeneExpression</span>
<span id="cb119-140"><a href="#cb119-140" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-141"><a href="#cb119-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-142"><a href="#cb119-142" aria-hidden="true" tabindex="-1"></a>Translation efficiency:</span>
<span id="cb119-143"><a href="#cb119-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-146"><a href="#cb119-146" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-147"><a href="#cb119-147" aria-hidden="true" tabindex="-1"></a><span class="co"># making Translation Matrix</span></span>
<span id="cb119-148"><a href="#cb119-148" aria-hidden="true" tabindex="-1"></a>TranslationMatrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb119-149"><a href="#cb119-149" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="dv">0</span> , <span class="dv">0</span>,</span>
<span id="cb119-150"><a href="#cb119-150" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span>, <span class="fl">1.2</span>, <span class="dv">0</span>,</span>
<span id="cb119-151"><a href="#cb119-151" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">1.8</span>),</span>
<span id="cb119-152"><a href="#cb119-152" aria-hidden="true" tabindex="-1"></a><span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb119-153"><a href="#cb119-153" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-154"><a href="#cb119-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-155"><a href="#cb119-155" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(TranslationMatrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb119-156"><a href="#cb119-156" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TranslationMatrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"protA"</span>, <span class="st">"protB"</span>, <span class="st">"protC"</span>)</span>
<span id="cb119-157"><a href="#cb119-157" aria-hidden="true" tabindex="-1"></a>TranslationMatrix</span>
<span id="cb119-158"><a href="#cb119-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-159"><a href="#cb119-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-160"><a href="#cb119-160" aria-hidden="true" tabindex="-1"></a><span class="fu">### Protein Output</span></span>
<span id="cb119-161"><a href="#cb119-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-164"><a href="#cb119-164" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-165"><a href="#cb119-165" aria-hidden="true" tabindex="-1"></a><span class="co"># computing Protein matrix</span></span>
<span id="cb119-166"><a href="#cb119-166" aria-hidden="true" tabindex="-1"></a>Protein_matrix <span class="ot">&lt;-</span> GeneExpression <span class="sc">%*%</span> TranslationMatrix</span>
<span id="cb119-167"><a href="#cb119-167" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Protein_matrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"total_protA"</span>, <span class="st">"total_protB"</span>, <span class="st">"total_protC"</span>)</span>
<span id="cb119-168"><a href="#cb119-168" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Protein_matrix)</span>
<span id="cb119-169"><a href="#cb119-169" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-170"><a href="#cb119-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-171"><a href="#cb119-171" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2. Transpose</span></span>
<span id="cb119-172"><a href="#cb119-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-175"><a href="#cb119-175" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-176"><a href="#cb119-176" aria-hidden="true" tabindex="-1"></a><span class="co"># Transpose of GeneExpression matrix</span></span>
<span id="cb119-177"><a href="#cb119-177" aria-hidden="true" tabindex="-1"></a>GeneExpression_Transpose <span class="ot">&lt;-</span> <span class="fu">t</span>(GeneExpression)</span>
<span id="cb119-178"><a href="#cb119-178" aria-hidden="true" tabindex="-1"></a>GeneExpression_Transpose</span>
<span id="cb119-179"><a href="#cb119-179" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-180"><a href="#cb119-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-181"><a href="#cb119-181" aria-hidden="true" tabindex="-1"></a>The new matrix represnts a matrix where the rows and columns of GeneExpression matrix have been interchanged.</span>
<span id="cb119-182"><a href="#cb119-182" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-183"><a href="#cb119-183" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3. Identity matrix and multipliccation</span></span>
<span id="cb119-184"><a href="#cb119-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-187"><a href="#cb119-187" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-188"><a href="#cb119-188" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating Identity matrix</span></span>
<span id="cb119-189"><a href="#cb119-189" aria-hidden="true" tabindex="-1"></a>I <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb119-190"><a href="#cb119-190" aria-hidden="true" tabindex="-1"></a>I</span>
<span id="cb119-191"><a href="#cb119-191" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-192"><a href="#cb119-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-193"><a href="#cb119-193" aria-hidden="true" tabindex="-1"></a>Now, multiply:</span>
<span id="cb119-194"><a href="#cb119-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-197"><a href="#cb119-197" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-198"><a href="#cb119-198" aria-hidden="true" tabindex="-1"></a>Product_matrix <span class="ot">=</span> TranslationMatrix <span class="sc">%*%</span>  I </span>
<span id="cb119-199"><a href="#cb119-199" aria-hidden="true" tabindex="-1"></a>Product_matrix</span>
<span id="cb119-200"><a href="#cb119-200" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-201"><a href="#cb119-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-202"><a href="#cb119-202" aria-hidden="true" tabindex="-1"></a>The product is identical to TranslationMatrix</span>
<span id="cb119-203"><a href="#cb119-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-204"><a href="#cb119-204" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4. Sub-matrix:</span></span>
<span id="cb119-205"><a href="#cb119-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-208"><a href="#cb119-208" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-209"><a href="#cb119-209" aria-hidden="true" tabindex="-1"></a><span class="co"># making submatrix A</span></span>
<span id="cb119-210"><a href="#cb119-210" aria-hidden="true" tabindex="-1"></a>A <span class="ot">=</span> <span class="fu">matrix</span>(</span>
<span id="cb119-211"><a href="#cb119-211" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">8</span>,</span>
<span id="cb119-212"><a href="#cb119-212" aria-hidden="true" tabindex="-1"></a>    <span class="dv">15</span>, <span class="dv">12</span>), <span class="at">nrow=</span><span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb119-213"><a href="#cb119-213" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-214"><a href="#cb119-214" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(A) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"sample1"</span>, <span class="st">"sample2"</span>)</span>
<span id="cb119-215"><a href="#cb119-215" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(A) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>)</span>
<span id="cb119-216"><a href="#cb119-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-217"><a href="#cb119-217" aria-hidden="true" tabindex="-1"></a>A</span>
<span id="cb119-218"><a href="#cb119-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-219"><a href="#cb119-219" aria-hidden="true" tabindex="-1"></a><span class="co"># finding inverse of A</span></span>
<span id="cb119-220"><a href="#cb119-220" aria-hidden="true" tabindex="-1"></a><span class="co">#inv_A &lt;- solve(A)</span></span>
<span id="cb119-221"><a href="#cb119-221" aria-hidden="true" tabindex="-1"></a><span class="co">#inv_A</span></span>
<span id="cb119-222"><a href="#cb119-222" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-223"><a href="#cb119-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-224"><a href="#cb119-224" aria-hidden="true" tabindex="-1"></a>The inverse matrix could not be calculated since A is a singular matrix. So, A <span class="sc">\*</span> A\^-1 is also not possible.</span>
<span id="cb119-225"><a href="#cb119-225" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-226"><a href="#cb119-226" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualization tasks:</span></span>
<span id="cb119-227"><a href="#cb119-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-228"><a href="#cb119-228" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span></span>
<span id="cb119-229"><a href="#cb119-229" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-230"><a href="#cb119-230" aria-hidden="true" tabindex="-1"></a><span class="ss">    1.  </span>MARplot</span>
<span id="cb119-231"><a href="#cb119-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-234"><a href="#cb119-234" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-235"><a href="#cb119-235" aria-hidden="true" tabindex="-1"></a><span class="co"># generating MARplot-style scatter plot</span></span>
<span id="cb119-236"><a href="#cb119-236" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(GeneExpression, Protein_matrix, <span class="at">type=</span><span class="st">"p"</span>, <span class="at">main=</span><span class="st">"Protein level vs. Gene Expression level"</span>)</span>
<span id="cb119-237"><a href="#cb119-237" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="st">"Sample-Gene"</span></span>
<span id="cb119-238"><a href="#cb119-238" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(GeneExpression, Protein_matrix, <span class="at">labels =</span> labels, <span class="at">pos=</span><span class="dv">3</span>)</span>
<span id="cb119-239"><a href="#cb119-239" aria-hidden="true" tabindex="-1"></a><span class="co"># generating a heatmap</span></span>
<span id="cb119-240"><a href="#cb119-240" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(Protein_matrix, <span class="at">main=</span> <span class="st">"Heatmap of Protein Level"</span>, <span class="at">Rowv =</span> <span class="cn">TRUE</span>, <span class="at">Colv =</span> <span class="cn">TRUE</span>, <span class="at">labRow=</span> <span class="fu">rownames</span>(Protein_matrix), <span class="at">labCol=</span> <span class="fu">c</span>(<span class="st">"ProteinA"</span>, <span class="st">"ProteinB"</span>, <span class="st">"ProteinC"</span>), <span class="at">col=</span><span class="fu">topo.colors</span>(<span class="dv">256</span>) )</span>
<span id="cb119-241"><a href="#cb119-241" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-242"><a href="#cb119-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-243"><a href="#cb119-243" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span></span>
<span id="cb119-244"><a href="#cb119-244" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-245"><a href="#cb119-245" aria-hidden="true" tabindex="-1"></a><span class="ss">    2.  </span>Heatmap of Expression:</span>
<span id="cb119-246"><a href="#cb119-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-249"><a href="#cb119-249" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-250"><a href="#cb119-250" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(GeneExpression, <span class="at">col =</span> <span class="fu">terrain.colors</span>(<span class="dv">10</span>), <span class="at">scale =</span> <span class="st">"column"</span>)</span>
<span id="cb119-251"><a href="#cb119-251" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-252"><a href="#cb119-252" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-253"><a href="#cb119-253" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation</span></span>
<span id="cb119-254"><a href="#cb119-254" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-255"><a href="#cb119-255" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>Matrix multiplication allows each gene in both samples to be multiplied to their respective translation efficiency. So, the product shows how successfully each gene is translated")</span>
<span id="cb119-256"><a href="#cb119-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-257"><a href="#cb119-257" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>The diagonal TranslationMatrix make sense biologically because they show translation efficiency of each gene and there is no other interaction between them. Although there could be interaction in real-world scenarios.</span>
<span id="cb119-258"><a href="#cb119-258" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-259"><a href="#cb119-259" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>If Sample2 has higher protein levels even with similar gene expression, it means that more mRNAs are translated to proteins compared to Sample1"</span>
<span id="cb119-260"><a href="#cb119-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-261"><a href="#cb119-261" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>The upward trend in MARplot may indicate an increase in translation efficacy and downward trend may indicate a decline in translation efficacy"</span>
<span id="cb119-262"><a href="#cb119-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-263"><a href="#cb119-263" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Clustering in the heatmap may suggest which samples are most similar to each other based on their prot.</span>
<span id="cb119-264"><a href="#cb119-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-265"><a href="#cb119-265" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 3: Animal Breeding – Bull Ranking by Economic Traits</span></span>
<span id="cb119-266"><a href="#cb119-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-267"><a href="#cb119-267" aria-hidden="true" tabindex="-1"></a><span class="fu">### Define Data</span></span>
<span id="cb119-268"><a href="#cb119-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-271"><a href="#cb119-271" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-272"><a href="#cb119-272" aria-hidden="true" tabindex="-1"></a><span class="co"># Define Bull EBVs</span></span>
<span id="cb119-273"><a href="#cb119-273" aria-hidden="true" tabindex="-1"></a>BullEBVs <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb119-274"><a href="#cb119-274" aria-hidden="true" tabindex="-1"></a>  <span class="dv">400</span>, <span class="fl">1.2</span>, <span class="fl">0.8</span>,</span>
<span id="cb119-275"><a href="#cb119-275" aria-hidden="true" tabindex="-1"></a>  <span class="dv">500</span>, <span class="fl">1.5</span>, <span class="fl">0.6</span></span>
<span id="cb119-276"><a href="#cb119-276" aria-hidden="true" tabindex="-1"></a>), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb119-277"><a href="#cb119-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-278"><a href="#cb119-278" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(BullEBVs) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Bull1"</span>, <span class="st">"Bull2"</span>)</span>
<span id="cb119-279"><a href="#cb119-279" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(BullEBVs) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Milk_yield"</span>, <span class="st">"Growth_rate"</span>, <span class="st">"Fertility"</span>)</span>
<span id="cb119-280"><a href="#cb119-280" aria-hidden="true" tabindex="-1"></a>BullEBVs</span>
<span id="cb119-281"><a href="#cb119-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-282"><a href="#cb119-282" aria-hidden="true" tabindex="-1"></a><span class="co"># Define Economic Weights</span></span>
<span id="cb119-283"><a href="#cb119-283" aria-hidden="true" tabindex="-1"></a>EconomicWeights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.002</span>, <span class="dv">50</span>, <span class="dv">100</span>), <span class="at">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb119-284"><a href="#cb119-284" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(EconomicWeights) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(BullEBVs)</span>
<span id="cb119-285"><a href="#cb119-285" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(EconomicWeights) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb119-286"><a href="#cb119-286" aria-hidden="true" tabindex="-1"></a>EconomicWeights</span>
<span id="cb119-287"><a href="#cb119-287" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-288"><a href="#cb119-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-289"><a href="#cb119-289" aria-hidden="true" tabindex="-1"></a><span class="fu">### Compute Total Economic Value</span></span>
<span id="cb119-290"><a href="#cb119-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-293"><a href="#cb119-293" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-294"><a href="#cb119-294" aria-hidden="true" tabindex="-1"></a>TotalValue <span class="ot">&lt;-</span> BullEBVs <span class="sc">%*%</span> EconomicWeights</span>
<span id="cb119-295"><a href="#cb119-295" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TotalValue) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Merit"</span>)</span>
<span id="cb119-296"><a href="#cb119-296" aria-hidden="true" tabindex="-1"></a>TotalValue</span>
<span id="cb119-297"><a href="#cb119-297" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-298"><a href="#cb119-298" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-299"><a href="#cb119-299" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb119-300"><a href="#cb119-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-301"><a href="#cb119-301" aria-hidden="true" tabindex="-1"></a>Bull1: (400 × 0.002) + (1.2 × 50) + (0.8 × 100) = 140.8</span>
<span id="cb119-302"><a href="#cb119-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-303"><a href="#cb119-303" aria-hidden="true" tabindex="-1"></a>Bull2: (500 × 0.002) + (1.5 × 50) + (0.6 × 100) = 136.0</span>
<span id="cb119-304"><a href="#cb119-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-305"><a href="#cb119-305" aria-hidden="true" tabindex="-1"></a>Bull1 is more valuable economically.</span>
<span id="cb119-306"><a href="#cb119-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-307"><a href="#cb119-307" aria-hidden="true" tabindex="-1"></a>**Biological Interpretation**</span>
<span id="cb119-308"><a href="#cb119-308" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-309"><a href="#cb119-309" aria-hidden="true" tabindex="-1"></a>Economic weights convert genetic merit (EBVs, Estimated Breeding Values) into economic merit. Traits with higher financial importance have a larger impact, regardless of absolute EBV values.</span>
<span id="cb119-310"><a href="#cb119-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-311"><a href="#cb119-311" aria-hidden="true" tabindex="-1"></a><span class="fu">### Multiply with Identity Matrix</span></span>
<span id="cb119-312"><a href="#cb119-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-315"><a href="#cb119-315" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-316"><a href="#cb119-316" aria-hidden="true" tabindex="-1"></a>I3 <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb119-317"><a href="#cb119-317" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(I3) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(BullEBVs)</span>
<span id="cb119-318"><a href="#cb119-318" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(I3) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(BullEBVs)</span>
<span id="cb119-319"><a href="#cb119-319" aria-hidden="true" tabindex="-1"></a>I3</span>
<span id="cb119-320"><a href="#cb119-320" aria-hidden="true" tabindex="-1"></a>BullEBVs_identity <span class="ot">&lt;-</span> BullEBVs <span class="sc">%*%</span> I3</span>
<span id="cb119-321"><a href="#cb119-321" aria-hidden="true" tabindex="-1"></a>BullEBVs_identity</span>
<span id="cb119-322"><a href="#cb119-322" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-323"><a href="#cb119-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-324"><a href="#cb119-324" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb119-325"><a href="#cb119-325" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-326"><a href="#cb119-326" aria-hidden="true" tabindex="-1"></a>Multiplying by identity matrix returns the original matrix. It confirms that EBV structure is preserved.</span>
<span id="cb119-327"><a href="#cb119-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-328"><a href="#cb119-328" aria-hidden="true" tabindex="-1"></a><span class="fu">### Remove Milk Yield and Recalculate Total Value</span></span>
<span id="cb119-329"><a href="#cb119-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-332"><a href="#cb119-332" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-333"><a href="#cb119-333" aria-hidden="true" tabindex="-1"></a>BullEBVs_noMilk <span class="ot">&lt;-</span> BullEBVs[, <span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb119-334"><a href="#cb119-334" aria-hidden="true" tabindex="-1"></a>EconomicWeights_noMilk <span class="ot">&lt;-</span> EconomicWeights[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb119-335"><a href="#cb119-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-336"><a href="#cb119-336" aria-hidden="true" tabindex="-1"></a>TotalValue_noMilk <span class="ot">&lt;-</span> BullEBVs_noMilk <span class="sc">%*%</span> EconomicWeights_noMilk</span>
<span id="cb119-337"><a href="#cb119-337" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TotalValue_noMilk) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"New_Merit"</span>)</span>
<span id="cb119-338"><a href="#cb119-338" aria-hidden="true" tabindex="-1"></a>TotalValue_noMilk</span>
<span id="cb119-339"><a href="#cb119-339" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-340"><a href="#cb119-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-341"><a href="#cb119-341" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb119-342"><a href="#cb119-342" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-343"><a href="#cb119-343" aria-hidden="true" tabindex="-1"></a>Bull1: (1.2 × 50) + (0.8 × 100) = 140</span>
<span id="cb119-344"><a href="#cb119-344" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-345"><a href="#cb119-345" aria-hidden="true" tabindex="-1"></a>Bull2: (1.5 × 50) + (0.6 × 100) = 135</span>
<span id="cb119-346"><a href="#cb119-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-347"><a href="#cb119-347" aria-hidden="true" tabindex="-1"></a>Bull1 still ranks higher, but by a smaller margin.</span>
<span id="cb119-348"><a href="#cb119-348" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-349"><a href="#cb119-349" aria-hidden="true" tabindex="-1"></a><span class="fu">### Bar Plot: Total Economic Value</span></span>
<span id="cb119-350"><a href="#cb119-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-353"><a href="#cb119-353" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-354"><a href="#cb119-354" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb119-355"><a href="#cb119-355" aria-hidden="true" tabindex="-1"></a>  TotalValue,</span>
<span id="cb119-356"><a href="#cb119-356" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb119-357"><a href="#cb119-357" aria-hidden="true" tabindex="-1"></a>  <span class="at">names.arg =</span> <span class="fu">rownames</span>(BullEBVs),</span>
<span id="cb119-358"><a href="#cb119-358" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"orange"</span>),</span>
<span id="cb119-359"><a href="#cb119-359" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Total Economic Value of Bulls"</span>,</span>
<span id="cb119-360"><a href="#cb119-360" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Total Value"</span></span>
<span id="cb119-361"><a href="#cb119-361" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-362"><a href="#cb119-362" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-363"><a href="#cb119-363" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-364"><a href="#cb119-364" aria-hidden="true" tabindex="-1"></a><span class="fu">### Heatmap of EBVs</span></span>
<span id="cb119-365"><a href="#cb119-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-368"><a href="#cb119-368" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-369"><a href="#cb119-369" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb119-370"><a href="#cb119-370" aria-hidden="true" tabindex="-1"></a>  BullEBVs,</span>
<span id="cb119-371"><a href="#cb119-371" aria-hidden="true" tabindex="-1"></a>  <span class="at">Rowv =</span> <span class="cn">NA</span>,</span>
<span id="cb119-372"><a href="#cb119-372" aria-hidden="true" tabindex="-1"></a>  <span class="at">Colv =</span> <span class="cn">NA</span>,</span>
<span id="cb119-373"><a href="#cb119-373" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb119-374"><a href="#cb119-374" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">256</span>),</span>
<span id="cb119-375"><a href="#cb119-375" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Heatmap of Bull EBVs"</span></span>
<span id="cb119-376"><a href="#cb119-376" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-377"><a href="#cb119-377" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-378"><a href="#cb119-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-379"><a href="#cb119-379" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation Questions</span></span>
<span id="cb119-380"><a href="#cb119-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-381"><a href="#cb119-381" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How do economic weights affect trait importance?</span>
<span id="cb119-382"><a href="#cb119-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-383"><a href="#cb119-383" aria-hidden="true" tabindex="-1"></a>Traits with higher weights contribute more to the total economic value. This makes them more influential in ranking and selection.</span>
<span id="cb119-384"><a href="#cb119-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-385"><a href="#cb119-385" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Why might you ignore milk yield?</span>
<span id="cb119-386"><a href="#cb119-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-387"><a href="#cb119-387" aria-hidden="true" tabindex="-1"></a>Milk yield may be excluded in systems focusing on fertility, growth, or when it is no longer a limiting factor. Environmental or economic contexts may also shift trait priorities.</span>
<span id="cb119-388"><a href="#cb119-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-389"><a href="#cb119-389" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What is the value of heatmaps?</span>
<span id="cb119-390"><a href="#cb119-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-391"><a href="#cb119-391" aria-hidden="true" tabindex="-1"></a>Heatmaps visually compare EBVs across bulls and traits.They help detect patterns, outliers, and clusters easily in multivariate data.</span>
<span id="cb119-392"><a href="#cb119-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-393"><a href="#cb119-393" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Can this method be extended to more bulls and traits?</span>
<span id="cb119-394"><a href="#cb119-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-395"><a href="#cb119-395" aria-hidden="true" tabindex="-1"></a>Yes. This method scales to any number of bulls or traits. Just ensure the EBVs matrix and economic weights are dimensionally compatible.</span>
<span id="cb119-396"><a href="#cb119-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-397"><a href="#cb119-397" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 4: Plant Breeding – Trait Contributions from Parental Lines</span></span>
<span id="cb119-398"><a href="#cb119-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-399"><a href="#cb119-399" aria-hidden="true" tabindex="-1"></a><span class="fu">### Define Data</span></span>
<span id="cb119-400"><a href="#cb119-400" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-401"><a href="#cb119-401" aria-hidden="true" tabindex="-1"></a>**Parent trait values (normalized 1–10)**</span>
<span id="cb119-402"><a href="#cb119-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-405"><a href="#cb119-405" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-406"><a href="#cb119-406" aria-hidden="true" tabindex="-1"></a>ParentTraits <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb119-407"><a href="#cb119-407" aria-hidden="true" tabindex="-1"></a>  <span class="dv">7</span>, <span class="dv">5</span>, <span class="dv">3</span>,</span>
<span id="cb119-408"><a href="#cb119-408" aria-hidden="true" tabindex="-1"></a>  <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">4</span>,</span>
<span id="cb119-409"><a href="#cb119-409" aria-hidden="true" tabindex="-1"></a>  <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">6</span></span>
<span id="cb119-410"><a href="#cb119-410" aria-hidden="true" tabindex="-1"></a>), <span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb119-411"><a href="#cb119-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-412"><a href="#cb119-412" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ParentTraits) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"P1"</span>, <span class="st">"P2"</span>, <span class="st">"P3"</span>)</span>
<span id="cb119-413"><a href="#cb119-413" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ParentTraits) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Drought_resistance"</span>, <span class="st">"Yield"</span>, <span class="st">"Maturation_time"</span>)</span>
<span id="cb119-414"><a href="#cb119-414" aria-hidden="true" tabindex="-1"></a>ParentTraits</span>
<span id="cb119-415"><a href="#cb119-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-416"><a href="#cb119-416" aria-hidden="true" tabindex="-1"></a><span class="co"># Define Hybrid Weights</span></span>
<span id="cb119-417"><a href="#cb119-417" aria-hidden="true" tabindex="-1"></a>HybridWeights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span>), <span class="at">nrow =</span> <span class="dv">1</span>)</span>
<span id="cb119-418"><a href="#cb119-418" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(HybridWeights) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ParentTraits)</span>
<span id="cb119-419"><a href="#cb119-419" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(HybridWeights) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb119-420"><a href="#cb119-420" aria-hidden="true" tabindex="-1"></a>HybridWeights</span>
<span id="cb119-421"><a href="#cb119-421" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-422"><a href="#cb119-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-423"><a href="#cb119-423" aria-hidden="true" tabindex="-1"></a><span class="fu">### Compute HybridTraits Vector</span></span>
<span id="cb119-424"><a href="#cb119-424" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-427"><a href="#cb119-427" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-428"><a href="#cb119-428" aria-hidden="true" tabindex="-1"></a>HybridTraits <span class="ot">&lt;-</span> HybridWeights <span class="sc">%*%</span> ParentTraits</span>
<span id="cb119-429"><a href="#cb119-429" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(HybridTraits) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Contribution"</span>)</span>
<span id="cb119-430"><a href="#cb119-430" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(HybridTraits) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(ParentTraits)</span>
<span id="cb119-431"><a href="#cb119-431" aria-hidden="true" tabindex="-1"></a>HybridTraits</span>
<span id="cb119-432"><a href="#cb119-432" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-433"><a href="#cb119-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-434"><a href="#cb119-434" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb119-435"><a href="#cb119-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-436"><a href="#cb119-436" aria-hidden="true" tabindex="-1"></a>HybridTraits = (0.5 × P1) + (0.3 × P2) + (0.2 × P3)</span>
<span id="cb119-437"><a href="#cb119-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-438"><a href="#cb119-438" aria-hidden="true" tabindex="-1"></a>Drought_resistance = (0.5 × 7) + (0.3 × 6) + (0.2 × 5) = 6.3</span>
<span id="cb119-439"><a href="#cb119-439" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-440"><a href="#cb119-440" aria-hidden="true" tabindex="-1"></a>Yield = (0.5 × 5) + (0.3 × 8) + (0.2 × 6) = 6.1</span>
<span id="cb119-441"><a href="#cb119-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-442"><a href="#cb119-442" aria-hidden="true" tabindex="-1"></a>Maturation_time = (0.5 × 3) + (0.3 × 4) + (0.2 × 6) = 3.9</span>
<span id="cb119-443"><a href="#cb119-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-444"><a href="#cb119-444" aria-hidden="true" tabindex="-1"></a>The hybrid is moderately strong in drought resistance and yield, and has a relatively shorter maturation time.</span>
<span id="cb119-445"><a href="#cb119-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-446"><a href="#cb119-446" aria-hidden="true" tabindex="-1"></a>**Biological Meaning of Unequal Contribution**</span>
<span id="cb119-447"><a href="#cb119-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-448"><a href="#cb119-448" aria-hidden="true" tabindex="-1"></a>When one parent contributes more to a trait, it suggests that the trait's heritable strength comes disproportionately from that parent. Breeders can use this knowledge to amplify desirable traits using the best parent.</span>
<span id="cb119-449"><a href="#cb119-449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-450"><a href="#cb119-450" aria-hidden="true" tabindex="-1"></a><span class="fu">### Multiply with Identity Matrix</span></span>
<span id="cb119-451"><a href="#cb119-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-454"><a href="#cb119-454" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-455"><a href="#cb119-455" aria-hidden="true" tabindex="-1"></a>I3 <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb119-456"><a href="#cb119-456" aria-hidden="true" tabindex="-1"></a>ParentTraits_identity <span class="ot">&lt;-</span> ParentTraits <span class="sc">%*%</span> I3</span>
<span id="cb119-457"><a href="#cb119-457" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ParentTraits_identity) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(ParentTraits)</span>
<span id="cb119-458"><a href="#cb119-458" aria-hidden="true" tabindex="-1"></a>ParentTraits_identity</span>
<span id="cb119-459"><a href="#cb119-459" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-460"><a href="#cb119-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-461"><a href="#cb119-461" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb119-462"><a href="#cb119-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-463"><a href="#cb119-463" aria-hidden="true" tabindex="-1"></a>Multiplying by identity matrix returns the original matrix. This operation verifies structural consistency and dimensionality.</span>
<span id="cb119-464"><a href="#cb119-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-465"><a href="#cb119-465" aria-hidden="true" tabindex="-1"></a><span class="fu">### Remove T3 (Maturation Time) and Recalculate</span></span>
<span id="cb119-466"><a href="#cb119-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-469"><a href="#cb119-469" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-470"><a href="#cb119-470" aria-hidden="true" tabindex="-1"></a>ParentTraits_T1T2 <span class="ot">&lt;-</span> ParentTraits[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb119-471"><a href="#cb119-471" aria-hidden="true" tabindex="-1"></a>ParentTraits_T1T2</span>
<span id="cb119-472"><a href="#cb119-472" aria-hidden="true" tabindex="-1"></a>HybridTraits_T1T2 <span class="ot">&lt;-</span> HybridWeights <span class="sc">%*%</span> ParentTraits_T1T2</span>
<span id="cb119-473"><a href="#cb119-473" aria-hidden="true" tabindex="-1"></a>HybridTraits_T1T2</span>
<span id="cb119-474"><a href="#cb119-474" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-475"><a href="#cb119-475" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-476"><a href="#cb119-476" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb119-477"><a href="#cb119-477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-478"><a href="#cb119-478" aria-hidden="true" tabindex="-1"></a>Drought_resistance = (0.5 × 7) + (0.3 × 6) + (0.2 × 5) = 6.3</span>
<span id="cb119-479"><a href="#cb119-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-480"><a href="#cb119-480" aria-hidden="true" tabindex="-1"></a>Yield = (0.5 × 5) + (0.3 × 8) + (0.2 × 6) = 6.1</span>
<span id="cb119-481"><a href="#cb119-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-482"><a href="#cb119-482" aria-hidden="true" tabindex="-1"></a>Removing a trait (T3) changes the trait profile. Hybrid selection may now favor traits that remain.</span>
<span id="cb119-483"><a href="#cb119-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-484"><a href="#cb119-484" aria-hidden="true" tabindex="-1"></a><span class="fu">### Heatmap of ParentTraits</span></span>
<span id="cb119-485"><a href="#cb119-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-488"><a href="#cb119-488" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-489"><a href="#cb119-489" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb119-490"><a href="#cb119-490" aria-hidden="true" tabindex="-1"></a>  ParentTraits,</span>
<span id="cb119-491"><a href="#cb119-491" aria-hidden="true" tabindex="-1"></a>  <span class="at">Rowv =</span> <span class="cn">NA</span>,</span>
<span id="cb119-492"><a href="#cb119-492" aria-hidden="true" tabindex="-1"></a>  <span class="at">Colv =</span> <span class="cn">NA</span>,</span>
<span id="cb119-493"><a href="#cb119-493" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb119-494"><a href="#cb119-494" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">256</span>),</span>
<span id="cb119-495"><a href="#cb119-495" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Heatmap of Parent Traits"</span></span>
<span id="cb119-496"><a href="#cb119-496" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-497"><a href="#cb119-497" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-498"><a href="#cb119-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-499"><a href="#cb119-499" aria-hidden="true" tabindex="-1"></a><span class="fu">### Bar Plot of Hybrid Traits</span></span>
<span id="cb119-500"><a href="#cb119-500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-503"><a href="#cb119-503" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-504"><a href="#cb119-504" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb119-505"><a href="#cb119-505" aria-hidden="true" tabindex="-1"></a>  HybridTraits,</span>
<span id="cb119-506"><a href="#cb119-506" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb119-507"><a href="#cb119-507" aria-hidden="true" tabindex="-1"></a>  <span class="at">names.arg =</span> <span class="fu">colnames</span>(ParentTraits),</span>
<span id="cb119-508"><a href="#cb119-508" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#66c2a5"</span>, <span class="st">"#fc8d62"</span>, <span class="st">"#8da0cb"</span>),</span>
<span id="cb119-509"><a href="#cb119-509" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Hybrid Trait Profile"</span>,</span>
<span id="cb119-510"><a href="#cb119-510" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Trait Value"</span></span>
<span id="cb119-511"><a href="#cb119-511" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-512"><a href="#cb119-512" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-513"><a href="#cb119-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-514"><a href="#cb119-514" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation Questions</span></span>
<span id="cb119-515"><a href="#cb119-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-516"><a href="#cb119-516" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How does the weighting of parents affect the hybrid's performance?</span>
<span id="cb119-517"><a href="#cb119-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-518"><a href="#cb119-518" aria-hidden="true" tabindex="-1"></a>    Stronger weights mean more genetic contribution. Traits from highly weighted parents dominate the hybrid profile.</span>
<span id="cb119-519"><a href="#cb119-519" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-520"><a href="#cb119-520" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What does the identity matrix represent here?</span>
<span id="cb119-521"><a href="#cb119-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-522"><a href="#cb119-522" aria-hidden="true" tabindex="-1"></a>    It represents a neutral transformation. It confirms data integrity when used in matrix multiplication.</span>
<span id="cb119-523"><a href="#cb119-523" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-524"><a href="#cb119-524" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>If you used equal weights (⅓ for each), how would the hybrid traits change?</span>
<span id="cb119-525"><a href="#cb119-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-526"><a href="#cb119-526" aria-hidden="true" tabindex="-1"></a>    Traits would reflect an even mix, potentially leading to balanced but less specialized performance.</span>
<span id="cb119-527"><a href="#cb119-527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-528"><a href="#cb119-528" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What real-world limitations does this simplified model ignore?</span>
<span id="cb119-529"><a href="#cb119-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-530"><a href="#cb119-530" aria-hidden="true" tabindex="-1"></a>    i.  Non-additive genetic effects (dominance, epistasis)</span>
<span id="cb119-531"><a href="#cb119-531" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-532"><a href="#cb119-532" aria-hidden="true" tabindex="-1"></a>    ii. Environmental interactions</span>
<span id="cb119-533"><a href="#cb119-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-534"><a href="#cb119-534" aria-hidden="true" tabindex="-1"></a>    iii. Trait heritability and correlations</span>
<span id="cb119-535"><a href="#cb119-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-536"><a href="#cb119-536" aria-hidden="true" tabindex="-1"></a>    iv. Breeding feasibility and cost</span>
<span id="cb119-537"><a href="#cb119-537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-538"><a href="#cb119-538" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 5: Managing Matrices and Weight Vectors Using Lists in R</span></span>
<span id="cb119-539"><a href="#cb119-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-540"><a href="#cb119-540" aria-hidden="true" tabindex="-1"></a>Now that we've explored trait-based decisions using matrices, it's time to organize our work using R's list structure. Lists help bundle related objects like matrices and weight vectors, keeping the analysis modular and scalable.</span>
<span id="cb119-541"><a href="#cb119-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-542"><a href="#cb119-542" aria-hidden="true" tabindex="-1"></a><span class="fu">### Create a Master List</span></span>
<span id="cb119-543"><a href="#cb119-543" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-546"><a href="#cb119-546" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-547"><a href="#cb119-547" aria-hidden="true" tabindex="-1"></a><span class="co"># Assuming previous matrices and weights are already defined:</span></span>
<span id="cb119-548"><a href="#cb119-548" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-549"><a href="#cb119-549" aria-hidden="true" tabindex="-1"></a><span class="co"># Making a MasterList</span></span>
<span id="cb119-550"><a href="#cb119-550" aria-hidden="true" tabindex="-1"></a>bioList <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb119-551"><a href="#cb119-551" aria-hidden="true" tabindex="-1"></a>  <span class="at">ProteinConc =</span> <span class="fu">list</span>(<span class="at">matrix =</span> ProteinMatrix, <span class="at">weights =</span> WeightVector),</span>
<span id="cb119-552"><a href="#cb119-552" aria-hidden="true" tabindex="-1"></a>  <span class="at">ProteinMap  =</span> <span class="fu">list</span>(<span class="at">matrix =</span> GeneExpression, <span class="at">weights =</span> TranslationMatrix),</span>
<span id="cb119-553"><a href="#cb119-553" aria-hidden="true" tabindex="-1"></a>  <span class="at">Animal =</span> <span class="fu">list</span>(<span class="at">matrix =</span> BullEBVs, <span class="at">weights =</span> EconomicWeights),</span>
<span id="cb119-554"><a href="#cb119-554" aria-hidden="true" tabindex="-1"></a>  <span class="at">Plant =</span> <span class="fu">list</span>(<span class="at">matrix =</span> ParentTraits, <span class="at">weights =</span> HybridWeights)</span>
<span id="cb119-555"><a href="#cb119-555" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-556"><a href="#cb119-556" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-557"><a href="#cb119-557" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(bioList)</span>
<span id="cb119-558"><a href="#cb119-558" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-559"><a href="#cb119-559" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-560"><a href="#cb119-560" aria-hidden="true" tabindex="-1"></a><span class="fu">### List the Structure</span></span>
<span id="cb119-561"><a href="#cb119-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-564"><a href="#cb119-564" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-565"><a href="#cb119-565" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bioList)               <span class="co"># Top-level list names</span></span>
<span id="cb119-566"><a href="#cb119-566" aria-hidden="true" tabindex="-1"></a><span class="fu">lengths</span>(bioList)             <span class="co"># Number of components in each sublist</span></span>
<span id="cb119-567"><a href="#cb119-567" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-568"><a href="#cb119-568" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-569"><a href="#cb119-569" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb119-570"><a href="#cb119-570" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-571"><a href="#cb119-571" aria-hidden="true" tabindex="-1"></a>Each top-level entry (e.g., ProteinConc, Plant) contains two components:</span>
<span id="cb119-572"><a href="#cb119-572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-573"><a href="#cb119-573" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>A matrix (e.g., ProteinMatrix)</span>
<span id="cb119-574"><a href="#cb119-574" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>A corresponding weight vector or matrix</span>
<span id="cb119-575"><a href="#cb119-575" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-576"><a href="#cb119-576" aria-hidden="true" tabindex="-1"></a><span class="fu">### Indexing Elements from Lists</span></span>
<span id="cb119-577"><a href="#cb119-577" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-580"><a href="#cb119-580" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-581"><a href="#cb119-581" aria-hidden="true" tabindex="-1"></a><span class="co"># Access the trait matrix for Plant</span></span>
<span id="cb119-582"><a href="#cb119-582" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Plant[[<span class="dv">1</span>]]</span>
<span id="cb119-583"><a href="#cb119-583" aria-hidden="true" tabindex="-1"></a><span class="co">#or</span></span>
<span id="cb119-584"><a href="#cb119-584" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix</span>
<span id="cb119-585"><a href="#cb119-585" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-586"><a href="#cb119-586" aria-hidden="true" tabindex="-1"></a><span class="co"># Access the weight vector for ProteinConc</span></span>
<span id="cb119-587"><a href="#cb119-587" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinConc[[<span class="dv">2</span>]]</span>
<span id="cb119-588"><a href="#cb119-588" aria-hidden="true" tabindex="-1"></a><span class="co">#or</span></span>
<span id="cb119-589"><a href="#cb119-589" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinConc<span class="sc">$</span>weights</span>
<span id="cb119-590"><a href="#cb119-590" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-591"><a href="#cb119-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-592"><a href="#cb119-592" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb119-593"><a href="#cb119-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-594"><a href="#cb119-594" aria-hidden="true" tabindex="-1"></a>Use double brackets <span class="in">`[[ ]]`</span> to extract unnamed list elements by position. But we named our list, so they are easily extractable using the <span class="in">`$`</span> notation.</span>
<span id="cb119-595"><a href="#cb119-595" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-596"><a href="#cb119-596" aria-hidden="true" tabindex="-1"></a><span class="fu">### Perform Weighted Calculations</span></span>
<span id="cb119-597"><a href="#cb119-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-600"><a href="#cb119-600" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-601"><a href="#cb119-601" aria-hidden="true" tabindex="-1"></a><span class="co"># Protein concentration score</span></span>
<span id="cb119-602"><a href="#cb119-602" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinConc<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>ProteinConc<span class="sc">$</span>weights</span>
<span id="cb119-603"><a href="#cb119-603" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-604"><a href="#cb119-604" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene → Protein contribution</span></span>
<span id="cb119-605"><a href="#cb119-605" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinMap<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>ProteinMap<span class="sc">$</span>weights</span>
<span id="cb119-606"><a href="#cb119-606" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-607"><a href="#cb119-607" aria-hidden="true" tabindex="-1"></a><span class="co"># Bull economic value</span></span>
<span id="cb119-608"><a href="#cb119-608" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Animal<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>Animal<span class="sc">$</span>weights</span>
<span id="cb119-609"><a href="#cb119-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-610"><a href="#cb119-610" aria-hidden="true" tabindex="-1"></a><span class="co"># Hybrid trait value</span></span>
<span id="cb119-611"><a href="#cb119-611" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Plant<span class="sc">$</span>weights <span class="sc">%*%</span> bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix</span>
<span id="cb119-612"><a href="#cb119-612" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-613"><a href="#cb119-613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-614"><a href="#cb119-614" aria-hidden="true" tabindex="-1"></a><span class="fu">### Subset and Recalculate</span></span>
<span id="cb119-615"><a href="#cb119-615" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-618"><a href="#cb119-618" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-619"><a href="#cb119-619" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove last trait from ParentTraits</span></span>
<span id="cb119-620"><a href="#cb119-620" aria-hidden="true" tabindex="-1"></a>ParentSubset <span class="ot">&lt;-</span> bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb119-621"><a href="#cb119-621" aria-hidden="true" tabindex="-1"></a>NewWeights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.6</span>, <span class="fl">0.4</span>), <span class="at">nrow =</span> <span class="dv">2</span>)</span>
<span id="cb119-622"><a href="#cb119-622" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-623"><a href="#cb119-623" aria-hidden="true" tabindex="-1"></a><span class="co"># Recalculated hybrid score</span></span>
<span id="cb119-624"><a href="#cb119-624" aria-hidden="true" tabindex="-1"></a>SubsetHybridScore <span class="ot">&lt;-</span> ParentSubset <span class="sc">%*%</span> NewWeights</span>
<span id="cb119-625"><a href="#cb119-625" aria-hidden="true" tabindex="-1"></a>SubsetHybridScore</span>
<span id="cb119-626"><a href="#cb119-626" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-627"><a href="#cb119-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-628"><a href="#cb119-628" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb119-629"><a href="#cb119-629" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-630"><a href="#cb119-630" aria-hidden="true" tabindex="-1"></a>Dropping a trait and reweighting highlights its influence in trait aggregation and selection.</span>
<span id="cb119-631"><a href="#cb119-631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-632"><a href="#cb119-632" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualization Tasks</span></span>
<span id="cb119-633"><a href="#cb119-633" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-634"><a href="#cb119-634" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Heatmap: Gene Expression</span></span>
<span id="cb119-635"><a href="#cb119-635" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-638"><a href="#cb119-638" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-639"><a href="#cb119-639" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb119-640"><a href="#cb119-640" aria-hidden="true" tabindex="-1"></a>  bioList<span class="sc">$</span>ProteinMap<span class="sc">$</span>matrix,</span>
<span id="cb119-641"><a href="#cb119-641" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb119-642"><a href="#cb119-642" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">256</span>),</span>
<span id="cb119-643"><a href="#cb119-643" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Gene Expression Heatmap"</span>,</span>
<span id="cb119-644"><a href="#cb119-644" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Proteins"</span>,</span>
<span id="cb119-645"><a href="#cb119-645" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Genes"</span></span>
<span id="cb119-646"><a href="#cb119-646" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-647"><a href="#cb119-647" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-648"><a href="#cb119-648" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-649"><a href="#cb119-649" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Bar Plots</span></span>
<span id="cb119-650"><a href="#cb119-650" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-651"><a href="#cb119-651" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Hybrid traits</span></span>
<span id="cb119-652"><a href="#cb119-652" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-655"><a href="#cb119-655" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-656"><a href="#cb119-656" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb119-657"><a href="#cb119-657" aria-hidden="true" tabindex="-1"></a>  bioList<span class="sc">$</span>Plant<span class="sc">$</span>weights <span class="sc">%*%</span> bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix,</span>
<span id="cb119-658"><a href="#cb119-658" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb119-659"><a href="#cb119-659" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Hybrid Trait Contributions"</span>,</span>
<span id="cb119-660"><a href="#cb119-660" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="st">"#66c2a5"</span>,</span>
<span id="cb119-661"><a href="#cb119-661" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Score"</span></span>
<span id="cb119-662"><a href="#cb119-662" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-663"><a href="#cb119-663" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-664"><a href="#cb119-664" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-665"><a href="#cb119-665" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Bull EBV (Economic Breeding Values)</span></span>
<span id="cb119-666"><a href="#cb119-666" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-669"><a href="#cb119-669" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-670"><a href="#cb119-670" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb119-671"><a href="#cb119-671" aria-hidden="true" tabindex="-1"></a>  bioList<span class="sc">$</span>Animal<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>Animal<span class="sc">$</span>weights,</span>
<span id="cb119-672"><a href="#cb119-672" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb119-673"><a href="#cb119-673" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Bull EBVs (Economic Values)"</span>,</span>
<span id="cb119-674"><a href="#cb119-674" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="st">"#fc8d62"</span>,</span>
<span id="cb119-675"><a href="#cb119-675" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Score"</span></span>
<span id="cb119-676"><a href="#cb119-676" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-677"><a href="#cb119-677" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-678"><a href="#cb119-678" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-679"><a href="#cb119-679" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation Questions</span></span>
<span id="cb119-680"><a href="#cb119-680" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-681"><a href="#cb119-681" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>Why use a list structure?</span>
<span id="cb119-682"><a href="#cb119-682" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-683"><a href="#cb119-683" aria-hidden="true" tabindex="-1"></a>Keeps each dataset and its weights together. Facilitates automated workflows and reuse.</span>
<span id="cb119-684"><a href="#cb119-684" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-685"><a href="#cb119-685" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>What’s tricky about <span class="in">`[[ ]]`</span> access?</span>
<span id="cb119-686"><a href="#cb119-686" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-687"><a href="#cb119-687" aria-hidden="true" tabindex="-1"></a>You must remember the order (<span class="in">`[[1]]`</span> = <span class="in">`matrix`</span>, <span class="in">`[[2]]`</span> = <span class="in">`weights`</span>). No names means you can't use <span class="in">`$matrix`</span>, only positional access.</span>
<span id="cb119-688"><a href="#cb119-688" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-689"><a href="#cb119-689" aria-hidden="true" tabindex="-1"></a>Loop across all list entries</span>
<span id="cb119-690"><a href="#cb119-690" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-691"><a href="#cb119-691" aria-hidden="true" tabindex="-1"></a>Weighted scores for all entries <span class="in">`lapply(bioList, function(x) x[[2]] %*% x[[1]])`</span></span>
<span id="cb119-692"><a href="#cb119-692" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-693"><a href="#cb119-693" aria-hidden="true" tabindex="-1"></a><span class="ss">-   </span>How does this help in large-scale pipelines?</span>
<span id="cb119-694"><a href="#cb119-694" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-695"><a href="#cb119-695" aria-hidden="true" tabindex="-1"></a>You can use this format with <span class="in">`lapply()`</span>, <span class="in">`purrr::map()`</span>, or in targets pipelines for reproducibility and modular processing.</span>
<span id="cb119-696"><a href="#cb119-696" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-697"><a href="#cb119-697" aria-hidden="true" tabindex="-1"></a><span class="fu">## 🧩 Homework solutions: Factors, Subsetting, and Biological Insight</span></span>
<span id="cb119-698"><a href="#cb119-698" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-699"><a href="#cb119-699" aria-hidden="true" tabindex="-1"></a><span class="ss">1.  </span>**Character vs Factor**</span>
<span id="cb119-700"><a href="#cb119-700" aria-hidden="true" tabindex="-1"></a>    A character vector simply holds string values, but a factor is a categorical variable with fixed levels, used especially in modeling.  </span>
<span id="cb119-701"><a href="#cb119-701" aria-hidden="true" tabindex="-1"></a>    For <span class="in">`mutation_status`</span>, a factor ensures consistent categories (e.g., <span class="in">`"Yes"`</span> or <span class="in">`"No"`</span>) and helps control level order and statistical reference groups.  </span>
<span id="cb119-702"><a href="#cb119-702" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-703"><a href="#cb119-703" aria-hidden="true" tabindex="-1"></a><span class="ss">2.  </span>**Factor Levels**</span>
<span id="cb119-704"><a href="#cb119-704" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-707"><a href="#cb119-707" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-708"><a href="#cb119-708" aria-hidden="true" tabindex="-1"></a>species <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Lactobacillus"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Bacteroides"</span>, <span class="st">"Lactobacillus"</span>)</span>
<span id="cb119-709"><a href="#cb119-709" aria-hidden="true" tabindex="-1"></a>species_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(species, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Bacteroides"</span>, <span class="st">"Escherichia"</span>, <span class="st">"Lactobacillus"</span>))</span>
<span id="cb119-710"><a href="#cb119-710" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(species_factor)</span>
<span id="cb119-711"><a href="#cb119-711" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-712"><a href="#cb119-712" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-713"><a href="#cb119-713" aria-hidden="true" tabindex="-1"></a>Because we defined the level order explicitly, R maintains that order regardless of data input.</span>
<span id="cb119-714"><a href="#cb119-714" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-715"><a href="#cb119-715" aria-hidden="true" tabindex="-1"></a><span class="ss">3.  </span>Ordered Factor Comparison</span>
<span id="cb119-716"><a href="#cb119-716" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-719"><a href="#cb119-719" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-720"><a href="#cb119-720" aria-hidden="true" tabindex="-1"></a>disease_severity <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Mild"</span>, <span class="st">"Severe"</span>, <span class="st">"Moderate"</span>), <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Mild"</span>, <span class="st">"Moderate"</span>, <span class="st">"Severe"</span>, <span class="st">"Critical"</span>), <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb119-721"><a href="#cb119-721" aria-hidden="true" tabindex="-1"></a>disease_severity[<span class="dv">1</span>] <span class="sc">&lt;</span> disease_severity[<span class="dv">2</span>]</span>
<span id="cb119-722"><a href="#cb119-722" aria-hidden="true" tabindex="-1"></a><span class="co"># TRUE</span></span>
<span id="cb119-723"><a href="#cb119-723" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-724"><a href="#cb119-724" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-725"><a href="#cb119-725" aria-hidden="true" tabindex="-1"></a>"Mild" is less severe than "Severe" based on the defined order.</span>
<span id="cb119-726"><a href="#cb119-726" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-727"><a href="#cb119-727" aria-hidden="true" tabindex="-1"></a><span class="ss">4.  </span>Proportion Extraction</span>
<span id="cb119-728"><a href="#cb119-728" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-731"><a href="#cb119-731" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-732"><a href="#cb119-732" aria-hidden="true" tabindex="-1"></a>prop <span class="ot">&lt;-</span> <span class="fu">prop.table</span>(<span class="fu">table</span>(species_factor))</span>
<span id="cb119-733"><a href="#cb119-733" aria-hidden="true" tabindex="-1"></a>prop[<span class="st">"Escherichia"</span>]</span>
<span id="cb119-734"><a href="#cb119-734" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-735"><a href="#cb119-735" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-736"><a href="#cb119-736" aria-hidden="true" tabindex="-1"></a><span class="in">`prop$Escherichia`</span> won't work — named numeric vectors require bracket-based access.</span>
<span id="cb119-737"><a href="#cb119-737" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-738"><a href="#cb119-738" aria-hidden="true" tabindex="-1"></a><span class="ss">5.  </span>Subsetting by Conditions</span>
<span id="cb119-739"><a href="#cb119-739" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-742"><a href="#cb119-742" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-743"><a href="#cb119-743" aria-hidden="true" tabindex="-1"></a>gene_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb119-744"><a href="#cb119-744" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_id =</span> <span class="fu">c</span>(<span class="st">"BRCA1"</span>, <span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"EGFR"</span>, <span class="st">"GAPDH"</span>),</span>
<span id="cb119-745"><a href="#cb119-745" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression =</span> <span class="fu">c</span>(<span class="fl">8.2</span>, <span class="fl">6.1</span>, <span class="fl">9.5</span>, <span class="fl">7.0</span>, <span class="fl">10.0</span>),</span>
<span id="cb119-746"><a href="#cb119-746" aria-hidden="true" tabindex="-1"></a>  <span class="at">mutation =</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"Yes"</span>, <span class="st">"No"</span>, <span class="st">"No"</span>)),</span>
<span id="cb119-747"><a href="#cb119-747" aria-hidden="true" tabindex="-1"></a>  <span class="at">pathway =</span> <span class="fu">c</span>(<span class="st">"DNA Repair"</span>, <span class="st">"Apoptosis"</span>, <span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>, <span class="st">"Metabolism"</span>)</span>
<span id="cb119-748"><a href="#cb119-748" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-749"><a href="#cb119-749" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(gene_df) <span class="ot">&lt;-</span> gene_df<span class="sc">$</span>gene_id <span class="co">#name the rows by the gene IDs</span></span>
<span id="cb119-750"><a href="#cb119-750" aria-hidden="true" tabindex="-1"></a>gene_df <span class="ot">&lt;-</span> gene_df[, <span class="sc">-</span><span class="dv">1</span>] <span class="co">#remove the first column which is not needed anymore</span></span>
<span id="cb119-751"><a href="#cb119-751" aria-hidden="true" tabindex="-1"></a><span class="co">#gene_df</span></span>
<span id="cb119-752"><a href="#cb119-752" aria-hidden="true" tabindex="-1"></a>gene_df[gene_df<span class="sc">$</span>expression <span class="sc">&gt;</span> <span class="dv">7</span> <span class="sc">&amp;</span> gene_df<span class="sc">$</span>mutation <span class="sc">==</span> <span class="st">"No"</span>, ]</span>
<span id="cb119-753"><a href="#cb119-753" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-754"><a href="#cb119-754" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-755"><a href="#cb119-755" aria-hidden="true" tabindex="-1"></a>Returns genes with **high expression (\&gt;7)** `and` **no mutation** — potentially highly active but wild-type genes.</span>
<span id="cb119-756"><a href="#cb119-756" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-757"><a href="#cb119-757" aria-hidden="true" tabindex="-1"></a><span class="ss">6.  </span>Group-wise Expression Summary</span>
<span id="cb119-758"><a href="#cb119-758" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-759"><a href="#cb119-759" aria-hidden="true" tabindex="-1"></a>The given vectors are:</span>
<span id="cb119-760"><a href="#cb119-760" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-763"><a href="#cb119-763" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-764"><a href="#cb119-764" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span>, <span class="st">"KO"</span>, <span class="st">"WT"</span>)</span>
<span id="cb119-765"><a href="#cb119-765" aria-hidden="true" tabindex="-1"></a>expression <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">5.2</span>, <span class="fl">8.1</span>, <span class="fl">4.3</span>, <span class="fl">9.0</span>, <span class="fl">5.7</span>)</span>
<span id="cb119-766"><a href="#cb119-766" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-767"><a href="#cb119-767" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-768"><a href="#cb119-768" aria-hidden="true" tabindex="-1"></a>The solution would be:</span>
<span id="cb119-769"><a href="#cb119-769" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-772"><a href="#cb119-772" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-773"><a href="#cb119-773" aria-hidden="true" tabindex="-1"></a>group_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(samples)</span>
<span id="cb119-774"><a href="#cb119-774" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-775"><a href="#cb119-775" aria-hidden="true" tabindex="-1"></a><span class="co"># Mean expression</span></span>
<span id="cb119-776"><a href="#cb119-776" aria-hidden="true" tabindex="-1"></a><span class="fu">tapply</span>(expression, group_factor, mean) <span class="do">## KO: 8.55, WT: 5.07</span></span>
<span id="cb119-777"><a href="#cb119-777" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-778"><a href="#cb119-778" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb119-779"><a href="#cb119-779" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">tapply</span>(expression, group_factor, mean), </span>
<span id="cb119-780"><a href="#cb119-780" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"salmon"</span>),</span>
<span id="cb119-781"><a href="#cb119-781" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Mean Expression"</span>,</span>
<span id="cb119-782"><a href="#cb119-782" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Group-wise Expression"</span>)</span>
<span id="cb119-783"><a href="#cb119-783" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-784"><a href="#cb119-784" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-785"><a href="#cb119-785" aria-hidden="true" tabindex="-1"></a><span class="ss">7.  </span>Gene Subsetting</span>
<span id="cb119-786"><a href="#cb119-786" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-787"><a href="#cb119-787" aria-hidden="true" tabindex="-1"></a><span class="in">```{r gene_df_subset}</span></span>
<span id="cb119-788"><a href="#cb119-788" aria-hidden="true" tabindex="-1"></a>gene_df[gene_df<span class="sc">$</span>expression <span class="sc">&gt;</span> <span class="dv">8</span> <span class="sc">&amp;</span> </span>
<span id="cb119-789"><a href="#cb119-789" aria-hidden="true" tabindex="-1"></a>        gene_df<span class="sc">$</span>pathway <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Cell Cycle"</span>, <span class="st">"Signaling"</span>), ]</span>
<span id="cb119-790"><a href="#cb119-790" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-791"><a href="#cb119-791" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-792"><a href="#cb119-792" aria-hidden="true" tabindex="-1"></a>It filters for genes highly expressed and involved in key biological pathways.</span>
<span id="cb119-793"><a href="#cb119-793" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-794"><a href="#cb119-794" aria-hidden="true" tabindex="-1"></a><span class="ss">8.  </span>Disease Stage Visualization</span>
<span id="cb119-795"><a href="#cb119-795" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-798"><a href="#cb119-798" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-799"><a href="#cb119-799" aria-hidden="true" tabindex="-1"></a>stages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Stage I"</span>, <span class="st">"Stage III"</span>, <span class="st">"Stage II"</span>, <span class="st">"Stage IV"</span>, <span class="st">"Stage I"</span>)</span>
<span id="cb119-800"><a href="#cb119-800" aria-hidden="true" tabindex="-1"></a>disease_stage <span class="ot">&lt;-</span> <span class="fu">factor</span>(stages, </span>
<span id="cb119-801"><a href="#cb119-801" aria-hidden="true" tabindex="-1"></a>                        <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"Stage I"</span>, <span class="st">"Stage II"</span>, <span class="st">"Stage III"</span>, <span class="st">"Stage IV"</span>), </span>
<span id="cb119-802"><a href="#cb119-802" aria-hidden="true" tabindex="-1"></a>                        <span class="at">ordered =</span> <span class="cn">TRUE</span>)</span>
<span id="cb119-803"><a href="#cb119-803" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-804"><a href="#cb119-804" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(<span class="fu">table</span>(disease_stage), </span>
<span id="cb119-805"><a href="#cb119-805" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="st">"lightgreen"</span>, </span>
<span id="cb119-806"><a href="#cb119-806" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Patient Count by Disease Stage"</span>,</span>
<span id="cb119-807"><a href="#cb119-807" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Count"</span>)</span>
<span id="cb119-808"><a href="#cb119-808" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-809"><a href="#cb119-809" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-810"><a href="#cb119-810" aria-hidden="true" tabindex="-1"></a>Let's do the severity order check:</span>
<span id="cb119-811"><a href="#cb119-811" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-814"><a href="#cb119-814" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-815"><a href="#cb119-815" aria-hidden="true" tabindex="-1"></a><span class="co"># Comparison</span></span>
<span id="cb119-816"><a href="#cb119-816" aria-hidden="true" tabindex="-1"></a>disease_stage[<span class="dv">2</span>] <span class="sc">&gt;</span> disease_stage[<span class="dv">1</span>]  <span class="co"># TRUE</span></span>
<span id="cb119-817"><a href="#cb119-817" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-818"><a href="#cb119-818" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-819"><a href="#cb119-819" aria-hidden="true" tabindex="-1"></a>So, "Stage III" is more sever than "Stage I".</span>
<span id="cb119-820"><a href="#cb119-820" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-821"><a href="#cb119-821" aria-hidden="true" tabindex="-1"></a><span class="ss">9.  </span>Oncogene Subsetting and Releveling</span>
<span id="cb119-822"><a href="#cb119-822" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-825"><a href="#cb119-825" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-826"><a href="#cb119-826" aria-hidden="true" tabindex="-1"></a><span class="co"># Define a small gene dataset</span></span>
<span id="cb119-827"><a href="#cb119-827" aria-hidden="true" tabindex="-1"></a>gene_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb119-828"><a href="#cb119-828" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene =</span> <span class="fu">c</span>(<span class="st">"TP53"</span>, <span class="st">"BRCA1"</span>, <span class="st">"MYC"</span>, <span class="st">"GAPDH"</span>, <span class="st">"EGFR"</span>),</span>
<span id="cb119-829"><a href="#cb119-829" aria-hidden="true" tabindex="-1"></a>  <span class="at">expression =</span> <span class="fu">c</span>(<span class="fl">9.1</span>, <span class="fl">7.3</span>, <span class="fl">10.5</span>, <span class="fl">5.2</span>, <span class="fl">8.6</span>),</span>
<span id="cb119-830"><a href="#cb119-830" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="fu">factor</span>(<span class="fu">c</span>(<span class="st">"Tumor Suppressor"</span>, <span class="st">"Oncogene"</span>, <span class="st">"Oncogene"</span>, <span class="st">"Housekeeping"</span>, <span class="st">"Oncogene"</span>))</span>
<span id="cb119-831"><a href="#cb119-831" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb119-832"><a href="#cb119-832" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-833"><a href="#cb119-833" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset: Oncogene rows with expression &gt; 8</span></span>
<span id="cb119-834"><a href="#cb119-834" aria-hidden="true" tabindex="-1"></a>gene_data[gene_data<span class="sc">$</span>type <span class="sc">==</span> <span class="st">"Oncogene"</span> <span class="sc">&amp;</span> gene_data<span class="sc">$</span>expression <span class="sc">&gt;</span> <span class="dv">8</span>, ]</span>
<span id="cb119-835"><a href="#cb119-835" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-836"><a href="#cb119-836" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-837"><a href="#cb119-837" aria-hidden="true" tabindex="-1"></a>Let's relevel now, "Housekeeping" is the reference:</span>
<span id="cb119-838"><a href="#cb119-838" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-841"><a href="#cb119-841" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-842"><a href="#cb119-842" aria-hidden="true" tabindex="-1"></a><span class="co"># Relevel: make "Housekeeping" the reference level</span></span>
<span id="cb119-843"><a href="#cb119-843" aria-hidden="true" tabindex="-1"></a>gene_data<span class="sc">$</span>type <span class="ot">&lt;-</span> <span class="fu">relevel</span>(gene_data<span class="sc">$</span>type, <span class="at">ref =</span> <span class="st">"Housekeeping"</span>)</span>
<span id="cb119-844"><a href="#cb119-844" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-845"><a href="#cb119-845" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the new levels</span></span>
<span id="cb119-846"><a href="#cb119-846" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(gene_data<span class="sc">$</span>type)</span>
<span id="cb119-847"><a href="#cb119-847" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-848"><a href="#cb119-848" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-849"><a href="#cb119-849" aria-hidden="true" tabindex="-1"></a>But the "Housekeeping" was already a reference by default (using alphabetic ordering by R). Making something else the reference would make more sense. Our code above work, but nothing new is done.</span>
<span id="cb119-850"><a href="#cb119-850" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-851"><a href="#cb119-851" aria-hidden="true" tabindex="-1"></a><span class="ss">10. </span>Simulated Expression by Tissue</span>
<span id="cb119-852"><a href="#cb119-852" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-855"><a href="#cb119-855" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-856"><a href="#cb119-856" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb119-857"><a href="#cb119-857" aria-hidden="true" tabindex="-1"></a>gene_expr <span class="ot">&lt;-</span> <span class="fu">rnorm</span>(<span class="dv">45</span>, <span class="at">mean =</span> <span class="dv">8</span>, <span class="at">sd =</span> <span class="dv">2</span>)</span>
<span id="cb119-858"><a href="#cb119-858" aria-hidden="true" tabindex="-1"></a>tissue <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"brain"</span>, <span class="st">"liver"</span>, <span class="st">"kidney"</span>), <span class="at">each =</span> <span class="dv">15</span>)</span>
<span id="cb119-859"><a href="#cb119-859" aria-hidden="true" tabindex="-1"></a>tissue_factor <span class="ot">&lt;-</span> <span class="fu">factor</span>(tissue, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"liver"</span>, <span class="st">"brain"</span>, <span class="st">"kidney"</span>))</span>
<span id="cb119-860"><a href="#cb119-860" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-861"><a href="#cb119-861" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(gene_expr <span class="sc">~</span> tissue_factor, </span>
<span id="cb119-862"><a href="#cb119-862" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"orange"</span>, <span class="st">"skyblue"</span>, <span class="st">"lightgreen"</span>), </span>
<span id="cb119-863"><a href="#cb119-863" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Expression by Tissue"</span>,</span>
<span id="cb119-864"><a href="#cb119-864" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Expression Level"</span>)</span>
<span id="cb119-865"><a href="#cb119-865" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-866"><a href="#cb119-866" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-867"><a href="#cb119-867" aria-hidden="true" tabindex="-1"></a>Let's calculate variability per tissue type now:</span>
<span id="cb119-868"><a href="#cb119-868" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-871"><a href="#cb119-871" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-872"><a href="#cb119-872" aria-hidden="true" tabindex="-1"></a><span class="co"># Variability</span></span>
<span id="cb119-873"><a href="#cb119-873" aria-hidden="true" tabindex="-1"></a><span class="fu">tapply</span>(gene_expr, tissue_factor, sd)</span>
<span id="cb119-874"><a href="#cb119-874" aria-hidden="true" tabindex="-1"></a><span class="co"># Returns standard deviation per tissue group</span></span>
<span id="cb119-875"><a href="#cb119-875" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-876"><a href="#cb119-876" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-877"><a href="#cb119-877" aria-hidden="true" tabindex="-1"></a><span class="fu">## 🧩 Homework Solution tidyverse</span></span>
<span id="cb119-878"><a href="#cb119-878" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-881"><a href="#cb119-881" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-882"><a href="#cb119-882" aria-hidden="true" tabindex="-1"></a><span class="co"># Load necessary package</span></span>
<span id="cb119-883"><a href="#cb119-883" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb119-884"><a href="#cb119-884" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-885"><a href="#cb119-885" aria-hidden="true" tabindex="-1"></a><span class="co"># Use the built-in iris dataset</span></span>
<span id="cb119-886"><a href="#cb119-886" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(iris)</span>
<span id="cb119-887"><a href="#cb119-887" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-888"><a href="#cb119-888" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the plot</span></span>
<span id="cb119-889"><a href="#cb119-889" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iris, <span class="fu">aes</span>(<span class="at">x =</span> Petal.Length, <span class="at">y =</span> Petal.Width, <span class="at">color =</span> Species)) <span class="sc">+</span></span>
<span id="cb119-890"><a href="#cb119-890" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span>  <span class="co"># scatter plot points</span></span>
<span id="cb119-891"><a href="#cb119-891" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>) <span class="sc">+</span>  <span class="co"># linear regression lines</span></span>
<span id="cb119-892"><a href="#cb119-892" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb119-893"><a href="#cb119-893" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Relationship between Petal Length and Width by Species"</span>,</span>
<span id="cb119-894"><a href="#cb119-894" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Petal Length (cm)"</span>,</span>
<span id="cb119-895"><a href="#cb119-895" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Petal Width (cm)"</span>,</span>
<span id="cb119-896"><a href="#cb119-896" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">"Iris Species"</span></span>
<span id="cb119-897"><a href="#cb119-897" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb119-898"><a href="#cb119-898" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()  <span class="co"># apply a clean minimal theme</span></span>
<span id="cb119-899"><a href="#cb119-899" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb119-900"><a href="#cb119-900" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-903"><a href="#cb119-903" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb119-904"><a href="#cb119-904" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb119-905"><a href="#cb119-905" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-906"><a href="#cb119-906" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(iris, <span class="fu">aes</span>(<span class="at">x =</span> Petal.Length, <span class="at">y =</span> Petal.Width)) <span class="sc">+</span></span>
<span id="cb119-907"><a href="#cb119-907" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> Species), <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span>  <span class="co"># color points by Species</span></span>
<span id="cb119-908"><a href="#cb119-908" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span>  <span class="co"># single regression line</span></span>
<span id="cb119-909"><a href="#cb119-909" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb119-910"><a href="#cb119-910" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Overall Regression: Petal Length vs Width (Iris Dataset)"</span>,</span>
<span id="cb119-911"><a href="#cb119-911" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Petal Length (cm)"</span>,</span>
<span id="cb119-912"><a href="#cb119-912" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Petal Width (cm)"</span>,</span>
<span id="cb119-913"><a href="#cb119-913" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="st">"Iris Species"</span></span>
<span id="cb119-914"><a href="#cb119-914" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb119-915"><a href="#cb119-915" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb119-916"><a href="#cb119-916" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb119-917"><a href="#cb119-917" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>