<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="dcterms.date" content="2025-05-08">
<title>pythrash - HW solutions</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="../../_include/styles.css">
</head>
<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg navbar-dark "><div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">PythRaSh</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../../index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-r" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">R</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-r">
<li>
    <a class="dropdown-item" href="../../ch/rbasics/git.html" rel="" target="">
 <span class="dropdown-text">git</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/firststeps.html" rel="" target="">
 <span class="dropdown-text">Basic R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/tidyverse.html" rel="" target="">
 <span class="dropdown-text">The tidyverse</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/correlation_regression.html" rel="" target="">
 <span class="dropdown-text">Advanced R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/baddata.html" rel="" target="">
 <span class="dropdown-text">Data Wrangling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/rbasics/solutions.html" rel="" target="">
 <span class="dropdown-text">Problem Solutions</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-python" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Python</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-python">
<li>
    <a class="dropdown-item" href="../../ch/python/basics.html" rel="" target="">
 <span class="dropdown-text">Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/panda.html" rel="" target="">
 <span class="dropdown-text">Advanced Python</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/MLA.html" rel="" target="">
 <span class="dropdown-text">Machine Learning Algorithms</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/python/AI.html" rel="" target="">
 <span class="dropdown-text">AI</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-linux-and-clt" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Linux and CLT</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-linux-and-clt">
<li>
    <a class="dropdown-item" href="../../ch/linux-and-ctl/basics.html" rel="" target="">
 <span class="dropdown-text">Linux Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/linux-and-ctl/advanced.html" rel="" target="">
 <span class="dropdown-text">Advanced Command Line</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-probability-and-statistics" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Probability and Statistics</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-probability-and-statistics">
<li>
    <a class="dropdown-item" href="../../ch/prob_stat/prob_basic.html" rel="" target="">
 <span class="dropdown-text">Probability Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/prob_stat/stat.html" rel="" target="">
 <span class="dropdown-text">Statistics</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-genomics" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Genomics</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-genomics">
<li>
    <a class="dropdown-item" href="../../ch/genomics/basics.html" rel="" target="">
 <span class="dropdown-text">Genomics Basics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/genomics/advanced.html" rel="" target="">
 <span class="dropdown-text">Advanced Genomics</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/genomics/pipeline.html" rel="" target="">
 <span class="dropdown-text">Pipelines</span></a>
  </li>  
    </ul>
</li>
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-blogs" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Blogs</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-blogs">
<li>
    <a class="dropdown-item" href="../../ch/blogs/en/index.html" rel="" target="">
 <span class="dropdown-text">English</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../ch/blogs/bn/index.html" rel="" target="">
 <span class="dropdown-text">Bangla</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/MdRasheduz-zaman" rel="" target="">
 <span class="menu-text">Contact</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/MdRasheduz-zaman/PythRaSh" rel="" target=""><i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
 <span class="menu-text"></span></a>
  </li>  
</ul>
<div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#l3-matrices-and-lists" id="toc-l3-matrices-and-lists" class="nav-link active" data-scroll-target="#l3-matrices-and-lists">L3: Matrices and Lists</a>
  <ul class="collapse">
<li>
<a href="#task-1-protein-quantification-in-biological-samples" id="toc-task-1-protein-quantification-in-biological-samples" class="nav-link" data-scroll-target="#task-1-protein-quantification-in-biological-samples">Task 1: Protein Quantification in Biological Samples</a>
  <ul class="collapse">
<li><a href="#matrices" id="toc-matrices" class="nav-link" data-scroll-target="#matrices">1. Matrices</a></li>
  <li><a href="#making-transpose-of-proteinmatrix" id="toc-making-transpose-of-proteinmatrix" class="nav-link" data-scroll-target="#making-transpose-of-proteinmatrix">2. Making transpose of ProteinMatrix</a></li>
  <li><a href="#identity-matrix" id="toc-identity-matrix" class="nav-link" data-scroll-target="#identity-matrix">3. Identity Matrix</a></li>
  <li><a href="#total-protein-per-sample" id="toc-total-protein-per-sample" class="nav-link" data-scroll-target="#total-protein-per-sample">4.1. Total Protein per Sample</a></li>
  <li><a href="#total-protein-per-protein-type" id="toc-total-protein-per-protein-type" class="nav-link" data-scroll-target="#total-protein-per-protein-type">4.2. Total Protein per Protein Type</a></li>
  <li><a href="#heatmap-of-protein-concentrations" id="toc-heatmap-of-protein-concentrations" class="nav-link" data-scroll-target="#heatmap-of-protein-concentrations">4.3. Heatmap of Protein Concentrations</a></li>
  <li><a href="#interpretation" id="toc-interpretation" class="nav-link" data-scroll-target="#interpretation">🧠 Interpretation</a></li>
  </ul>
</li>
  <li>
<a href="#task-2-gene-to-protein-translation" id="toc-task-2-gene-to-protein-translation" class="nav-link" data-scroll-target="#task-2-gene-to-protein-translation">Task 2: Gene-to-Protein Translation</a>
  <ul class="collapse">
<li><a href="#protein-output" id="toc-protein-output" class="nav-link" data-scroll-target="#protein-output">Protein Output</a></li>
  <li><a href="#transpose" id="toc-transpose" class="nav-link" data-scroll-target="#transpose">2. Transpose</a></li>
  <li><a href="#identity-matrix-and-multipliccation" id="toc-identity-matrix-and-multipliccation" class="nav-link" data-scroll-target="#identity-matrix-and-multipliccation">3. Identity matrix and multipliccation</a></li>
  <li><a href="#sub-matrix" id="toc-sub-matrix" class="nav-link" data-scroll-target="#sub-matrix">4. Sub-matrix:</a></li>
  <li><a href="#visualization-tasks" id="toc-visualization-tasks" class="nav-link" data-scroll-target="#visualization-tasks">Visualization tasks:</a></li>
  <li><a href="#interpretation-1" id="toc-interpretation-1" class="nav-link" data-scroll-target="#interpretation-1">🧠 Interpretation</a></li>
  </ul>
</li>
  <li>
<a href="#task-3-animal-breeding-bull-ranking-by-economic-traits" id="toc-task-3-animal-breeding-bull-ranking-by-economic-traits" class="nav-link" data-scroll-target="#task-3-animal-breeding-bull-ranking-by-economic-traits">Task 3: Animal Breeding – Bull Ranking by Economic Traits</a>
  <ul class="collapse">
<li><a href="#define-data" id="toc-define-data" class="nav-link" data-scroll-target="#define-data">Define Data</a></li>
  <li><a href="#compute-total-economic-value" id="toc-compute-total-economic-value" class="nav-link" data-scroll-target="#compute-total-economic-value">Compute Total Economic Value</a></li>
  <li><a href="#multiply-with-identity-matrix" id="toc-multiply-with-identity-matrix" class="nav-link" data-scroll-target="#multiply-with-identity-matrix">Multiply with Identity Matrix</a></li>
  <li><a href="#remove-milk-yield-and-recalculate-total-value" id="toc-remove-milk-yield-and-recalculate-total-value" class="nav-link" data-scroll-target="#remove-milk-yield-and-recalculate-total-value">Remove Milk Yield and Recalculate Total Value</a></li>
  <li><a href="#bar-plot-total-economic-value" id="toc-bar-plot-total-economic-value" class="nav-link" data-scroll-target="#bar-plot-total-economic-value">Bar Plot: Total Economic Value</a></li>
  <li><a href="#heatmap-of-ebvs" id="toc-heatmap-of-ebvs" class="nav-link" data-scroll-target="#heatmap-of-ebvs">Heatmap of EBVs</a></li>
  <li><a href="#interpretation-questions" id="toc-interpretation-questions" class="nav-link" data-scroll-target="#interpretation-questions">🧠 Interpretation Questions</a></li>
  </ul>
</li>
  <li>
<a href="#task-4-plant-breeding-trait-contributions-from-parental-lines" id="toc-task-4-plant-breeding-trait-contributions-from-parental-lines" class="nav-link" data-scroll-target="#task-4-plant-breeding-trait-contributions-from-parental-lines">Task 4: Plant Breeding – Trait Contributions from Parental Lines</a>
  <ul class="collapse">
<li><a href="#define-data-1" id="toc-define-data-1" class="nav-link" data-scroll-target="#define-data-1">Define Data</a></li>
  <li><a href="#compute-hybridtraits-vector" id="toc-compute-hybridtraits-vector" class="nav-link" data-scroll-target="#compute-hybridtraits-vector">Compute HybridTraits Vector</a></li>
  <li><a href="#multiply-with-identity-matrix-1" id="toc-multiply-with-identity-matrix-1" class="nav-link" data-scroll-target="#multiply-with-identity-matrix-1">Multiply with Identity Matrix</a></li>
  <li><a href="#remove-t3-maturation-time-and-recalculate" id="toc-remove-t3-maturation-time-and-recalculate" class="nav-link" data-scroll-target="#remove-t3-maturation-time-and-recalculate">Remove T3 (Maturation Time) and Recalculate</a></li>
  <li><a href="#heatmap-of-parenttraits" id="toc-heatmap-of-parenttraits" class="nav-link" data-scroll-target="#heatmap-of-parenttraits">Heatmap of ParentTraits</a></li>
  <li><a href="#bar-plot-of-hybrid-traits" id="toc-bar-plot-of-hybrid-traits" class="nav-link" data-scroll-target="#bar-plot-of-hybrid-traits">Bar Plot of Hybrid Traits</a></li>
  <li><a href="#interpretation-questions-1" id="toc-interpretation-questions-1" class="nav-link" data-scroll-target="#interpretation-questions-1">🧠 Interpretation Questions</a></li>
  </ul>
</li>
  <li>
<a href="#task-5-managing-matrices-and-weight-vectors-using-lists-in-r" id="toc-task-5-managing-matrices-and-weight-vectors-using-lists-in-r" class="nav-link" data-scroll-target="#task-5-managing-matrices-and-weight-vectors-using-lists-in-r">Task 5: Managing Matrices and Weight Vectors Using Lists in R</a>
  <ul class="collapse">
<li><a href="#create-a-master-list" id="toc-create-a-master-list" class="nav-link" data-scroll-target="#create-a-master-list">Create a Master List</a></li>
  <li><a href="#list-the-structure" id="toc-list-the-structure" class="nav-link" data-scroll-target="#list-the-structure">List the Structure</a></li>
  <li><a href="#indexing-elements-from-lists" id="toc-indexing-elements-from-lists" class="nav-link" data-scroll-target="#indexing-elements-from-lists">Indexing Elements from Lists</a></li>
  <li><a href="#perform-weighted-calculations" id="toc-perform-weighted-calculations" class="nav-link" data-scroll-target="#perform-weighted-calculations">Perform Weighted Calculations</a></li>
  <li><a href="#subset-and-recalculate" id="toc-subset-and-recalculate" class="nav-link" data-scroll-target="#subset-and-recalculate">Subset and Recalculate</a></li>
  <li><a href="#visualization-tasks-1" id="toc-visualization-tasks-1" class="nav-link" data-scroll-target="#visualization-tasks-1">Visualization Tasks</a></li>
  <li><a href="#interpretation-questions-2" id="toc-interpretation-questions-2" class="nav-link" data-scroll-target="#interpretation-questions-2">🧠 Interpretation Questions</a></li>
  </ul>
</li>
  </ul>
</li>
  </ul><div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="solutions.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">HW solutions</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Author</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Md Rasheduzzaman </p>
  </div>
    <div class="quarto-title-meta-contents">
        <p class="affiliation">
            <a href="https://www.fli.de/de/institute/institut-fuer-infektionsmedizin-imed/">
            FLI Insel Riems
            </a>
          </p>
        <p class="affiliation">
            <a href="https://github.com/MdRasheduz-zaman">
            &amp; Freelancer
            </a>
          </p>
      </div>
    </div>

<div class="quarto-title-meta">

      
    <div>
    <div class="quarto-title-meta-heading">Last updated</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 8, 2025</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="abstract-title">Content summary</div>
    dataframe, matrices, list, factor, vector, etc.
  </div>
</div>

</header><style>
  .content {
    text-align: justify !important;
  }
  p {
    text-align: justify !important;
  }
  .quarto-body {
    text-align: justify !important;
  }
</style>
<section id="l3-matrices-and-lists" class="level1"><h1>L3: Matrices and Lists</h1>
<section id="task-1-protein-quantification-in-biological-samples" class="level2"><h2 class="anchored" data-anchor-id="task-1-protein-quantification-in-biological-samples">Task 1: Protein Quantification in Biological Samples</h2>
<p>We measured the concentration (in µg/µL) of three proteins (P1, P2, P3) in four samples (S1–S4):</p>
<section id="matrices" class="level3"><h3 class="anchored" data-anchor-id="matrices">1. Matrices</h3>
<div class="cell">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Making Protein Matrix</span></span>
<span><span class="va">ProteinMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">3</span>, <span class="fl">2</span>,</span>
<span>    <span class="fl">7</span>, <span class="fl">6</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample1"</span>, <span class="st">"Sample2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="va">ProteinMatrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        ProteinX ProteinY ProteinZ
Sample1        5        3        2
Sample2        7        6        4</code></pre>
</div>
</div>
<p>Now goes the weight matrix</p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Making weight matrix</span></span>
<span><span class="va">WeightVector</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span>  nrow<span class="op">=</span><span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">WeightVector</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5</code></pre>
</div>
</div>
<p>Now, multiply them.</p>
<div class="cell">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Multiplying Matrices</span></span>
<span><span class="va">TotalConc</span> <span class="op">=</span> <span class="va">ProteinMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">WeightVector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">TotalConc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">TotalConc</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Total_Protein_Conc
Sample1                8.5
Sample2               15.5</code></pre>
</div>
</div>
</section><section id="making-transpose-of-proteinmatrix" class="level3"><h3 class="anchored" data-anchor-id="making-transpose-of-proteinmatrix">2. Making transpose of ProteinMatrix</h3>
<div class="cell">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ProteinMatTranspose</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span></span>
<span><span class="va">ProteinMatTranspose</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Sample1 Sample2
ProteinX       5       7
ProteinY       3       6
ProteinZ       2       4</code></pre>
</div>
</div>
</section><section id="identity-matrix" class="level3"><h3 class="anchored" data-anchor-id="identity-matrix">3. Identity Matrix</h3>
<div class="cell">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">I</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">Identitycheck</span> <span class="op">=</span> <span class="va">ProteinMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">Identitycheck</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="va">Identitycheck</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        ProteinX ProteinY ProteinZ
Sample1        5        3        2
Sample2        7        6        4</code></pre>
</div>
</div>
</section><section id="total-protein-per-sample" class="level3"><h3 class="anchored" data-anchor-id="total-protein-per-sample">4.1. Total Protein per Sample</h3>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Sample1 Sample2 
     10      17 </code></pre>
</div>
</div>
</section><section id="total-protein-per-protein-type" class="level3"><h3 class="anchored" data-anchor-id="total-protein-per-protein-type">4.2. Total Protein per Protein Type</h3>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">ProteinMatrix</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ProteinX ProteinY ProteinZ 
      12        9        6 </code></pre>
</div>
</div>
</section><section id="heatmap-of-protein-concentrations" class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-protein-concentrations">4.3. Heatmap of Protein Concentrations</h3>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">ProteinMatrix</span>, scale <span class="op">=</span> <span class="st">"none"</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation" class="level3"><h3 class="anchored" data-anchor-id="interpretation">🧠 Interpretation</h3>
<ul>
<li>Multiplying the protein levels by the weight vector shows how much each protein contributes in a sample. The result shows total protein concentration per sample.</li>
<li>The result shows that sample S2 has the highest protein burden.</li>
<li>The identity matrix represents no protein interactions or measurement biases. It is a simple matrix calculation.</li>
<li>New calculation:</li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># changing the weight of ProteinZ to 3.0</span></span>
<span><span class="va">newweightvector</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">3.0</span><span class="op">)</span>,</span>
<span>  nrow<span class="op">=</span><span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">WeightVector</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Weight"</span><span class="op">)</span></span>
<span><span class="va">newTotalconc</span> <span class="op">=</span> <span class="va">ProteinMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">newweightvector</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">newTotalconc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span><span class="va">newTotalconc</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Total_Protein_Conc
Sample1               11.5
Sample2               21.5</code></pre>
</div>
</div>
<p>Still, S2 has more protein burden.</p>
<p>Bonus:</p>
<ul>
<li>Heatmap reveals PX is most abundant across all samples.</li>
</ul>
<hr></section></section><section id="task-2-gene-to-protein-translation" class="level2"><h2 class="anchored" data-anchor-id="task-2-gene-to-protein-translation">Task 2: Gene-to-Protein Translation</h2>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># making  Gene Expression matrix</span></span>
<span><span class="va">GeneExpression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">8</span>, <span class="fl">5</span>,</span>
<span>    <span class="fl">15</span>, <span class="fl">12</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>  nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">GeneExpression</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Sample1"</span>, <span class="st">"Sample2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">GeneExpression</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span><span class="va">GeneExpression</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        GeneA GeneB GeneC
Sample1    10     8     5
Sample2    15    12    10</code></pre>
</div>
</div>
<p>Translation efficiency:</p>
<div class="cell">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># making Translation Matrix</span></span>
<span><span class="va">TranslationMatrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">0</span> , <span class="fl">0</span>,</span>
<span>  <span class="fl">0</span>, <span class="fl">1.2</span>, <span class="fl">0</span>,</span>
<span>  <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">1.8</span><span class="op">)</span>,</span>
<span>nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">TranslationMatrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span><span class="op">)</span></span>
<span><span class="va">TranslationMatrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1] [,2] [,3]
GeneA  1.5  0.0  0.0
GeneB  0.0  1.2  0.0
GeneC  0.0  0.0  1.8</code></pre>
</div>
</div>
<section id="protein-output" class="level3"><h3 class="anchored" data-anchor-id="protein-output">Protein Output</h3>
<div class="cell">
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># computing Protein matrix</span></span>
<span><span class="va">Protein_matrix</span> <span class="op">&lt;-</span> <span class="va">GeneExpression</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">TranslationMatrix</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Protein_matrix</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        [,1] [,2] [,3]
Sample1 15.0  9.6    9
Sample2 22.5 14.4   18</code></pre>
</div>
</div>
</section><section id="transpose" class="level3"><h3 class="anchored" data-anchor-id="transpose">2. Transpose</h3>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Transpose of GeneExpression matrix</span></span>
<span><span class="va">GeneExpression_Transpose</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">GeneExpression</span><span class="op">)</span></span>
<span><span class="va">GeneExpression_Transpose</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Sample1 Sample2
GeneA      10      15
GeneB       8      12
GeneC       5      10</code></pre>
</div>
</div>
<p>The new matrix represnts a matrix where the rows and columns of GeneExpression matrix have been interchanged.</p>
</section><section id="identity-matrix-and-multipliccation" class="level3"><h3 class="anchored" data-anchor-id="identity-matrix-and-multipliccation">3. Identity matrix and multipliccation</h3>
<div class="cell">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Creating Identity matrix</span></span>
<span><span class="va">I</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">I</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3]
[1,]    1    0    0
[2,]    0    1    0
[3,]    0    0    1</code></pre>
</div>
</div>
<p>Now, multiply:</p>
<div class="cell">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">Product_matrix</span> <span class="op">=</span> <span class="va">TranslationMatrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span>  <span class="va">I</span> </span>
<span><span class="va">Product_matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1] [,2] [,3]
GeneA  1.5  0.0  0.0
GeneB  0.0  1.2  0.0
GeneC  0.0  0.0  1.8</code></pre>
</div>
</div>
<p>The product is identical to TranslationMatrix</p>
</section><section id="sub-matrix" class="level3"><h3 class="anchored" data-anchor-id="sub-matrix">4. Sub-matrix:</h3>
<div class="cell">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># making submatrix A</span></span>
<span><span class="va">A</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">8</span>,</span>
<span>    <span class="fl">15</span>, <span class="fl">12</span><span class="op">)</span>, nrow<span class="op">=</span><span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sample1"</span>, <span class="st">"sample2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GeneA"</span>, <span class="st">"GeneB"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">A</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        GeneA GeneB
sample1    10     8
sample2    15    12</code></pre>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># finding inverse of A</span></span>
<span><span class="co">#inv_A &lt;- solve(A)</span></span>
<span><span class="co">#inv_A</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The inverse matrix could not be calculated since A is a singular matrix. So, A * A^-1 is also not possible.</p>
</section><section id="visualization-tasks" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks">Visualization tasks:</h3>
<ul>
<li><ol type="1">
<li>MARplot</li>
</ol></li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># generating MARplot-style scatter plot</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">GeneExpression</span>, <span class="va">Protein_matrix</span>, type<span class="op">=</span><span class="st">"p"</span>, main<span class="op">=</span><span class="st">"Protein level vs. Gene Expression level"</span><span class="op">)</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="st">"Sample-Gene"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html">text</a></span><span class="op">(</span><span class="va">GeneExpression</span>, <span class="va">Protein_matrix</span>, labels <span class="op">=</span> <span class="va">labels</span>, pos<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="1650"></p>
</div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># generating a heatmap</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">Protein_matrix</span>, main<span class="op">=</span> <span class="st">"Heatmap of Protein Level"</span>, Rowv <span class="op">=</span> <span class="cn">TRUE</span>, Colv <span class="op">=</span> <span class="cn">TRUE</span>, labRow<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Protein_matrix</span><span class="op">)</span>, labCol<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ProteinA"</span>, <span class="st">"ProteinB"</span>, <span class="st">"ProteinC"</span><span class="op">)</span>, col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">topo.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span> <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid" width="1650"></p>
</div>
</div>
<ul>
<li><ol start="2" type="1">
<li>Heatmap of Expression:</li>
</ol></li>
</ul>
<div class="cell">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span><span class="va">GeneExpression</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">terrain.colors</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>, scale <span class="op">=</span> <span class="st">"column"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation-1" class="level3"><h3 class="anchored" data-anchor-id="interpretation-1">🧠 Interpretation</h3>
<ol type="1">
<li><p>Matrix multiplication allows each gene in both samples to be multiplied to their respective translation efficiency. So, the product shows how successfully each gene is translated”)</p></li>
<li><p>The diagonal TranslationMatrix make sense biologically because they show translation efficiency of each gene and there is no other interaction between them. Although there could be interaction in real-world scenarios.</p></li>
<li><p>If Sample2 has higher protein levels even with similar gene expression, it means that more mRNAs are translated to proteins compared to Sample1”</p></li>
<li><p>The upward trend in MARplot may indicate an increase in translation efficacy and downward trend may indicate a decline in translation efficacy”</p></li>
<li><p>Clustering in the heatmap may suggest which samples are most similar to each other based on their prot.</p></li>
</ol>
<hr></section></section><section id="task-3-animal-breeding-bull-ranking-by-economic-traits" class="level2"><h2 class="anchored" data-anchor-id="task-3-animal-breeding-bull-ranking-by-economic-traits">Task 3: Animal Breeding – Bull Ranking by Economic Traits</h2>
<section id="define-data" class="level3"><h3 class="anchored" data-anchor-id="define-data">Define Data</h3>
<div class="cell">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define Bull EBVs</span></span>
<span><span class="va">BullEBVs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="fl">400</span>, <span class="fl">1.2</span>, <span class="fl">0.8</span>,</span>
<span>  <span class="fl">500</span>, <span class="fl">1.5</span>, <span class="fl">0.6</span></span>
<span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Bull1"</span>, <span class="st">"Bull2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Milk_yield"</span>, <span class="st">"Growth_rate"</span>, <span class="st">"Fertility"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define Economic Weights</span></span>
<span><span class="va">EconomicWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.002</span>, <span class="fl">50</span>, <span class="fl">100</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="compute-total-economic-value" class="level3"><h3 class="anchored" data-anchor-id="compute-total-economic-value">Compute Total Economic Value</h3>
<div class="cell">
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">TotalValue</span> <span class="op">&lt;-</span> <span class="va">BullEBVs</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">EconomicWeights</span></span>
<span><span class="va">TotalValue</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       [,1]
Bull1 140.8
Bull2 136.0</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Bull1: (400 × 0.002) + (1.2 × 50) + (0.8 × 100) = 140.8</p>
<p>Bull2: (500 × 0.002) + (1.5 × 50) + (0.6 × 100) = 136.0</p>
<p>Bull1 is more valuable economically.</p>
<p><strong>Biological Interpretation</strong></p>
<p>Economic weights convert genetic merit (EBVs) into economic merit. Traits with higher financial importance have a larger impact, regardless of absolute EBV values.</p>
</section><section id="multiply-with-identity-matrix" class="level3"><h3 class="anchored" data-anchor-id="multiply-with-identity-matrix">Multiply with Identity Matrix</h3>
<div class="cell">
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">I3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">BullEBVs_identity</span> <span class="op">&lt;-</span> <span class="va">BullEBVs</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I3</span></span>
<span><span class="va">BullEBVs_identity</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1] [,2] [,3]
Bull1  400  1.2  0.8
Bull2  500  1.5  0.6</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Multiplying by identity matrix returns the original matrix. It confirms that EBV structure is preserved.</p>
</section><section id="remove-milk-yield-and-recalculate-total-value" class="level3"><h3 class="anchored" data-anchor-id="remove-milk-yield-and-recalculate-total-value">Remove Milk Yield and Recalculate Total Value</h3>
<div class="cell">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">BullEBVs_noMilk</span> <span class="op">&lt;-</span> <span class="va">BullEBVs</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">EconomicWeights_noMilk</span> <span class="op">&lt;-</span> <span class="va">EconomicWeights</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span>, , drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span></span>
<span><span class="va">TotalValue_noMilk</span> <span class="op">&lt;-</span> <span class="va">BullEBVs_noMilk</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">EconomicWeights_noMilk</span></span>
<span><span class="va">TotalValue_noMilk</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1]
Bull1  140
Bull2  135</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Bull1: (1.2 × 50) + (0.8 × 100) = 140</p>
<p>Bull2: (1.5 × 50) + (0.6 × 100) = 135</p>
<p>Bull1 still ranks higher, but by a smaller margin.</p>
</section><section id="bar-plot-total-economic-value" class="level3"><h3 class="anchored" data-anchor-id="bar-plot-total-economic-value">Bar Plot: Total Economic Value</h3>
<div class="cell">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">TotalValue</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  names.arg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">BullEBVs</span><span class="op">)</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"skyblue"</span>, <span class="st">"orange"</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Total Economic Value of Bulls"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Total Value"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="heatmap-of-ebvs" class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-ebvs">Heatmap of EBVs</h3>
<div class="cell">
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">BullEBVs</span>,</span>
<span>  Rowv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  Colv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Heatmap of Bull EBVs"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation-questions" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions">🧠 Interpretation Questions</h3>
<ul>
<li>How do economic weights affect trait importance?</li>
</ul>
<p>Traits with higher weights contribute more to the total economic value. This makes them more influential in ranking and selection.</p>
<ul>
<li>Why might you ignore milk yield?</li>
</ul>
<p>Milk yield may be excluded in systems focusing on fertility, growth, or when it is no longer a limiting factor. Environmental or economic contexts may also shift trait priorities.</p>
<ul>
<li>What is the value of heatmaps?</li>
</ul>
<p>Heatmaps visually compare EBVs across bulls and traits.They help detect patterns, outliers, and clusters easily in multivariate data.</p>
<ul>
<li>Can this method be extended to more bulls and traits?</li>
</ul>
<p>Yes. This method scales to any number of bulls or traits. Just ensure the EBVs matrix and economic weights are dimensionally compatible.</p>
</section></section><section id="task-4-plant-breeding-trait-contributions-from-parental-lines" class="level2"><h2 class="anchored" data-anchor-id="task-4-plant-breeding-trait-contributions-from-parental-lines">Task 4: Plant Breeding – Trait Contributions from Parental Lines</h2>
<section id="define-data-1" class="level3"><h3 class="anchored" data-anchor-id="define-data-1">Define Data</h3>
<p><strong>Parent trait values (normalized 1–10)</strong></p>
<div class="cell">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ParentTraits</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>  <span class="fl">7</span>, <span class="fl">5</span>, <span class="fl">3</span>,</span>
<span>  <span class="fl">6</span>, <span class="fl">8</span>, <span class="fl">4</span>,</span>
<span>  <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">6</span></span>
<span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"P1"</span>, <span class="st">"P2"</span>, <span class="st">"P3"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Drought_resistance"</span>, <span class="st">"Yield"</span>, <span class="st">"Maturation_time"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define Hybrid Weights</span></span>
<span><span class="va">HybridWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="compute-hybridtraits-vector" class="level3"><h3 class="anchored" data-anchor-id="compute-hybridtraits-vector">Compute HybridTraits Vector</h3>
<div class="cell">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">HybridTraits</span> <span class="op">&lt;-</span> <span class="va">HybridWeights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">ParentTraits</span></span>
<span><span class="va">HybridTraits</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Drought_resistance Yield Maturation_time
[1,]                6.3   6.1             3.9</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>HybridTraits = (0.5 × P1) + (0.3 × P2) + (0.2 × P3)</p>
<p>Drought_resistance = (0.5 × 7) + (0.3 × 6) + (0.2 × 5) = 6.2</p>
<p>Yield = (0.5 × 5) + (0.3 × 8) + (0.2 × 6) = 6.1</p>
<p>Maturation_time = (0.5 × 3) + (0.3 × 4) + (0.2 × 6) = 3.7</p>
<p>The hybrid is moderately strong in drought resistance and yield, but has a relatively shorter maturation time.</p>
<p><strong>Biological Meaning of Unequal Contribution</strong></p>
<p>When one parent contributes more to a trait, it suggests that the trait’s heritable strength comes disproportionately from that parent. Breeders can use this knowledge to amplify desirable traits.</p>
</section><section id="multiply-with-identity-matrix-1" class="level3"><h3 class="anchored" data-anchor-id="multiply-with-identity-matrix-1">Multiply with Identity Matrix</h3>
<div class="cell">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">I3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">ParentTraits_identity</span> <span class="op">&lt;-</span> <span class="va">ParentTraits</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">I3</span></span>
<span><span class="va">ParentTraits_identity</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   [,1] [,2] [,3]
P1    7    5    3
P2    6    8    4
P3    5    6    6</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Multiplying by identity matrix returns the original matrix. This operation verifies structural consistency and dimensionality.</p>
</section><section id="remove-t3-maturation-time-and-recalculate" class="level3"><h3 class="anchored" data-anchor-id="remove-t3-maturation-time-and-recalculate">Remove T3 (Maturation Time) and Recalculate</h3>
<div class="cell">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ParentTraits_T1T2</span> <span class="op">&lt;-</span> <span class="va">ParentTraits</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">HybridTraits_T1T2</span> <span class="op">&lt;-</span> <span class="va">HybridWeights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">ParentTraits_T1T2</span></span>
<span><span class="va">HybridTraits_T1T2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Drought_resistance Yield
[1,]                6.3   6.1</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Drought_resistance = (0.5 × 7) + (0.3 × 6) + (0.2 × 5) = 6.2</p>
<p>Yield = (0.5 × 5) + (0.3 × 8) + (0.2 × 6) = 6.1</p>
<p>Removing a trait (T3) changes the trait profile. Hybrid selection may now favor traits that remain.</p>
</section><section id="heatmap-of-parenttraits" class="level3"><h3 class="anchored" data-anchor-id="heatmap-of-parenttraits">Heatmap of ParentTraits</h3>
<div class="cell">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">ParentTraits</span>,</span>
<span>  Rowv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  Colv <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Heatmap of Parent Traits"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="bar-plot-of-hybrid-traits" class="level3"><h3 class="anchored" data-anchor-id="bar-plot-of-hybrid-traits">Bar Plot of Hybrid Traits</h3>
<div class="cell">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">HybridTraits</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  names.arg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">ParentTraits</span><span class="op">)</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#66c2a5"</span>, <span class="st">"#fc8d62"</span>, <span class="st">"#8da0cb"</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Hybrid Trait Profile"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Trait Value"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="interpretation-questions-1" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-1">🧠 Interpretation Questions</h3>
<ul>
<li>
<p>How does the weighting of parents affect the hybrid’s performance?</p>
<p>Stronger weights mean more genetic contribution. Traits from highly weighted parents dominate the hybrid profile.</p>
</li>
<li>
<p>What does the identity matrix represent here?</p>
<p>It represents a neutral transformation. It confirms data integrity when used in matrix multiplication.</p>
</li>
<li>
<p>If you used equal weights (⅓ for each), how would the hybrid traits change?</p>
<p>Traits would reflect an even mix, potentially leading to balanced but less specialized performance.</p>
</li>
<li>
<p>What real-world limitations does this simplified model ignore?</p>
<ol type="i">
<li><p>Non-additive genetic effects (dominance, epistasis)</p></li>
<li><p>Environmental interactions</p></li>
<li><p>Trait heritability and correlations</p></li>
<li><p>Breeding feasibility and cost</p></li>
</ol>
</li>
</ul></section></section><section id="task-5-managing-matrices-and-weight-vectors-using-lists-in-r" class="level2"><h2 class="anchored" data-anchor-id="task-5-managing-matrices-and-weight-vectors-using-lists-in-r">Task 5: Managing Matrices and Weight Vectors Using Lists in R</h2>
<p>Now that we’ve explored trait-based decisions using matrices, it’s time to organize our work using R’s list structure. Lists help bundle related objects like matrices and weight vectors, keeping the analysis modular and scalable.</p>
<section id="create-a-master-list" class="level3"><h3 class="anchored" data-anchor-id="create-a-master-list">Create a Master List</h3>
<div class="cell">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Assuming previous matrices and weights are already defined:</span></span>
<span></span>
<span><span class="co"># Making a MasterList</span></span>
<span><span class="va">bioList</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  ProteinConc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">ProteinMatrix</span>, weights <span class="op">=</span> <span class="va">WeightVector</span><span class="op">)</span>,</span>
<span>  ProteinMap  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">GeneExpression</span>, weights <span class="op">=</span> <span class="va">TranslationMatrix</span><span class="op">)</span>,</span>
<span>  Animal <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">BullEBVs</span>, weights <span class="op">=</span> <span class="va">EconomicWeights</span><span class="op">)</span>,</span>
<span>  Plant <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>matrix <span class="op">=</span> <span class="va">ParentTraits</span>, weights <span class="op">=</span> <span class="va">HybridWeights</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">bioList</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$ProteinConc
$ProteinConc$matrix
        ProteinX ProteinY ProteinZ
Sample1        5        3        2
Sample2        7        6        4

$ProteinConc$weights
         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5


$ProteinMap
$ProteinMap$matrix
        GeneA GeneB GeneC
Sample1    10     8     5
Sample2    15    12    10

$ProteinMap$weights
      [,1] [,2] [,3]
GeneA  1.5  0.0  0.0
GeneB  0.0  1.2  0.0
GeneC  0.0  0.0  1.8


$Animal
$Animal$matrix
      Milk_yield Growth_rate Fertility
Bull1        400         1.2       0.8
Bull2        500         1.5       0.6

$Animal$weights
      [,1]
[1,] 2e-03
[2,] 5e+01
[3,] 1e+02


$Plant
$Plant$matrix
   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6

$Plant$weights
     [,1] [,2] [,3]
[1,]  0.5  0.3  0.2</code></pre>
</div>
</div>
</section><section id="list-the-structure" class="level3"><h3 class="anchored" data-anchor-id="list-the-structure">List the Structure</h3>
<div class="cell">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">bioList</span><span class="op">)</span>               <span class="co"># Top-level list names</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ProteinConc" "ProteinMap"  "Animal"      "Plant"      </code></pre>
</div>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span><span class="op">(</span><span class="va">bioList</span><span class="op">)</span>             <span class="co"># Number of components in each sublist</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ProteinConc  ProteinMap      Animal       Plant 
          2           2           2           2 </code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Each top-level entry (e.g., ProteinConc, Plant) contains two components:</p>
<ul>
<li>A matrix (e.g., ProteinMatrix)</li>
<li>A corresponding weight vector or matrix</li>
</ul></section><section id="indexing-elements-from-lists" class="level3"><h3 class="anchored" data-anchor-id="indexing-elements-from-lists">Indexing Elements from Lists</h3>
<div class="cell">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Access the trait matrix for Plant</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#or</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Drought_resistance Yield Maturation_time
P1                  7     5               3
P2                  6     8               4
P3                  5     6               6</code></pre>
</div>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Access the weight vector for ProteinConc</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5</code></pre>
</div>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#or</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Weight
ProteinX    0.5
ProteinY    1.0
ProteinZ    1.5</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Use double brackets [[ ]] to extract unnamed list elements by position. But we named our list, so they are easily extractable using the <code>$</code> notation.</p>
</section><section id="perform-weighted-calculations" class="level3"><h3 class="anchored" data-anchor-id="perform-weighted-calculations">Perform Weighted Calculations</h3>
<div class="cell">
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Protein concentration score</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">ProteinConc</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Weight
Sample1    8.5
Sample2   15.5</code></pre>
</div>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Gene → Protein contribution</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">ProteinMap</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">ProteinMap</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        [,1] [,2] [,3]
Sample1 15.0  9.6    9
Sample2 22.5 14.4   18</code></pre>
</div>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Bull economic value</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">weights</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       [,1]
Bull1 140.8
Bull2 136.0</code></pre>
</div>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Hybrid trait value</span></span>
<span><span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">weights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Drought_resistance Yield Maturation_time
[1,]                6.3   6.1             3.9</code></pre>
</div>
</div>
</section><section id="subset-and-recalculate" class="level3"><h3 class="anchored" data-anchor-id="subset-and-recalculate">Subset and Recalculate</h3>
<div class="cell">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Remove last trait from ParentTraits</span></span>
<span><span class="va">ParentSubset</span> <span class="op">&lt;-</span> <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="va">NewWeights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.4</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Recalculated hybrid score</span></span>
<span><span class="va">SubsetHybridScore</span> <span class="op">&lt;-</span> <span class="va">ParentSubset</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">NewWeights</span></span>
<span><span class="va">SubsetHybridScore</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   [,1]
P1  6.2
P2  6.8
P3  5.4</code></pre>
</div>
</div>
<p><strong>Interpretation</strong></p>
<p>Dropping a trait and reweighting highlights its influence in trait aggregation and selection.</p>
</section><section id="visualization-tasks-1" class="level3"><h3 class="anchored" data-anchor-id="visualization-tasks-1">Visualization Tasks</h3>
<section id="heatmap-gene-expression" class="level4"><h4 class="anchored" data-anchor-id="heatmap-gene-expression">Heatmap: Gene Expression</h4>
<div class="cell">
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html">heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">bioList</span><span class="op">$</span><span class="va">ProteinMap</span><span class="op">$</span><span class="va">matrix</span>,</span>
<span>  scale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html">heat.colors</a></span><span class="op">(</span><span class="fl">256</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Gene Expression Heatmap"</span>,</span>
<span>  xlab <span class="op">=</span> <span class="st">"Proteins"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Genes"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="bar-plots" class="level4"><h4 class="anchored" data-anchor-id="bar-plots">Bar Plots</h4>
<section id="hybrid-traits" class="level5"><h5 class="anchored" data-anchor-id="hybrid-traits">Hybrid traits</h5>
<div class="cell">
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">weights</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Plant</span><span class="op">$</span><span class="va">matrix</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Hybrid Trait Contributions"</span>,</span>
<span>  col <span class="op">=</span> <span class="st">"#66c2a5"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Score"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section><section id="bull-ebv-economic-breeding-values" class="level5"><h5 class="anchored" data-anchor-id="bull-ebv-economic-breeding-values">Bull EBV (Economic Breeding Values)</h5>
<div class="cell">
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span></span>
<span>  <span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">matrix</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">bioList</span><span class="op">$</span><span class="va">Animal</span><span class="op">$</span><span class="va">weights</span>,</span>
<span>  beside <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  main <span class="op">=</span> <span class="st">"Bull EBVs (Economic Values)"</span>,</span>
<span>  col <span class="op">=</span> <span class="st">"#fc8d62"</span>,</span>
<span>  ylab <span class="op">=</span> <span class="st">"Score"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="solutions_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="1650"></p>
</div>
</div>
</section></section></section><section id="interpretation-questions-2" class="level3"><h3 class="anchored" data-anchor-id="interpretation-questions-2">🧠 Interpretation Questions</h3>
<ul>
<li>Why use a list structure?</li>
</ul>
<p>Keeps each dataset and its weights together. Facilitates automated workflows and reuse.</p>
<ul>
<li>What’s tricky about <code>[[ ]]</code> access?</li>
</ul>
<p>You must remember the order (<code>[[1]]</code> = <code>matrix</code>, <code>[[2]]</code> = <code>weights</code>). No names means you can’t use $matrix, only positional access.</p>
<p>Loop across all list entries</p>
<p>Weighted scores for all entries <code>lapply(bioList, function(x) x[[2]] %*% x[[1]])</code></p>
<ul>
<li>How does this help in large-scale pipelines?</li>
</ul>
<p>You can use this format with <code><a href="https://rdrr.io/r/base/lapply.html">lapply()</a></code>, <code><a href="https://purrr.tidyverse.org/reference/map.html">purrr::map()</a></code>, or in targets pipelines for reproducibility and modular processing.</p>


<!-- -->

</section></section></section><div id="quarto-appendix" class="default"><section class="quarto-appendix-contents"><h2 class="anchored quarto-appendix-heading">Citation</h2><div><div class="quarto-appendix-secondary-label">BibTeX citation:</div><pre class="sourceCode code-with-copy quarto-appendix-bibtex"><code class="sourceCode bibtex">@online{rasheduzzaman2025,
  author = {Md Rasheduzzaman},
  title = {HW Solutions},
  date = {2025-05-08},
  langid = {en},
  abstract = {dataframe, matrices, list, factor, vector, etc.}
}
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre><div class="quarto-appendix-secondary-label">For attribution, please cite this work as:</div><div id="ref-rasheduzzaman2025" class="csl-entry quarto-appendix-citeas" role="listitem">
Md Rasheduzzaman. 2025. <span>“HW Solutions.”</span> May 8, 2025.
</div></div></section></div></main><!-- /main --><hr>
<div style="margin-top: 2em;">
  <p style="font-weight: 500; font-size: 1.1em;">
    💬 Have thoughts or questions? Join the discussion below using your GitHub account!
  </p>
  <p style="font-size: 0.95em; color: #666;">
    You can <strong>edit or delete</strong> your own comments. Reactions like 👍 ❤️ 🚀 are also supported.
  </p>
</div>

<div id="giscus_thread" style="margin-top: 1.5em;"></div>

<script src="https://giscus.app/client.js" data-repo="MdRasheduz-zaman/PythRaSh" data-repo-id="R_kgDOOZN9Xw" data-category-id="DIC_kwDOOZN9X84CpE6N" data-mapping="pathname" data-strict="0" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="preferred_color_scheme" data-lang="en" data-loading="lazy" crossorigin="anonymous" async="">
</script><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb81" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "HW solutions"</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="an">abstract:</span><span class="co"> "dataframe, matrices, list, factor, vector, etc."</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;style&gt;</span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">.content</span> {</span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>    <span class="kw">text-align</span>: <span class="dv">justify</span> <span class="at">!important</span><span class="op">;</span></span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a>  p {</span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a>    <span class="kw">text-align</span>: <span class="dv">justify</span> <span class="at">!important</span><span class="op">;</span></span>
<span id="cb81-12"><a href="#cb81-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb81-13"><a href="#cb81-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">.quarto-body</span> {</span>
<span id="cb81-14"><a href="#cb81-14" aria-hidden="true" tabindex="-1"></a>    <span class="kw">text-align</span>: <span class="dv">justify</span> <span class="at">!important</span><span class="op">;</span></span>
<span id="cb81-15"><a href="#cb81-15" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb81-16"><a href="#cb81-16" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/style&gt;</span></span>
<span id="cb81-17"><a href="#cb81-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-18"><a href="#cb81-18" aria-hidden="true" tabindex="-1"></a><span class="fu"># L3: Matrices and Lists</span></span>
<span id="cb81-19"><a href="#cb81-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-20"><a href="#cb81-20" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 1: Protein Quantification in Biological Samples</span></span>
<span id="cb81-21"><a href="#cb81-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-22"><a href="#cb81-22" aria-hidden="true" tabindex="-1"></a>We measured the concentration (in µg/µL) of three proteins (P1, P2, P3) in four samples (S1–S4):</span>
<span id="cb81-23"><a href="#cb81-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-24"><a href="#cb81-24" aria-hidden="true" tabindex="-1"></a><span class="fu">### 1. Matrices</span></span>
<span id="cb81-25"><a href="#cb81-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-28"><a href="#cb81-28" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-29"><a href="#cb81-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Making Protein Matrix</span></span>
<span id="cb81-30"><a href="#cb81-30" aria-hidden="true" tabindex="-1"></a>ProteinMatrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb81-31"><a href="#cb81-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">2</span>,</span>
<span id="cb81-32"><a href="#cb81-32" aria-hidden="true" tabindex="-1"></a>    <span class="dv">7</span>, <span class="dv">6</span>, <span class="dv">4</span>),</span>
<span id="cb81-33"><a href="#cb81-33" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb81-34"><a href="#cb81-34" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-35"><a href="#cb81-35" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ProteinMatrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Sample1"</span>, <span class="st">"Sample2"</span>)</span>
<span id="cb81-36"><a href="#cb81-36" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ProteinMatrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb81-37"><a href="#cb81-37" aria-hidden="true" tabindex="-1"></a>ProteinMatrix</span>
<span id="cb81-38"><a href="#cb81-38" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-39"><a href="#cb81-39" aria-hidden="true" tabindex="-1"></a>Now goes the weight matrix</span>
<span id="cb81-40"><a href="#cb81-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-43"><a href="#cb81-43" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-44"><a href="#cb81-44" aria-hidden="true" tabindex="-1"></a><span class="co"># Making weight matrix</span></span>
<span id="cb81-45"><a href="#cb81-45" aria-hidden="true" tabindex="-1"></a>WeightVector <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb81-46"><a href="#cb81-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">1.5</span>),</span>
<span id="cb81-47"><a href="#cb81-47" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow=</span><span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb81-48"><a href="#cb81-48" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-49"><a href="#cb81-49" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb81-50"><a href="#cb81-50" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb81-51"><a href="#cb81-51" aria-hidden="true" tabindex="-1"></a>WeightVector</span>
<span id="cb81-52"><a href="#cb81-52" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-53"><a href="#cb81-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-54"><a href="#cb81-54" aria-hidden="true" tabindex="-1"></a>Now, multiply them.</span>
<span id="cb81-55"><a href="#cb81-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-58"><a href="#cb81-58" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-59"><a href="#cb81-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Multiplying Matrices</span></span>
<span id="cb81-60"><a href="#cb81-60" aria-hidden="true" tabindex="-1"></a>TotalConc <span class="ot">=</span> ProteinMatrix <span class="sc">%*%</span> WeightVector</span>
<span id="cb81-61"><a href="#cb81-61" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(TotalConc) <span class="ot">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span id="cb81-62"><a href="#cb81-62" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(TotalConc)</span>
<span id="cb81-63"><a href="#cb81-63" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-64"><a href="#cb81-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-65"><a href="#cb81-65" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2. Making transpose of ProteinMatrix</span></span>
<span id="cb81-66"><a href="#cb81-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-69"><a href="#cb81-69" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-70"><a href="#cb81-70" aria-hidden="true" tabindex="-1"></a>ProteinMatTranspose <span class="ot">=</span> <span class="fu">t</span>(ProteinMatrix)</span>
<span id="cb81-71"><a href="#cb81-71" aria-hidden="true" tabindex="-1"></a>ProteinMatTranspose</span>
<span id="cb81-72"><a href="#cb81-72" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-73"><a href="#cb81-73" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3. Identity Matrix</span></span>
<span id="cb81-74"><a href="#cb81-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-77"><a href="#cb81-77" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-78"><a href="#cb81-78" aria-hidden="true" tabindex="-1"></a>I <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb81-79"><a href="#cb81-79" aria-hidden="true" tabindex="-1"></a>Identitycheck <span class="ot">=</span> ProteinMatrix <span class="sc">%*%</span> I</span>
<span id="cb81-80"><a href="#cb81-80" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Identitycheck) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb81-81"><a href="#cb81-81" aria-hidden="true" tabindex="-1"></a>Identitycheck</span>
<span id="cb81-82"><a href="#cb81-82" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-83"><a href="#cb81-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-84"><a href="#cb81-84" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4.1. Total Protein per Sample</span></span>
<span id="cb81-85"><a href="#cb81-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-88"><a href="#cb81-88" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-89"><a href="#cb81-89" aria-hidden="true" tabindex="-1"></a><span class="fu">rowSums</span>(ProteinMatrix)</span>
<span id="cb81-90"><a href="#cb81-90" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-91"><a href="#cb81-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-92"><a href="#cb81-92" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4.2. Total Protein per Protein Type</span></span>
<span id="cb81-93"><a href="#cb81-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-96"><a href="#cb81-96" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-97"><a href="#cb81-97" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(ProteinMatrix)</span>
<span id="cb81-98"><a href="#cb81-98" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-99"><a href="#cb81-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-100"><a href="#cb81-100" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4.3. Heatmap of Protein Concentrations</span></span>
<span id="cb81-101"><a href="#cb81-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-104"><a href="#cb81-104" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-105"><a href="#cb81-105" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(ProteinMatrix, <span class="at">scale =</span> <span class="st">"none"</span>, <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">10</span>))</span>
<span id="cb81-106"><a href="#cb81-106" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-107"><a href="#cb81-107" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-108"><a href="#cb81-108" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation</span></span>
<span id="cb81-109"><a href="#cb81-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-110"><a href="#cb81-110" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Multiplying the protein levels by the weight vector shows how much each protein contributes in a sample. The result shows total protein concentration per sample.</span>
<span id="cb81-111"><a href="#cb81-111" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The result shows that sample S2 has the highest protein burden.</span>
<span id="cb81-112"><a href="#cb81-112" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The identity matrix represents no protein interactions or measurement biases. It is a simple matrix calculation.</span>
<span id="cb81-113"><a href="#cb81-113" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>New calculation:</span>
<span id="cb81-114"><a href="#cb81-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-117"><a href="#cb81-117" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-118"><a href="#cb81-118" aria-hidden="true" tabindex="-1"></a><span class="co"># changing the weight of ProteinZ to 3.0</span></span>
<span id="cb81-119"><a href="#cb81-119" aria-hidden="true" tabindex="-1"></a>newweightvector <span class="ot">=</span> <span class="fu">matrix</span>(</span>
<span id="cb81-120"><a href="#cb81-120" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">3.0</span>),</span>
<span id="cb81-121"><a href="#cb81-121" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow=</span><span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb81-122"><a href="#cb81-122" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-123"><a href="#cb81-123" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"ProteinX"</span>, <span class="st">"ProteinY"</span>, <span class="st">"ProteinZ"</span>)</span>
<span id="cb81-124"><a href="#cb81-124" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(WeightVector) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Weight"</span>)</span>
<span id="cb81-125"><a href="#cb81-125" aria-hidden="true" tabindex="-1"></a>newTotalconc <span class="ot">=</span> ProteinMatrix <span class="sc">%*%</span> newweightvector</span>
<span id="cb81-126"><a href="#cb81-126" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(newTotalconc) <span class="ot">&lt;-</span> <span class="st">"Total_Protein_Conc"</span></span>
<span id="cb81-127"><a href="#cb81-127" aria-hidden="true" tabindex="-1"></a>newTotalconc</span>
<span id="cb81-128"><a href="#cb81-128" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-129"><a href="#cb81-129" aria-hidden="true" tabindex="-1"></a>Still, S2 has more protein burden.</span>
<span id="cb81-130"><a href="#cb81-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-131"><a href="#cb81-131" aria-hidden="true" tabindex="-1"></a>Bonus:</span>
<span id="cb81-132"><a href="#cb81-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-133"><a href="#cb81-133" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Heatmap reveals PX is most abundant across all samples.</span>
<span id="cb81-134"><a href="#cb81-134" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-135"><a href="#cb81-135" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb81-136"><a href="#cb81-136" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-137"><a href="#cb81-137" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 2: Gene-to-Protein Translation</span></span>
<span id="cb81-138"><a href="#cb81-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-141"><a href="#cb81-141" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-142"><a href="#cb81-142" aria-hidden="true" tabindex="-1"></a><span class="co"># making  Gene Expression matrix</span></span>
<span id="cb81-143"><a href="#cb81-143" aria-hidden="true" tabindex="-1"></a>GeneExpression <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb81-144"><a href="#cb81-144" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">8</span>, <span class="dv">5</span>,</span>
<span id="cb81-145"><a href="#cb81-145" aria-hidden="true" tabindex="-1"></a>    <span class="dv">15</span>, <span class="dv">12</span>, <span class="dv">10</span>),</span>
<span id="cb81-146"><a href="#cb81-146" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb81-147"><a href="#cb81-147" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-148"><a href="#cb81-148" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(GeneExpression) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Sample1"</span>, <span class="st">"Sample2"</span>)</span>
<span id="cb81-149"><a href="#cb81-149" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(GeneExpression) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb81-150"><a href="#cb81-150" aria-hidden="true" tabindex="-1"></a>GeneExpression</span>
<span id="cb81-151"><a href="#cb81-151" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-152"><a href="#cb81-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-153"><a href="#cb81-153" aria-hidden="true" tabindex="-1"></a>Translation efficiency:</span>
<span id="cb81-154"><a href="#cb81-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-157"><a href="#cb81-157" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-158"><a href="#cb81-158" aria-hidden="true" tabindex="-1"></a><span class="co"># making Translation Matrix</span></span>
<span id="cb81-159"><a href="#cb81-159" aria-hidden="true" tabindex="-1"></a>TranslationMatrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb81-160"><a href="#cb81-160" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="dv">0</span> , <span class="dv">0</span>,</span>
<span id="cb81-161"><a href="#cb81-161" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span>, <span class="fl">1.2</span>, <span class="dv">0</span>,</span>
<span id="cb81-162"><a href="#cb81-162" aria-hidden="true" tabindex="-1"></a>  <span class="dv">0</span>, <span class="dv">0</span>, <span class="fl">1.8</span>),</span>
<span id="cb81-163"><a href="#cb81-163" aria-hidden="true" tabindex="-1"></a><span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb81-164"><a href="#cb81-164" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-165"><a href="#cb81-165" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-166"><a href="#cb81-166" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(TranslationMatrix) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>, <span class="st">"GeneC"</span>)</span>
<span id="cb81-167"><a href="#cb81-167" aria-hidden="true" tabindex="-1"></a>TranslationMatrix</span>
<span id="cb81-168"><a href="#cb81-168" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-169"><a href="#cb81-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-170"><a href="#cb81-170" aria-hidden="true" tabindex="-1"></a><span class="fu">### Protein Output</span></span>
<span id="cb81-171"><a href="#cb81-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-174"><a href="#cb81-174" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-175"><a href="#cb81-175" aria-hidden="true" tabindex="-1"></a><span class="co"># computing Protein matrix</span></span>
<span id="cb81-176"><a href="#cb81-176" aria-hidden="true" tabindex="-1"></a>Protein_matrix <span class="ot">&lt;-</span> GeneExpression <span class="sc">%*%</span> TranslationMatrix</span>
<span id="cb81-177"><a href="#cb81-177" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(Protein_matrix)</span>
<span id="cb81-178"><a href="#cb81-178" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-179"><a href="#cb81-179" aria-hidden="true" tabindex="-1"></a><span class="fu">### 2. Transpose</span></span>
<span id="cb81-182"><a href="#cb81-182" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-183"><a href="#cb81-183" aria-hidden="true" tabindex="-1"></a><span class="co"># Transpose of GeneExpression matrix</span></span>
<span id="cb81-184"><a href="#cb81-184" aria-hidden="true" tabindex="-1"></a>GeneExpression_Transpose <span class="ot">&lt;-</span> <span class="fu">t</span>(GeneExpression)</span>
<span id="cb81-185"><a href="#cb81-185" aria-hidden="true" tabindex="-1"></a>GeneExpression_Transpose</span>
<span id="cb81-186"><a href="#cb81-186" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-187"><a href="#cb81-187" aria-hidden="true" tabindex="-1"></a>The new matrix represnts a matrix where the rows and columns of GeneExpression matrix have been interchanged.</span>
<span id="cb81-188"><a href="#cb81-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-189"><a href="#cb81-189" aria-hidden="true" tabindex="-1"></a><span class="fu">### 3. Identity matrix and multipliccation</span></span>
<span id="cb81-192"><a href="#cb81-192" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-193"><a href="#cb81-193" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating Identity matrix</span></span>
<span id="cb81-194"><a href="#cb81-194" aria-hidden="true" tabindex="-1"></a>I <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb81-195"><a href="#cb81-195" aria-hidden="true" tabindex="-1"></a>I</span>
<span id="cb81-196"><a href="#cb81-196" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-197"><a href="#cb81-197" aria-hidden="true" tabindex="-1"></a>Now, multiply:</span>
<span id="cb81-200"><a href="#cb81-200" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-201"><a href="#cb81-201" aria-hidden="true" tabindex="-1"></a>Product_matrix <span class="ot">=</span> TranslationMatrix <span class="sc">%*%</span>  I </span>
<span id="cb81-202"><a href="#cb81-202" aria-hidden="true" tabindex="-1"></a>Product_matrix</span>
<span id="cb81-203"><a href="#cb81-203" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-204"><a href="#cb81-204" aria-hidden="true" tabindex="-1"></a>The product is identical to TranslationMatrix</span>
<span id="cb81-205"><a href="#cb81-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-206"><a href="#cb81-206" aria-hidden="true" tabindex="-1"></a><span class="fu">### 4. Sub-matrix:</span></span>
<span id="cb81-209"><a href="#cb81-209" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-210"><a href="#cb81-210" aria-hidden="true" tabindex="-1"></a><span class="co"># making submatrix A</span></span>
<span id="cb81-211"><a href="#cb81-211" aria-hidden="true" tabindex="-1"></a>A <span class="ot">=</span> <span class="fu">matrix</span>(</span>
<span id="cb81-212"><a href="#cb81-212" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">8</span>,</span>
<span id="cb81-213"><a href="#cb81-213" aria-hidden="true" tabindex="-1"></a>    <span class="dv">15</span>, <span class="dv">12</span>), <span class="at">nrow=</span><span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span></span>
<span id="cb81-214"><a href="#cb81-214" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-215"><a href="#cb81-215" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(A) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"sample1"</span>, <span class="st">"sample2"</span>)</span>
<span id="cb81-216"><a href="#cb81-216" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(A) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"GeneA"</span>, <span class="st">"GeneB"</span>)</span>
<span id="cb81-217"><a href="#cb81-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-218"><a href="#cb81-218" aria-hidden="true" tabindex="-1"></a>A</span>
<span id="cb81-219"><a href="#cb81-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-220"><a href="#cb81-220" aria-hidden="true" tabindex="-1"></a><span class="co"># finding inverse of A</span></span>
<span id="cb81-221"><a href="#cb81-221" aria-hidden="true" tabindex="-1"></a><span class="co">#inv_A &lt;- solve(A)</span></span>
<span id="cb81-222"><a href="#cb81-222" aria-hidden="true" tabindex="-1"></a><span class="co">#inv_A</span></span>
<span id="cb81-223"><a href="#cb81-223" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-224"><a href="#cb81-224" aria-hidden="true" tabindex="-1"></a>The inverse matrix could not be calculated since A is a singular matrix.</span>
<span id="cb81-225"><a href="#cb81-225" aria-hidden="true" tabindex="-1"></a>So, A * A^-1 is also not possible.</span>
<span id="cb81-226"><a href="#cb81-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-227"><a href="#cb81-227" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualization tasks:</span></span>
<span id="cb81-228"><a href="#cb81-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-229"><a href="#cb81-229" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>1. MARplot</span>
<span id="cb81-232"><a href="#cb81-232" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-233"><a href="#cb81-233" aria-hidden="true" tabindex="-1"></a><span class="co"># generating MARplot-style scatter plot</span></span>
<span id="cb81-234"><a href="#cb81-234" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(GeneExpression, Protein_matrix, <span class="at">type=</span><span class="st">"p"</span>, <span class="at">main=</span><span class="st">"Protein level vs. Gene Expression level"</span>)</span>
<span id="cb81-235"><a href="#cb81-235" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> <span class="st">"Sample-Gene"</span></span>
<span id="cb81-236"><a href="#cb81-236" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(GeneExpression, Protein_matrix, <span class="at">labels =</span> labels, <span class="at">pos=</span><span class="dv">3</span>)</span>
<span id="cb81-237"><a href="#cb81-237" aria-hidden="true" tabindex="-1"></a><span class="co"># generating a heatmap</span></span>
<span id="cb81-238"><a href="#cb81-238" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(Protein_matrix, <span class="at">main=</span> <span class="st">"Heatmap of Protein Level"</span>, <span class="at">Rowv =</span> <span class="cn">TRUE</span>, <span class="at">Colv =</span> <span class="cn">TRUE</span>, <span class="at">labRow=</span> <span class="fu">rownames</span>(Protein_matrix), <span class="at">labCol=</span> <span class="fu">c</span>(<span class="st">"ProteinA"</span>, <span class="st">"ProteinB"</span>, <span class="st">"ProteinC"</span>), <span class="at">col=</span><span class="fu">topo.colors</span>(<span class="dv">256</span>) )</span>
<span id="cb81-239"><a href="#cb81-239" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-240"><a href="#cb81-240" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-241"><a href="#cb81-241" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>2. Heatmap of Expression:</span>
<span id="cb81-242"><a href="#cb81-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-245"><a href="#cb81-245" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-246"><a href="#cb81-246" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(GeneExpression, <span class="at">col =</span> <span class="fu">terrain.colors</span>(<span class="dv">10</span>), <span class="at">scale =</span> <span class="st">"column"</span>)</span>
<span id="cb81-247"><a href="#cb81-247" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-248"><a href="#cb81-248" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation</span></span>
<span id="cb81-249"><a href="#cb81-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-250"><a href="#cb81-250" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Matrix multiplication allows each gene in both samples to be multiplied to their respective translation efficiency. So, the product shows how successfully each gene is translated")</span>
<span id="cb81-251"><a href="#cb81-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-252"><a href="#cb81-252" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>The diagonal TranslationMatrix make sense biologically because they show translation efficiency of each gene and there is no other interaction between them. Although there could be interaction in real-world scenarios.</span>
<span id="cb81-253"><a href="#cb81-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-254"><a href="#cb81-254" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>If Sample2 has higher protein levels even with similar gene expression, it means that more mRNAs are translated to proteins compared to Sample1"</span>
<span id="cb81-255"><a href="#cb81-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-256"><a href="#cb81-256" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>The upward trend in MARplot may indicate an increase in translation efficacy and downward trend may indicate a decline in translation efficacy"</span>
<span id="cb81-257"><a href="#cb81-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-258"><a href="#cb81-258" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>Clustering in the heatmap may suggest which samples are most similar to each other based on their prot.</span>
<span id="cb81-259"><a href="#cb81-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-260"><a href="#cb81-260" aria-hidden="true" tabindex="-1"></a>---</span>
<span id="cb81-261"><a href="#cb81-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-262"><a href="#cb81-262" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 3: Animal Breeding – Bull Ranking by Economic Traits</span></span>
<span id="cb81-263"><a href="#cb81-263" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-264"><a href="#cb81-264" aria-hidden="true" tabindex="-1"></a><span class="fu">### Define Data</span></span>
<span id="cb81-265"><a href="#cb81-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-268"><a href="#cb81-268" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-269"><a href="#cb81-269" aria-hidden="true" tabindex="-1"></a><span class="co"># Define Bull EBVs</span></span>
<span id="cb81-270"><a href="#cb81-270" aria-hidden="true" tabindex="-1"></a>BullEBVs <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb81-271"><a href="#cb81-271" aria-hidden="true" tabindex="-1"></a>  <span class="dv">400</span>, <span class="fl">1.2</span>, <span class="fl">0.8</span>,</span>
<span id="cb81-272"><a href="#cb81-272" aria-hidden="true" tabindex="-1"></a>  <span class="dv">500</span>, <span class="fl">1.5</span>, <span class="fl">0.6</span></span>
<span id="cb81-273"><a href="#cb81-273" aria-hidden="true" tabindex="-1"></a>), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb81-274"><a href="#cb81-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-275"><a href="#cb81-275" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(BullEBVs) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Bull1"</span>, <span class="st">"Bull2"</span>)</span>
<span id="cb81-276"><a href="#cb81-276" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(BullEBVs) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Milk_yield"</span>, <span class="st">"Growth_rate"</span>, <span class="st">"Fertility"</span>)</span>
<span id="cb81-277"><a href="#cb81-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-278"><a href="#cb81-278" aria-hidden="true" tabindex="-1"></a><span class="co"># Define Economic Weights</span></span>
<span id="cb81-279"><a href="#cb81-279" aria-hidden="true" tabindex="-1"></a>EconomicWeights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.002</span>, <span class="dv">50</span>, <span class="dv">100</span>), <span class="at">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb81-280"><a href="#cb81-280" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-281"><a href="#cb81-281" aria-hidden="true" tabindex="-1"></a><span class="fu">### Compute Total Economic Value</span></span>
<span id="cb81-282"><a href="#cb81-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-285"><a href="#cb81-285" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-286"><a href="#cb81-286" aria-hidden="true" tabindex="-1"></a>TotalValue <span class="ot">&lt;-</span> BullEBVs <span class="sc">%*%</span> EconomicWeights</span>
<span id="cb81-287"><a href="#cb81-287" aria-hidden="true" tabindex="-1"></a>TotalValue</span>
<span id="cb81-288"><a href="#cb81-288" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-289"><a href="#cb81-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-290"><a href="#cb81-290" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb81-291"><a href="#cb81-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-292"><a href="#cb81-292" aria-hidden="true" tabindex="-1"></a>Bull1: (400 × 0.002) + (1.2 × 50) + (0.8 × 100) = 140.8</span>
<span id="cb81-293"><a href="#cb81-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-294"><a href="#cb81-294" aria-hidden="true" tabindex="-1"></a>Bull2: (500 × 0.002) + (1.5 × 50) + (0.6 × 100) = 136.0</span>
<span id="cb81-295"><a href="#cb81-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-296"><a href="#cb81-296" aria-hidden="true" tabindex="-1"></a>Bull1 is more valuable economically.</span>
<span id="cb81-297"><a href="#cb81-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-298"><a href="#cb81-298" aria-hidden="true" tabindex="-1"></a>**Biological Interpretation**</span>
<span id="cb81-299"><a href="#cb81-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-300"><a href="#cb81-300" aria-hidden="true" tabindex="-1"></a>Economic weights convert genetic merit (EBVs) into economic merit. Traits with higher financial importance have a larger impact, regardless of absolute EBV values.</span>
<span id="cb81-301"><a href="#cb81-301" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-302"><a href="#cb81-302" aria-hidden="true" tabindex="-1"></a><span class="fu">### Multiply with Identity Matrix</span></span>
<span id="cb81-303"><a href="#cb81-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-306"><a href="#cb81-306" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-307"><a href="#cb81-307" aria-hidden="true" tabindex="-1"></a>I3 <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb81-308"><a href="#cb81-308" aria-hidden="true" tabindex="-1"></a>BullEBVs_identity <span class="ot">&lt;-</span> BullEBVs <span class="sc">%*%</span> I3</span>
<span id="cb81-309"><a href="#cb81-309" aria-hidden="true" tabindex="-1"></a>BullEBVs_identity</span>
<span id="cb81-310"><a href="#cb81-310" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-311"><a href="#cb81-311" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-312"><a href="#cb81-312" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb81-313"><a href="#cb81-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-314"><a href="#cb81-314" aria-hidden="true" tabindex="-1"></a>Multiplying by identity matrix returns the original matrix. It confirms that EBV structure is preserved.</span>
<span id="cb81-315"><a href="#cb81-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-316"><a href="#cb81-316" aria-hidden="true" tabindex="-1"></a><span class="fu">### Remove Milk Yield and Recalculate Total Value</span></span>
<span id="cb81-317"><a href="#cb81-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-320"><a href="#cb81-320" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-321"><a href="#cb81-321" aria-hidden="true" tabindex="-1"></a>BullEBVs_noMilk <span class="ot">&lt;-</span> BullEBVs[, <span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb81-322"><a href="#cb81-322" aria-hidden="true" tabindex="-1"></a>EconomicWeights_noMilk <span class="ot">&lt;-</span> EconomicWeights[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb81-323"><a href="#cb81-323" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-324"><a href="#cb81-324" aria-hidden="true" tabindex="-1"></a>TotalValue_noMilk <span class="ot">&lt;-</span> BullEBVs_noMilk <span class="sc">%*%</span> EconomicWeights_noMilk</span>
<span id="cb81-325"><a href="#cb81-325" aria-hidden="true" tabindex="-1"></a>TotalValue_noMilk</span>
<span id="cb81-326"><a href="#cb81-326" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-327"><a href="#cb81-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-328"><a href="#cb81-328" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb81-329"><a href="#cb81-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-330"><a href="#cb81-330" aria-hidden="true" tabindex="-1"></a>Bull1: (1.2 × 50) + (0.8 × 100) = 140</span>
<span id="cb81-331"><a href="#cb81-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-332"><a href="#cb81-332" aria-hidden="true" tabindex="-1"></a>Bull2: (1.5 × 50) + (0.6 × 100) = 135</span>
<span id="cb81-333"><a href="#cb81-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-334"><a href="#cb81-334" aria-hidden="true" tabindex="-1"></a>Bull1 still ranks higher, but by a smaller margin.</span>
<span id="cb81-335"><a href="#cb81-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-336"><a href="#cb81-336" aria-hidden="true" tabindex="-1"></a><span class="fu">### Bar Plot: Total Economic Value</span></span>
<span id="cb81-337"><a href="#cb81-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-340"><a href="#cb81-340" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-341"><a href="#cb81-341" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb81-342"><a href="#cb81-342" aria-hidden="true" tabindex="-1"></a>  TotalValue,</span>
<span id="cb81-343"><a href="#cb81-343" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb81-344"><a href="#cb81-344" aria-hidden="true" tabindex="-1"></a>  <span class="at">names.arg =</span> <span class="fu">rownames</span>(BullEBVs),</span>
<span id="cb81-345"><a href="#cb81-345" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"skyblue"</span>, <span class="st">"orange"</span>),</span>
<span id="cb81-346"><a href="#cb81-346" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Total Economic Value of Bulls"</span>,</span>
<span id="cb81-347"><a href="#cb81-347" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Total Value"</span></span>
<span id="cb81-348"><a href="#cb81-348" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-349"><a href="#cb81-349" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-350"><a href="#cb81-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-351"><a href="#cb81-351" aria-hidden="true" tabindex="-1"></a><span class="fu">### Heatmap of EBVs</span></span>
<span id="cb81-352"><a href="#cb81-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-355"><a href="#cb81-355" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-356"><a href="#cb81-356" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb81-357"><a href="#cb81-357" aria-hidden="true" tabindex="-1"></a>  BullEBVs,</span>
<span id="cb81-358"><a href="#cb81-358" aria-hidden="true" tabindex="-1"></a>  <span class="at">Rowv =</span> <span class="cn">NA</span>,</span>
<span id="cb81-359"><a href="#cb81-359" aria-hidden="true" tabindex="-1"></a>  <span class="at">Colv =</span> <span class="cn">NA</span>,</span>
<span id="cb81-360"><a href="#cb81-360" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb81-361"><a href="#cb81-361" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">256</span>),</span>
<span id="cb81-362"><a href="#cb81-362" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Heatmap of Bull EBVs"</span></span>
<span id="cb81-363"><a href="#cb81-363" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-364"><a href="#cb81-364" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-365"><a href="#cb81-365" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-366"><a href="#cb81-366" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation Questions</span></span>
<span id="cb81-367"><a href="#cb81-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-368"><a href="#cb81-368" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How do economic weights affect trait importance?</span>
<span id="cb81-369"><a href="#cb81-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-370"><a href="#cb81-370" aria-hidden="true" tabindex="-1"></a>Traits with higher weights contribute more to the total economic value. This makes them more influential in ranking and selection.</span>
<span id="cb81-371"><a href="#cb81-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-372"><a href="#cb81-372" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Why might you ignore milk yield?</span>
<span id="cb81-373"><a href="#cb81-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-374"><a href="#cb81-374" aria-hidden="true" tabindex="-1"></a>Milk yield may be excluded in systems focusing on fertility, growth, or when it is no longer a limiting factor. Environmental or economic contexts may also shift trait priorities.</span>
<span id="cb81-375"><a href="#cb81-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-376"><a href="#cb81-376" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What is the value of heatmaps?</span>
<span id="cb81-377"><a href="#cb81-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-378"><a href="#cb81-378" aria-hidden="true" tabindex="-1"></a>Heatmaps visually compare EBVs across bulls and traits.They help detect patterns, outliers, and clusters easily in multivariate data.</span>
<span id="cb81-379"><a href="#cb81-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-380"><a href="#cb81-380" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Can this method be extended to more bulls and traits?</span>
<span id="cb81-381"><a href="#cb81-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-382"><a href="#cb81-382" aria-hidden="true" tabindex="-1"></a>Yes. This method scales to any number of bulls or traits. Just ensure the EBVs matrix and economic weights are dimensionally compatible.</span>
<span id="cb81-383"><a href="#cb81-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-384"><a href="#cb81-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-385"><a href="#cb81-385" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 4: Plant Breeding – Trait Contributions from Parental Lines</span></span>
<span id="cb81-386"><a href="#cb81-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-387"><a href="#cb81-387" aria-hidden="true" tabindex="-1"></a><span class="fu">### Define Data</span></span>
<span id="cb81-388"><a href="#cb81-388" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-389"><a href="#cb81-389" aria-hidden="true" tabindex="-1"></a>**Parent trait values (normalized 1–10)**</span>
<span id="cb81-392"><a href="#cb81-392" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-393"><a href="#cb81-393" aria-hidden="true" tabindex="-1"></a>ParentTraits <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(</span>
<span id="cb81-394"><a href="#cb81-394" aria-hidden="true" tabindex="-1"></a>  <span class="dv">7</span>, <span class="dv">5</span>, <span class="dv">3</span>,</span>
<span id="cb81-395"><a href="#cb81-395" aria-hidden="true" tabindex="-1"></a>  <span class="dv">6</span>, <span class="dv">8</span>, <span class="dv">4</span>,</span>
<span id="cb81-396"><a href="#cb81-396" aria-hidden="true" tabindex="-1"></a>  <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">6</span></span>
<span id="cb81-397"><a href="#cb81-397" aria-hidden="true" tabindex="-1"></a>), <span class="at">nrow =</span> <span class="dv">3</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>)</span>
<span id="cb81-398"><a href="#cb81-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-399"><a href="#cb81-399" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ParentTraits) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"P1"</span>, <span class="st">"P2"</span>, <span class="st">"P3"</span>)</span>
<span id="cb81-400"><a href="#cb81-400" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(ParentTraits) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Drought_resistance"</span>, <span class="st">"Yield"</span>, <span class="st">"Maturation_time"</span>)</span>
<span id="cb81-401"><a href="#cb81-401" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-402"><a href="#cb81-402" aria-hidden="true" tabindex="-1"></a><span class="co"># Define Hybrid Weights</span></span>
<span id="cb81-403"><a href="#cb81-403" aria-hidden="true" tabindex="-1"></a>HybridWeights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span>), <span class="at">nrow =</span> <span class="dv">1</span>)</span>
<span id="cb81-404"><a href="#cb81-404" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-405"><a href="#cb81-405" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-406"><a href="#cb81-406" aria-hidden="true" tabindex="-1"></a><span class="fu">### Compute HybridTraits Vector</span></span>
<span id="cb81-407"><a href="#cb81-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-410"><a href="#cb81-410" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-411"><a href="#cb81-411" aria-hidden="true" tabindex="-1"></a>HybridTraits <span class="ot">&lt;-</span> HybridWeights <span class="sc">%*%</span> ParentTraits</span>
<span id="cb81-412"><a href="#cb81-412" aria-hidden="true" tabindex="-1"></a>HybridTraits</span>
<span id="cb81-413"><a href="#cb81-413" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-414"><a href="#cb81-414" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb81-415"><a href="#cb81-415" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-416"><a href="#cb81-416" aria-hidden="true" tabindex="-1"></a>HybridTraits = (0.5 × P1) + (0.3 × P2) + (0.2 × P3)</span>
<span id="cb81-417"><a href="#cb81-417" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-418"><a href="#cb81-418" aria-hidden="true" tabindex="-1"></a>Drought_resistance = (0.5 × 7) + (0.3 × 6) + (0.2 × 5) = 6.2</span>
<span id="cb81-419"><a href="#cb81-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-420"><a href="#cb81-420" aria-hidden="true" tabindex="-1"></a>Yield = (0.5 × 5) + (0.3 × 8) + (0.2 × 6) = 6.1</span>
<span id="cb81-421"><a href="#cb81-421" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-422"><a href="#cb81-422" aria-hidden="true" tabindex="-1"></a>Maturation_time = (0.5 × 3) + (0.3 × 4) + (0.2 × 6) = 3.7</span>
<span id="cb81-423"><a href="#cb81-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-424"><a href="#cb81-424" aria-hidden="true" tabindex="-1"></a>The hybrid is moderately strong in drought resistance and yield, but has a relatively shorter maturation time.</span>
<span id="cb81-425"><a href="#cb81-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-426"><a href="#cb81-426" aria-hidden="true" tabindex="-1"></a>**Biological Meaning of Unequal Contribution**</span>
<span id="cb81-427"><a href="#cb81-427" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-428"><a href="#cb81-428" aria-hidden="true" tabindex="-1"></a>When one parent contributes more to a trait, it suggests that the trait's heritable strength comes disproportionately from that parent. Breeders can use this knowledge to amplify desirable traits.</span>
<span id="cb81-429"><a href="#cb81-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-430"><a href="#cb81-430" aria-hidden="true" tabindex="-1"></a><span class="fu">### Multiply with Identity Matrix</span></span>
<span id="cb81-431"><a href="#cb81-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-434"><a href="#cb81-434" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-435"><a href="#cb81-435" aria-hidden="true" tabindex="-1"></a>I3 <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="dv">3</span>)</span>
<span id="cb81-436"><a href="#cb81-436" aria-hidden="true" tabindex="-1"></a>ParentTraits_identity <span class="ot">&lt;-</span> ParentTraits <span class="sc">%*%</span> I3</span>
<span id="cb81-437"><a href="#cb81-437" aria-hidden="true" tabindex="-1"></a>ParentTraits_identity</span>
<span id="cb81-438"><a href="#cb81-438" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-439"><a href="#cb81-439" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-440"><a href="#cb81-440" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb81-441"><a href="#cb81-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-442"><a href="#cb81-442" aria-hidden="true" tabindex="-1"></a>Multiplying by identity matrix returns the original matrix. This operation verifies structural consistency and dimensionality.</span>
<span id="cb81-443"><a href="#cb81-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-444"><a href="#cb81-444" aria-hidden="true" tabindex="-1"></a><span class="fu">### Remove T3 (Maturation Time) and Recalculate</span></span>
<span id="cb81-445"><a href="#cb81-445" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-448"><a href="#cb81-448" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-449"><a href="#cb81-449" aria-hidden="true" tabindex="-1"></a>ParentTraits_T1T2 <span class="ot">&lt;-</span> ParentTraits[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb81-450"><a href="#cb81-450" aria-hidden="true" tabindex="-1"></a>HybridTraits_T1T2 <span class="ot">&lt;-</span> HybridWeights <span class="sc">%*%</span> ParentTraits_T1T2</span>
<span id="cb81-451"><a href="#cb81-451" aria-hidden="true" tabindex="-1"></a>HybridTraits_T1T2</span>
<span id="cb81-452"><a href="#cb81-452" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-453"><a href="#cb81-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-454"><a href="#cb81-454" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb81-455"><a href="#cb81-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-456"><a href="#cb81-456" aria-hidden="true" tabindex="-1"></a>Drought_resistance = (0.5 × 7) + (0.3 × 6) + (0.2 × 5) = 6.2</span>
<span id="cb81-457"><a href="#cb81-457" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-458"><a href="#cb81-458" aria-hidden="true" tabindex="-1"></a>Yield = (0.5 × 5) + (0.3 × 8) + (0.2 × 6) = 6.1</span>
<span id="cb81-459"><a href="#cb81-459" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-460"><a href="#cb81-460" aria-hidden="true" tabindex="-1"></a>Removing a trait (T3) changes the trait profile. Hybrid selection may now favor traits that remain.</span>
<span id="cb81-461"><a href="#cb81-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-462"><a href="#cb81-462" aria-hidden="true" tabindex="-1"></a><span class="fu">### Heatmap of ParentTraits</span></span>
<span id="cb81-463"><a href="#cb81-463" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-466"><a href="#cb81-466" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-467"><a href="#cb81-467" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb81-468"><a href="#cb81-468" aria-hidden="true" tabindex="-1"></a>  ParentTraits,</span>
<span id="cb81-469"><a href="#cb81-469" aria-hidden="true" tabindex="-1"></a>  <span class="at">Rowv =</span> <span class="cn">NA</span>,</span>
<span id="cb81-470"><a href="#cb81-470" aria-hidden="true" tabindex="-1"></a>  <span class="at">Colv =</span> <span class="cn">NA</span>,</span>
<span id="cb81-471"><a href="#cb81-471" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb81-472"><a href="#cb81-472" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">256</span>),</span>
<span id="cb81-473"><a href="#cb81-473" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Heatmap of Parent Traits"</span></span>
<span id="cb81-474"><a href="#cb81-474" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-475"><a href="#cb81-475" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-476"><a href="#cb81-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-477"><a href="#cb81-477" aria-hidden="true" tabindex="-1"></a><span class="fu">### Bar Plot of Hybrid Traits</span></span>
<span id="cb81-478"><a href="#cb81-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-481"><a href="#cb81-481" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-482"><a href="#cb81-482" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb81-483"><a href="#cb81-483" aria-hidden="true" tabindex="-1"></a>  HybridTraits,</span>
<span id="cb81-484"><a href="#cb81-484" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb81-485"><a href="#cb81-485" aria-hidden="true" tabindex="-1"></a>  <span class="at">names.arg =</span> <span class="fu">colnames</span>(ParentTraits),</span>
<span id="cb81-486"><a href="#cb81-486" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"#66c2a5"</span>, <span class="st">"#fc8d62"</span>, <span class="st">"#8da0cb"</span>),</span>
<span id="cb81-487"><a href="#cb81-487" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Hybrid Trait Profile"</span>,</span>
<span id="cb81-488"><a href="#cb81-488" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Trait Value"</span></span>
<span id="cb81-489"><a href="#cb81-489" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-490"><a href="#cb81-490" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-491"><a href="#cb81-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-492"><a href="#cb81-492" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation Questions</span></span>
<span id="cb81-493"><a href="#cb81-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-494"><a href="#cb81-494" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How does the weighting of parents affect the hybrid's performance?</span>
<span id="cb81-495"><a href="#cb81-495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-496"><a href="#cb81-496" aria-hidden="true" tabindex="-1"></a>  Stronger weights mean more genetic contribution. Traits from highly weighted parents dominate the hybrid profile.</span>
<span id="cb81-497"><a href="#cb81-497" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-498"><a href="#cb81-498" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What does the identity matrix represent here?</span>
<span id="cb81-499"><a href="#cb81-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-500"><a href="#cb81-500" aria-hidden="true" tabindex="-1"></a>  It represents a neutral transformation. It confirms data integrity when used in matrix multiplication.</span>
<span id="cb81-501"><a href="#cb81-501" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-502"><a href="#cb81-502" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>If you used equal weights (⅓ for each), how would the hybrid traits change?</span>
<span id="cb81-503"><a href="#cb81-503" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-504"><a href="#cb81-504" aria-hidden="true" tabindex="-1"></a>  Traits would reflect an even mix, potentially leading to balanced but less specialized performance.</span>
<span id="cb81-505"><a href="#cb81-505" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-506"><a href="#cb81-506" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What real-world limitations does this simplified model ignore?</span>
<span id="cb81-507"><a href="#cb81-507" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-508"><a href="#cb81-508" aria-hidden="true" tabindex="-1"></a>  i. Non-additive genetic effects (dominance, epistasis)</span>
<span id="cb81-509"><a href="#cb81-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-510"><a href="#cb81-510" aria-hidden="true" tabindex="-1"></a>  ii. Environmental interactions</span>
<span id="cb81-511"><a href="#cb81-511" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-512"><a href="#cb81-512" aria-hidden="true" tabindex="-1"></a>  iii. Trait heritability and correlations</span>
<span id="cb81-513"><a href="#cb81-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-514"><a href="#cb81-514" aria-hidden="true" tabindex="-1"></a>  iv. Breeding feasibility and cost</span>
<span id="cb81-515"><a href="#cb81-515" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-516"><a href="#cb81-516" aria-hidden="true" tabindex="-1"></a><span class="fu">## Task 5: Managing Matrices and Weight Vectors Using Lists in R</span></span>
<span id="cb81-517"><a href="#cb81-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-518"><a href="#cb81-518" aria-hidden="true" tabindex="-1"></a>Now that we've explored trait-based decisions using matrices, it's time to organize our work using R's list structure. Lists help bundle related objects like matrices and weight vectors, keeping the analysis modular and scalable.</span>
<span id="cb81-519"><a href="#cb81-519" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-520"><a href="#cb81-520" aria-hidden="true" tabindex="-1"></a><span class="fu">### Create a Master List</span></span>
<span id="cb81-521"><a href="#cb81-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-524"><a href="#cb81-524" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-525"><a href="#cb81-525" aria-hidden="true" tabindex="-1"></a><span class="co"># Assuming previous matrices and weights are already defined:</span></span>
<span id="cb81-526"><a href="#cb81-526" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-527"><a href="#cb81-527" aria-hidden="true" tabindex="-1"></a><span class="co"># Making a MasterList</span></span>
<span id="cb81-528"><a href="#cb81-528" aria-hidden="true" tabindex="-1"></a>bioList <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb81-529"><a href="#cb81-529" aria-hidden="true" tabindex="-1"></a>  <span class="at">ProteinConc =</span> <span class="fu">list</span>(<span class="at">matrix =</span> ProteinMatrix, <span class="at">weights =</span> WeightVector),</span>
<span id="cb81-530"><a href="#cb81-530" aria-hidden="true" tabindex="-1"></a>  <span class="at">ProteinMap  =</span> <span class="fu">list</span>(<span class="at">matrix =</span> GeneExpression, <span class="at">weights =</span> TranslationMatrix),</span>
<span id="cb81-531"><a href="#cb81-531" aria-hidden="true" tabindex="-1"></a>  <span class="at">Animal =</span> <span class="fu">list</span>(<span class="at">matrix =</span> BullEBVs, <span class="at">weights =</span> EconomicWeights),</span>
<span id="cb81-532"><a href="#cb81-532" aria-hidden="true" tabindex="-1"></a>  <span class="at">Plant =</span> <span class="fu">list</span>(<span class="at">matrix =</span> ParentTraits, <span class="at">weights =</span> HybridWeights)</span>
<span id="cb81-533"><a href="#cb81-533" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-534"><a href="#cb81-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-535"><a href="#cb81-535" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(bioList)</span>
<span id="cb81-536"><a href="#cb81-536" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-537"><a href="#cb81-537" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-538"><a href="#cb81-538" aria-hidden="true" tabindex="-1"></a><span class="fu">### List the Structure</span></span>
<span id="cb81-539"><a href="#cb81-539" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-542"><a href="#cb81-542" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-543"><a href="#cb81-543" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bioList)               <span class="co"># Top-level list names</span></span>
<span id="cb81-544"><a href="#cb81-544" aria-hidden="true" tabindex="-1"></a><span class="fu">lengths</span>(bioList)             <span class="co"># Number of components in each sublist</span></span>
<span id="cb81-545"><a href="#cb81-545" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-546"><a href="#cb81-546" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-547"><a href="#cb81-547" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb81-548"><a href="#cb81-548" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-549"><a href="#cb81-549" aria-hidden="true" tabindex="-1"></a>Each top-level entry (e.g., ProteinConc, Plant) contains two components:</span>
<span id="cb81-550"><a href="#cb81-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-551"><a href="#cb81-551" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>A matrix (e.g., ProteinMatrix)</span>
<span id="cb81-552"><a href="#cb81-552" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>A corresponding weight vector or matrix</span>
<span id="cb81-553"><a href="#cb81-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-554"><a href="#cb81-554" aria-hidden="true" tabindex="-1"></a><span class="fu">### Indexing Elements from Lists</span></span>
<span id="cb81-555"><a href="#cb81-555" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-558"><a href="#cb81-558" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-559"><a href="#cb81-559" aria-hidden="true" tabindex="-1"></a><span class="co"># Access the trait matrix for Plant</span></span>
<span id="cb81-560"><a href="#cb81-560" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Plant[[<span class="dv">1</span>]]</span>
<span id="cb81-561"><a href="#cb81-561" aria-hidden="true" tabindex="-1"></a><span class="co">#or</span></span>
<span id="cb81-562"><a href="#cb81-562" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix</span>
<span id="cb81-563"><a href="#cb81-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-564"><a href="#cb81-564" aria-hidden="true" tabindex="-1"></a><span class="co"># Access the weight vector for ProteinConc</span></span>
<span id="cb81-565"><a href="#cb81-565" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinConc[[<span class="dv">2</span>]]</span>
<span id="cb81-566"><a href="#cb81-566" aria-hidden="true" tabindex="-1"></a><span class="co">#or</span></span>
<span id="cb81-567"><a href="#cb81-567" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinConc<span class="sc">$</span>weights</span>
<span id="cb81-568"><a href="#cb81-568" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-569"><a href="#cb81-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-570"><a href="#cb81-570" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb81-571"><a href="#cb81-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-572"><a href="#cb81-572" aria-hidden="true" tabindex="-1"></a>Use double brackets [<span class="co">[ ]</span>] to extract unnamed list elements by position. But we named our list, so they are easily extractable using the <span class="in">`$`</span> notation.</span>
<span id="cb81-573"><a href="#cb81-573" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-574"><a href="#cb81-574" aria-hidden="true" tabindex="-1"></a><span class="fu">### Perform Weighted Calculations</span></span>
<span id="cb81-575"><a href="#cb81-575" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-578"><a href="#cb81-578" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-579"><a href="#cb81-579" aria-hidden="true" tabindex="-1"></a><span class="co"># Protein concentration score</span></span>
<span id="cb81-580"><a href="#cb81-580" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinConc<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>ProteinConc<span class="sc">$</span>weights</span>
<span id="cb81-581"><a href="#cb81-581" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-582"><a href="#cb81-582" aria-hidden="true" tabindex="-1"></a><span class="co"># Gene → Protein contribution</span></span>
<span id="cb81-583"><a href="#cb81-583" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>ProteinMap<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>ProteinMap<span class="sc">$</span>weights</span>
<span id="cb81-584"><a href="#cb81-584" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-585"><a href="#cb81-585" aria-hidden="true" tabindex="-1"></a><span class="co"># Bull economic value</span></span>
<span id="cb81-586"><a href="#cb81-586" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Animal<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>Animal<span class="sc">$</span>weights</span>
<span id="cb81-587"><a href="#cb81-587" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-588"><a href="#cb81-588" aria-hidden="true" tabindex="-1"></a><span class="co"># Hybrid trait value</span></span>
<span id="cb81-589"><a href="#cb81-589" aria-hidden="true" tabindex="-1"></a>bioList<span class="sc">$</span>Plant<span class="sc">$</span>weights <span class="sc">%*%</span> bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix</span>
<span id="cb81-590"><a href="#cb81-590" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-591"><a href="#cb81-591" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-592"><a href="#cb81-592" aria-hidden="true" tabindex="-1"></a><span class="fu">### Subset and Recalculate</span></span>
<span id="cb81-593"><a href="#cb81-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-596"><a href="#cb81-596" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-597"><a href="#cb81-597" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove last trait from ParentTraits</span></span>
<span id="cb81-598"><a href="#cb81-598" aria-hidden="true" tabindex="-1"></a>ParentSubset <span class="ot">&lt;-</span> bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb81-599"><a href="#cb81-599" aria-hidden="true" tabindex="-1"></a>NewWeights <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">c</span>(<span class="fl">0.6</span>, <span class="fl">0.4</span>), <span class="at">nrow =</span> <span class="dv">2</span>)</span>
<span id="cb81-600"><a href="#cb81-600" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-601"><a href="#cb81-601" aria-hidden="true" tabindex="-1"></a><span class="co"># Recalculated hybrid score</span></span>
<span id="cb81-602"><a href="#cb81-602" aria-hidden="true" tabindex="-1"></a>SubsetHybridScore <span class="ot">&lt;-</span> ParentSubset <span class="sc">%*%</span> NewWeights</span>
<span id="cb81-603"><a href="#cb81-603" aria-hidden="true" tabindex="-1"></a>SubsetHybridScore</span>
<span id="cb81-604"><a href="#cb81-604" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-605"><a href="#cb81-605" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-606"><a href="#cb81-606" aria-hidden="true" tabindex="-1"></a>**Interpretation**</span>
<span id="cb81-607"><a href="#cb81-607" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-608"><a href="#cb81-608" aria-hidden="true" tabindex="-1"></a>Dropping a trait and reweighting highlights its influence in trait aggregation and selection.</span>
<span id="cb81-609"><a href="#cb81-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-610"><a href="#cb81-610" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualization Tasks</span></span>
<span id="cb81-611"><a href="#cb81-611" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-612"><a href="#cb81-612" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Heatmap: Gene Expression</span></span>
<span id="cb81-613"><a href="#cb81-613" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-616"><a href="#cb81-616" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-617"><a href="#cb81-617" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(</span>
<span id="cb81-618"><a href="#cb81-618" aria-hidden="true" tabindex="-1"></a>  bioList<span class="sc">$</span>ProteinMap<span class="sc">$</span>matrix,</span>
<span id="cb81-619"><a href="#cb81-619" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"none"</span>,</span>
<span id="cb81-620"><a href="#cb81-620" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">256</span>),</span>
<span id="cb81-621"><a href="#cb81-621" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Gene Expression Heatmap"</span>,</span>
<span id="cb81-622"><a href="#cb81-622" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Proteins"</span>,</span>
<span id="cb81-623"><a href="#cb81-623" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Genes"</span></span>
<span id="cb81-624"><a href="#cb81-624" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-625"><a href="#cb81-625" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-626"><a href="#cb81-626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-627"><a href="#cb81-627" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Bar Plots</span></span>
<span id="cb81-628"><a href="#cb81-628" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-629"><a href="#cb81-629" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Hybrid traits</span></span>
<span id="cb81-630"><a href="#cb81-630" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-633"><a href="#cb81-633" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-634"><a href="#cb81-634" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb81-635"><a href="#cb81-635" aria-hidden="true" tabindex="-1"></a>  bioList<span class="sc">$</span>Plant<span class="sc">$</span>weights <span class="sc">%*%</span> bioList<span class="sc">$</span>Plant<span class="sc">$</span>matrix,</span>
<span id="cb81-636"><a href="#cb81-636" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb81-637"><a href="#cb81-637" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Hybrid Trait Contributions"</span>,</span>
<span id="cb81-638"><a href="#cb81-638" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="st">"#66c2a5"</span>,</span>
<span id="cb81-639"><a href="#cb81-639" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Score"</span></span>
<span id="cb81-640"><a href="#cb81-640" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-641"><a href="#cb81-641" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-642"><a href="#cb81-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-643"><a href="#cb81-643" aria-hidden="true" tabindex="-1"></a><span class="fu">##### Bull EBV (Economic Breeding Values)</span></span>
<span id="cb81-644"><a href="#cb81-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-647"><a href="#cb81-647" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb81-648"><a href="#cb81-648" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(</span>
<span id="cb81-649"><a href="#cb81-649" aria-hidden="true" tabindex="-1"></a>  bioList<span class="sc">$</span>Animal<span class="sc">$</span>matrix <span class="sc">%*%</span> bioList<span class="sc">$</span>Animal<span class="sc">$</span>weights,</span>
<span id="cb81-650"><a href="#cb81-650" aria-hidden="true" tabindex="-1"></a>  <span class="at">beside =</span> <span class="cn">TRUE</span>,</span>
<span id="cb81-651"><a href="#cb81-651" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Bull EBVs (Economic Values)"</span>,</span>
<span id="cb81-652"><a href="#cb81-652" aria-hidden="true" tabindex="-1"></a>  <span class="at">col =</span> <span class="st">"#fc8d62"</span>,</span>
<span id="cb81-653"><a href="#cb81-653" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Score"</span></span>
<span id="cb81-654"><a href="#cb81-654" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb81-655"><a href="#cb81-655" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb81-656"><a href="#cb81-656" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-657"><a href="#cb81-657" aria-hidden="true" tabindex="-1"></a><span class="fu">### 🧠 Interpretation Questions</span></span>
<span id="cb81-658"><a href="#cb81-658" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-659"><a href="#cb81-659" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Why use a list structure?</span>
<span id="cb81-660"><a href="#cb81-660" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-661"><a href="#cb81-661" aria-hidden="true" tabindex="-1"></a>Keeps each dataset and its weights together.</span>
<span id="cb81-662"><a href="#cb81-662" aria-hidden="true" tabindex="-1"></a>Facilitates automated workflows and reuse.</span>
<span id="cb81-663"><a href="#cb81-663" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-664"><a href="#cb81-664" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What’s tricky about <span class="in">`[[ ]]`</span> access?</span>
<span id="cb81-665"><a href="#cb81-665" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-666"><a href="#cb81-666" aria-hidden="true" tabindex="-1"></a>You must remember the order (<span class="in">`[[1]]`</span> = <span class="in">`matrix`</span>, <span class="in">`[[2]]`</span> = <span class="in">`weights`</span>).</span>
<span id="cb81-667"><a href="#cb81-667" aria-hidden="true" tabindex="-1"></a>No names means you can't use $matrix, only positional access.</span>
<span id="cb81-668"><a href="#cb81-668" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-669"><a href="#cb81-669" aria-hidden="true" tabindex="-1"></a>Loop across all list entries</span>
<span id="cb81-670"><a href="#cb81-670" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-671"><a href="#cb81-671" aria-hidden="true" tabindex="-1"></a>Weighted scores for all entries</span>
<span id="cb81-672"><a href="#cb81-672" aria-hidden="true" tabindex="-1"></a><span class="in">`lapply(bioList, function(x) x[[2]] %*% x[[1]])`</span></span>
<span id="cb81-673"><a href="#cb81-673" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-674"><a href="#cb81-674" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How does this help in large-scale pipelines?</span>
<span id="cb81-675"><a href="#cb81-675" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-676"><a href="#cb81-676" aria-hidden="true" tabindex="-1"></a>You can use this format with <span class="in">`lapply()`</span>, <span class="in">`purrr::map()`</span>, or in targets pipelines for reproducibility and modular processing.</span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>